{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Capítulo 4 - Objetos de anotação de sequência"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "No Capítulo 3 (https://github.com/GTL98/biopython_br/tree/main/Cap%203%20-%20Objetos%20de%20Sequ%C3%AAncia) foi introduzida as classes de sequência. \"Acima\" das classes **Seq** está a classe de Registro de Sequência, ou **SeqRecord**, definida pelo módulo `Bio.SeqRecord`. Essa classe tem ferramentas de nível superior como identificadores e recursos (como os objetos **SeqFeature**) associados a sequência e usado ao longo da interface de entrada/saída do módulo `Bio.SeqIO` (mais detalhes no Capítulo 5: https://github.com/GTL98/biopython_br/tree/main/Cap%205%20-%20Sequ%C3%AAncias%20Input_Output).\n",
    "\n",
    "Se você trabalhará somente com arquivos FASTA, você provavelmente pulará essa parte por agora. Mas se por outro lado você usar arquivos de sequências com ricas anotações, como GenBank ou EMBL, as informações a seguir são de suma importância.\n",
    "\n",
    "Esse capítulo deverá explicar muitas coisas que os objetos **SeqRecord** e **SeqFeature** podem fazer. Você também está convidado a ler documentação online do objeto **SeqRecord** (https://biopython.org/docs/1.79/api/Bio.SeqRecord.html) e do objeto **SeqFeature** (https://biopython.org/docs/1.79/api/Bio.SeqFeature.html).\n",
    "\n",
    "Caso tenha alguma dúvida, basta chamar o método **help** do Python:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on class SeqRecord in module Bio.SeqRecord:\n",
      "\n",
      "class SeqRecord(builtins.object)\n",
      " |  SeqRecord(seq, id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=None, features=None, annotations=None, letter_annotations=None)\n",
      " |  \n",
      " |  A SeqRecord object holds a sequence and information about it.\n",
      " |  \n",
      " |  Main attributes:\n",
      " |   - id          - Identifier such as a locus tag (string)\n",
      " |   - seq         - The sequence itself (Seq object or similar)\n",
      " |  \n",
      " |  Additional attributes:\n",
      " |   - name        - Sequence name, e.g. gene name (string)\n",
      " |   - description - Additional text (string)\n",
      " |   - dbxrefs     - List of database cross references (list of strings)\n",
      " |   - features    - Any (sub)features defined (list of SeqFeature objects)\n",
      " |   - annotations - Further information about the whole sequence (dictionary).\n",
      " |     Most entries are strings, or lists of strings.\n",
      " |   - letter_annotations - Per letter/symbol annotation (restricted\n",
      " |     dictionary). This holds Python sequences (lists, strings\n",
      " |     or tuples) whose length matches that of the sequence.\n",
      " |     A typical use would be to hold a list of integers\n",
      " |     representing sequencing quality scores, or a string\n",
      " |     representing the secondary structure.\n",
      " |  \n",
      " |  You will typically use Bio.SeqIO to read in sequences from files as\n",
      " |  SeqRecord objects.  However, you may want to create your own SeqRecord\n",
      " |  objects directly (see the __init__ method for further details):\n",
      " |  \n",
      " |  >>> from Bio.Seq import Seq\n",
      " |  >>> from Bio.SeqRecord import SeqRecord\n",
      " |  >>> record = SeqRecord(Seq(\"MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\"),\n",
      " |  ...                    id=\"YP_025292.1\", name=\"HokC\",\n",
      " |  ...                    description=\"toxic membrane protein\")\n",
      " |  >>> print(record)\n",
      " |  ID: YP_025292.1\n",
      " |  Name: HokC\n",
      " |  Description: toxic membrane protein\n",
      " |  Number of features: 0\n",
      " |  Seq('MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF')\n",
      " |  \n",
      " |  If you want to save SeqRecord objects to a sequence file, use Bio.SeqIO\n",
      " |  for this.  For the special case where you want the SeqRecord turned into\n",
      " |  a string in a particular file format there is a format method which uses\n",
      " |  Bio.SeqIO internally:\n",
      " |  \n",
      " |  >>> print(record.format(\"fasta\"))\n",
      " |  >YP_025292.1 toxic membrane protein\n",
      " |  MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\n",
      " |  <BLANKLINE>\n",
      " |  \n",
      " |  You can also do things like slicing a SeqRecord, checking its length, etc\n",
      " |  \n",
      " |  >>> len(record)\n",
      " |  44\n",
      " |  >>> edited = record[:10] + record[11:]\n",
      " |  >>> print(edited.seq)\n",
      " |  MKQHKAMIVAIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\n",
      " |  >>> print(record.seq)\n",
      " |  MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __add__(self, other)\n",
      " |      Add another sequence or string to this sequence.\n",
      " |      \n",
      " |      The other sequence can be a SeqRecord object, a Seq object (or\n",
      " |      similar, e.g. a MutableSeq) or a plain Python string. If you add\n",
      " |      a plain string or a Seq (like) object, the new SeqRecord will simply\n",
      " |      have this appended to the existing data. However, any per letter\n",
      " |      annotation will be lost:\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Quality/solexa_faked.fastq\", \"fastq-solexa\")\n",
      " |      >>> print(\"%s %s\" % (record.id, record.seq))\n",
      " |      slxa_0001_1_0001_01 ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTNNNNNN\n",
      " |      >>> print(list(record.letter_annotations))\n",
      " |      ['solexa_quality']\n",
      " |      \n",
      " |      >>> new = record + \"ACT\"\n",
      " |      >>> print(\"%s %s\" % (new.id, new.seq))\n",
      " |      slxa_0001_1_0001_01 ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTNNNNNNACT\n",
      " |      >>> print(list(new.letter_annotations))\n",
      " |      []\n",
      " |      \n",
      " |      The new record will attempt to combine the annotation, but for any\n",
      " |      ambiguities (e.g. different names) it defaults to omitting that\n",
      " |      annotation.\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> with open(\"GenBank/pBAD30.gb\") as handle:\n",
      " |      ...     plasmid = SeqIO.read(handle, \"gb\")\n",
      " |      >>> print(\"%s %i\" % (plasmid.id, len(plasmid)))\n",
      " |      pBAD30 4923\n",
      " |      \n",
      " |      Now let's cut the plasmid into two pieces, and join them back up the\n",
      " |      other way round (i.e. shift the starting point on this plasmid, have\n",
      " |      a look at the annotated features in the original file to see why this\n",
      " |      particular split point might make sense):\n",
      " |      \n",
      " |      >>> left = plasmid[:3765]\n",
      " |      >>> right = plasmid[3765:]\n",
      " |      >>> new = right + left\n",
      " |      >>> print(\"%s %i\" % (new.id, len(new)))\n",
      " |      pBAD30 4923\n",
      " |      >>> str(new.seq) == str(right.seq + left.seq)\n",
      " |      True\n",
      " |      >>> len(new.features) == len(left.features) + len(right.features)\n",
      " |      True\n",
      " |      \n",
      " |      When we add the left and right SeqRecord objects, their annotation\n",
      " |      is all consistent, so it is all conserved in the new SeqRecord:\n",
      " |      \n",
      " |      >>> new.id == left.id == right.id == plasmid.id\n",
      " |      True\n",
      " |      >>> new.name == left.name == right.name == plasmid.name\n",
      " |      True\n",
      " |      >>> new.description == plasmid.description\n",
      " |      True\n",
      " |      >>> new.annotations == left.annotations == right.annotations\n",
      " |      True\n",
      " |      >>> new.letter_annotations == plasmid.letter_annotations\n",
      " |      True\n",
      " |      >>> new.dbxrefs == left.dbxrefs == right.dbxrefs\n",
      " |      True\n",
      " |      \n",
      " |      However, we should point out that when we sliced the SeqRecord,\n",
      " |      any annotations dictionary or dbxrefs list entries were lost.\n",
      " |      You can explicitly copy them like this:\n",
      " |      \n",
      " |      >>> new.annotations = plasmid.annotations.copy()\n",
      " |      >>> new.dbxrefs = plasmid.dbxrefs[:]\n",
      " |  \n",
      " |  __bool__(self)\n",
      " |      Boolean value of an instance of this class (True).\n",
      " |      \n",
      " |      This behaviour is for backwards compatibility, since until the\n",
      " |      __len__ method was added, a SeqRecord always evaluated as True.\n",
      " |      \n",
      " |      Note that in comparison, a Seq object will evaluate to False if it\n",
      " |      has a zero length sequence.\n",
      " |      \n",
      " |      WARNING: The SeqRecord may in future evaluate to False when its\n",
      " |      sequence is of zero length (in order to better match the Seq\n",
      " |      object behaviour)!\n",
      " |  \n",
      " |  __contains__(self, char)\n",
      " |      Implement the 'in' keyword, searches the sequence.\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Fasta/sweetpea.nu\", \"fasta\")\n",
      " |      >>> \"GAATTC\" in record\n",
      " |      False\n",
      " |      >>> \"AAA\" in record\n",
      " |      True\n",
      " |      \n",
      " |      This essentially acts as a proxy for using \"in\" on the sequence:\n",
      " |      \n",
      " |      >>> \"GAATTC\" in record.seq\n",
      " |      False\n",
      " |      >>> \"AAA\" in record.seq\n",
      " |      True\n",
      " |      \n",
      " |      Note that you can also use Seq objects as the query,\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> Seq(\"AAA\") in record\n",
      " |      True\n",
      " |      \n",
      " |      See also the Seq object's __contains__ method.\n",
      " |  \n",
      " |  __eq__(self, other)\n",
      " |      Define the equal-to operand (not implemented).\n",
      " |  \n",
      " |  __format__(self, format_spec)\n",
      " |      Return the record as a string in the specified file format.\n",
      " |      \n",
      " |      This method supports the Python format() function and f-strings.\n",
      " |      The format_spec should be a lower case string supported by\n",
      " |      Bio.SeqIO as a text output file format. Requesting a binary file\n",
      " |      format raises a ValueError. e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> record = SeqRecord(Seq(\"MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\"),\n",
      " |      ...                    id=\"YP_025292.1\", name=\"HokC\",\n",
      " |      ...                    description=\"toxic membrane protein\")\n",
      " |      ...\n",
      " |      >>> format(record, \"fasta\")\n",
      " |      '>YP_025292.1 toxic membrane protein\\nMKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\\n'\n",
      " |      >>> print(f\"Here is {record.id} in FASTA format:\\n{record:fasta}\")\n",
      " |      Here is YP_025292.1 in FASTA format:\n",
      " |      >YP_025292.1 toxic membrane protein\n",
      " |      MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\n",
      " |      <BLANKLINE>\n",
      " |      \n",
      " |      See also the SeqRecord's format() method.\n",
      " |  \n",
      " |  __ge__(self, other)\n",
      " |      Define the greater-than-or-equal-to operand (not implemented).\n",
      " |  \n",
      " |  __getitem__(self, index)\n",
      " |      Return a sub-sequence or an individual letter.\n",
      " |      \n",
      " |      Slicing, e.g. my_record[5:10], returns a new SeqRecord for\n",
      " |      that sub-sequence with some annotation preserved as follows:\n",
      " |      \n",
      " |      * The name, id and description are kept as-is.\n",
      " |      * Any per-letter-annotations are sliced to match the requested\n",
      " |        sub-sequence.\n",
      " |      * Unless a stride is used, all those features which fall fully\n",
      " |        within the subsequence are included (with their locations\n",
      " |        adjusted accordingly). If you want to preserve any truncated\n",
      " |        features (e.g. GenBank/EMBL source features), you must\n",
      " |        explicitly add them to the new SeqRecord yourself.\n",
      " |      * With the exception of any molecule type, the annotations\n",
      " |        dictionary and the dbxrefs list are not used for the new\n",
      " |        SeqRecord, as in general they may not apply to the\n",
      " |        subsequence. If you want to preserve them, you must explicitly\n",
      " |        copy them to the new SeqRecord yourself.\n",
      " |      \n",
      " |      Using an integer index, e.g. my_record[5] is shorthand for\n",
      " |      extracting that letter from the sequence, my_record.seq[5].\n",
      " |      \n",
      " |      For example, consider this short protein and its secondary\n",
      " |      structure as encoded by the PDB (e.g. H for alpha helices),\n",
      " |      plus a simple feature for its histidine self phosphorylation\n",
      " |      site:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
      " |      >>> rec = SeqRecord(Seq(\"MAAGVKQLADDRTLLMAGVSHDLRTPLTRIRLAT\"\n",
      " |      ...                     \"EMMSEQDGYLAESINKDIEECNAIIEQFIDYLR\"),\n",
      " |      ...                 id=\"1JOY\", name=\"EnvZ\",\n",
      " |      ...                 description=\"Homodimeric domain of EnvZ from E. coli\")\n",
      " |      >>> rec.letter_annotations[\"secondary_structure\"] = \"  S  SSSSSSHHHHHTTTHHHHHHHHHHHHHHHHHHHHHHTHHHHHHHHHHHHHHHHHHHHHTT  \"\n",
      " |      >>> rec.features.append(SeqFeature(FeatureLocation(20, 21),\n",
      " |      ...                     type = \"Site\"))\n",
      " |      \n",
      " |      Now let's have a quick look at the full record,\n",
      " |      \n",
      " |      >>> print(rec)\n",
      " |      ID: 1JOY\n",
      " |      Name: EnvZ\n",
      " |      Description: Homodimeric domain of EnvZ from E. coli\n",
      " |      Number of features: 1\n",
      " |      Per letter annotation for: secondary_structure\n",
      " |      Seq('MAAGVKQLADDRTLLMAGVSHDLRTPLTRIRLATEMMSEQDGYLAESINKDIEE...YLR')\n",
      " |      >>> rec.letter_annotations[\"secondary_structure\"]\n",
      " |      '  S  SSSSSSHHHHHTTTHHHHHHHHHHHHHHHHHHHHHHTHHHHHHHHHHHHHHHHHHHHHTT  '\n",
      " |      >>> print(rec.features[0].location)\n",
      " |      [20:21]\n",
      " |      \n",
      " |      Now let's take a sub sequence, here chosen as the first (fractured)\n",
      " |      alpha helix which includes the histidine phosphorylation site:\n",
      " |      \n",
      " |      >>> sub = rec[11:41]\n",
      " |      >>> print(sub)\n",
      " |      ID: 1JOY\n",
      " |      Name: EnvZ\n",
      " |      Description: Homodimeric domain of EnvZ from E. coli\n",
      " |      Number of features: 1\n",
      " |      Per letter annotation for: secondary_structure\n",
      " |      Seq('RTLLMAGVSHDLRTPLTRIRLATEMMSEQD')\n",
      " |      >>> sub.letter_annotations[\"secondary_structure\"]\n",
      " |      'HHHHHTTTHHHHHHHHHHHHHHHHHHHHHH'\n",
      " |      >>> print(sub.features[0].location)\n",
      " |      [9:10]\n",
      " |      \n",
      " |      You can also of course omit the start or end values, for\n",
      " |      example to get the first ten letters only:\n",
      " |      \n",
      " |      >>> print(rec[:10])\n",
      " |      ID: 1JOY\n",
      " |      Name: EnvZ\n",
      " |      Description: Homodimeric domain of EnvZ from E. coli\n",
      " |      Number of features: 0\n",
      " |      Per letter annotation for: secondary_structure\n",
      " |      Seq('MAAGVKQLAD')\n",
      " |      \n",
      " |      Or for the last ten letters:\n",
      " |      \n",
      " |      >>> print(rec[-10:])\n",
      " |      ID: 1JOY\n",
      " |      Name: EnvZ\n",
      " |      Description: Homodimeric domain of EnvZ from E. coli\n",
      " |      Number of features: 0\n",
      " |      Per letter annotation for: secondary_structure\n",
      " |      Seq('IIEQFIDYLR')\n",
      " |      \n",
      " |      If you omit both, then you get a copy of the original record (although\n",
      " |      lacking the annotations and dbxrefs):\n",
      " |      \n",
      " |      >>> print(rec[:])\n",
      " |      ID: 1JOY\n",
      " |      Name: EnvZ\n",
      " |      Description: Homodimeric domain of EnvZ from E. coli\n",
      " |      Number of features: 1\n",
      " |      Per letter annotation for: secondary_structure\n",
      " |      Seq('MAAGVKQLADDRTLLMAGVSHDLRTPLTRIRLATEMMSEQDGYLAESINKDIEE...YLR')\n",
      " |      \n",
      " |      Finally, indexing with a simple integer is shorthand for pulling out\n",
      " |      that letter from the sequence directly:\n",
      " |      \n",
      " |      >>> rec[5]\n",
      " |      'K'\n",
      " |      >>> rec.seq[5]\n",
      " |      'K'\n",
      " |  \n",
      " |  __gt__(self, other)\n",
      " |      Define the greater-than operand (not implemented).\n",
      " |  \n",
      " |  __init__(self, seq, id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=None, features=None, annotations=None, letter_annotations=None)\n",
      " |      Create a SeqRecord.\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - seq         - Sequence, required (Seq or MutableSeq)\n",
      " |       - id          - Sequence identifier, recommended (string)\n",
      " |       - name        - Sequence name, optional (string)\n",
      " |       - description - Sequence description, optional (string)\n",
      " |       - dbxrefs     - Database cross references, optional (list of strings)\n",
      " |       - features    - Any (sub)features, optional (list of SeqFeature objects)\n",
      " |       - annotations - Dictionary of annotations for the whole sequence\n",
      " |       - letter_annotations - Dictionary of per-letter-annotations, values\n",
      " |         should be strings, list or tuples of the same length as the full\n",
      " |         sequence.\n",
      " |      \n",
      " |      You will typically use Bio.SeqIO to read in sequences from files as\n",
      " |      SeqRecord objects.  However, you may want to create your own SeqRecord\n",
      " |      objects directly.\n",
      " |      \n",
      " |      Note that while an id is optional, we strongly recommend you supply a\n",
      " |      unique id string for each record.  This is especially important\n",
      " |      if you wish to write your sequences to a file.\n",
      " |      \n",
      " |      You can create a 'blank' SeqRecord object, and then populate the\n",
      " |      attributes later.\n",
      " |  \n",
      " |  __iter__(self)\n",
      " |      Iterate over the letters in the sequence.\n",
      " |      \n",
      " |      For example, using Bio.SeqIO to read in a protein FASTA file:\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Fasta/loveliesbleeding.pro\", \"fasta\")\n",
      " |      >>> for amino in record:\n",
      " |      ...     print(amino)\n",
      " |      ...     if amino == \"L\": break\n",
      " |      X\n",
      " |      A\n",
      " |      G\n",
      " |      L\n",
      " |      >>> print(record.seq[3])\n",
      " |      L\n",
      " |      \n",
      " |      This is just a shortcut for iterating over the sequence directly:\n",
      " |      \n",
      " |      >>> for amino in record.seq:\n",
      " |      ...     print(amino)\n",
      " |      ...     if amino == \"L\": break\n",
      " |      X\n",
      " |      A\n",
      " |      G\n",
      " |      L\n",
      " |      >>> print(record.seq[3])\n",
      " |      L\n",
      " |      \n",
      " |      Note that this does not facilitate iteration together with any\n",
      " |      per-letter-annotation.  However, you can achieve that using the\n",
      " |      python zip function on the record (or its sequence) and the relevant\n",
      " |      per-letter-annotation:\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> rec = SeqIO.read(\"Quality/solexa_faked.fastq\", \"fastq-solexa\")\n",
      " |      >>> print(\"%s %s\" % (rec.id, rec.seq))\n",
      " |      slxa_0001_1_0001_01 ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTNNNNNN\n",
      " |      >>> print(list(rec.letter_annotations))\n",
      " |      ['solexa_quality']\n",
      " |      >>> for nuc, qual in zip(rec, rec.letter_annotations[\"solexa_quality\"]):\n",
      " |      ...     if qual > 35:\n",
      " |      ...         print(\"%s %i\" % (nuc, qual))\n",
      " |      A 40\n",
      " |      C 39\n",
      " |      G 38\n",
      " |      T 37\n",
      " |      A 36\n",
      " |      \n",
      " |      You may agree that using zip(rec.seq, ...) is more explicit than using\n",
      " |      zip(rec, ...) as shown above.\n",
      " |  \n",
      " |  __le___(self, other)\n",
      " |      Define the less-than-or-equal-to operand (not implemented).\n",
      " |  \n",
      " |  __len__(self)\n",
      " |      Return the length of the sequence.\n",
      " |      \n",
      " |      For example, using Bio.SeqIO to read in a FASTA nucleotide file:\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Fasta/sweetpea.nu\", \"fasta\")\n",
      " |      >>> len(record)\n",
      " |      309\n",
      " |      >>> len(record.seq)\n",
      " |      309\n",
      " |  \n",
      " |  __lt__(self, other)\n",
      " |      Define the less-than operand (not implemented).\n",
      " |  \n",
      " |  __ne__(self, other)\n",
      " |      Define the not-equal-to operand (not implemented).\n",
      " |  \n",
      " |  __radd__(self, other)\n",
      " |      Add another sequence or string to this sequence (from the left).\n",
      " |      \n",
      " |      This method handles adding a Seq object (or similar, e.g. MutableSeq)\n",
      " |      or a plain Python string (on the left) to a SeqRecord (on the right).\n",
      " |      See the __add__ method for more details, but for example:\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Quality/solexa_faked.fastq\", \"fastq-solexa\")\n",
      " |      >>> print(\"%s %s\" % (record.id, record.seq))\n",
      " |      slxa_0001_1_0001_01 ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTNNNNNN\n",
      " |      >>> print(list(record.letter_annotations))\n",
      " |      ['solexa_quality']\n",
      " |      \n",
      " |      >>> new = \"ACT\" + record\n",
      " |      >>> print(\"%s %s\" % (new.id, new.seq))\n",
      " |      slxa_0001_1_0001_01 ACTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTNNNNNN\n",
      " |      >>> print(list(new.letter_annotations))\n",
      " |      []\n",
      " |  \n",
      " |  __repr__(self)\n",
      " |      Return a concise summary of the record for debugging (string).\n",
      " |      \n",
      " |      The python built in function repr works by calling the object's ___repr__\n",
      " |      method.  e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> rec = SeqRecord(Seq(\"MASRGVNKVILVGNLGQDPEVRYMPNGGAVANITLATSESWRDKAT\"\n",
      " |      ...                     \"GEMKEQTEWHRVVLFGKLAEVASEYLRKGSQVYIEGQLRTRKWTDQ\"\n",
      " |      ...                     \"SGQDRYTTEVVVNVGGTMQMLGGRQGGGAPAGGNIGGGQPQGGWGQ\"\n",
      " |      ...                     \"PQQPQGGNQFSGGAQSRPQQSAPAAPSNEPPMDFDDDIPF\"),\n",
      " |      ...                 id=\"NP_418483.1\", name=\"b4059\",\n",
      " |      ...                 description=\"ssDNA-binding protein\",\n",
      " |      ...                 dbxrefs=[\"ASAP:13298\", \"GI:16131885\", \"GeneID:948570\"])\n",
      " |      >>> print(repr(rec))\n",
      " |      SeqRecord(seq=Seq('MASRGVNKVILVGNLGQDPEVRYMPNGGAVANITLATSESWRDKATGEMKEQTE...IPF'), id='NP_418483.1', name='b4059', description='ssDNA-binding protein', dbxrefs=['ASAP:13298', 'GI:16131885', 'GeneID:948570'])\n",
      " |      \n",
      " |      At the python prompt you can also use this shorthand:\n",
      " |      \n",
      " |      >>> rec\n",
      " |      SeqRecord(seq=Seq('MASRGVNKVILVGNLGQDPEVRYMPNGGAVANITLATSESWRDKATGEMKEQTE...IPF'), id='NP_418483.1', name='b4059', description='ssDNA-binding protein', dbxrefs=['ASAP:13298', 'GI:16131885', 'GeneID:948570'])\n",
      " |      \n",
      " |      Note that long sequences are shown truncated. Also note that any\n",
      " |      annotations, letter_annotations and features are not shown (as they\n",
      " |      would lead to a very long string).\n",
      " |  \n",
      " |  __str__(self)\n",
      " |      Return a human readable summary of the record and its annotation (string).\n",
      " |      \n",
      " |      The python built in function str works by calling the object's ___str__\n",
      " |      method.  e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> record = SeqRecord(Seq(\"MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\"),\n",
      " |      ...                    id=\"YP_025292.1\", name=\"HokC\",\n",
      " |      ...                    description=\"toxic membrane protein, small\")\n",
      " |      >>> print(str(record))\n",
      " |      ID: YP_025292.1\n",
      " |      Name: HokC\n",
      " |      Description: toxic membrane protein, small\n",
      " |      Number of features: 0\n",
      " |      Seq('MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF')\n",
      " |      \n",
      " |      In this example you don't actually need to call str explicity, as the\n",
      " |      print command does this automatically:\n",
      " |      \n",
      " |      >>> print(record)\n",
      " |      ID: YP_025292.1\n",
      " |      Name: HokC\n",
      " |      Description: toxic membrane protein, small\n",
      " |      Number of features: 0\n",
      " |      Seq('MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF')\n",
      " |      \n",
      " |      Note that long sequences are shown truncated.\n",
      " |  \n",
      " |  format(self, format)\n",
      " |      Return the record as a string in the specified file format.\n",
      " |      \n",
      " |      The format should be a lower case string supported as an output\n",
      " |      format by Bio.SeqIO, which is used to turn the SeqRecord into a\n",
      " |      string.  e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> record = SeqRecord(Seq(\"MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\"),\n",
      " |      ...                    id=\"YP_025292.1\", name=\"HokC\",\n",
      " |      ...                    description=\"toxic membrane protein\")\n",
      " |      >>> record.format(\"fasta\")\n",
      " |      '>YP_025292.1 toxic membrane protein\\nMKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\\n'\n",
      " |      >>> print(record.format(\"fasta\"))\n",
      " |      >YP_025292.1 toxic membrane protein\n",
      " |      MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\n",
      " |      <BLANKLINE>\n",
      " |      \n",
      " |      The Python print function automatically appends a new line, meaning\n",
      " |      in this example a blank line is shown.  If you look at the string\n",
      " |      representation you can see there is a trailing new line (shown as\n",
      " |      slash n) which is important when writing to a file or if\n",
      " |      concatenating multiple sequence strings together.\n",
      " |      \n",
      " |      Note that this method will NOT work on every possible file format\n",
      " |      supported by Bio.SeqIO (e.g. some are for multiple sequences only,\n",
      " |      and binary formats are not supported).\n",
      " |  \n",
      " |  lower(self)\n",
      " |      Return a copy of the record with a lower case sequence.\n",
      " |      \n",
      " |      All the annotation is preserved unchanged. e.g.\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Fasta/aster.pro\", \"fasta\")\n",
      " |      >>> print(record.format(\"fasta\"))\n",
      " |      >gi|3298468|dbj|BAA31520.1| SAMIPF\n",
      " |      GGHVNPAVTFGAFVGGNITLLRGIVYIIAQLLGSTVACLLLKFVTNDMAVGVFSLSAGVG\n",
      " |      VTNALVFEIVMTFGLVYTVYATAIDPKKGSLGTIAPIAIGFIVGANI\n",
      " |      <BLANKLINE>\n",
      " |      >>> print(record.lower().format(\"fasta\"))\n",
      " |      >gi|3298468|dbj|BAA31520.1| SAMIPF\n",
      " |      gghvnpavtfgafvggnitllrgivyiiaqllgstvaclllkfvtndmavgvfslsagvg\n",
      " |      vtnalvfeivmtfglvytvyataidpkkgslgtiapiaigfivgani\n",
      " |      <BLANKLINE>\n",
      " |      \n",
      " |      To take a more annotation rich example,\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> old = SeqIO.read(\"EMBL/TRBG361.embl\", \"embl\")\n",
      " |      >>> len(old.features)\n",
      " |      3\n",
      " |      >>> new = old.lower()\n",
      " |      >>> len(old.features) == len(new.features)\n",
      " |      True\n",
      " |      >>> old.annotations[\"organism\"] == new.annotations[\"organism\"]\n",
      " |      True\n",
      " |      >>> old.dbxrefs == new.dbxrefs\n",
      " |      True\n",
      " |  \n",
      " |  reverse_complement(self, id=False, name=False, description=False, features=True, annotations=False, letter_annotations=True, dbxrefs=False)\n",
      " |      Return new SeqRecord with reverse complement sequence.\n",
      " |      \n",
      " |      By default the new record does NOT preserve the sequence identifier,\n",
      " |      name, description, general annotation or database cross-references -\n",
      " |      these are unlikely to apply to the reversed sequence.\n",
      " |      \n",
      " |      You can specify the returned record's id, name and description as\n",
      " |      strings, or True to keep that of the parent, or False for a default.\n",
      " |      \n",
      " |      You can specify the returned record's features with a list of\n",
      " |      SeqFeature objects, or True to keep that of the parent, or False to\n",
      " |      omit them. The default is to keep the original features (with the\n",
      " |      strand and locations adjusted).\n",
      " |      \n",
      " |      You can also specify both the returned record's annotations and\n",
      " |      letter_annotations as dictionaries, True to keep that of the parent,\n",
      " |      or False to omit them. The default is to keep the original\n",
      " |      annotations (with the letter annotations reversed).\n",
      " |      \n",
      " |      To show what happens to the pre-letter annotations, consider an\n",
      " |      example Solexa variant FASTQ file with a single entry, which we'll\n",
      " |      read in as a SeqRecord:\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Quality/solexa_faked.fastq\", \"fastq-solexa\")\n",
      " |      >>> print(\"%s %s\" % (record.id, record.seq))\n",
      " |      slxa_0001_1_0001_01 ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTNNNNNN\n",
      " |      >>> print(list(record.letter_annotations))\n",
      " |      ['solexa_quality']\n",
      " |      >>> print(record.letter_annotations[\"solexa_quality\"])\n",
      " |      [40, 39, 38, 37, 36, 35, 34, 33, 32, 31, 30, 29, 28, 27, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0, -1, -2, -3, -4, -5]\n",
      " |      \n",
      " |      Now take the reverse complement, here we explicitly give a new\n",
      " |      identifier (the old identifier with a suffix):\n",
      " |      \n",
      " |      >>> rc_record = record.reverse_complement(id=record.id + \"_rc\")\n",
      " |      >>> print(\"%s %s\" % (rc_record.id, rc_record.seq))\n",
      " |      slxa_0001_1_0001_01_rc NNNNNNACGTACGTACGTACGTACGTACGTACGTACGTACGTACGT\n",
      " |      \n",
      " |      Notice that the per-letter-annotations have also been reversed,\n",
      " |      although this may not be appropriate for all cases.\n",
      " |      \n",
      " |      >>> print(rc_record.letter_annotations[\"solexa_quality\"])\n",
      " |      [-5, -4, -3, -2, -1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40]\n",
      " |      \n",
      " |      Now for the features, we need a different example. Parsing a GenBank\n",
      " |      file is probably the easiest way to get an nice example with features\n",
      " |      in it...\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> with open(\"GenBank/pBAD30.gb\") as handle:\n",
      " |      ...     plasmid = SeqIO.read(handle, \"gb\")\n",
      " |      >>> print(\"%s %i\" % (plasmid.id, len(plasmid)))\n",
      " |      pBAD30 4923\n",
      " |      >>> plasmid.seq\n",
      " |      Seq('GCTAGCGGAGTGTATACTGGCTTACTATGTTGGCACTGATGAGGGTGTCAGTGA...ATG')\n",
      " |      >>> len(plasmid.features)\n",
      " |      13\n",
      " |      \n",
      " |      Now, let's take the reverse complement of this whole plasmid:\n",
      " |      \n",
      " |      >>> rc_plasmid = plasmid.reverse_complement(id=plasmid.id+\"_rc\")\n",
      " |      >>> print(\"%s %i\" % (rc_plasmid.id, len(rc_plasmid)))\n",
      " |      pBAD30_rc 4923\n",
      " |      >>> rc_plasmid.seq\n",
      " |      Seq('CATGGGCAAATATTATACGCAAGGCGACAAGGTGCTGATGCCGCTGGCGATTCA...AGC')\n",
      " |      >>> len(rc_plasmid.features)\n",
      " |      13\n",
      " |      \n",
      " |      Let's compare the first CDS feature - it has gone from being the\n",
      " |      second feature (index 1) to the second last feature (index -2), its\n",
      " |      strand has changed, and the location switched round.\n",
      " |      \n",
      " |      >>> print(plasmid.features[1])\n",
      " |      type: CDS\n",
      " |      location: [1081:1960](-)\n",
      " |      qualifiers:\n",
      " |          Key: label, Value: ['araC']\n",
      " |          Key: note, Value: ['araC regulator of the arabinose BAD promoter']\n",
      " |          Key: vntifkey, Value: ['4']\n",
      " |      <BLANKLINE>\n",
      " |      >>> print(rc_plasmid.features[-2])\n",
      " |      type: CDS\n",
      " |      location: [2963:3842](+)\n",
      " |      qualifiers:\n",
      " |          Key: label, Value: ['araC']\n",
      " |          Key: note, Value: ['araC regulator of the arabinose BAD promoter']\n",
      " |          Key: vntifkey, Value: ['4']\n",
      " |      <BLANKLINE>\n",
      " |      \n",
      " |      You can check this new location, based on the length of the plasmid:\n",
      " |      \n",
      " |      >>> len(plasmid) - 1081\n",
      " |      3842\n",
      " |      >>> len(plasmid) - 1960\n",
      " |      2963\n",
      " |      \n",
      " |      Note that if the SeqFeature annotation includes any strand specific\n",
      " |      information (e.g. base changes for a SNP), this information is not\n",
      " |      amended, and would need correction after the reverse complement.\n",
      " |      \n",
      " |      Note trying to reverse complement a protein SeqRecord raises an\n",
      " |      exception:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> protein_rec = SeqRecord(Seq(\"MAIVMGR\"), id=\"Test\",\n",
      " |      ...                         annotations={\"molecule_type\": \"protein\"})\n",
      " |      >>> protein_rec.reverse_complement()\n",
      " |      Traceback (most recent call last):\n",
      " |         ...\n",
      " |      ValueError: Proteins do not have complements!\n",
      " |      \n",
      " |      If you have RNA without any U bases, it must be annotated as RNA\n",
      " |      otherwise it will be treated as DNA by default with A mapped to T:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> rna1 = SeqRecord(Seq(\"ACG\"), id=\"Test\")\n",
      " |      >>> rna2 = SeqRecord(Seq(\"ACG\"), id=\"Test\", annotations={\"molecule_type\": \"RNA\"})\n",
      " |      >>> print(rna1.reverse_complement(id=\"RC\", description=\"unk\").format(\"fasta\"))\n",
      " |      >RC unk\n",
      " |      CGT\n",
      " |      <BLANKLINE>\n",
      " |      >>> print(rna2.reverse_complement(id=\"RC\", description=\"RNA\").format(\"fasta\"))\n",
      " |      >RC RNA\n",
      " |      CGU\n",
      " |      <BLANKLINE>\n",
      " |      \n",
      " |      Also note you can reverse complement a SeqRecord using a MutableSeq:\n",
      " |      \n",
      " |      >>> from Bio.Seq import MutableSeq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> rec = SeqRecord(MutableSeq(\"ACGT\"), id=\"Test\")\n",
      " |      >>> rec.seq[0] = \"T\"\n",
      " |      >>> print(\"%s %s\" % (rec.id, rec.seq))\n",
      " |      Test TCGT\n",
      " |      >>> rc = rec.reverse_complement(id=True)\n",
      " |      >>> print(\"%s %s\" % (rc.id, rc.seq))\n",
      " |      Test ACGA\n",
      " |  \n",
      " |  translate(self, table='Standard', stop_symbol='*', to_stop=False, cds=False, gap=None, id=False, name=False, description=False, features=False, annotations=False, letter_annotations=False, dbxrefs=False)\n",
      " |      Return new SeqRecord with translated sequence.\n",
      " |      \n",
      " |      This calls the record's .seq.translate() method (which describes\n",
      " |      the translation related arguments, like table for the genetic code),\n",
      " |      \n",
      " |      By default the new record does NOT preserve the sequence identifier,\n",
      " |      name, description, general annotation or database cross-references -\n",
      " |      these are unlikely to apply to the translated sequence.\n",
      " |      \n",
      " |      You can specify the returned record's id, name and description as\n",
      " |      strings, or True to keep that of the parent, or False for a default.\n",
      " |      \n",
      " |      You can specify the returned record's features with a list of\n",
      " |      SeqFeature objects, or False (default) to omit them.\n",
      " |      \n",
      " |      You can also specify both the returned record's annotations and\n",
      " |      letter_annotations as dictionaries, True to keep that of the parent\n",
      " |      (annotations only), or False (default) to omit them.\n",
      " |      \n",
      " |      e.g. Loading a FASTA gene and translating it,\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> gene_record = SeqIO.read(\"Fasta/sweetpea.nu\", \"fasta\")\n",
      " |      >>> print(gene_record.format(\"fasta\"))\n",
      " |      >gi|3176602|gb|U78617.1|LOU78617 Lathyrus odoratus phytochrome A (PHYA) gene, partial cds\n",
      " |      CAGGCTGCGCGGTTTCTATTTATGAAGAACAAGGTCCGTATGATAGTTGATTGTCATGCA\n",
      " |      AAACATGTGAAGGTTCTTCAAGACGAAAAACTCCCATTTGATTTGACTCTGTGCGGTTCG\n",
      " |      ACCTTAAGAGCTCCACATAGTTGCCATTTGCAGTACATGGCTAACATGGATTCAATTGCT\n",
      " |      TCATTGGTTATGGCAGTGGTCGTCAATGACAGCGATGAAGATGGAGATAGCCGTGACGCA\n",
      " |      GTTCTACCACAAAAGAAAAAGAGACTTTGGGGTTTGGTAGTTTGTCATAACACTACTCCG\n",
      " |      AGGTTTGTT\n",
      " |      <BLANKLINE>\n",
      " |      \n",
      " |      And now translating the record, specifying the new ID and description:\n",
      " |      \n",
      " |      >>> protein_record = gene_record.translate(table=11,\n",
      " |      ...                                        id=\"phya\",\n",
      " |      ...                                        description=\"translation\")\n",
      " |      >>> print(protein_record.format(\"fasta\"))\n",
      " |      >phya translation\n",
      " |      QAARFLFMKNKVRMIVDCHAKHVKVLQDEKLPFDLTLCGSTLRAPHSCHLQYMANMDSIA\n",
      " |      SLVMAVVVNDSDEDGDSRDAVLPQKKKRLWGLVVCHNTTPRFV\n",
      " |      <BLANKLINE>\n",
      " |  \n",
      " |  upper(self)\n",
      " |      Return a copy of the record with an upper case sequence.\n",
      " |      \n",
      " |      All the annotation is preserved unchanged. e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> record = SeqRecord(Seq(\"acgtACGT\"), id=\"Test\",\n",
      " |      ...                    description = \"Made up for this example\")\n",
      " |      >>> record.letter_annotations[\"phred_quality\"] = [1, 2, 3, 4, 5, 6, 7, 8]\n",
      " |      >>> print(record.upper().format(\"fastq\"))\n",
      " |      @Test Made up for this example\n",
      " |      ACGTACGT\n",
      " |      +\n",
      " |      \"#$%&'()\n",
      " |      <BLANKLINE>\n",
      " |      \n",
      " |      Naturally, there is a matching lower method:\n",
      " |      \n",
      " |      >>> print(record.lower().format(\"fastq\"))\n",
      " |      @Test Made up for this example\n",
      " |      acgtacgt\n",
      " |      +\n",
      " |      \"#$%&'()\n",
      " |      <BLANKLINE>\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors defined here:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      " |  \n",
      " |  letter_annotations\n",
      " |      Dictionary of per-letter-annotation for the sequence.\n",
      " |      \n",
      " |      For example, this can hold quality scores used in FASTQ or QUAL files.\n",
      " |      Consider this example using Bio.SeqIO to read in an example Solexa\n",
      " |      variant FASTQ file as a SeqRecord:\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Quality/solexa_faked.fastq\", \"fastq-solexa\")\n",
      " |      >>> print(\"%s %s\" % (record.id, record.seq))\n",
      " |      slxa_0001_1_0001_01 ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTNNNNNN\n",
      " |      >>> print(list(record.letter_annotations))\n",
      " |      ['solexa_quality']\n",
      " |      >>> print(record.letter_annotations[\"solexa_quality\"])\n",
      " |      [40, 39, 38, 37, 36, 35, 34, 33, 32, 31, 30, 29, 28, 27, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0, -1, -2, -3, -4, -5]\n",
      " |      \n",
      " |      The letter_annotations get sliced automatically if you slice the\n",
      " |      parent SeqRecord, for example taking the last ten bases:\n",
      " |      \n",
      " |      >>> sub_record = record[-10:]\n",
      " |      >>> print(\"%s %s\" % (sub_record.id, sub_record.seq))\n",
      " |      slxa_0001_1_0001_01 ACGTNNNNNN\n",
      " |      >>> print(sub_record.letter_annotations[\"solexa_quality\"])\n",
      " |      [4, 3, 2, 1, 0, -1, -2, -3, -4, -5]\n",
      " |      \n",
      " |      Any python sequence (i.e. list, tuple or string) can be recorded in\n",
      " |      the SeqRecord's letter_annotations dictionary as long as the length\n",
      " |      matches that of the SeqRecord's sequence.  e.g.\n",
      " |      \n",
      " |      >>> len(sub_record.letter_annotations)\n",
      " |      1\n",
      " |      >>> sub_record.letter_annotations[\"dummy\"] = \"abcdefghij\"\n",
      " |      >>> len(sub_record.letter_annotations)\n",
      " |      2\n",
      " |      \n",
      " |      You can delete entries from the letter_annotations dictionary as usual:\n",
      " |      \n",
      " |      >>> del sub_record.letter_annotations[\"solexa_quality\"]\n",
      " |      >>> sub_record.letter_annotations\n",
      " |      {'dummy': 'abcdefghij'}\n",
      " |      \n",
      " |      You can completely clear the dictionary easily as follows:\n",
      " |      \n",
      " |      >>> sub_record.letter_annotations = {}\n",
      " |      >>> sub_record.letter_annotations\n",
      " |      {}\n",
      " |      \n",
      " |      Note that if replacing the record's sequence with a sequence of a\n",
      " |      different length you must first clear the letter_annotations dict.\n",
      " |  \n",
      " |  seq\n",
      " |      The sequence itself, as a Seq or MutableSeq object.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes defined here:\n",
      " |  \n",
      " |  __hash__ = None\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from Bio.SeqRecord import SeqRecord\n",
    "help(SeqRecord)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4.1 O objeto SeqRecord"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A classe **SeqRecord** é definida pelo módulo `Bio.SeqRecord`. Essa classe possui ferramentas de alto nível, como os identificadores e recursos (ver o Capítulo 3: https://github.com/GTL98/biopython_br/tree/main/Cap%203%20-%20Objetos%20de%20Sequ%C3%AAncia) e é o tipo básico de dado para as sequência de entrada/saída do módulo `Bio.SeqIO` (ver o Capítulo 5: https://github.com/GTL98/biopython_br/tree/main/Cap%205%20-%20Sequ%C3%AAncias%20Input_Output).\n",
    "\n",
    "A classe **SeqRecord** é bem simples e fornece as seguintes informações como atributos:\n",
    "- **.seq**: A sequência por si própria, geralmente um objeto **Seq**;\n",
    "\n",
    "\n",
    "- **.id**: A identificação primária da sequência (uma string). Na maioria dos casos é um número de acesso;\n",
    "\n",
    "\n",
    "- **.name**: O nome/identificação \"comum\" da sequência (uma string). Em alguns casos é o mesmo que o número de acesso, mas também pode ser o nome de um clone. Pode ser análogo a identificação LOCUS em arquivos GenBank;\n",
    "\n",
    "\n",
    "- **.description**: Uma descrição ou o nome da sequência que o humano consiga ler (uma string);\n",
    "\n",
    "\n",
    "- **.letter_annotations**: Guarda as anotações por letra usando um restrito dicionário de informações adicionais sobre as letras na sequência. As chaves são o nome das informações e as informações são os valores (i.e. uma tupla, uma lista ou uma string), com o mesmo tamanho da sequência. Esse atributo é frequentemente usado para pontuações de qualidade (ver a **Seção 20.1.6** no Capítulo 20: ***link do capítulo 20***), ou informações da estrutura secundária (e.g. arquivos de alinhamento Stockholm/PFAM);\n",
    "\n",
    "\n",
    "- **.annotations**: Um dicionário com informações adicionais da sequência. As chaves são o nome da informação e o valor é a informação. Isso permite a adição de informações \"não relacionadas\" à sequência;\n",
    "\n",
    "\n",
    "- **.features**: Uma lista de objetos **SeqFeatures** com mais informações relacionadas sobre as características presentes na sequência (e.g. domínios em uma proteína ou posição dos genes em um genoma). A estrutura dos recursos da sequência serão melhores detalhadas na **Seção 4.3** e;\n",
    "\n",
    "\n",
    "- **.dbxrefs**: Uma lista de referências cruzadas de bancos de dados como strings."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4.2 Criando o SeqRecord"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Usar o objeto **SeqRecord** não é muito complicado, uma vez que as informações são apresentadas como atributos da classe. Normalmente você não precisará criar um objeto **SeqRecord** \"na mão\", mas usará o módulo `Bio.SeqIO` para ler o arquivo de sequência para você (ver os exemplos do Capítulo 5: https://github.com/GTL98/biopython_br/tree/main/Cap%205%20-%20Sequ%C3%AAncias%20Input_Output). Entretanto, criar um objeto **SeqRecord** é bem simples:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4.2.1 Objetos SeqRecord do zero"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Para criar o mínimo necessário do objeto **SeqRecord**, você precisa somente de um objeto **Seq**:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "simples_seq = Seq('GATC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.SeqRecord import SeqRecord\n",
    "simples_seq_r = SeqRecord(simples_seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Além disso, você pode informar a identificação, nome e descrição para a função de inicialização. Caso contrário, esas informações serão desconhecidas e podem ser modificadas posteriormente:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'<unknown id>'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "simples_seq_r.id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "simples_seq_r.id = 'AC12345'\n",
    "simples_seq_r.description = 'Sequência que eu gostaria de escrever um artigo'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequência que eu gostaria de escrever um artigo\n"
     ]
    }
   ],
   "source": [
    "print(simples_seq_r.description)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('GATC')"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "simples_seq_r.seq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Incluir a identificação da sequência é vital se você estiver querendo exportar o objeto **SeqRecord** em um arquivo. Você deve incluir a identificação quando estiver criando o objeto **SeqRecord**:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "simples_seq = Seq('GATC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.SeqRecord import SeqRecord\n",
    "simples_seq_r = SeqRecord(simples_seq, id='AC12345')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Como mencionado mais acima, o objeto **SeqRecord** possui o atributo **annotations** como um dicionário. Ele é usado para qualquer anotação que não se encaixe nos outros atributos mais específicos. Adicinar anotações é fácil e envolve apenas saber como adicionar informações em um dicionário Python:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "simples_seq_r.annotations['evidence'] = 'Nenhuma. Acabei de inventar.'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'evidence': 'Nenhuma. Acabei de inventar.'}\n"
     ]
    }
   ],
   "source": [
    "print(simples_seq_r.annotations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nenhuma. Acabei de inventar.\n"
     ]
    }
   ],
   "source": [
    "print(simples_seq_r.annotations['evidence'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Trabalhar com anotações por letra segue a mesma lógica. **letter_annotations** é um dicionário Python onde a chave é o nome da informação e o valor é a informação, mas sempre respeitando o tamanho da sequência; nesse exemplo o tamanho é de 4 pares de base, então deverá ter 4 valores de informação:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "simples_seq_r.letter_annotations['phred_quality'] = [40, 40, 38, 30]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'phred_quality': [40, 40, 38, 30]}\n"
     ]
    }
   ],
   "source": [
    "print(simples_seq_r.letter_annotations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[40, 40, 38, 30]\n"
     ]
    }
   ],
   "source": [
    "print(simples_seq_r.letter_annotations['phred_quality'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Os atributos **dbxrefs** e **features** são uma lista Python e deverão ser usados para armazenar stings e objetos **SeqFeature** (discutidos mais a frente nesse capítulo), respectivamente."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4.2.2 Objetos SeqRecord de arquivos FASTA"
   ]
  },
  {
   "attachments": {
    "image.png": {
     "image/png": "iVBORw0KGgoAAAANSUhEUgAABKoAAABaCAYAAACYPb4ZAAAgAElEQVR4Ae2Yi3J0x22E9f4v7RRc6eRTC8Bg9iyXpH5MlYNLX4AD0pLDv/6zby+wF9gL7AX2AnuBvcBeYC+wF9gL7AX2AnuBvcBeYC/wAy7w1w/YYVfYC+wF9gJ7gb3AXmAvsBfYC+wF9gJ7gb3AXmAvsBfYC/xn/1C1vwR7gb1Ae4G//vrrP/GfT7/vmvvp7/zN8/60n9Gf9r2/8Xdzf0a/8ae2O+8F9gJ/wgX2n89/wk95v3Ev8L4LfP7/+3zf7uu0F9gLfPEF9D8qFL943P/Za57i/wGb/JgL6Gej+GMWwyLaTRHQdSoPxWuDFXz5BfSzUfzygTtgL7AX2AsML6B/LlVxaPNraf7dTz7knV5P9ljtXuCrL+C/615/9fzv9n/rH6r8eFFnz3kZJ3rOU13x39HXDI8Tb9eorrTCJ/EVj0rDfjab+CTPPKL3VS+b9+5ZPuPkf8s/+f0U/Lu+67vm/pS7f2IP3viVedRH/hPfO3d8p9dPvNV37DS56YSj3W+40vzWyG/9rd/wqb3/pFvxW5lPb02N8ifa8Jg8zfL4qnai+zTHv83rT+/z6Xnv/N53er3rDr5TVk9nZdroZa/isp/pvEe+cuds/fkL6GdRxU9sxNmfmMcZ+W89GRc5P4S5WxCL3J/jWe2ad9TZHO9Vc5yX1a7NOFXvndrwquZEf/I6vbCJzw1Hvlm88ZlyfU6lm/Iq/U/v6/s+ved3zf30d37HPN1W8dUdnupfnTvVaT/Fqa7ivcun8v/T+rqnYvb9whQzDntTHjW/Ldc3Kv62/T+5r26k+MnZn56lb+xitdMrGnp1emHkMxfeRfKV3/Kl+67Y7RvYn/B0g6ffKh/Fp37v0GuXSazmvaKdaMTJ5grLYsbf3mcvkP1c2PvqbTgr8k+/48SbpfxjVPtHqa9IXL1TpOYd+WkecZ9H7JRTe+ISf1Unjxs9uVkuz1PMtK/2PjlLO/pM9T1Oea7bei/wXRf4U35n/5Tv/K7fo6dz/ecTNd8JJ/dPyv0uf9K3337rn3Ir/86u9ht23MBO76Qn7l7ETjm1N1zqflLu3/CTdvvpu/zE2/lOp9pvfOILf1WX6dWros/a+vsv4D+rr97o0/P8e9p/A3E5F2Y1+cydSyxyPe8TEydi1SfnJj/NdZzejvlujndaYl3eeUrXcYgFn3Xk3XPuid953WCcGzrWX7nDac4Jv/nG5e4FPnWBP+X39k/5zk/93rx7jv98ouY74eT+Sbnf5U/69ttv/RNu5d8YNZ/jxCIn7rV7dVr5kKOeYof5LGkUK230xVEk96fm2lXxp+75E/fSzRR/wo7aRdF3Up9RHPaUC2MMzJ/4iiecHtKox1o999v6ey/w6Z/Rp+f5df/5Gw/G7XLOZw3b8l8o5Ef+qXea2+EdFvt3eId13z7RnTiB65244kW84VL3jvzVnZ/MPn3vCdds56kW7lG4onDViupXUTyPt3zpK130xVHMeoFlT5oqZhr2Kp365L4rl7eifFUrql9F8TxWfPZdo5oc5cKmUTrFk068U6x8TrpXcJ8lj6ovPKJzvCa3y12nOtMIi9i9jkcsyytf54pX9YU/ifIOD+byZI+548I8ijeJrlWdaYUpiqNaUf0qiucx4zvnVLuH829x8d1HtfCviprjMZvnnFPtHs4XXvWFf1ec7HXiBK534ooX8cTt8A6bemuXk5d4T2M1p+p38241zve6myXMNaqFMwpTDEx5RL2s51xylEt/iuJ38eQxxX2GdFVfeMQnnImWs5hPtB0nML2OJ85Pi76z6tOe4nnMdBmHPWmyXmDsR65X9YVn8RVN+LhO9WSGuFXMPNR7RSNtxP+/FrvIfQCgf6TOZU0y+5HHy3rUfFU+meucm50r7ZNvdk+/jePa13mqna++xynPdV9Rf3KXalbV5/c6J6vJV+686HtPtTSMwrpIvvKOH1j3Tlri7kMsy53POuNnPWrekWczup7P7LjCXKNaeBfFVey4GSadYsZhT7wqklvllfbVfjWn6nNOxVGf3CwX7xSpdS4x5uRVfXI8p0Z5xvGeammeRvopjxivqjmTnCwnt8oznfdcm+HeU+3aqIWdIrUnruPUZjOf4j4v6q942Rzv+VzHT/VJH3jl4drvqH23bIcJR7opd8JzTtTxvK/ZjM6RlhzlzlX/3dHnnOpuvmtvuK5VXXkI76JrO25g8TKOfDKMPfGqSO4przxu+6c5jtO/w8TLOFlP/Emc6CecmDXlTfb6ao7vmtXZDhkv61Gb4ewFl7VyeaieRumy6B4Zhz3nZzX5kWecruf6qUem897of1n4cm6iOuOxN+WF5lOP+/lcx1jHfqxvtZne/dyTN6m4VZ9az13juOqM572oP/F87lfP5DzNynrCIhI/5dTdasOb7zRLODXKhVVRvCxWmqpPj4qjPrmeizOJrn1ST+aRw1nsn3LqIj/xhb+qe1Xv81jL8xSpeUd+muc4ZzrmNbmeO/dUS+889Rk7jmNdTc/IO26Guf6VWr4+n7V8yfWeMI/iVdH5Ve36ilf1qa84VV/aCq/60ik6T33FDnesquX1rljN8b7Pc/xUv1vvfl9d+/dl8yYc6abcjucY65jDOnI+x1iTx5wc9yPvae5zTnU3z7U3XNeydh9iXf6qzj3l432vxcuic0915vFK7zTHcc7oMPEyTtYTfxIn+gknZk15k72+kuN7VrXvUPGqvvQVfuq/qpcuiz4z46jn3K6WJmLHyzBqb/Wu9frv/3ZwFPUri4WcOtmxF7nz1BP/K2O2S7ZPxst6U+2ER//sBsSrPNN5z7WOq3beqZbuK6LP/ooZ9PR5XpOrvOM4FjVfhpPjeKclppxe6mWxm+N85/oMx13P+pZLLfMbH+omuXtHzdfhjk11weu08vFd1FeceIhbxRuPCfe0c7VH1/e5PuOE09u5xDw/cTucmPtG3eGBVY+6jOe4avmpVlT/SXQv1sxjhtfZXHEUMw574jESjzwwf+QzF48913eY5pEjT0bi7k8e85OmwztMM6Z7iD+J75g78fBdXKNaPNWK6n9X1B6MvguxyLs35VY873sds7Ne1q94vv+U57rb2udEzXfCOy4xz30OcZ9JLPITLs5El3llPffSDHIzjnrkRe7vhDt/Wruvz+5wx7KZGSfrZdqqN9FPOOE/5VW7fKrve0btr+pRm2kynD35Tnsxw7ny0PwTLl7mRcxz9yXumO8krvPU72KncayaK/9//mSFFNEHkFZh7AefdeRVj95flU93eTcv+2afwbr6fnKyvNKx7zpizJ13qql9d+6z3+2f+flM1RPuLUfejPRgP3I+xxwn95S7V8d3rs91/Mar43bYzczOJ8Mm3hkn67l/x3Es6tvnHrf64N94OPeVnb9qR9+tmvMp3mkO8WrXrE9d5P4cd47jrn+llqe0qhkdU51F6iLvnnNPfHqdtI5LW/WFK554J1w+jCdNhzsW9SfeO+a6x2Rv1/j3Oj7x/GqO73Squ31cW3Ez3id62T7Z3Iz3tDeZM+HEHlPeaeeTj+NRT57rqp0zXub/Tt7UK9uj6018K07V17wKr/rSneJJf8Lpf8Ol7pO57xj15Lmu0mS8J72Yk+l9/oQz9ZryJjMnHH7LhD/hyHP20xX7f2M1YNIPi4yX9Wzsl5Q+12sNzfre87rTCquie0XtL+NkPdd57RrHVTsvar4TTu7T3Gc99ZvofabqTCvsJtLHdcROuWu9PumJ32hP3BP+6lzp3D+rxX1HdP/MM+N4b1LT+8Qnt8rdo+J1/RsP53rdzXmCTeZMOLHDJ3mcxe+v+uRku1Kn/KRx/Ctq30W1ImdmPeKRi6PoOGtxFImdcmkUT3zh4iuq71G44i3u/Kjlpegc9RWJq1dFct+ZV/PUn8wSV/GrNBPfr+To+6ax28U9Kq7zvJYu63vP604rjLHSk/OOfDJnwoldpjzu7ZqsJj+b4xrnq8540548GDMtceUT3oQjv5s48a043j/V2st56k+j60915+vajvtd2Ks7TnUZ70kv7pTp/X4TztQr47l/VvtOmU/GYS/zPfWo9/zvf2lwNKl9GCkVxn7wWUde9ej9Vbnvwpoz2a92JuekJV7l9NNMcW8w18pDsfMSJ+KEN+HQ89X8U3N8v+lc501qznI+sUnu+qx+xafT+AznnnDyn3Bdy5oznub0jTx7Gcd7k9q9X9HQw/XEpvmth/Ozejp7yvMZmW7CCd0neZzFnau+OMRPuTSKzlf/K6NmcoZ6EfnUZ89zcRQdZy2OIrFTLo3iiS9cfEX1PQpXvMWdH7W8FJ2jvuIJF4/RNe+o6V/l3RzXdFxhr2ik/e7ou0cdz/vdnlOu81jTn/3Is33IOWmJK6deM4S9M07mTDix05SXcV3LOvte4lXuOudle1Q993o3L9stm3nbm/hWHO93NfdyHrFJ7vquPvm59sT/DvzVHae6jPekFzfK9H67CWfqlfHcP6t9p8wn47CX+Z561Hv+9/8l6CjqbAjg/6bOIe4Y6+CxjvxTz+dWs52X7XyjnXxfNjObm3lV2ifcieeEk+1w2/vUHN9rOnfKc3/VT/XyieherMnLcnIj796Je8LpPeU6L2o+x4k9zSfeGSfrPdnF/VhXvuRE/sp74uFa1q/sUmnoG3n2JpzQfZKXzcp6/B7Ho+ZznFjkJ9z576g1k17qReRTnz3PxVF0nLU4isROuTSKJ75w8RXV9yhc8RZ3ftTyUnSO+oqOsxYni+S9O8/mqVfNEq5Y8dgXV5HYb831LYrdd4ijWHGFM2Zc4pHH8576rnee46qnPPFfjZM5E07Mf5UXOr6pjzTOZy1ORPYjv+nRR3nmJ4xxwptw6DnNJ74Vx/tZne3hvIzT9Vyf1Z2emGuJ/ZT81R2nuoz3pBd3y/R+zwln6nXD8z28nu4l3S1fuir+/Z92BWs6tOM5xlpj2Yv8U+80t8M7LPY/4d03VtqqT68JR/wpd8KbcLK56k3jzZyp54Q3nTvlVTOf6t/le7PHiXvCufOUe+KdcM68zSfeGSfr3c6u+FPvKa+aE/13eLzTJ9t1suOEc7Pnu/zo4/Nf+Vb6Re7vhDv/HbVmTrwmXHEUO19xGDs+MWoin76p7sQ74dk+ncaxqKfPtVPdU9507pTHfV7RUM+cXux/MucOkZ/eDf/E7fAOix1POL/jhkvdbT6ZM+HcfN/J74R339hpM2zay2Zm2ld5U6/Mv+tNfCtO1e/mCfsureYrvmsP+X1F9B2jnjzXVZqM96QXczK9z59wpl43PN/D6+le0t3ypavi8ad7M/DEdVy1llPNKIwx8Hc+zlMuf9WMwiKyr1y4akZhp0hN5HwdJt6E826uz/S9NS/iDZc65a5X/6vjdK7zoq5ehrm+0mb9zE+8W98b/ol7wrVjxCm34zkW9Tuf+7t3hXu/2yvDsp5mu7f6Hqc817G+8XjHzpw9zU87nnDOeSd34kVO7OE1d3M8uHzUKid+0jv3XXW1S+Y/4YqjmPmwJx4j8cgD80d+hjuf9Ul7wsNrwuHMk8b9oubz2jHqiT3N3zGXu3V+3PUVDfXK3Wc6X/p3xFd2cE23h3P5jR0Wnh3eYdk+zs847+id5pxw7jDldjzHovaX9cRxvfoRM2zao4/yTCuM0XlR851wcm9z93Z9h3eY+3jt2qizl/Vdm+mmvVe9XtVN93Kez4vaX9WjNtNkOHvynfZiRsbl7BP+Ctc9o67eCaNX5aE+ucqFeezmiltvPTisTBS1kKL6iup7FB7Rsaqm5h15Ncf72SznVLVrK17WpzbDux61kXfcDKM+w6sedZ5nGuewzvhdj9p35j6z83ZuV7uPcx3vatd2tft0XMdO2hvcvbuavh0vw6h9mmf+XY/zOp5j1EXueFe7VnWnESauovqTKI3iRCOONO+I8pxGzpxqgufvRpvp5Zf5CPOYcbveSe/4V9Tab+KdcdWbxGzGRBccf65zvKtde6ozr5MmcH8TDTnUs3/KqXuan2YRr2aRU+WudZ7j09p9ov6ql83Ketn8jNf13KPjEnNd1MS73LUdN8Nc/2qdeXc9zul4jr2qCx9/7t3V1DovsGkv47qWNefeasPnXY87TXLOdT6xSe76qnYv5zne1a491ZVXpqu47+hn87Kez8o4XU9650R/2su4rvVac2+11D3Vyst3y2pxFTNO1ZOmiu1/u2laGbBPfuT+HFc95YmfebvHbU3vKq88Kz77mZZ4lz/Rhq+/blaGvaJ3jdeTOdRk/K5H7Ttzn3nydn5Vu4/zHO9q11Z15lFxs77rnXODu7arT743Wve6qbs5jmW+zqlq11Y877vOa+d7fcun/lWt657W3OmU+6wTn7hroybe5ZlWvUwnLIsZv+q53nmOf0WtmRPvjKveJFYzXtG6pvKu+q6v6kof/UqjfqYVNonUT/jBefd719yTj+/tfMentftE/VUvm+W9arbzTrX7nPiBV+9V7URHTjX/tk/PU+7eJz7x79CeZgbOHSOvelnftaz/a4T/Q2ySQ/ooncwSxwepr+j4pJa2i+7jXMe72rWnuvLKdBX3Xf1spveyWc6pamqdE9i0l3Fdy5pzn2pv9D6XNffLcnKVZ7ysJ34V6397/K8iTKfPF8h0zqn8M556me+7eprhceLvGtWVVngWK433M616zlUtfBql85jpndPV1He8wMid5Ce/V3GfPfVxnepKL1yx4lV96bJYaaKf8aue+zjvBndtV7tvtXfWz7Sv9rodhU28xfXYaZ3LutM5Rp3nN9yTNrycw9pnvbPmnCyvZmXcqld5RP8VjfwyrbAqVhrvu/6EO/8dtWZOvDKuepN4mlF5ZDrnZpxJz31UT7TBET+LlYdzxav6FU6+OF8ROcfzm3muZe0+xCJ/8uj1xOek5RzPn2jdK+rqZdyOT59bbcWv+pz1JK/82a/8yTnlmUemCZ73p1rpMn7lK41ixcv60mTxZofMu9Lf9rPdvFd5TnmVXn33YS0OI/HIb55rT3Xn7dqO+07M56o+zRDPY6bLONNe+Dk3q7O5U638Ko/Op9MQ04wskud5xo/e9M2ZU8fl7QX2AnuBP+AC/g/fP+CT9xP3AnuBvcBeYC/wx15g/73/7/7R78/32c+X93vm9O9S8y6R75tfYK81v9Uy9wJ7gb3A/11g/8Xzf6fYZC+wF9gL7AX2Av/6C+y/9//dP+L9+b7+8+XtXnf5dyp5m8j3zS/w66/lP/yben6mZe4F9gJ7gb9fwP9Z83d0q73AXmAvsBfYC+wF/k0X2H/v/5t+mv/8lv35/vMmk87erb/S3qe/T4fuH6q66yy2F9gL7AWKC+y/eIrDbHsvsBfYC+wF9gL/wgvsv/f/hT9UfNL+fHGMQer3inrfPy/gd/onYzvVBX79b5T/8G/q6ijb3wvsBfYCpwv4P2tO/MX3AnuBvcBeYC+wF/i9F9h/7//en91k8/35Tq70/xze6/+7m/kFeKfI980vsNea32qZe4G9wF5gL7AX2AvsBfYCe4G9wF5gL7AX2AvsBfYCX3iB/UPVFx53rfcCe4G9wF5gL7AX2AvsBfYCe4G9wF5gL7AX2AvsBeYX2D9UzW+1zL3AXmAvsBfYC+wF9gJ7gb3AXmAvsBfYC+wF9gJ7gS+8wP6h6guPu9Z7gb3AXmAvsBfYC+wF9gJ7gb3AXmAvsBfYC+wF9gLzC+wfqua3WuZeYC+wF9gL7AX2AnuBvcBeYC+wF9gL7AX2AnuBvcAXXmD/UPWFx13rvcBeYC+wF9gL7AX2AnuBvcBeYC+wF9gL7AX2AnuB+QX2D1XzWy1zL7AX2AvsBfYCe4G9wF5gL7AX2AvsBfYCe4G9wF7gCy+wf6j6wuOu9V5gL7AX2AvsBfYCe4G9wF5gL7AX2AvsBfYCe4G9wPwC+4eq+a2WuRfYC+wF9gJ7gb3AXmAvsBfYC+wF9gJ7gb3AXmAv8IUX2D9UfeFx13ovsBfYC+wF9gJ7gb3AXmAvsBfYC+wF9gJ7gb3AXmB+gf1D1fxW/wrmX3/99Z9X/lN9fOdFTcfrMHooz/jCTvGJNrwzffQmr9Ke9JluMq/ad6qt9Kd95Z/trZ44VRQvi9Rk+KRHD+adlrwsr7QZl71MR7zLb7QZd9LL5me6jFf1Mr161Kh3G+nBvPMhz/NM5xzVGXfSk95jpXVeVr+qzXSZf9XL9NE7vUx30hDP9JO54fEd2mwmv6fLn2i/6nsnt872nuhudq5mnPrVvTtdpan27fjEspnEu/xGm3EnvWp+pa346mc6YZOY6dXr9OIwVnxybvJbv4rPfjWfnCzPdBmv6mV69SqN+uJlUZwqZhr1Ko364mVRnCpmGvWoUe820kN55iHsFG+0GXfSq3aotM6veKe++6judOJUsdI6v+Kd+u6jutIJV6x4p770Hiud87L6iTbzm/bO/8tx6rS8X3GB6hft1PePO/GJh5b1Te5zK6+Ml/Wy2RnPe5nOe65R7byuloYx4xPv8le1mc571VzndbV7dFzHQuu9af10LvWTmeR7numdU9U32ow76WWzM13G816mq3qhrbBT/+ncid45qk+7Vbj0ihWPfXE9klPlrmGdaYhXeabzXqWNvnOjnrxM573Kx3lZ/RXa8LyZ5Tu8qs103vNZqp3X1dIodlxh4noU3kVqOl6H0SPyjuuYayt9xst67h/19N1oM+6k57u8oqFHpide5Zmu6mUeGTfjRS/jTnru94pGHk+01TfIu4uTueK4j/qT+Fu1k2/LOP69UU95T7XZnEnP595qJvyM88pc+byipUY+t5EekU/01Ez4GYcer8ylPvP3Hvnvzuf/5nv35PX7lgv4L9e05rJTjXihVX4bOVe5PNxXeBdf0Upzitnck8bxziMw8jOu98S/0Upzij7LZ5z0gfNN+OS8Mk/6p3Oll98pip9FaQNTHnHyxJ9oxb2N2R7ymMyVXpppdO+pLnh8NzrXykcevpNwRnFv4yse1CifzhU/i/IITHnE7pHX5ROPT83t9iSW7Uy8yzOtetLdfO8TreadomYwnjSOv6KlRrn7VrX4ESvOqf/Eg1rlmuc7Ce/ip7Sacxu5+1RLjefyiL7yiN0jb5JnXtJN5op7Gzl3qqVG+ROte0RNP+FZJG+S02PCJ+e3avkNNzm/V7n0USuPOHniT7Ti3kbuMdW+onHvJx6vaF/RvGPnnzBXO/j3VLX4XxFnv/lfMXk9f8wF/BevW8y5qitN4NWTVrHisS+ufL0m13Pneu38qMlRXvG8L75H56kOnj9qA/Pa+ayd6zW5yslRLowxMH/ie3Seanq4RrW4HqnNMOk7XujIY+6equlHvnLxGKlhP3LpxPHa+ayd6zW5VU5N5JNHTfC9zjzIYZ5x5dlhEw/5kKu883ZMmoju6dyqpod8ptyKn/V9TsbRN0zmiyvfiabjVvtkc+Qz1Yif7Zh5iM/4Ca1m+Fyvxcuic70+acSveN4X36PzVAdPzzXExImY9Z9o6S1/+jnOmjzm5DAPjj/Xee181s71mlzPneu187Oamsi759yKX/XDmx5Znc2nhnnGladjrvPa+VlNTeTdc27Fz/pPtNqJHtHzWjxGcpiTwzw4euQzF+7xt2uz75l8N3XO95pcz53rtfOzmprIu+fcil/16e1exDx3rmrnqeZ8cRnFY6SGfeb0OPGdW/Gr/nfMfefO3P82738L7R9kXHoyiHzmT7SVD/uR+zvh4jtPtXCPwhUDV87ouqwmn3nGZY9c5uR0OTWRd++G2/kE9ooXNe5xM2+q5bzIb94TrebQY7rzEy3nffJ7n8zV9yreeN1w5a/4RJt5RI+e4lTRuV5XOvapiXzyqAm+15kHOdM5mY/PO3m9Yy49fH61o/fp8ZU738zxHVXTI3pei8dITuSvPHqE3uvMk5zbud+l1Xdw/vR7n2g577fc6snOupXijdcNV/4e6RGY185n7VyvyfXcuV47P6upibx7N9zKhx7B8TrTkRP5K48e07k+hx6nPW64T+a4VjXnR89r8RjJifzm/Wlav80r309N+HntM1g712tyq5yayLt3w+18ArvxuuH63CfaJ16/ce47d/bb3dTlb6EvmNXVoIzrvSdaedFDPUVikauveMLFY3RN5ku+55l+6pFp3T+rM533XOe4audF/eS538SLmuCzPu1D7kRL/snbd3ftrV5+9JnsLJ1zvc724awMp7fnrr3RP9H6HpPvlObJXNfK8zbS52Z353odvpPH+a9oJnN9xnROtb/7TXmvzuW8yfdm+9Cj22PKezIj07LHHaLPOtv9hNO7y+nz1XM5K/um6Z63Wvpyh8n3PtFy1u3Orr3Ru5bfcMqfaDPvqZ/zon7l0Sf0rE+e5H5Sq+/k/G7XKU++VaTP5HvJV155d31pI07mZl70kM8T3ru19OOu0Wed7e54xqE/8z9Ny29X7jdQv4vUBI915N0j91YrX3p086Y8+Z7i1M953Y4+07WO39ZTvylvOn/qN+Vlc59oM78nvfS33hes6mxwxc36rs84XY965xGLvMMd6+qTb6cNzN+JL/xVXTbz5OW4au2iqP6rUT6KE5+Mq17E7olHjnqZlliG08fzJ1p6ySfrnXa61YqvyJmnXBrFE5+4NIrEXsnlo1h5CFeseFlfGsWMM+llevUidk88ctQ7aaUh/1Yjj4j0Yd+x6Qz3YM1Znd+UR+8slw8x9br5Fb/T0Lfj0Vv5E608Ison62U7ia9I3U2e6dWL6I9Yhjuf9Xdpsx2y3ul7tP9UK74idadcGsUTn7g0isROuTSKJ/4Jl49ixReuWPFO/UyvXsTuiUeOel+p1TzO6uZNefKtonyIq5fNJ5bh9Oly+ZCj3tSX/E4z5XEX5U+08ogon6yX7S6+InWnXBrFE5+4NIrETrk0iic+cWkUib2Sy0dx4pFx1YvYPfHIUe+klYb8TjPlyfcUp35TXjbvifaJ32+c+8LrMvUAACAASURBVO6ds/tNe+lv/WTB4GSv0zrmHo67f4d3WPh0+A0WXD7XnnBqs71cL85JR/y0E7nyp8Zx1eT4no6xlt4jOe7n3KjJJ171J5xOSyzym9dpHWPNGbf9p1rOi/zmdVrHWMcM1pHzOcaaPObkuN+U5x6sw4N1N4PzPKcHsao/4Uy0lc/kOyr/qh+ziPkMx1hzT+bkuN+U5x6sK49Jnxzm9H91Z/pl+XRGplWPHupFrPqOdd9GP88r/6r/dC59b3d+otV300M9/yb2mb+ipSbym9dpHWMdM1hHfvOeaLM5U7+O5xhrzrzt/wRttUN8S/X4nR2v0kefHuRVfdf4XOo8n/hTQ36Vkx959aa8TP9EKz96qBex6jsWPD7qPP8TtbyN8uwuwrJIPvGqP+FMtJVPaKtH345X6b0/9et4jrGOeawjf/qmflPedJ+p35SXzX2izfye9NKflC8Y9eS5LtNUnKpPj47TYeFR4VX/1bnhx9f5O+Za+jB3HTHlE84tt/N0jLXmeCQn8tMjn1z2Kx9yplpqKl96Me+0jrGuPKp+tdfE07XUOMb5Wd5pHWMdXqwj53OMNXnMyXG/Kc89WIcH624G53lOD2LsV97k3GrfxZ/6cFf/HsdY0585Oe435bkH68qj6nc7SEP/jj/lyZfxiVY+9FAvIvuR83UYeV1OD/LYf+fczpfzs/yJVn70UC8i+5Fnjxzi7Lu2w+iR5Z3WMdbhxTrym/dEm82Z+nU8x1hz5qQfnOx9l5a7cIdqz+BPefT2nB7E2PcdbjByK/+q73PJU07/jj/lyZfxiVY+9FAvIvuR891g5N76/hu0vJtyflfkp0c+uexXPuTcal/hc161E31P+dSv4znGOuazjvzpm/pNedN9pn5TXjb3iTbze9JLf1K+oNfVQOdNank5V33GjtNh4VHh3p/U0526uSeMMzyf7Ogc92A95XY8x1hzFnNyIj898p3bYcHt8ApjP/Kb12kdY80ZVf/0PSe88mU/8pvXaR1j7bv6XHI9r/Z7B889WJ92rvbyvnsS7zCfT90J67gx8/S6vSqMfZ/hGOtqF3Lcj5qO5xjryoP9yCuN85wbuurRs+Nl+ida+dFDPcUKYz/yVx49XF9h7N/O/S6tvo3z1VPssOB0eIWxH/nN67SOsfZd3zn3Zn9xuVu3S8dzjLXm+Hezf8JOeDVPMzq8w6RXJDfy6k15lT769HBehbEfOZ9jrCse+5FXGuc513chn54djxrlT7SZh3qK9FcvIvu+s2Os/0Qt76acN/H7icNIPvuRd9gJP2k5i9zIqzflVXrvT/06nmOsYx7ryJ++qd+UN91n6jflZXOfaDO/J73yJ+VLZrUPzjinnjycpz5jx+mw8Khw70/q6U7d3BPGGZ5PdnSOe7Cecr+LF7tyNndX3uEd1nlTF/nNm2orHvvZ3A7vsJ/4vdw38ur9JN50l8m3ZBz6O95hwT3h9CM38u6Rm/EqnP1uxk/jcZ+b7z1x33GDJzMybfRe/V7qum/7SXOf7PxE++TOT7RPdp5qM17Wq34PvP9E615+u/Cu3nRuxWM/m9HhHebf4N5PtJ1X+FaPMzveRJ9x6E+c/W5uxWOfvspPuHgRyY28elNepn+i9R1P/sSnczNe1qO38oyX9cRnzHhZjxrlGS/ric/4bl5405OzlHd4h028NcO54Vs9zux4ld77U78nvKnWd6vqqd+UV83x/tRvynP/qJ9oM78nvfq3EK6+MGvQHn0YPSPPXsfpsPCq8Kqfzc96J32Hd1g2S71XddJ7nPp9Fy/29dmnmt944joubdUX3sWptuJ5/1RzlxPXcWmrvvAuTrUZL+tls34Sb7pL9h3Rc/2pps+J6zi1nj/hutZrzar6whV/Gs/3OdX6jiy6NuNEb8rL9E+02Wz381o7VH3hp+j6Uy0/56k/id+ljd189qnm95y4jktb9YV3carNeFmvm0XsiZY+yqd+T3muP9XaL+KJ6/i7tPTJ9nBcdbePOF10/amWl/PU91jxvH+q3Ze1a4kxn/KoUf5EGx6uP9W3c90vmylPj/8WrX/XzQ0yrt/Fa85z7FRT67lrHVc95Yl/ilO/J7yp9rSr8KnflCffU5z6TXnZvCfazO9JL/+LUOPYLd9hjeV/oYm249xgwdXrdOJ08aTvcMeinjzXTTQd58Zvwp1wYp8pL+O61mt+r2OnutMSO+U+J+NXHO+fanqfuI53WmKnvPOVtuJUfekiTjg3vCm3mlv1uXOVu/ZU0+fEdZxaz59wXes1Z3WYeBNOcKe8Kbfy8/6p1ndk0bUZR70brjSK79S6l9eaGbHDyMty155qejiX2Cn/KVrfw2t+h2OnutMSO+U+J+NXnKqfeXjvifaJ12RuxfH+qeaeJ67j79LSJ/JuzhPuSetzvaa+w8SrON4/1fLLomszjno3XGkU36l1L681M2KHiVdxqr50nf9v1PK7um9zXsb17/eaHo6damo9d63jrG+41GX5jdeEW3GqfrbTqXfjdcP9KXPfufPpmzo8/ctILFe9bnHHTj6ccdLe4lPvk6/7sI7c9U/x8PPnPZ/pOPUdJp77qZ9F52b+zsl8ojflZVzXes2Zjp3qk5a48vD0l805cYRn2q4nXcSOl2EnLXHl4ePv5B1858jD+xN/aT26l+OsnXszd6KNWRPPzIu9087kek6t50+4rvWasxyL2p9zHFc95QXfuTdzM23X035ZdF3GUc+5UWcv679bm/mpx53UYySuPHB/1Exy6jM+ceXTueIzfkKbfYd6vov6k3jSElcevv6yWSeO8Ik2uK/OrbSar+h7qJ9F5052k0+m7XrSRex4GfYuLX2yPRxnfdpL3OD5y7Rdj/qMRzxy5wj3/qmWLouuzTjqOTfq7GX9d2szP/W4k3qMxCMnFrme94lVnKr/G7TaXdG/X/0sOvdU0+PEdZxaz59wQ5u9qk/uu+dWft6vdqv6X7lzeP+kue+8Fe92m6e/VdlyVc8HVrys/0QbfnyZf9d7VUtd5D7jKe5+qk++4mXRtV67xnGvnX+qXa/adep7JM8x1uRFHo89cj0nT1pxHKtq8RkrbtUPLTF6eU5e5E+18nffqhafseJW/XdpOx9iWV7tVvXpUXG8X2nY9zzzYM/5rMmLvHqv8Cqv6Hd+jp3qao7rKp76zj/V/h3yyaJ7ZRz3C83puW9VZz4V1/vUEmPfc/Ii53OsqisN+567F3HHqpoa5RXX++IzOqeqKw37nrtX4Ow5nzV5kfM5VtXUKK+4VV+6iBXH+9Qod05Vi19F11U89Z1/qkNHjnyySF7k36md7pfx1PPvqWrxI5LDvufkRc7n2KkOLTn08py8yKs35Unv/KoWn7Hier/SsO/51MN5WU3vDO96v13b7U+MOe/BvufkRR6PPeezJk9a4sqnvIrvetXiV1E8xYqnvnjTKF3EVzTUK3cf9avo/Kqu9Oq7Tv0qOr+qM33F9X6mfVcv/aevL1DV1RIV3/uud/xUP9G/qj3pbvHgn74z8OxNdJWWfu5DrMpd09VTjwmv4kTfd/DerZZ8985q8pln3KoXOmL08Zy8yJ9q6e/eWU0+84xb9aiLvOJlfdeqdq76XXRNV7tPxxVGjXoRu0eeuOzdajM+/TTjxMtw9U5+jne1PD26xvGsdk1Xh5545qceeZFXb8qj3jVZTT7zjOu9is++551HcB3PanoSZ99z8iL353hWu0Z1xvWeuB6dl9XUEGffc/Iij8ee81mTJ22HOz/TSJ9xq540ihWPfXE9klPlrvHadY5ntWu6OvTEMz/1yIv8O7XaiTHbj3iWuyarqSPOvufkRe7P8a4OLXH3Yk1e5NWb8qh3TVaTzzzjeq/is+955xFcx7v65P1v1+r7/TvV90ieY6zJizwee+R6Tp60znG/jkete2c1+VnumozjPdd09Sta13jt8xzPatdkdaZjzzXEqtw1Wf0V2srzpl/+0zf7CPWmA8T3eNJX/KpPv4pT9amN3HmqnadauKL6iuorqp9FcTxmXPacr5qcLhdfseM6Jk0Wneu1axxXTZ56VXSu15Uu+icucc87X2GuYS3OZI+OW3lSo/zEJe65PLroGtadLjByPb/VnvjEfRZr8jwnz/OO65jX9ArMa+eznnDJibx65FUc9SdccjyXTxVv+fRxLeuKx36WVx7kkhP59LmO9cmDXM9dS9wxr09c4p7fepFPL/aVE/dcnCo6n3WlUZ9cz8VRJK5eFZ3rdaWL/olL3PPOV5hrWIuTRfI8z/jsOZ81eVVOfuQ3z7Ws6VP1yVHuXK/Fy6Jzvc406k245EQ+fa5j7R4d9gqXfp7Tjxj7WT7hkhP59LmO9cmDXM9dS9wxrydccjx3P6+dz9q5XpPruXO9dj5r53pNrufO9XrKJ889vHau185nPeGSE/n0uY71xIP8m7nh7VrW3WzyPO90wl7RPNn3O+f6t7LWXl8V57+FX7XBhS8PE/m+vcBeYC+wF9gL7AX2AnuBvcBeYC+wF9gL7AX2AnuBf88FftVfe/YPVf+eX7z9kr3AXmAvsBfYC+wF9gJ7gb3AXmAvsBfYC+wF9gJ+gf1DlV9k673AXmAvsBfYC+wF9gJ7gb3AXmAvsBfYC+wF9gJ7gW+5wP6h6lvOvkP3AnuBvcBeYC+wF9gL7AX2AnuBvcBeYC+wF9gL7AX8AvuHKr/I1nuBvcBeYC+wF9gL7AX2AnuBvcBeYC+wF9gL7AX2At9ygV/1h6pvudAO3QvsBfYCe4G9wF5gL7AX2AvsBfYCe4G9wF5gL7AX+MgF9g9VHznzDtkL7AX2AnuBvcBeYC+wF9gL7AX2AnuBvcBeYC+wFzhdYP9QdbrQ4nuBvcBeYC+wF9gL7AX2AnuBvcBeYC+wF9gL7AX2Ah+5wP6h6iNn3iF7gb3AXmAvsBfYC+wF9gJ7gb3AXmAvsBfYC+wF9gKnC+wfqk4XWnwvsBfYC+wF9gJ7gb3AXmAvsBfYC+wF9gJ7gb3AXuAjF9g/VH3kzDtkL7AX2AvsBfYCe4G9wF5gL7AX2AvsBfYCe4G9wF7gdIH9Q9XpQovvBfYCe4G9wF5gL7AX2AvsBfYCe4G9wF5gL7AX2At85AL7h6qPnHmH7AX2AnuBvcBeYC+wF9gL7AX2AnuBvcBeYC+wF9gLnC6wf6g6XWjxvcBeYC+wF9gL7AX2AnuBvcBeYC+wF9gL7AX2AnuBj1xg/1D1kTPvkL3AXmAvsBfYC+wF9gJ7gb3AXmAvsBfYC+wF9gJ7gdMF/vGHqr/++us/r/ynGtR5VZroZ7qKn3EnvWrOVJvtU2kzLnuZjvgpz/TqfZVW/oynWcSp85w8z50b9c3L9OrRR73bGB63GvGfarl/5eUc1drhNkrvsfNxrtdTbcfrMJ+nutIIr2Kmq7hZP9NHz1/FO/XD58Sp8O/U+vefdsn46mXfJ2wSM716nV4cxo7vGHXMnec1ucqd09XSZJG6DJ/0wmPCyzicr3zKe/fck5/2q2K2t3qVpprZ8YnJn5F4l1OjvOILv43hd6sRP9tFGGPGy3rUKM94WU98xoyX9ahRnvGiJ/w23vpVfParHciJvOKd+u6jutOJU8VKW/HVz3TCJjHTR+/0Mt1JI/yJNjwyvXqaUUXxPFZ89Z0f9c3L9OqdfMTL4ldpX5mlXW60GXfS0yyPnda5W+8Fphf4x3/bu1+0DvOBHdcx10btnKirl3EnvWrOVMt9XtGc9MSrfDJXHPdQfxJdG3Wmy3jey3RVz7WfnFvtdOpXO550gT/V/tcA/yebCfhvacad9P5mUvxeVD5PtZXvqf/KXNewzuYRr/JM5z1qHZvW4THlOu87ta9+O3XK/buinrxMV/Uyv4yb8byX6bznGtbOjXryMl3VC78KO/Wfav1bsnnOUZ1xJz3pFScacaRRVH8SpWHMdMS7/FPabM6kF7tPeBkn++4p77dqs++b9Px7X9HI41Y74WcczVPMOFVPGsWKx764WSRPecbznrhddA3rTEe8y1/VZrqq5/MrHvuuYU2ecuJVLu4kusdEI847teElX0afUdXUKL/hStNF9+u4jrl2673A6QL/+F+u/ks1rTloqhGPWuXElEfMHvGbPLxu+ORyD/a7nBrPpfOdnMdamml8l1Y+mhu18ojdI2+SZ17SffVczbmNvteN/qnW76XZ7us8x6WbRHpN+OQ81dLrJn9lLjWea3b0lUfsHnldTo+O12G+V8d17Du1+nbf6VRLxyiNfw85nkszja7nLObh170n8+Qrj6+YW3mrP4m+10Qjjr6RkZjyiNkjfpPT60bnezzRagd5RK3c54jrUfyv1mrObfS9bvT+rfRiHp6Tp9k/WasdbyO/f6qlRvkr2qnGeZoZ0bFT/YqWGs81z3dxHmtpTpEaz6WNvvKIkyf+jVaaaeQer2io557Mw7d707ni0Uu9aXyXVj6aG7XyiJMn/kQr7m3kHk+09Nl8L1BdYPSb77+IlVn0nau60gTuTxphXjs/q6mRT8bLejda51azqn7Mp0dWT3aUR8aVpzBxPQr3GDx/1AbmtfOdM+W7j+u8dv675sqX8yK/eZ/Wcp7fYbo3PU7f61zV1Sz6ietxonXO1CN0zo06e1XfPbK68vPZFS/rszfxIZ/5b9D6jqr5HcwD9yeNMK+dHzU5zDOu+I65zmvny4e8yLNX9d0jqys/n3uakfn4vM4j0/sOGcd71Ph851Y1PU47O1d15y1MXI/CPQbPH7WBee181s71mlzPneu187Oamshv3itaamKW191853r9E7XaibtG3j3nVvys/0TrO7mX46ydq5oc5oHricsojJEa9iN3rdfOd434FS/ru0dWV7qMqx0iVo8c5h1fGPnKhTEGVj3pxPE605HDPONGT97KqVH+1Vr5a5528lq8LDrX60zjPWoi755zVVeak1+l2/6ffYH+t/B/b6NfPsXuZOIodtwKkzZiPK8rHfvUyId4l99ob7jVTHoEx+tMR07kN++JVnPoET2vxWMkJ/JXHj1C73XmSc6rc+X7xOvTWs6b3krfqUiPyLt3w3WfJ9onXu+YS4/YxWvfzznBf/I479brN2if7Ki70sPvL45HaiJ/5dEj9F5nnuT8trn6niff8IqWmumdtasiPSLv3g3XfZ5o5UWP6HktXhad63WmUc+5XovXRWoiv3mvaKmJWV53853r9U/UaifuGnn3brju80T7xOvJ3Cda7UyP6HktHiM5kb/y6BF6rztP53qdacmJ/OY90WoOPaLntXiM5ER+875Lqx05P3pei5dF53qdabxHTeTdu+F2PovtBboL9L+F/6uc/jI67/RLXi1Gn+Cwnnq+otE+U+2UJ98q0mfyveQrr7y9Lz6jcya161lH7s/xjOOarKZP4KwzT8czTjan6rlfxcv6n9ZyXuzDOvLJm2qcN/XP9rrRZt/gu2ScbG7FO/U5b+JL/tNvnczr9n+yyye0PuPVe9FncjPylXd3rDBpI74yt/I99Z/O1b6nORXO+bder2ipiZ1YT+dPNc6b+md73Wh5a+6Q+ZLr+XdptQfn337/K1pqfsutnux8e2fOivzmPdFmc6Z+zrvZ27XZHpMefYLPOtvnhE9m+hyvs7n05Q4TLfnK6dfl4it23A6TPmI81upR73jGIZ/5d2mrHaLvO5HrObm3WnnRI/LqOa/jVh7b3wtMLlD/FkLtv5CA/pZOeX8TJYV8CKkXcfLIn2rkO9VOefKtonyIqxfRH7EMdz7rJ9rMJ+tlO/32ufrOJ9/xaa3mafeI6kWcPPI7zZSXzXyifeL3rrny4S7qRfRHLMOdf6qf+P107ZP9eDf5ZL3A/Imv6Pi0zvTqRfRHLMOdX9XyIa5e5kssw+kzyZ/4vaKVhrupF3HyyO80U14284mWfvLJet3uwf8urXbVfEX1J1EaxRsNudJH7J545Kj3U7XalXt2u0558mV8oqWP8qnflCdfxifazCfrxQx/3zWXe2iHrPfunTVLkTNv8kyvXkR/xDLc+ay/S5vtkPVO36P9X9FKIw9F9T0KV3R8673Auy7wz/+WJ876RVRMKP9tCVckT70sVrxJnxzmPofYKZ9qp7xuHj3Iq/rBIRY5n2Osn2o1xz1P/d8+V993+g7yspx3i/zm3WrJ55yqTw5z8iOvXsdzjHX4sY6czzHW5DEnx/1e4VHjOWcRq/rBIdbtR78uf+L307Xdfo6x5r1u+6efEf08/8q59O5y7kRe1Q8Oscj5HGNNHnNy3I+8LL/Vkk+/qk8Oc/Ijr17Hc4x1+LGOnM8x1hVv0p9wqlnv0lY+fgPyspx7TrTk06/qTzg/WVvt392K39Px6K38iVYejFO/jucY65jFOvJXHj2or/rfOVf7VbtV/dPO1Hl+0mqnU6QvuVX/NJc6z79Tq2/jTur5Xuwzf6KtfMKzepznPMdYV37b3wtUF6h/C6HgL5n/QoLW/gvAPVhXHlW/20Ea+k/40kWcaqc8entOD2LsR853g5EbHqxPvuRW86v+ybvT/bS53a7ETjm/y+/zbi1n0Zv9yQ5TfsdzjHXsxjpyPsdYk8ecHPd7hUeN55xFjH3focPoMc2f+P10bbefY6x5u0k/OHzUnDByK4+qf/KmbppP9rmZSz/Pq52mvEx/qyWffuxHfnpTfsdzjHXMZx05n2OsK17Vd2/xJp5fodX8iNyhmkU+81st+bc+v1F7+43B53dGfvOeaLM5U7+O5xjrp9+rnd0z6weHjxrHyOtyepDHfuVNzlRLjfs6xjr8WbuW87ucHuSx7943GLmnncn1/KlW30Zf9U7e4j3RymM668TjLp5z1uZ7gckF/v5P0kIx/UXreI6x5tiqH5wOo4dzQ3fzOKfTTnndbHo4r8LYj5zPMdbBY32r1Rx6qKdYYez/trn6tojdd5CX5Z/Ucpbv0mEdN3TVo6fzHGMdfqxvtdk+nR/5Ux41ntNjilET+dP3xO+na7v9HGPNm1b94FQY+5HzOca64rEf+UTjc92jqivvV+fSz/PJDrffMZ2h2eSrp9hh4iiSG3n1Op5jrMOPdeR8jrGueOxHXmnE6/AOO3mftJrvPqG7eZwz0ZLvczrM9/wtWu7J7+tuNeXRW/kTrTwYp34dzzHWMYt15K88eri+wtj/5Fztx/nqKVYY+76zY6zDl7VrNfcU6eHcCmPf5zrG+rQzuZ4/1erb6KueYof5fGkUT1rx3Cd01aOn8xxjXfltfy9QXaD+LYSCv2T+Cwna+B9MlR/79FV+wsWLSG7kN2+qnfKq2dRnnApnP/LqZbysl+krHvsnHXHqIq9exWM/01Y4+5FX7928bM50xlMt55y8Mpw9ekVevSe8J9psn3f7ZTOixzkZp8LZj/zpe+L307XT/Soe+9mdK5z9yKtX8djPtBXOfuS3j/pMW+Hsd3PfzTvt2O0SWu5z8spw9ujVzX3Ce6J9+r2cze9W3uEdNtlLM5wbvjePe5y05GYzOrzD/Bvc+7u03R6xU/W4b8fL9E+0T/ymczNe1st2qXrUZ5wKZz/y20d9pu3wDguvCmc/8uplvKxX6bM+9Tc4dZFXL+NlvUyf8bLeVBs86k86x59oO6/wrR5nvoNXzdn+XqD+LcRt3v0LWfl5/1RjxX+krv0HoWlMtVNeNcr1p1o+zlPfY8bLeq6LuuJ5/1TL23nqe6x43j/V8nWe+h7fzXP/qKcznmp9zqnO5qnnWvU9PuE90foeUb/bL5uRzfG5Xsun6gu/jU/8frp2ul/F8/6p1u2dp77Hiuf9Uy1f56k/ja4/1fJ1nvoe381z/6inMzKua73O5qk35T7hPdE+/V6ffap1l6dz6ZN5Od7VvvMN17Ve08uxU/0TtNwhct/ZcdVTnviMT7T0UT71e8KbarWTR9efaumdp/40uv5U0/fEdVzaqi9cMeNlPfEn0fWnWp7OU99jxst6ros642W9qTbzdD+v6e3YqabWc9c6rvrdPPlu3Av4Bd76h6own/zyVhzvn2r/GNauJXbKb7Q3XJ/r2lNNvXOJKa84VV+6iBXH+6d64jnhnOY4PvG85QS/m0O/LP+U1uec6mxX9VyrfhYn3IpT9Tlnwgn+lHfL5S6Z1ud6TX2HkTfJn3j9Bu1kx4rj/VPNezuXmPKK4/1TLb+IziV2yl17qunnXGLKJ5zgTnnyZbzROvdUc47nrnWc9YRbcar+xN+1p7rz/JSWO0Tucx3v6hutc0815564jv8ELXeIvNvxCfed2idek++rOFXf98lq155qejiX2Cl37amm34nreKclprzSV33puujaU00v5xJTXnGqvnQRK07Vf0XrXl53ns71mlrP38298fNdtt4LxAW+/A9V8Uvqr/rF9f6pdl/WriV2ym+0zo06e1k/03Y9+mY84pE7R7j3o/bnHOHeP9XSRcy4xDOO8Ezb9aTLPEPnz70cVz3lic/4Ka3POdXc0XPXOs7auVH7c45w799o5aHoXupn0bnZ3NBl/Uzb9Tg/4xFXns0Vpuhe6k/ib9D6jtlNnKNv9/6pli5ixiWecYRn2q4nXeYZuuxl/W5GhtH3hGe7Uc/cvYid8hutc091N9u1N9zQ+qv8vP9Um/mpx53Um8Z3aekTuc93vKtvtM491Zx74jr+E7TcIfJuxxM3tNnL+j4n42ifzJM99yLmuXOzuc6Rh/czbXCzfqbtepopP+cSV/6Jub4Ha+0x3Zla7u59Yj6DdTU381OPevUYiWf+wqlRLkxRfcWqH7g/aRSFq55G6SJONeJR67k4io6zFoeReOTEIt+3F7i9wOi35vYXzfmnOpYmp/sI8iKv3pSX6W+1zq9qziKHfc/Ji5zPsVP9VEt/enlOXuR8jp3q0JJDL8/Ji5zPsVNNLXPXETvln9ByRrcPeZFXb8qT3vmnWrqIJ67j1DKf8qRxflWLH5Ec9j0nL3I+x6qamix3Xcaper9F63ue6vheUkMRfQAAA+RJREFUcqrvd15o+Ogxyd2PXp67H3HHqrrSsO+5exF37FRTy9x1xE75VEte50le5NWb8qR3/qmWLuKJ67hr6OX5T9L6bv4dsevNy74t05OX4eqRp13YEy+L5H23drpfxlPPv6eqxWesuN6nJsuf8l3vNWc6VtWVhn3P3Yu4Y1Vdadj33L0CZ8/5rMmLnM+xU/2KttKw77nvQdyxU/2dWu7GPTwnL/J47DmfNXnSElc+5VV813st3ca9wPQCf/+nUaF65RfNNV0dY4kXa/y3TV7k1ZvyMv0rWtdkNWcRZ99z8iL353hXP9XS271Ykxe5P8e7OrTE3Ys1eZH7c7yrXavaNepP4ie0nNHtRF7k1ZvyqHdNV1MXecd1zLWqpzzxp3MrPvuen3ZxPKvd02vXON7Vv0nru3Z1fDPxJzegzyn/9Fx+F3dj33PyIvfneFe7VrVr1J/EqZa8zpe8yKs35VHvmq6mLvKO65jz3Yv1T9JyL+XZfsJOcaolr/MkL/J47P0WbbYnv0PflvHYc01Wk88843qP/Cy/5YeHa7raZ3ZcYdSoF7F75GVcx7Oa/sTZ95y8yOOx53zW5Enb4c5nTV3kxKqcGnLY95y8yP053tXfqeVevgdr8iKPxx65npMnrXPcr+NR695dTd3me4HJBf753+xE5b90CaVsuZY1RVWfHOUTLjmR37xXta5j7fM77BUu/Tx3P6+dz5rcqk+O8gmXHM/lE5EY+1k+4ZLjeebJ3i3/01rux9lZPuGSE/nNcy3rkw+5nt9qT3zhPoe1OIodJo7iiUvcc3l08RWN/P4NWn6Dviti1SdH+YRLjufyefdcn8OaM9891/04N/LTu+XTb6olj/osn3DJifzmuZb1yYdcz6klxn6WO9frTKOec70WL4sTLjmR37yplryTv3O97vTO9fpT2mwOd4l8+lzH+uRBrucnbeCvaOTrWtbiZJE8z51P3DGvT1zint96kU+v6HtNrucTLjmeux9r57ImL/IOe4VLP8/dz2vns3au1+R6Ti4x9rPcuV5nGvUmXHIiv3muZX3js9y9AC9w91tI5eZ7gb3AXmAvsBfYC+wF9gJ7gb3AXmAvsBfYC+wF9gJ7gTdeYP9Q9cZjrtVeYC+wF9gL7AX2AnuBvcBeYC+wF9gL7AX2AnuBvcDrF9g/VL1+u1XuBfYCe4G9wF5gL7AX2AvsBfYCe4G9wF5gL7AX2Au88QL7h6o3HnOt9gJ7gb3AXmAvsBfYC+wF9gJ7gb3AXmAvsBfYC+wFXr/A/qHq9dutci+wF9gL7AX2AnuBvcBeYC+wF9gL7AX2AnuBvcBe4I0X+B9b4cUbRSoukgAAAABJRU5ErkJggg=="
    }
   },
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Para essa seção, usaremos um arquivo FASTA que contém o a sequência completa do plasmídeo pPCP1 da bactéria *Yersinia pestis biovar Microtus*, podendo ser baixado pelo NCBI (https://www.ncbi.nlm.nih.gov/nuccore/NC_005816.1) ou no diretório deste mesmo documento no GitHub.\n",
    "\n",
    "O arquivo deve começar da seguinte maneira (você pode notar que há somente um registro presente, uma única linha após o sinal de maior (>)):\n",
    "\n",
    "![image.png](attachment:image.png) ...\n",
    "\n",
    "Se precisar, volte ao Capítulo 2 (https://github.com/GTL98/biopython_br/tree/main/Cap%202%20-%20O%20que%20podemos%20fazer%20com%20o%20Biopython) e veja como funciona o `Bio.SeqIO.parse(...)` e o loop para percorrer todos os registros em um arquivo como os objetos **SeqRecords**. O módulo `Bio.SeqIO` possui uma função irmã para uso em arquivos que possuem somente uma linha de registro, como esse que estamos vendo agora:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "registro = SeqIO.read('NC_005816.1.fna', 'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SeqRecord(seq=Seq('TGTAACGAACGGTGCAATAGTGATCCACACCCAACGCCTGAAATCAGATCCAGG...CTG'), id='NC_005816.1', name='NC_005816.1', description='NC_005816.1 Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, complete sequence', dbxrefs=[])"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "registro"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Agora, vamos olhar os atributos chaves individualmente do objeto **SeqRecord** (começando com o atributo **seq** que retornará um objeto **Seq**):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('TGTAACGAACGGTGCAATAGTGATCCACACCCAACGCCTGAAATCAGATCCAGG...CTG')"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "registro.seq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Os identificadores e descrição:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'NC_005816.1'"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "registro.id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'NC_005816.1'"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "registro.name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'NC_005816.1 Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, complete sequence'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "registro.description"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Observe nenhum dos outros atributos de anotação são preenchidos ao ler um arquivo FASTA:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "registro.dbxrefs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{}"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "registro.annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{}"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "registro.letter_annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "registro.features"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Nesse nosso exemplo, o arquivo FASTA veio do NCBI; que por sua vez possui um conjunto de convenções bem estabelecidas para formatar as linhas de um arquivo FASTA. Porém, os arquivos FASTA de outras fontes variam, o que não torna possível generalizá-lo."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4.2.3 Objetos SeqRecord de arquivos GenBank"
   ]
  },
  {
   "attachments": {
    "image-2.png": {
     "image/png": "iVBORw0KGgoAAAANSUhEUgAAAoEAAAE4CAYAAAAkfimYAAAgAElEQVR4Ae2cgZIjSY5j9/9/es7YN28ahSI9PKQISZmCzLQkAZB0hzJTYXNX/X//ySsOxIE4EAfiQByIA3Hg6xz4v6+7cS4cB+JAHIgDcSAOxIE48J88BOaHIA7EgTgQB+JAHIgDX+hAHgK/8EPPleNAHIgDcSAOxIE4kIfA/AzEgTgQB+JAHIgDceALHchD4Bd+6LlyHIgDcSAOxIE4EAfyEJifgTgQB+JAHIgDcSAOfKEDeQj8wg89V44DcSAOxIE4EAfiQB4C8zMQB+JAHIgDcSAOxIEvdOCvh8D/+7+/oD9sKV7ff5D/LSa+m+2Y9jrX7fpNGHfXO3UeoOs47Z3yo/6JV5zcd4ATnacufnrRu9JMvcF/tgP62U+fv2q626545ab53czCzuqnOe/C/e5+H3g9n2vwodNq30/N33kvdneev9LP1TmU83PC6Vldo1yXM4N4pOn4FcZcYqeFq8hLMc/R7MSdXtX4TOUqn14rznv+mrJq7jjHVrVzdRjFNPeDfkNd93cPztZHPp2dp3rNuz0dP2EdXjMnvNsX7Hc5UJ+9fv6ac9MOg6vovNaVe629R7n2Hmk/ke/Or1jlWtcdvFas4z7x3mfO1Hlwpv8K7Tt99d1aa849Fatc69J4TV8XXfvsPN/h87vzucbrrsf3TLXPqrrDtF95zTmHYytc52q+/RBYy7qFLNWoCzTv+hXTXPu+Ja/7uwdaa/6IJ1M/ONFngxOdp+54x6iJ9FbsMOWT/24Hus9fMc07JyYenKi9Haa85me02vcp+XR+8IrknNnrwsGIaBOvceBdvq72Fjfx4J0G7lFntF/zs/O6szGDuUTwKe7qpn7FdZbmk0bxyr2Hmuj6rj71ENgN0IMcLe54xTSfdv1mnPsT1VvPH/FB59JfGDgRjp3gROU1L141XrtW68q117nUv9+B7vNXTPPOjY4vDJyovR2mvOZoK/I+w6v2HTnn993gHksHpj1gFcmV//SccxM5L/V0J/AdXaehn30VdzHvYb7jzOx41XZ5dxZ0OxwaImdhxiPxqlk6x88BR3Te612d93W1ztIcbWEdrjy5xlWP6ir/qIfAfw40/IH1g//Gmg+OiB/cVXGwM1H7K6f2yF7H6SFOu4945nu/7tuZ4f2pf7YDfP7cwn8G4MGpVa85vMdOA7aK7EVztqbvXREffD+4x9KB0XNUo/vU6OfvzjlpCnfuqNb5ri1uF2OO67WuXOtpPrM8eq/yOxwa4tn9uo+8m1UYb3RHUee4Fo7ovNe7Ou/rap3lOTXxqF/5VY/qKv+4h0AOWJc4cxH6fnLU+5IT//mwFv8fQXfuzayK5DoXbOJ1h2vgmKFz4TSqDryb2enQJ/4+B/gZ4HMn1k3h9NbOdzrX6BzldG6Xd1rFNKe/w+BeHaezgBPrXOREznpUo/vE6GefzjjpOlwxzbvZHb+LdfMK037N0XcYnMeVdodTDTnRd+3U3lt1hz0yS3uYSVSuy3d1Xa9iPoe6InnpNV/173Kq+2f+X8DwoDEd5J8h/+1ZaVSnOx/p0f7flKsX5MTJvzP3r1k6j16wIx49kb6pLtw1Z7VTP3MSf7cD+vlrzq0Vq1zrTgNG7PRwHjutYprT22Fwr47TWcCJdS5yImc9qtF9YvSzT2ecdBPOnEf4rqfDdEfx+laO/JF4tHeaSR+xdOTEqXfCu74drDT6Zn7X69xKg7biSqe7j3Q6k7ldzy6m87oe5TV/+n8JrGUsJOoCzTu+w456lP9NuXpBTqx7av7Ivad+cKLP3sU7XYdNd+m0HebnS/17HdDPX3NurJjm8BUn/IjTGZNWZ2tOb4fBvTp2ZykMnKh3dQy9x1ff5ZF9epdV/6SbcGY9wnc9HVY7OlwxzTnTmbjqn7jC4Yh6VsV2zzL1dHiHdXsmXeFwxK5fsV2d9mg+9e/ik25nh2rItx8Cq6Fb7tiqLk55rzmURtUr/htzv2vnT6c548VR/xHPrulsZ/qZpVH7ux2qTf67HdCfBW6qWPfzoXz1eM2cI0515D7rbM2cd8XuvIppXmesWjHN9Q4TrppPyP0+05mm+0w4c47me/+kd53OJ6/o/VOf9hzlPkNrzZmjmObd+ehZRZ/hWuUr19q1Xndax47qmuka37Oqj3qdP6qnXd436QpvHwJrgL8ZMuHw/wyVfsWd84P6bOd91m+r/b744fcEd73rpvqof+InXPeopnJ/OU+NjpoInvgdDvC5V+xeyu9ofIb2O3dUa2+3+4g/mn837+fzO3S1YprrWSdcNZ+U13n1zdkU01x58lXU3sr1pVzhyiun+Zl+1T6aT7s578QXri90iq1y9B61Z8Wpbsp3+lXTzSn+kZfO1dxnTZzimmu/4pqrxvPHbuNTUseBOBAHvtyB+qObVxyIA3HgJzmQv1o/6dPKWeNAHPhYB/IQ+LEfTQ4WB+LA4EAeAgdjfhqs/9Nvl/+0++S8ceCnOZCHwJ/2ieW8cSAO5CEwPwNxIA7EgTgQB+JAHPhCB/IQ+IUfeq4cB+JAHIgDcSAOxIE8BOZnIA7EgTgQB+JAHIgDX+hAHgK/8EPPleNAHIgDcSAOxIE4kIfA/AzEgTgQB+JAHIgDceALHfjjIfDoX5Ue8eXfSrPitFc/h+rhtervOLCd/kkDXtHnganm0bxmd68J77RXYN0dz8x9tv/Mrndo77zfzmd95/5X+Ol39Pt4/YozZce+A3w+xKmz+OlF76Q54mvu1Dvt3MF39u7MeUbDGXZnXOkDuzV255h4xcm7/sKK/20v7vzT7vbHJ9EdXjHNpw/wGU31er/XOz9AXY+ed+I7XLHKtd45i+49yu+cfbQbvrsj3E58tn9nx50a/wx81533O9pdZ7lzv9/16trPzn2Jn3A/PcvV979i3jvPV7t1v+Z6N9c590xdvav5OvvRfLrXo/PO9J2925Vn9VndWToN9+s4x0rbzWXGb4jdnXfv9Uzv7g7XHT4EVgMHI/oQrZ/RVK/3e63n0b2adz1HfLebHuZ1Gji0z0Sf5fUzs9O758A7PX/n7j13nlNxvyk+N/2abs52zbTrp7zzfN1ux6iJ6kCHFQ9O1B7lNZ+03vtIfefsR86z6rnyrN0sxTTvztTxjlETuzk/HXvmbs/0Purbxz0E1kXUCM25ZIfBeb/i5F1/h7keDXFnFzN2I7OJ3le4vpWnZ+JLqxx6ZigHphH9pJtwnVH5rq7rO+rX2ZX7a+Id15oZHQZHVE3lZ17odYb2T7hqKlcdM8Fd2+Har3pmTbxqu1z7da/izO76wdAQFWeuc2icL12H0Q+P5h+x/B/l0XS90rJMtVdnO641A9F3HJpnIvN1RocV3+ETBk70+ROuup2cORV5d33oOq4wejudcso7rjV7OgxOo+oq332h1X7thZ+wjp+04FPPhNP3TKzZ+vZZynEOx7Smv9PCaUSnGLnOVZ3jWtNbUXHtV83Z/I+foGkouB8AXJd2mPKVTxpw4qRV3mdPParr+juMHjiPO7uYsRu7HfQWB8/uVY1G+8lXUWeqzvcX12k7jDnOeY2ui6VV/SN1N1cxna+45pNmwrV3lR/dh97VHud0puY6S3s0L43WUz+zjiKzfA44/V6DVzzinN+pVaO57p12q15z792pd/pXmuKc93rnHJPGZ3X76HVt4YppLzgRLTWR2fBa7+Q1R2d5zQzVgBGd01pz9MTi4D1HQ0RHrdE5r1Xrue4trqu154hXLfMU837ntL4qr52rl/OckVi9njNPcdehAdeafKe/NNNrp3/qXeF/bJwOAE5cDvzvDzsH7no6rGYqTk7UnR32LL+aCUfUsyqmZ3gmn2Z2uGKas18xzeG7OOk6fBfr9hTW9Z/Rar/mzFBMc3iPz2h2en2f1l3/LlZzOq3jrvFaz3PU67z3es2uiuTdDOV0xoSj6XjFNKfH908a19Gves3hz8Sd/pWm4zrszJlcW/N4FzfN73Aw7dcZR7yeBa1iR3nXs4tNs7Vfc9crN+X0KA9WscM7THs077SKVa5v7Z32q0Z7da5qyI94dGfi0cyJL1y5nZxzqXaFFbej7TSruSs9fUfxlofAw6XD065eiJyoMzvsWX41E45Yu8iJuv/ZfJpZePdm39QHX1H7Fdd8mtPhuxjzdT853FE82sU8jzpXOcXJux1wxJXmaD4zutjN3cVqXqd13DVdXZi+OatrwXcj/cymD3yqj/Bned2vOXOJHedY1bzpOxPp9bnMmPDiVxz9V8dpZ4cXNuGcf8Xr2Tud8l3e9exizCu9v+EqKuc4te7UvOPBKp7Rah/5UX/H0zvtP8M/qtW+Vf7o+atPe3dyzqHaFVYcezzSh0Zrzb2PWjWP5IcPgbqo8qPXMxrtJSfq3g57lp9mFg5HrF3kRN3/bD7NnHD2HfHoiJP+DN5pO6x2dniHcT6PnVYxzb23qzt9h3nvjqZ6dnXM7/S72Gqfz6Am7u53PX27cep33GvmT/izvM7VnLnEjuuwlR5uJ3azO4xZKw7N1XHa2eEdVucBJ/oZO7zDvM/rrmcX03Pq3K4fXrmdvOsDq6gzwDsMzmOnVUxz7636WV5nHs1S7W5+NHPiC1duJ+dMql1hxXVaeogrzYqj/5H4x1OdL6laMc2nZc9ovLdqx2pvh+l5HuW7PsU05xyO6TkezaeZhU8c5zmzc5p1Bu+0Hdadr3STtruHa7vasW4OWKctrMPpqXjEo93VTfqpf4U757We37muVkxzznwmTv2Oe82OwidO74XeY9fv8zoNc85od87D3Cn6PmZ2ONw0S/HVHVV3lE/nqL6Jc/xszZm8D3wVvcdrenfx0k3amqXcTr6z3+doTf8UXXtUd3NWPc51/WArbXErnhkej/o6nj3EmrmTu07Pov2OTxy64ifNiqP/kfjXQyCLiDoUTKPylRc3vbSPXLXe6xpqjdpP7nMU195Ot+Jdj5b5V0Xfo3PZqRF+1Vca7em0zlOv5usc9B7p9zNQK7/KV3Ppc03VvJwD9zjpHKemn5oIvhvpI3ofuEfVrTh0aKg1wlWsF9Fz7dnNdZb2gFfs3qqt3DXwzKHu4tSr2pVGuerRncoprrNX+W7/pNvdSf/qLBNH77RLec113oSjWfHKaU7vKqq+cn85T606MPqJpVFOcTjmKOe5z1CeOWiomXsU6SO63nc5T02/671Gr1F7Ne80ip3Jda6faeJUt8q138+knOaqU5xc+crBicqDaVT+kfzv34RHpqQnDrzAgfrBzysOxIE48IgD3/734yfd/xPP+olneuT3wHvyreqOpP5YB37rL+HHGp6DxYFf5MC3//34Kfevc37iWT/xTFf8euYh8AoXM+MlDvzWX8KXmJclceDLHfj2vx8/4f51xk8956ee69lf6zwEPutg+uNAHIgDcSAOxIE48AMdyEPgD/zQcuQ4EAfiQByIA3EgDjzrQB4Cn3Uw/XEgDsSBOBAH4kAc+IEO5CHwB35oOXIciANxIA7EgTgQB551IA+BzzqY/jgQB+JAHIgDcSAO/EAH8hD4Az+0HDkOxIE4EAfiQByIA886kIfAZx1MfxyIA3EgDsSBOBAHfqADeQj8gR9ajhwH4kAciANxIA7EgWcdyEPgsw5e3M9/LJPo48GJzlcNR0RTtb8co4eoerRw1KqpXHnXrDifkzoOxIE4EAfiQBy4z4G/nwru25XJBw74A5PLeYAC97pwn6G15jpDc9VUvqqP9jGXuDMPbWIciANxIA7EgThwrwN5CLzX31PT9YGra+x4xTSnXzHNX8Wzp+LRftUmjwNxIA7EgTgQB+51IA+B9/p7eno9KPH2ZnCP6Ar3l2Kao1Os8u7daTtMZ8Fr7GYf9Wh/8jgQB+JAHIgDceA6B/5+arhudiY96YA/IHnt4zteMc3pVUxzeI0dr5jm2kd+xKNLjANxIA7EgTgQB+53IA+B93v88AZ/aKraMR3ufFef0au28m63Yr6v61e989RHc9AlxoE4EAfiQByIA487kIfAx727vJOHH2K3AE6j6hyvWl9neLT0+6zCJ4xe5xUnZz5xwuET40AciANxIA7Egecd+PMJ4fl5mfBhDvhD2IcdL8eJA3EgDsSBOBAH3uRAHgLfZPyr1uYh8FVOZ08ciANxIA7EgZ/lQB4Cf9bndfq0eQg8bVka4kAciANxIA58hQN5CPyKjzmXjANxIA7EgTgQB+LAnw7kIfBPP1LFgTgQB+JAHIgDceArHMhD4Fd8zLlkHIgDcSAOxIE4EAf+dCAPgX/6kSoOxIE4EAfiQByIA1/hQB4Cv+JjziXjQByIA3EgDsSBOPCnA+1DIP+x3ulflj7Kax/5n8f5///xYbiK/lLOeeecr1mu0flwjml9Z767H13FR15H/ROvOLnvByc6T1389KJ3pZl6g8eBOBAH4kAciAN7DrTfxHz5EnWUY2dq1+rcyp/lmTfN6XDFKtd650zsvCLu7H/2fEf9K945v3PHT1iHv9pvP3/qOBAH4kAciAPf5MCph8DpixvD3s2vzlFnm84H3mngmH1nPNr/7FmmfnCi3xGc6Dx1xztGTaS3YocpnzwOxIE4EAfiQBy4zoHDh0D9Yta8O8K7ec7UnaPDXI+GWLzm6O+K7CL6fsUfOUPXXxg4UWcf8ZOWs3cz4bR3wlyTOg7EgTgQB+JAHLjGgb8eAv1LW2vNu/U7fGn07XNWXGmPeDTdXMeoa2a9PCqG9s54tB/+0TNof+XUHrm34/QQp3Mc8cz3ft23M8P7U8eBOBAH4kAciAP7Drz8IXD3aEcPASuehwnd1WHwcMTCyYlo74y6i5yoZ3r0DMyqSK5zwSZe97oGjhk6F06j6sC7mZ0OfWIciANxIA7EgTjwuAMf+xDIlY4eAjp+F/Md2kdORHtn1F3kxNqr+SPnqP5uBtgR7zvpA/d6deZdbadjX2IciANxIA7EgTjwuAN/PATWF+70rhVHX8jP8t01HpnZ9XQYd4IjgmvsznY1drRf+Ud2T/3gRJ+9i3e6Dqv5Hb6L+flSx4E4EAfiQByIA+cd+OshsBuhX86al/ZM7dpul2NHPR3fYd1ZHfO+qh3z811Z+65uf6c5c4aj/iOeXdPZzvQzS6P2dztUmzwOxIE4EAfiQBx43IHTD4G1ii9n/cLWI0y84uRT3x28n7126KurHVP91bnvqtqx2gnecTtnOuqf+AnXnaqp3F/OU6OjJoInxoE4EAfiQByIA9c68Pe39LXzMy0OxIE4EAfiQByIA3HgAx3IQ+AHfig5UhyIA3EgDsSBOBAH7nYgD4F3O/yi+fw/n07xRcfImjgQB+JAHIgDceCHOJCHwB/yQeWYcSAOxIE4EAfiQBy40oE8BF7pZmbFgTgQB+JAHIgDceCHOJCHwB/yQeWYcSAOxIE4EAfiQBy40oE8BF7pZmbFgTgQB+JAHIgDceCHOJCHwB/yQeWYcSAOxIE4EAfiQBy40oE/HgLrX5Z2L/DuX56q3nnlyFca56r2l2uUh3NM60/Od8+PruIjr6P+iVec3PeDE52nLn560bvSTL3gz/QyIzEOxIE4EAfiwG924I9v4umLE5w4GeL8Ub07B53PK1yxyrV2njmfGnfO/+z9jvpXvHPuY8dPWIdf9XnV7Gm+nzl1HIgDcSAOxIFvdeCvh0D/8tQvVOfcNOePau+n9r7CC+twOKJrvGbHJ8bujnp+zR85/9QPTvTZ4ETnqTveMWoivRU7TPmdnBnEnZ5o4kAciANxIA58owN/PASWAf7lqbXmnVnOH9XdjO4ME0Y/ezwe9dH/KfHo/PCPnrfrLwycqPOP+ElbuPaqDm4Hc81u3d1ltze6OBAH4kAciAPf4MDph0C+2IlqEhhROXK4itOr4zqMfjiPxYOh/eTIWYl+fsUfuYf2V07tkb2O00OcznDEM9/7dd/ODO/XmlmKJY8DcSAOxIE4EAf+deCvJzH98tW8Wo6+WOG97991/2YrDXP+Va93oyfqWRXTeZ+Y61nJiXqnR8/OrIrkOhds4nWva+CYoXPhNKoOvJvZ6dCv4qN9q5nh4kAciANxIA78JgcOHwL1skdfrMprrjM873S7GLPQEwsnJ6L95KhnJSfqnR69Q83SecwBO+LRE+mb6sJdc1Y79TNnio/2TfOCx4E4EAfiQBz4bQ789RBYF+QLlMilvQYnKl+51mg8dppdrGbpHu0jJ/reT6z1rORE7vrMuXWWzgEnKrfa63qvz/RO2m6mn6+rH+3rZgWLA3EgDsSBOPAbHbjtIbDM2vki7jQdNs1TreboHfvkD9HPWnWH6R2cV67LXX+2Zmb1db0dRo9G18EpXrnWqulw+IpHvGqTx4E4EAfiQBz4RgfGh8DuS7Qwf6tp3oMWDbVGuIqKkyvfaZSvHn1NM1TzSfnu+bmX63fvctQ/8ROue1VTub+cp0ZHTQTXuMOhIWp/8jgQB+JAHIgDceA///n7WzquxIE4EAfiQByIA3EgDvx6B/IQ+Os/4lwwDsSBOBAH4kAciAN/O5CHwL89+ZEI/8+eU/yRl8qh40AciANxIA7EgdscyEPgbdZmcByIA3EgDsSBOBAHPteBPAR+7meTk8WBOBAH4kAciANx4DYH8hB4m7UZHAfiQByIA3EgDsSBz3UgD4Gf+9nkZHEgDsSBOBAH4kAcuM2BPATeZm0Gx4E4EAfiQByIA3Hgcx344yGw+5elevQjvrSlmV5H/fDar/M0R9NhcD8tcn+Nfgc4x3dqejXu9D2iYccjve/q4cwa9SyKV+4v510Dr32uKQ7dEfcIz252UHvsZqtmxTN7pdFZyeNAHIgDceA9DvzxTdb90VZMc47rmNfoNE6awp3TWvOa57Xu+Il5dx/HqnZs967e98yso513zfY7HJ3jDO+z/Q4dv5rf6TvMZ6Ahwne1YppXT9WO6axHuJ257EiMA3EgDsSBz3bg8CGwjs+XBVGv5JjXqiWfNIU7p/WUM/enR70fd+kwuLOxm9VhZ+e+Un/nebvZimled/bafXC+6g6b+lzrukfOoD3dfDCi7wQnKt9hyiePA3EgDsSBz3LgqYfA+qPvf/i97q47acCJ1dvlhSne7fiJmN/J70ntOr+r6lSrOT2KkVfkjU4jHBGOWiMcUbnKu1encUxrnwHn+FFdff5STPPSeb3qVb32aU4/WEVyOI9neZ256l1xehc9z1GPapPHgTgQB+LA+x3441tv+iMOXtHffgW0jms9acCJ1eN51Yrp3J+eczeN3Z1W93eOWe4ltepVy17l6YGjdg284z7fa+bR30WfOWl2dN7rPX4+aqL3e93NK43imjvX1Uc7jnjdp/lR3w7PvIq8vS91HIgDcSAOfI4Dpx8C9ejdH3q+CFTn+aRRnJxYMyrn7TN/Q6131fv63VwHf4TjHZE+YtevmOb0VNzFO51imut8zXc0qj+T12x9ey+70TjvNXpwrcmJnaYw59EdcR3vs7x+Zjb7fKbXuiN5HIgDcSAOvNeBpx4C6+j+R97r7nqTRnFyou4qTPFux0/EujvtYuqP350ZROepH+WnPser7t67+0vnM+m9Ih7NVl7zabdrtCYnMuOonnTgxG5Oh6H36NodvuvpMJ+VOg7EgTgQB97jwOFDYP0R5w85UY/qmNeqJZ80ipMTq3fKmfvTo96Pu+xi7g/9inezOp1imk/9u/ikY8cRX7odDfPOxqPZyleudbfLea3Jidyt6u6t87VHcfKO72YqRi+xmwFXseN3MZ2TPA7EgTgQB97nwPIhkC8Jjud/5J0vnWvo1ThpHPf5Ha9zf3re3c+xumOHgTunteadVzu8a7zWuc5V7ZjrV/x0R52xq/Ee+joczM92Ra0zNGenn2vSoD/i0flcxY+4Fa/7K9fad6SOA3EgDsSB9zrw10Mgf7iJejwwjcpXrpzmE6f9pdcX/WDOF95h6H9a5L4a9Q6Ka77SOKe15zXz6KV7j/Qd7/07Gj+Tz5h4x4/q7iza4zzncE2Hl+ao33nmgjPXo+qcoxeNxo7r+lV3ltd9yeNAHIgDceCzHDj+1v+s8+Y0P8gBfXj4QcfOUeNAHIgDcSAOfIUDeQj8io/5NZfUhz7+F6PXbM6WOBAH4kAciANx4KwDeQg861j0Swd4+NMHwmVDyDgQB+JAHIgDceAtDuQh8C22Z2kciANxIA7EgTgQB97rQB4C3+t/tseBOBAH4kAciANx4C0O5CHwLbZnaRyIA3EgDsSBOBAH3utAHgLf63+2x4E4EAfiQByIA3HgLQ7kIfAttmdpHIgDcSAOxIE4EAfe60AeAt/r/x/b9V/Wkv8hkP8Yt+O7NXOJu31X6NhZ8V0vPcN0DtX4Ob0Hbek0p8/1qjviOv6o/4jnjETOSQQngvtc+Ir+WnGuTR0H4kAciAPvc+Dvv+DvO8vXb56+UNUYvmAV2827+bu9V+redY7dveiI3L1qxcg1kmsPecUV33EdpvM8X+k7TrHKvfb5Wru+OO3vau1PHgfiQByIA+91IA+B7/X/j+3+BVpkh/3RdKK4ctaJtX9J33WO3b3oiFyAehXhvKdq5zoNGFF7NIfXOPETXr3Kac7cDpu4STvhzEmMA3EgDsSB9ziQh8D3+N5u9S/LqhWjVqwbpDrVat71gWl/hymvM8mVp18jOsXItXfSoaFnN07zvB+d71G8eo5q1XjOTt8BTmTH1I9u4p+dr/t1V7ev0x7t95mp40AciANx4HUO5CHwdV4fbuILU2PX1H3ZonOOWco7Blex6weH034w13R1YfXSnv9C/8OV0z0rnXJHOTN1j/Y4rjV5RfLqJfeoXJd3ej1L1wNWvfTTo7XyiqMlKqe57kGrUeeDa7/yiqNNjANxIA7Egfc7kIfA938G/zuBf1nqF+n/RPLQoVjl3g/f4dNseoj0En3PhHs/tfcf4Tpftc/mNbeb7ZjW5BXJ9T5nsE7rd1INezqMPriK5PSh0agacHrhiPDEDoziIQUAACAASURBVAerSF56zelPjANxIA7Egfc7kIfA938G/ztB92W5i9WQTrvCO65m+Nt1umfKq6deyv8XajG0vrvrZ84V0eevariK5Jxbo+au05rzX4XVnGdmcR6N07yzuM5MHgfiQByIA5/hQB4CP+Nz+OcU0xerH7HTleYs7j1dPxhx1aMazryL+Vz67456vsqn9+p8zCCqtsP8TqrR3iOda30O/Y57ja6LnbbD/Cw6a9KrJnkciANxIA683oE8BL7e83Gjf1lW7Vg1dxi4c177cuU193nK7eT0+74jXGd3vfTv6KZ+xXWO5p3mLF967zlb6zkq1/475us+3QXe7YTz83W1apPHgTgQB+LAex3IQ+B7/f9jO1+wGlWguOYrzYqrGf7yuWiIpV/l3q/zldP8SKM8+/UMzk+17iRX7TQTnKg9lYMT4at2DH3HgXlknvd2s12z6mXPpFGcfNoJX5G5O1rtSx4H4kAciAOvdeDvp4DX7s+2X+RAvvR/0YeZq8SBOBAH4sCvdyAPgb/+I37dBfMQ+DqvsykOxIE4EAfiwLMO5CHwWQfT/z8H8hD4PyuSxIE4EAfiQBz4eAfyEPjxH1EOGAfiQByIA3EgDsSB6x3IQ+D1nmZiHIgDcSAOxIE4EAc+3oE8BH78R5QDxoE4EAfiQByIA3HgegfyEHi9p5kYB+JAHIgDcSAOxIGPdyAPgR//EeWAcSAOxIE4EAfiQBy43oE8BF7v6dMT61/Z8n562IUDjv71L2c+0l14pL9G6Rmmc6jGByhH7pqjmr6K3etdvO7VXM+oeOX+WvHOdf0+b6qf6Z1mBo8DcSAOxIE/Hfj7r/yffKo3OMAXIPENR2hX7p5nV9cueQKsvb772frscZ7dd3e/3qd26T7N0Smm+S6P7kysPd2uMzOijQNxIA7EgWMH8hB47NHLFXwBEl9+gGHh7nl2dcOah+GjvUd8Ld7RTAecesGJ3g9OvIuf5oJ3+xXTvOvpeHS7kRnE3b7o4kAciANx4LwDeQg879ntHXwBViRnKRgRnAhOBCeCEyccviIv8o5DUxGdYmdw5nv/qp52as9VGp2peTdf7/JuXs9auZ9Hzwqvmh3edzxa695HZ6QvDsSBOBAH1g78+w2/1oV9kQP+5ae15t1xdnjVVE69ytmlmsK8Vh25xk6/i+mcLq85R6+rNNMena/3AidW/zt4PbfuV9zP5tyKZyax693FakZecSAOxIE4cK8D+Ut7r7+np/uXn9aad4Mf4ekh1tydnP2qXWET1/WjPRN9TtW8meMacI30aFR+lTOfXrSKF/YunvNwBq3JOeukOeJ1jmrBd+Mzvbs7oosDcSAOfLsDeQj8sJ8A//Lr6sIc5xpwHa+c5tWr+p1c95ETtR+M6JzX6M7GaY7imk/zdzSr3q4frCK5zgC7m2fn0R50FTmb52iUByOuODRTfKZ3mhk8DsSBOBAH/nQgD4F/+vHWqr74pnd3sKMvSue91pnK7eT0qnaFwVWkh6jco3nN6uYppvm0Z0dztpeZRO8HJ97FM/doD7qKqtUcTYftcGimuJo79QSPA3EgDsSBcw7kIfCcX7eqpy++sziH9L6qHeu0qpny6lOOOSscDX1EcGLhE4emi12fzzlbd3tW2Nn579D7Tu5TuHNaH/HMqdhp4VecasgT40AciANx4B4H8hB4j68PTdUvXB0AzpcnUTWVgxOd7zSlBUcP1uHMVo32KU8Or7Hrh1/1oZkivRpde8S5/my9ml+z3snX7tVLz9ZpV7xyOzs6jc7QvNMGiwNxIA7EgeccWH8jPDc73XGgdYAv95YMGAfiQByIA3EgDrzEgTwEvsTmLFEH6iEwrzgQB+JAHIgDceC9DuTb+L3+f812/te/n/oAqOfv8q/5IHPROBAH4kAc+DUO5CHw13yUuUgciANxIA7EgTgQB/YdyEPgvldRxoE4EAfiQByIA3Hg1ziQh8Bf81HmInEgDsSBOBAH4kAc2HcgD4H7XkUZB+JAHIgDcSAOxIFf40AeAn/NR5mLxIE4EAfiQByIA3Fg34E/HgK7f7mp2OpfRapO14MTlSNXrnJ9o6mouPagcX5HQ+8nRD//M2fSWTpH8c4f1SaPA3EgDsSBOBAHfq8Dy4dAf0jw2m1xXmvNp75Oo5jmNeOodo3r/Rzvrv18Xu+eb+qb8N250cWBOBAH4kAciAO/x4HxIbB7YOgwtcJ5rTXXnsqL4+0cvEY0PtNr1008undHP5/Xu+eb+iZ8d250cSAOxIE4EAfiwO9xoH0IrIeF7oGhw9wKNET4VV2c8/RVhCPCHdWucz38p0Q/n9d1zsL03Z2966O30yums1dzVKf9yeNAHIgDcSAOxIGf4cBfD4F8uXfHh9PouuLqRYSnBw4enBq9RjjVgnU6xSpXrc5w3btrPZuemXPBH9XoVjNW3DS/cO/zmt7EOBAH4kAciANx4LMdOP0QuHOd6cEAvKLmNZO6mw+nkVz1HTbNLu2k15mvzDnPdDZ4PdMupj2VdzuOZh3xviN1HIgDcSAOxIE48LkO/PUQWEftHhDAd67SPSxov85HS+zmwxF1luqV38GnOdr7ylzPrzlnKKx7wxO7XjiPqu1mO7/qdy51HIgDcSAOxIE48LkOtA+BdVz98uf4HQancdKBa9RcZ5AX32kUVy05sdPBVWS2Yu/K9SzduZVfnXFX5/c/6uv4DludLVwciANxIA7EgTjwGQ6MD4F1PP+C93q6wqQrXLmu9pmuV1656byu0f6ux/lX1n7WrnasO9+Ohj7VVq41GqLzXk868MQ4EAfiQByIA3HgcxxYPgTWMfWhgC99jd1VtEd5+sC8Zh+4z+lqxbSPnF3dbO1V3btyP8+jd/A53Id5GuGIypHDVQSrSK08GLxzqeNAHIgDcSAOxIHPcOCPh8DPOFJO8ZMcyMPeT/q0ctY4EAfiQByIA/86kIfAf71I9oADeQh8wLS0xIE4EAfiQBz4AAfyEPgBH8JPPkIeAn/yp5ezx4E4EAfiwDc7kIfAb/70c/c4EAfiQByIA3Hgax3IQ+DXfvS5eByIA3EgDsSBOPDNDuQh8Js//dw9DsSBOBAH4kAc+FoH8hD4tR99Lh4H4kAciANxIA58swN5CPzmTz93jwNxIA7EgTgQB77Wgb8eAutfe+rbnVHO/2Wo19Wr2KpX96huwtEo/5Nz9UnvMeGquTJ/1tdn+6+8yx2z7rzfzmd95/47/PKZfke/j9fen/q9DvD5EKfTFD+96J00R3zNnXqnncHjQBzoHfjjN/XoF6vjFdOcdYppXrzXHaYazZn/m6Lfz+tX3LV2PrP32f5X3HG14+jud97vaHed+879K1+u4Pzs3Jf4CffTs1xx56tnvPN8tVv3a673dJ1zz9TVu5qvs5PHgThw7MD2Q+DqF48/BkRdq5jmpTmqdU6nd/6n12f9+On3/cTz+2fwyjO+c/cr7sn9pviKMxzt4GxHunfx7zxft9sxaqL61GHFgxO1R3nNJ633po4DcWDtwKmHwGkUv5BE1SmmeWmOap3T6Z3/DTWeEP1OhetbeXomvrTKoWeGcmAa0U+6CdcZle/qur6jfp1dub8m3nGtmdFhcETVVH7mhV5naP+Eq6Zy1TET3LUdrv2qZ9bEq7bLtV/3Ks7srh8MDVFx5jqHxvnSdRj98Gj+Ecv/UR5N1ysty1R7dbbjWjMQfceheSYyX2d0WPEdPmHgRJ8/4apLHgfiwGMO/PUtVb9wvHVk94sID0cEr6hY5fpWnWudg9d+nd3pfyLGnYh6B+4OdlSXTudozowuTjrf5/OZNfV3+pWWeUTf/0jNrCnunGfSTPi0y/Gj+6Bf7XFOZ2qus7RH89JoPfUz6ygyy+eA0+81eMUjzvmdWjWa695pt+o1996deqd/pSnOea93zjFpfFa3j17XFq6Y9oIT0VITmQ2vdfI4EAcec+Cvh0DG1C+e/vJpjoYIRwSvqBh5RfJJqzh51wP3m+J0zw5XTHP8UExz+C5Oug7fxbo9hXX9Z7TarzkzFNMc3uMzmp1e36d117+L1ZxO67hrvNbzHPU6771es6sieTdDOZ0x4Wg6XjHN6fH9k8Z19Ktec/gzcad/pem4DjtzJtfWPN7FTfM7HEz7dcYRr2dBq1jyOBAHzjswPgQyil82IrhGOGLHFaa85ug7DM77Ff9t+eRD4d2b+0998BW1X3HNpzkdvosxX/eTwx3Fo13M86hzlVOcvNsBR1xpjuYzo4vd3F2s5nVax13T1YXpm7O6Fnw30s9s+sCn+gh/ltf9mjOX2HGOVc2bvjORXp/LjAkvfsXRf3WcdnZ4YRPO+Ve8nr3TKZ88DsSBPQeefgjUX+zuF1Mxz7Wu43rtVzjiXf9T6+meE849j3h0xEl/Bu+0HVY7O7zDOJ/HTquY5t7b1Z2+w7x3R1M9uzrmd/pdbLXPZ1ATd/e7nr7dOPU77jXzJ/xZXudqzlxix3XYSg+3E7vZHcasFYfm6jjt7PAOq/OAE/2MHd5h3pc6DsSBYwe2HwJrVPeLp1jlR7UeSbXgjmmtOfrfGKd7Fj5x5cOK63ya9GfwTtth3flKN2l3zuu9z87jjD7Xz3LEo9/VTfqpf4U75zV31Ljar/2a03MmTv2Oe82OwieuNCsO3jVd7ZjuJ5/mOa/12bw7R2Edznl2dqxm7PSjmc6xOov3nK13dqNJjANx4NiBPx4C+eOg0Uco57/ApV3xrke72qEceo3K/5bcfdJ76d3J4Vd9pUFPpI8I7lF5cmJpeXkfNXxFMPqIqply7Z36XKM6587u8X5q5lATwXcjfUTvA/eouhWHDg21RriK9SJ6rj27uc7SHvCK3Vu1lbsGnjnUXZx6VbvSKFc9ulM5xXX2Kt/tn3S7O+lfnWXi6J12Ka+5zptwNCteOc3pTYwDceCcA/9+g5/rizoOvNyB+qOfVxyIA3EgDsSBOHCNA/lWvcbHTHmBA3kIfIHJWREH4kAciANf40AeAr/mo/75F81D4M//DHODOBAH4kAc+BwH8hD4OZ9FThIH4kAciANxIA7EgZc5kIfAl1mdRXEgDsSBOBAH4kAc+BwH8hD4OZ9FThIH4kAciANxIA7EgZc5kIfAl1mdRXEgDsSBOBAH4kAc+BwH8hD4OZ9FThIH4kAciANxIA7EgZc5cPlDoP4HPP1fc644bnxWQ19Feh2jhtcIR1Sucn85ht51Z2ufS/+Ew18dn73Ps/1X3+fqeXfeb+ezvnP/1V518/yOfh+vuxnB3ucAnw9xOknx04veSXPE19ypd9oZPA7Egd6B+Te11y/R7hdTMc1rUFd3mC9FQ4SvusPgia45wpXXXnIiukejz/H60bln+mrnM3uf7T9z1ju0R3e/835Hu+u+d+6/w0+d6WfnvsRPuJ+eRc/+Kfk7z1e7db/m6o/rnHumrt7VfJ2dPA7EgWMHLnsIXP1i8seCyLGOanQe6SPCV91h8ETXHOHO0+8R3aORefR7DZ54nwPv9Pydu+9z9N/J3G+K/yrfl3G2951gvfmd5+t2O0ZN1Nt0WPHgRO1RXvNJ672p40AcWDtw6UPgtIpfWCI6rTWHnyLaiuSlJScqprOU38HR0DdFdM9En+2zite38tqLRvnKwYnKg1XsXuCTbsJ91q6u6yts1a9c5f6aeMe1ZkaHwRFVU/mZF3qdof0TrprKVcdMcNd2uParnlkTr9ou137dqzizu34wNETFmescGudL12H0w6P5Ryz/R3k0Xa+0LFPt1dmOa81A9B2H5pnIfJ3RYcV3+ISBE33+hKsueRyIA485cO5barGj+0VFDldR3/AV0SjW5a7Tmpw4zVXedxTHu+OYqTM0956zNbOI2l+Y4kc1Z2WG9oJ1cdL5Pp/PrKm/06+0zCP6/kdqZk1x5zyTZsKnXY4f3Qf9ao9zOlNznaU9mpdG66mfWUeRWT4HnH6vwSsecc7v1KrRXPdOu1Wvuffu1Dv9K01xznu9c45J47O6ffS6tnDFtBeciJaayGx4rZPHgTjwmAMvfwisY9Yvtf9id/WujqvrDHIiGvZr3eWr3c51O7qZu9g0r8MV05xdimkO38VJ1+G7WLensK7/jFb7NWeGYprDe3xGs9Pr+7Tu+nexmtNpHXeN13qeo17nvddrdlUk72YopzMmHE3HK6Y5Pb5/0riOftVrDn8m7vSvNB3XYWfO5Nqax7u4aX6Hg2m/zjji9SxoFUseB+LAeQfe8hBYx/RfYq+5iuOrWjlyIvO63cp5rv3kFcnPzvP5Xa2zlWevRzRTH3xF7VVc82lOh+9izNf95HBH8WgX8zzqXOUUJ+92wBFXmqP5zOhiN3cXq3md1nHXdHVh+uasrgXfjfQzmz7wqT7Cn+V1v+bMJXacY1Xzpu9MpNfnMmPCi19x9F8dp50dXtiEc/4Vr2fvdMonjwNxYM+B2x8C9Rdff3EVr6N6zfG7HrQameF92u8c9Spqv+baM+GqOZNP8yac2Uc8OuKkP4N32g6rnR3eYZzPY6dVTHPv7epO32Heu6Opnl0d8zv9Lrba5zOoibv7XU/fbpz6Hfea+RP+LK9zNWcuseM6bKWH24nd7A5j1opDc3WcdnZ4h9V5wIl+xg7vMO9LHQfiwLEDlz0E1qruF1MxzTt98SuNc1wPnKi4Y91e9F3Ufs1VO+GqOZNP8wqfuJq/4rr9k/4M3mk7rDtf6Sbtznm999l5nNHn+lmOePS7ukk/9a9w57zmjhpX+7Vfc3rOxKnfca/ZUfjElWbFwbumqx3T/eTTPOe1Ppt35yiswznPzo7VjJ1+NNM5VmfxnrP1zm40iXEgDhw7cOlDYK3jDwxRj9D9wk8Y/cpr3s11nhloqTXCVVSc3HmtyUt75Ws1j3NpZPeqrzTa02mdp17N1znoPdLvZ6BWfpWv5tLnmqp5OQfucdI5Tk0/NRF8N9JH9D5wj6pbcejQUGuEq1gvoufas5vrLO0Br9i9VVu5a+CZQ93FqVe1K41y1aM7lVNcZ6/y3f5Jt7uT/tVZJo7eaZfymuu8CUez4pXTnN7EOBAHzjnw7zfkub6o48DLHag/+nnFgTgQB+JAHIgD1ziQb9VrfMyUFziQh8AXmJwVcSAOxIE48DUO5CHwaz7qn3/RPAT+/M8wN4gDcSAOxIHPcSAPgZ/zWeQkcSAOxIE4EAfiQBx4mQN5CHyZ1VkUB+JAHIgDcSAOxIHPcSAPgZ/zWeQkcSAOxIE4EAfiQBx4mQN5CHyZ1VkUB+JAHIgDcSAOxIHPcSAPgZ/zWeQkcSAOxIE4EAfiQBx4mQN5CHyZ1ceL9D9+Sn7c9VmKd56b3Ro/y531aercP/2F9++6B/uJ7zrHb9ybn8/nPlV+JjU+N/G13b/h81859lM/l9Wddrif/62zc8sfoul+yTrsmetcPc/Pwi+S46+o/W53nMV3XHmvO2dzzrt33OE5Zz+K3e6773t0plfzd973ztn4dPeOmn/3Du7i8Yq9V8zwc+3Wr9j9ih1H9/2EMxyd8Uo+D4FXuvnkrO6Hr8OeWXP1vGfOcnVvd7cOe2bv1fP0LHfOZs8rdrDr1fE3323Xyzs9uHM293vFDna9Ol5xtytmPHrvV+x+xY6j+3/CGY7OeCWfh8Ar3Xxylv/wVe1YrQDvOOfRaI/neuwjzudPvYp7rjucqxq+41ZY9flLMXLmU3c9rtHac+1fcegmTeH1Up4ejWf4bqb2wzP/iFudT3uZ18UjHXzXO2HVs/tivvYo5nnNdUxr3ztxhfss7T3i0er8rqfjV71wxKm/20WPRu1XnFz5bmbHr3qVq1z74Va4anZ0zPe+o5q7Tjp45lP7mZRXTafTXWi1X3ly5elhNhENPRrhtHfi0WiP50e98N0sOI06X3Fy5lBr1N5Od8TXLDQ69535/l/Od57yS3bzw6HRr16cvo5q1VbueuXZC3ZUT/OOdjB/1b+aof2ae88j59+ZoTs1f3QfOx/t9zNo7Tm7HK/6aL/z9Pisox2q77TdHu3p8m7Ojo4+3el5zXFM+3QPOJjWOsNndjUYs6bad6he86P9R/Mf7fczaO253qXjlH/kPDVTZ3Q7FOu0vlf1q5w+jaoHd2xVO1czePk8r0un+qlG4/1eT/2cp+OVO+J931HdzasefXnd9aA/2tfNolejz1HuHfmfjrzjBNn5Pwf8h2jnh0V7NP/fUEtWmo5TTHPG7mLoPXb9rtmta5a+va/bpZjm2qu45qqpvOMU01x7wYkdp5jm2qO5ajRfaTpOMc2ZuYuh99j1u2an9jlV817100csbZd3mGu7PVMf2md5P4POYwex4xTTnB6d3/Ed1vXqHOU9X83rOMU0Z+4uht5j1++a3fpoVsc75rXu7jjFNKdPMc3hK4ITO04xzbVHc9VovtJ0nGKaM7PD4Cp2fIftaKc+3feJeR4CP+hT6X6IHKva31zBteAaVxqfS03/qhdNxZWOmRq195l8tffoXCte52ruZ9U7aY5u6gUnou/OpHPJ0Xf9cMSVhnked3rRVDy7Q3sfzaedjvvd4Il+fnDixHPu0vlbOXLiNBdeo2onfNKU3s9FzaypF5yInpleM5cI3/XDEVca5nnc6UVT8ewO7X0mX+2dzuU9Xut5iuveaFa9037Fu37Hntmvuzizxm627tdc+zTvZihf+TSn63Wtanzup9Z5CPygT8Z/oPwH8hHer9fNQLPi/Cz0dHGa0+Ed1s3cwY5mPcprn+Z+phVX2okHJ+pcxTRHo5jm8B5XmhVXc454dk26Du8w5pyJNaebpZjmzAYjFt7lHTZpmX0FP82acD2navwszq14ZhK1VzHN0SimObzHlWbF1Zwjnl2TrsM7jDln49GsjnfMaz3Diivdozx9xGnnI7zOOjpjN1/7H+G7ng47Opueg3yaA/8pMQ+Bn/JJNL+k9UOkP0ia17E73jV+Pe9RfsWxT/VTPp3B8WnfhE/7wH0+OHGHd01XO6bzJ640xTmvtebomT3V2lO51tpLvtKsuG4/Mz1OZ3B82jfhvsfrrk93al69qleuyzuMGZxDNXCKaQ5Pb1crB78zwzXMKXziduZ7706tmsq15lwaV5oVx/l11pRPZ3B82jfh0z5wnw9O7HjHVrtXXO3wWewldv3ao3k3r+MVq1xr9mpcaVZcdx6d2/HTvOmMk973UF81h3l3xTwE3uXsA3P5IdPoY5zrftBUc5b3Xu3X3M9VddfrPaqhx2ehcfyo9l2uP+JLz26izzjS0KdRZyheub5WHDrVFOYzwFxH/xGvfeT0drvgurne7xpqnQF2tMt7qNmpEU5nM98jGnp2edWzu5sFx1z6XKu45trfzWCO6lb9PkP7znDs0H7OAkdUje+gRzX0dZz2a6495DpTc3ifT6082NEu79G+1W7v6/Zov/POKa+576H2fvCKKw6dauiBI6qmcn+teOe0X3OfSa39hWmPcprTi165Vb/2aU6/Yu/M//4E3nma7I4D/3VAf7liyvc58Bs//994p+/7yfz/N85n+a2f/DX3/qSfnzwEXvOZZsqFDtQvyCf9klx4tYzacOC3fv75md748H+A5Lf+fP4A63/FET/t5ycPgb/ix+r3XOLTfkF+j7M/4ya/+fOvu+X1sx34zT+fP/uT+Rmn/8Sfn/xV+hk/OzllHIgDcSAOxIE4EAcudSAPgZfamWFxIA7EgTgQB+JAHPgZDuQh8Gd8TjllHIgDcSAOxIE4EAcudSAPgZfamWFxIA7EgTgQB+JAHPgZDuQh8Gd8TjllHIgDcSAOxIE4EAcudSAPgZfa+fiw6V8OTvjjm9adz/7rpWf716f7XHb6nCb8rps86/+z/XfdK3PjQByIA3HgegfyEHi9pw9P9AcGrx8efKLx2YeAZ/tPHPXjpP55ef2KAz/r/7P9r7hjdsSBOBAH4sA1DuQh8BofL5niDw1eX7IkQ25zwD8vr29bnMFxIA7EgTgQBx5wIA+BD5h2ZwsPDkTfVbi+ladn4kurHHpmKAemEf2km3CdUfmRTvnK9bXi0KkGjKicz3YN9ZnITKL3rvbTo5qz/fR6X9VH8+lF181gDtpJEzwOxIE4EAc+34E/v2E//7y//oR8ARP1wv7Fe1RXr87RXOd6Pul8n89nztTf6V3rNTPpVb5yrdF4D7XrvZ504DuR8xC1x/cd1dWrczTXuZ5POt/n85kz9Xf6lZZ5iXEgDsSBOPCZDuQh8AM/l+mLtcMV05xrKaY5fBcnXYfvYt2ewrRf807f8R2mvcprjqbD4B6N08wOV0xzdiumOXwXJ12H72LdnsK6/kkbPA7EgTgQBz7LgTwEftbn8c9ppi/Wwrs3V5j64Ctqv+KaT3M6fBdjvu4nV468i+g9qta5qnl1nPLono3TzKP9U5+eR2corvk0p8N3MebrfnK4xDgQB+JAHPhZDvz7Dfmzzv2rT9t9MdeFJxwzjnh0xEl/Bu+0HTadX7Wac0aNj/Dao7nOvTqf9kw4+494dMRJfwbvtB1WOzu8wzhfYhyIA3EgDny2A3kI/MDPZ/piLXzi6horrrvmpD+Dd9oO685XOtV67Wfe4bXH9V6rVvNdnfZoXv3d62ju1NfNKmzSn8E7bYd1+0rXaSd8ukfwOBAH4kAceI8D/bfVe86Srf91oPtixRy+YDUqR95F7el2OE/NrKlHeXo0wld0fJrpOmYoTg63M58ejdqvMxzfrWv29NK95GhXfaVBT6SPCO5ReXJiaXl5HzV8RTD6iJ1GseRxIA7EgTjweQ78+w3weWfLieLA2xzoHm7edpgsjgNxIA7EgThwgwN5CLzB1Iz82Q7wv3b97Fvk9HEgDsSBOBAH1g7kIXDtT9gvcyAPgF/2gee6cSAOxIEvdiAPgV/84efqcSAOxIE4EAfiwPc6kIfA7/3sc/M4EAfiQByIA3Hgix3IQ+AXf/i5ehyIA3EgDsSBOPC9DuQh8Hs/+9w8/TyYrgAAF05JREFUDsSBOBAH4kAc+GIH8hD4xR9+XT3/KZTv+gHI5/1dn3duGwfiQBxYOfDHQyD/MpLojeAaz2i0j1z7wYjKVQ6uUTXgjmntc5SjX6Pyr8zrDHe/fIfXd++/c/6n3qXOpe/Og2d47a28e014pw0WB+JAHIgDv9eBP74l/MvhqO5s8R7VrLjSOX9U62z6j3pWvHM+/1V1nePus3TzO+xVd75yzzP+3emBn6vb1WHqjfNaV+619mquOsWTx4E4EAfiwPc48FUPgUdffEf8K34sOAPxrp3d/A67a/9dc7kD8eyeR/t29nSzFdO8mzfx4ETt7bDiJ1x7k8eBOBAH4sDvduDXPQTWx6VfcFPefayq7fhXYneeZZoNXpF3d2c49KpR7g5ed63ybveR3s9O7X3gROenuvT+Ukxz11Xd8YWBE7W3w+BXHJrEOBAH4kAc+L0O/PGtVF8K+vZrK0feaRyjpkcjXEXFK/eX866hJjKTOYqDaSze38q/Mj866zNnmWZzd2Z7Xbj3aq05MzT6vK5W/TP50Vmm2Ud9zvsdprmFH/XCM5OamVqj0bnKgzvGLHitk8eBOBAH4sB3OfDHkxZfGPoFo3bAK+Y5vRrRHPXD00sfEZ7ao/LkxNJq7r07fNdzF3Z01mf2TrM7vMN0t/Kaq4a84xXTnJ5H46OzVn0TN+Hd2UvLu3jtVZxe5+npcOXgiczTuOJUlzwOxIE4EAd+pwPtQ2BdtfuC6DC3ZaVZcb6z03aY7leenOjztY9ctWDvineeZZrd4Y5V7W/1SDnFK1dOc9VNuGp28przyGvVN3ETvrNfezWnV7HKte40YMROv8OhSYwDcSAOxIHf68Af35T6hdF94Sg/WbLSrLiap3zlWjvf7Vc9OfFsfzf/lZie++q90+wOV0xzztRhE7fS0qPxrP6K3tXOiZtwPc+Ua6/m6BXTHL7ihD/D6fzkcSAOxIE48DsdGB8C67r+5eJ1Z8lKs+J29j3S7z2r2rnufq/C7jzLNNvxndo16o9zVTumes/PaK/qXZ2x464+o87b2af6sx6sen1W6jgQB+JAHPh9Dhw+BOoXBV9KGt0S1Xec9rq2qxXzXuVqV1c7ho5ZekYwjcq/Itfdml+9u2b7S/d1fOlVQ80c5Xb60Z/pRztFZnqc9B2+6l1x3SzHtN+5qpWvvHupxvkVp9pptmqSx4E4EAfiwO92oP+W+d13zu2aB+aY8l0O5CHwuz7v3DYOxIE40DmQh8DOlS/B8iDwJR+0XTOfuxmSMg7EgTjwpQ7kIfBLP3iunQcCnPiOmM/7Oz7n3DIOxIE4sONAHgJ3XIomDsSBOBAH4kAciAO/zIE8BP6yDzTXiQNxIA7EgTgQB+LAjgN5CNxxKZo4EAfiQByIA3EgDvwyB/IQ+Ms+0FwnDsSBOBAH4kAciAM7DuQhcMelaOJAHIgDcSAOxIE48MscuOUhsP4FIm/1q/uXiWDou8gMtNQVHfN+1yqvHLnyPhvNXdF3a107X30ev6eeh9w1d9Xs6+JdO8/O7c4GdnbWo3r2ER+dc3Vfnad7TXinfRbrdnXYs3u6/mnPhHcznsVql7/PzqT/bF/p6dX4yJxHe3Qv+aOzru6bzjPhV++veezqIvyze5n9yBx6NT4y55me2r16cbaVZsX5/GfnrXbBrW+E6mTkIkTavS58F9vRHs1y/qjm3O+MekbN33Gmbn+H3X22d+x89E6vPOsrdz3ih5/P60dmnunp9nXYmZlntL7L6zOzHtH6vqodO5r7SA8zu10dhv7q6LueucvVZ6t5fr4Ju2P3NFPPpPmkP8Kv9PyK8xyd1/mjnc/cz3vZRfSzXFX/modAN1ANwkQi3FGN7l3R7+TnffW5pv0Tftf5Xr3v0XvUOV951lfuesQTP5/Xj8w809Pt67AzM89ofZfXZ2Y9ou32ddgjs3d6pl0TvjPzjKbb02FnZl6t1fNofvWenXm1X8+g+U7/3Zp3nOfOncye4l1+3v4QyIXqAppzoV3sqL+b4ztcc1TT/644na9w3t3Z4IhoqvZXh7mGetJOOH1Xx6N9xeub/fR1nGqUJ4c/E9l3pucZ7e4+7kRkJ/3g1PAVldNcNaucmcSV9mqu2zlhhTtHDUd95pz0EM/0PqvtdjpWtb5154SrZpX7LrQTDn9V7PZMWOHOUcNR6/mU01w1q1x7uvmr3qs530+tZ+x2TvyEdzN2MM6zo71Ks9r57P2YPcWr7uBz/n4qcMXJmgvQprXmHb/Ciqv+7g1Hv0f2eq/rmIOu41+FcWbd153LdV0NRvSZWq/yrr/0E76a9Qy32udc1WCasx9uuofy9OzER/t2Zq803HHa7zj6mqk5O1SveceDrSIziCvt1Vzt7N66x8+lNb0rvXJdzjxip7kL851+n6Oac/kc8KM49U340byzvO+pusN0rvJHetUyo8PgptjtmbR34d25u3O57qiu87rm0TtcNefM/p2dO5puJ30Vya/0q9v5z/yJeBTXw/+zQP7XJ+ecZ2enO9JOPdqHpiI5Oz3uaLznyro73xHW8XUmcKKes8OU13zSTrj2XplP+47wjldMc87bYXCr+GjfauYZrvb7GbxmHjgRvKJimqPpMLgpPtIzzTqDd3s7TGcqrzmaDoOb4iM906wzeO3Vt/d259rFfFZXd7NKN+HdjGew2qPvnVl6Ns3pVUzzjgc7ijWnm3XUdyXf7d/F9ByP9Gj/Ku9mr/RXcDs7dzTdWeirSF46zbu+Z7Ef978E+oUxiOh81XBExTq9Ytqj+J35tLPDFdNczwdO7DjFprzrL+2ET3Oexad9R3jHK6Y5Z+wwuCk+0jPNehbXs2iuc8GJHVfYEa99q7ybs9JfxXV7Hava3+x3beEdhn6Kj/RMs87gR3uL796+42iO66mnvgmn76q4s6c0/mZ/16+Y5qseuC5OuzvtXVh3j9rV4Y75+Z2f5jxyl272I3PO9Ozs3NF0O+mrSF46zbu+Z7FLHwI5fBeny3QX7LCj/lUPHJFZWk9G7mim3kfw2jft7HDFNNfd4MSOU2zKp/4On2ZcgU/7jvCOV6xyf589L/1n++7S+/26PWiIqlGscn+rdjfXmbs9V+i6vYppzj7FNO94sKPYzTnquYI/2nvEc4ZdHXpi11dYh9NzZTza0/GKac65FKvc3+h2o8/b7btKx/m7eXo2eMU07/gVBncmdvvO9D+i3dm5o+l2T30T3s14BLv8IbA7BJeoSF46r+lVDRh6rR3r+hTT3Ht9LrX3gN8VV/s6TrHKta4zaq05nGNHuN67es/00zv1wR/Fbmf1dHNVqzk7FNMcvoulm7QTrnNW/aU74nXWUa7n6eY67/OOeNdX3e1Rnc5UnHyn/2gGszR2PYppXn1VK6Y5cyesw1c9cN1e5XZ411OvznRm7qNzvK9qx3bOMfVxzyl2u1TrvO9xnrMyo+PhNPpcOMe9nnTgxKkPfhWrd3p1nGKa14yqHQOfdsDv9HWaVT87p3PBr+K0U3uONNP+qc/xqZ8zHPHoiPMnjuJE9MPSqjgHJKLRqPoj3LXMJa76XUOtUfvvznWv5uwtzF+OaZ9z1as8dTfzqJc53svMrh/tqhfNKu7M7nZ0fYrR49HPAj/h8MRJ5zj11Ac/Rfo0ula5yvXldXGKeS+1zqBH+zreMa2nuWiOeHQeuzM5xmxwYs3SnNkT1uGrHjj2HPWveJ2l+U5PafzNDMep4YlHOHzF7gXfcYUd8au+iQNnNmcjshcd0XntJ0dL7PAOK32Hdxizpx7lp5y5HtEX7i/HtLe0yiuneTdT+5Rf9aFDQ+3xiHe91vR6LI1j1Nqvug53DL3i01w0Rzw64t+fKkziVztQP0h3vu6e/8jZpzN1eIc9snPquXv+tHeFT2fq8A5bzT7L3T3/7Hlcf/f57p7v9zlb332+u+efvW/ppzN1eIc9snPquXv+tPcq/O7z3z3/yIe795+Zf+83/ZET4T/SgfoBOvNDdPYSd88/ex70050dv/v8d8/nvmej+0C/43ef/+753OvRePf57p7/6L3pu/t8d8/nHmdjnat7OX73+e+e393xSuzu8989/8iLu/efnd//1B7dIvyvdeDsD9BZI+6ef/Y8rud8GlUDrtiV+d3znz0r59OoM8EVuzK/e/6zZ737fHfPz/2fc4DPR6NOBFfsyvzu+VeetZt19/nvnt/dSbG79z8yPw+B+gkljwNxIA7EgTgQB+LAlziQh8Av+aBzzTgQB+JAHIgDcSAOqAN5CFQ3kseBOBAH4kAciANx4EscyEPgl3zQuWYciANxIA7EgTgQB9SBPASqG8njQByIA3EgDsSBOPAlDuQh8AM/aP6FT8WzL3rP9pWeXo2PzHmm52g3/CM76O3iI/PO9uzsRXN29q6+5r/rxd00nj0LvWf7Sk8v8ZEZz/SwV+Mz87xX53ru2qvr2je9VtzUcwfunrzjXH6GO+45zfTdWk89V+Err1fcVft35qgf5Dt9P10z/+b+9Jv90PP7L4TXR9d65oe329VhR2d4lPddXtfcwjr8qp2Pznmkr7vH1ffzc3U7XXNX3e3usNX+Z/zxXV6v9l7Bdfs67IpdzLh7/s6eV52Bs+zGV5+r9vlOr3fPfpXuVftXe1bcVfd8ZM6nnuuRu6x68hC4cufF3Lt/6Lr9HXaHLdOeCb/iDDX7zvlHZ3zH7nfsxIdp94TTd1X0PV5ftWeaM+2b8GnOLl5z75rtZ1jtWXE+55X1q8/16n1HXtZ5XnWm1Z4Vd3SHO/lPPdfVd85D4NWOPjHv6Ieu4xWrnPcjx9BZ1f/MrLP7fTf9inMexdA9Eq+a88ju6vH9R/dDf6RbnYcZK81d3LTb8el+E7573m7Pbu8VOt/PTHAieMUOU36VP9O7mttxvktrzau3at46Cx0ctWuUJ1fNbt7N3+09q9vZ1WkmrHDeZ8+CvpsNd3X0XVprXnur5q3nQAdH7RrlyVWzm3fzd3t/ki4PgR/0aR390HX8LrZzzZrl752+KzTdPWpuh3fY2TNcMePsTtdPZ1jhznntO7w+q/f+Z+ppt+Ka1y6vJ2znXDVL3zs9V2q6u+h9Or7Dds70aN/O7E5T+9g55XpXZtADpzVYp11hcEfRdx3pn+F3dnUax7rasZ1zPtKzM3fS1D52Tnn1omGO1to38eBE7Qfbjc/07u74BF0eAj/hU/jvGY5+6Dp+F9u5ps+q2rGdOY9opj0d3mFnd14x4+xO109nOINPWt9FfVZP3xVx2j3htbPjOmznfPRVJN/pu0oz7QQn6r4OU37KH+2b5u3g7Kyo+aoXXWk0p0cxzTsebDd283Z7z+p8V9W8meWawhXTnB7XKL7Kp1mrnmc5dlbUfDUXXWk0p0cxzTsebDd283Z7f5IuD4Ef9Gkd/dB1/C62c80rZ+3sU023u/gO7zCddZQ/2380f5efznEGn7TTGc7qpzmP4NNuxSv3t+9SvXOrWvs0X/VcyU07wYm6s8OU7/JHero5ZzH2VtRc58BphKeHuqJimqPpMLij+Ezv0Wznp12Ka06/YprDV5xw1Wh+Vq+9z+Tsrai5zoTTCE8PdUXFNEfTYXBH8Zneo9mfxOch8IM+jaMfuo7fxXaueeWsnX2q6XYX3+EdprNWefU+07+afZabznEGn7TTWc7qpzmP4N3uwsCJOnsX054p11mVaz31XIl3+/QcE3/mDDrvTN8VWs6vkbzma84+xTSf+NLoG90jsdv3yJydHs7sWj2D5ugU0xy+4oSrhry0Z/T0XRHZq5G85mvOPsU0n/jS6BvdI7Hb98icT+/JQ+CHfUL+g6e15nXsqh0DX11rt2+l6/ayc+qDn6LP9Jq+CVd+0kw4vRVLs9Id8TpLc5/pNdozeKctrMO5G3umuOpnxjR/mkmf8r7HZzpPr+vAibt905yp/2g+/BR9n+854pnrfeAVfYZy5Kt+ZuzMYR7R53Y12m5Pt1MxzXWO574X3vu9Vt3ElWaaT/8Uuz7do3m356ifvZ1OOfIprvqr54jfnetzqtbXEc9Z6PF+cI8+F977vVbdxJVmmk//p8U/Xf+0033pefgh6n7QnFONcpq7jXATDl+xe8F3XGFH/NSnvd1u5nr0efATDk+cdI5TT33wq0hvRX8ppzm6qQeeSC+1RjiPnUYxzelVbCenT6P3OVc1L+U0hyfCURN1VmErnWuZsepTTZezT6PrnOvOgWbVi+ZMP/Popd6N3ud1zQGrSM18MOqO135y1dPTzVJu4lXjc6mnvfCrSK9G1Ttetb6Udw4dGmoiuEd4Ijy1xyPe9dTe53XpwCpSaz85ER1a7SdHS5xwn4He46pfZ3jfp9Z//oR96ilzrssdqB/kO1/PzH+mlztdMYNZXbx7frfzDHb3+e6ef+aunfbu8909v7uTYnfvv3u+3mU3n87U4R22u2dHd/f8nTOsNHef7+75q7tN3HSmDu+wae4j+N3zHznT1HPvk8C0NfhbHagf0Dt/SJ+dr2fTfNe0Z/cf7bl7/tH+I/7u8909/+h+R/zd57t7/rffb7p/+d69HL/787l7fnfHM9jd57t7/pm7qrbO1b0cv/v8d8/v7vgM1rv2zMT0frQDd/+AXjWfORXPvOg703NGe/f8M2fptHef7+753Z3OYHef7+75R3e9e//d84/ud8RzPo3aA67Ylfnd8589693nu3v+VffnnBX1Ba7Ylfnd8688K7P+dAg0MQ7EgTgQB+JAHIgDceBXO5CHwF/98eZycSAOxIE4EAfiQBzoHchDYO9L0DgQB+JAHIgDcSAO/GoH8hD4qz/eXC4OxIE4EAfiQByIA70DeQjsfQkaB+JAHIgDcSAOxIFf7UAeAj/o4/V/yVRH67APOvIfR+FfRmn8Q/CBxa6/k27C33FV9f2TzvUOL7IzDsSBOBAHjh3IQ+CxRy9TdF/cHfayA51c5Get2rGTI2+Xnzmfa72+/bAHC/w8Xh+0h44DcSAOxIEvcyAPgR/0gXdf2h32QUf+4yjdWTvsj6Y3F2fO51qv33yVvx64P+187/Yn++NAHIgDceBPB/IQ+Kcfb626L23FNOegE1a4c9Rw1MyqqJzmqpnyad6kfxSfzsX+iWef8vTAHUX0xE6v85WnZ+JLq5zmOmfKmQ/vNXhiHIgDcSAOxIFyIA+BH/RzoF/6mnPE7kvdsVXNTOb98wMg/0V173Ve+7rc+7t9Xd8ZrNtBf7dvpa8+55k1RfRE1zmudeVa+37nnPddXjOf6HzqOBAH4kAciAPqQB4C1Y0350cPAUd8d3zt0RytYpp3PNgUq1/fk+5KXM+sOTsU07zjwY5iN2fqUa3m6BXTvOPBpkh/RfJJGzwOxIE4EAfiQB4CP+hnoPviVkxzju1Y1f6etIVrv+arHjiPXb9rnq39brpTc/YopnnHgx3Fbg49xflbOXKiztK848GmqP2aT/rgcSAOxIE48N0O5CHwgz7/7otbMc05tmKaP8rXDH0zZyd2+3f6djXdfMU0Z6Zimnc82FHs5lRPhyumOTsUq9zf6HZiN2unL5o4EAfiQBz4TgfyEPhBn7t+iXMsxTQvvmrFNN/h0XS7wLroe9H4fnCPU7/rvPb5Psf56les0yvv+6Z66nG82+cztUdz12ntc+G832vVTRyaxDgQB+JAHPj9DuQh8IM+4+6L2bGqedfRz/Cu9X7menSL4Dvcsa6e+jutY/RWrBfRc/qUR8MMarS70WdqH7PREKddzms/uc5njvbBO3a2nzmJcSAOxIE48B0O5CHwOz7nrVv6QwRNHd5h6BMfc2DytMM77LGt6YoDcSAOxIFvdSAPgd/6yTf3nh4sHK/asWZcoJMOTJ46Hv9PGht5HIgDcSAOtA7kIbC15XtBHjA0qhvgiiW/zgH81ajTwRVLHgfiQByIA3HgEQfyEPiIa+mJA3EgDsSBOBAH4sAPdyAPgT/8A8zx40AciANxIA7EgTjwiAP/DyW8H5ldC3GtAAAAAElFTkSuQmCC"
    }
   },
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Na seção anterior, nós vimos a sequência completa do plasmídeo pPCP1 da bactéria *Yersinia pestis biovar Microtus* no formato FASTA, mas agora veremos essa mesma sequência em um arquivo GenBank. De novo, esse arquivo pode ser encontrado online no NCBI (https://www.ncbi.nlm.nih.gov/nuccore/NC_005816.1) ou no mesmo diretório desse documento no GitHub.\n",
    "\n",
    "O arquivo contém um único registro (i.e. uma linha LOCUS) e começa assim:\n",
    "\n",
    "![image-2.png](attachment:image-2.png)\n",
    "\n",
    "(Desculpe a qualidade de visualização, é que o meu editor de texto não consege formatar as informações de uma maneira bonita).\n",
    "\n",
    "Novamente, nós usaremos o módulo `Bio.SeqIO` para ler o arquivo. O código é idêntico ao usado com o arquivo FASTA:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "registro = SeqIO.read('NC_005816.1.gb', 'genbank')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SeqRecord(seq=Seq('TGTAACGAACGGTGCAATAGTGATCCACACCCAACGCCTGAAATCAGATCCAGG...CTG'), id='NC_005816.1', name='NC_005816', description='Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, complete sequence', dbxrefs=['BioProject:PRJNA224116', 'BioSample:SAMN02602970', 'Assembly:GCF_000007885.1'])"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "registro"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('TGTAACGAACGGTGCAATAGTGATCCACACCCAACGCCTGAAATCAGATCCAGG...CTG')"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "registro.seq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "O argumento **name** vem da linha LOCUS, enquanto o argumento **id** inclui o sufixo da versão. A descrição vem da linha DEFINITION:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'NC_005816.1'"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "registro.id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'NC_005816'"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "registro.name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, complete sequence'"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "registro.description"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Assim como o arquivo FASTA, o arquivo GenBank não vem com anotações por letra:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{}"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "registro.letter_annotations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Muitas informações estão registradas no argumento **annotations**, por exemplo:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "13"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(registro.annotations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Yersinia pestis biovar Microtus str. 91001'"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "registro.annotations['source']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "O argumento **dbxrefs**:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['BioProject:PRJNA224116',\n",
       " 'BioSample:SAMN02602970',\n",
       " 'Assembly:GCF_000007885.1']"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "registro.dbxrefs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Por último, mas não menos importante (se não o mais interessante), são as entradas da tabela de recursos (e.g. características da CDS ou dos genes) que possuem os registros como objetos **SeqFeature** no argumento **features**:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(registro.features)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Falaremos do objeto **SeqFeature** na próxima seção."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4.3 Objetos de característica, localização e posição"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4.3.1 Objetos SeqFeature"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As características das sequências são essenciais para descrevê-las. Depois de você analisar além da sequência em si, você necessita alguma maneira de organizar e obter um resumo fácil das informações conhecidas dessa sequência. Enquanto provavelmente é impossível desenvolver uma classe de recurso que generalize bem, a classe **SeqFeature** do Bipython tenta encapsular o máximo de informação possível da sequência. Essa classe foi desenhada com base nas tabelas de recursos do GenBank e EMBL, então caso você entenda como elas funcionam, você certamente terá mais facilidade em compreender a estrutura das classe do Biopython.\n",
    "\n",
    "A ideia por trás o objeto **SeqFeature** é descrever uma região em uma determinada sequência pai, normalmente um objeto **SeqRecord**. A região é descrita com um objeto de localização, geralmente um intervalo entre duas posições (ver a **Seção 4.3.2**).\n",
    "\n",
    "A classe **SeqFeature** possui alguns atributos, vamos olhar o que cada um faz agora e depois veremos como usá-los em exemplos da vida real. Os atributos de **SeqFeatures** são:\n",
    "\n",
    "- **.type**: É uma descrição textual do tipo da característica (como 'CDS' ou 'gene');\n",
    "\n",
    "\n",
    "- **.location**: A localização de um objeto **SeqFeature** na sequência que você está lidando. **SeqFeature** delega mais funcionalidades para o local selecionado, e inclui alguns atributos para analisar as propriedades da região selecionada:\n",
    "    - **.ref**: Abreviação para **.location.ref**. Qualquer sequência de referência (diferente) à qual o local está se referindo. Geralmente tem valor *None*;\n",
    "    - **.ref_db**: Abreviação para **.location.ref_db**. Especifíca o banco de dados o identificador usado em **.ref**. Geralmente tem o valor *None* e;\n",
    "    - **.strand**: Abreviação para **.location.strand**. A fita na sequência em que a característica está localizada. Para uma sequência de fita dupla, pode ser o valor 1 para a fita superior, -1 para a inferior, 0 se a fita é importante mas desconhecida e *None* se isso não tem importância. O valor *None* é usado para proteínas o sequências de fita simples.\n",
    "    \n",
    "\n",
    "- **.qualifiers**: É um dicionário Python com informações adicionais sobre o recurso. A chave do dicionário é a descrição da característica em uma palavra e o valor é a informação dessa característica. Por exemplo, uma chave bem comum é a \"evidence\" e o valor é \"(comptational (non-experimental)\". Esse é apenas uma maneira de permitir que a pessoa que esteja olhando esse recurso saiba que ainda não foram feitos testes laboratoriais. Observe que o outro valor será uma lista de strings (onde há uma única string no exemplo). Isso é um reflexo das tabelas de características dos arquivos GenBank e EMBL e;\n",
    "\n",
    "\n",
    "- **.sub_features**: Esse argumento é usaod para representar características que são complicadas de localizar, como 'joins' em arquivos GenBank e EMBL. Esse argumento foi descontinuado com a introdução do objeto **CompundLocation**, então podemos ignorá-lo."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4.3.2 Posições e localizações"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A ideia por trás da classe **SeqFeature** é descrever uma região da sequência pai, para isso nós usamos o objeto de localização, normalmente descrevendo o intervalo entre duas posições. Para deixar mais claro as terminologias, usaremos:\n",
    "\n",
    "- **posição**: Refere-se a posição na sequência, que pode ser confuso ou não. Por exemplo: 5, 20, <100 e >200 são todas posições e;\n",
    "\n",
    "\n",
    "- **localização**: A locarlização é uma região da sequência limitada por algumas posições. Por exemplo: 5..20 (i.e. 5 até 20) é uma localização.\n",
    "\n",
    "E sim, no começo é meio confuso e você certamente retornará aqui para esclarecer as dúvidas."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4.3.2.1 Objeto FeatureLocation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A não ser que você trabalhe com sequências genéticas de eucariotos, as localizações de **SeqFeature** são bem simples: você deve começar com um ponto de partida e chegada e uma fita. Isso é basicamente o que o objeto **FeatureLocation** faz.\n",
    "\n",
    "No mundo real, com certeza, as coisas podem ser um pouco mais complicadas. Primeiro, nós devemos lidar com os locais compostos formados por várias regiões. Segundo, as posições próprias posições podem ser confusas (sem exatidão)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4.3.2.2 Objeto CompoundLocation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Na versão 1.62 do Biopython foi introduzdo o **CompoundLocation** como parte para reestruturar como os locais complexos compostos por várias regiões são representados. Seu principal uso é em locais 'join' em arquivos GenBank e EMBL."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4.3.2.3 Posições confusas"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Até agora nós usamos posições simples. Uma complicação de lidar com as caracterísicas dos locais está na própia posição. Na biologia, muitas vezes as coisas não são totalmente certas. Por exemplo, você quer fazer um experimento com primers e descobre qe o começo da transcrição do mRNA ocorre em um dos dois sítios. Isso é ma informação sem função útil, mas complicação vem na hora de representar essa posição. Para nos ajudar com isso, temos o conceito de posições confusas. Basicamente há diversos tipode de posições confusas, logo o Biopython dispõede 5 classes para lidar com isso:\n",
    "\n",
    "- **ExactPosition**: Como o nome sgere, essa classe representa a posição que é especificada como exata ao longo da sequência. É representada por um número e você consegue esse valor olhando o atributo **position** do objeto;\n",
    "\n",
    "\n",
    "- **BeforePosition**: Essa classe representa a posição confusa que ocorre antes do sítio especificado. Na notação de arquivos GenBank e EMBL, essa representaão é algo como '<13', siginificando que a posição real está abaixo de 13. Para obter o valor superior ao especificado, olhe o atributo **position** do objeto;\n",
    "\n",
    "\n",
    "- **AfterPosition**: Ao contrário da **BeforePosition**, esta classe representa a posição que ocorre após o sítio especificado. É representado em arquivos GenBank como '>13', e igaul ao **BeforePositio**, você obtém o valor limite olhando o atributo **position** do objeto;\n",
    "\n",
    "\n",
    "- **WithinPosition**: Ocasionalmente usado em localizações dos arquivos GenBank e EMBL, esta classe modela uma posição que ocorre entre dois nucleotídeos especificados. Na notação do GenBank e EMBL, isso seria representado como '(1.5)', ou seja, a posição está entre 1 e 5. Para obter a infomação nessa classe, você tem que olhar para dois atributos. O atributo **position** especiica o limite inferior do intervalo que estamos procurando, que no nosso exemplo é 1. O atributo **extension** especifica o limite superior do intervalo, que no caso será o 4. Então, o **object.position** é o limite inferior e o **object.position + object.extension** é o limite superior;\n",
    "\n",
    "\n",
    "- **OneOfPosition**: Ocasionalmente em localizações dos arquivos GenBank e EMBL, esta classe lida com uma posição onde há vários valores possíveis. Por exemplo, você pode usar se o códon de inicío não está claro e possui dois candidatos para essa vaga. Alternativamente, pode ser tratado explicitamente como duas caraterísticas genéticas relacionadas e;\n",
    "\n",
    "\n",
    "- **UnknownPosition**: Essa classe lida com uma posição de localização desconhecida. Isso não é usado em arquivos do GenBank e EMBL, mas corresponde ao caractere \"?\" usado no UniProt."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Abaixo segue um exemplo de localização com os pontos confusos:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqFeature\n",
    "posicao_comeco = SeqFeature.AfterPosition(5)\n",
    "posicao_final = SeqFeature.BetweenPosition(9, left=8, right=9)\n",
    "minha_localizacao = SeqFeature.FeatureLocation(posicao_comeco, posicao_final)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Observe que os dedatlhes da localizações confusas mudaram na versão 1.59 do Biopython, em particlar as classes **BetweenPosition** e **WithinPosition** onde devemos deixar explícito a posição usada para o fatiamento. Para a posição inicial geralmente é usado um valor no parâmetro **left**, enquanto para o final o valor vai no parâmetro **right** (sempre um valor maior do que o que vai no **left**).\n",
    "\n",
    "Se você quiser mostrar o objeto **FeatureLocation**, basta chamá-lo dentro de um **print**:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[>5:(8^9)]\n"
     ]
    }
   ],
   "source": [
    "print(minha_localizacao)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Você pode acessar a posição inicial e final usando os astributos **start** e **end**, respectivamente, na localização:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AfterPosition(5)"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "minha_localizacao.start"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">5\n"
     ]
    }
   ],
   "source": [
    "print(minha_localizacao.start)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "BetweenPosition(9, left=8, right=9)"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "minha_localizacao.end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(8^9)\n"
     ]
    }
   ],
   "source": [
    "print(minha_localizacao.end)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Se você não quiser que retorne uma posição confusa mas sim somente os números, pode usar a classe **int**:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "int(minha_localizacao.start)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "int(minha_localizacao.end)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Para a compatibilidade com as versões mais antigas do Biopython, você pode usar os argumentos **nofuzzy_start** e **nofuzzy_end** para as localizações serem retornadas como inteiros:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "minha_localizacao.nofuzzy_start"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "minha_localizacao.nofuzzy_end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Observe que isso apenas retorna a posição das localizações.\n",
    "\n",
    "De maneira análoga, é fácil criar uma posição sem se importar com as posições confusas. Você pode apenas passar em números para o **FeaturePosition** e será retornado os objetos **ExactPosition**:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "posicao_exata = SeqFeature.FeatureLocation(5, 9)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[5:9]\n"
     ]
    }
   ],
   "source": [
    "print(posicao_exata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ExactPosition(5)"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "posicao_exata.start"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "int(posicao_exata.start)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "posicao_exata.nofuzzy_start"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Essa é a maior parte das questões essênciais sobre as posições confusas em Biopython. Esse mecanismo foi desenhado para lidar de uma maneira não muito complicada em comparação com as posições exatas."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4.3.2.4 Teste de localização"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Você pode usar a palavra reservada **in** com o **SeqFeature** ou a localização do objeto para ver se uma base/resíduo para uma coordenada pai está presente no recurso/localização ou não.\n",
    "\n",
    "Por exemplo, vamos supor que você tem um SNP de interesse e gostaria de saber as características desse SNP. Vamos supor que esse SNP esteja na posição 4350 (lembre-se que o Python começa a contar a partir do 0). Abaixo está uma maneira simples de solução usando a força bruta, onde é checado todas as características em um loop:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "source ['taxon:229193']\n",
      "gene None\n",
      "CDS None\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "meu_snp = 4390\n",
    "registro = SeqIO.read('NC_005816.1.gb', 'genbank')\n",
    "for caracteristica in registro.features:\n",
    "    if meu_snp in caracteristica:\n",
    "        print(caracteristica.type, caracteristica.qualifiers.get('db_xref'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4.3.3 Sequência descrita por uma característica ou localização"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "O **SeqFeature** ou o objeto de localização não contêm a sequência direramente, ao invés disso, a localização descreve como conseguir essas informações da sequência pai. Por exemplo, considere uma pequena sequência de um gene com a localização 5:18 na fita reversa, que nos arquivos GenBank e EMBL é usada a notação -1, mais o menos assim:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "seq = Seq('ACCGAGACGGCAAAGGCTAGCATAGGTATGAGACTTCCTTCCTGCCAGTGCTGAGGAACTGGGAGCCTAC')\n",
    "caracteristica = SeqFeature(FeatureLocation(5, 18), type='gene', strand=-1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Você pode pegar a sequência pai, fatiar entre 5:18 e fazer o reverso complementar. Se você estiver usando a versão 1.59 do Biopython ou posterior, a localização inicial e final da característica podem ser usadas como as posições de início e fim do pedaço desejado:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AGCCTTTGCCGTC\n"
     ]
    }
   ],
   "source": [
    "caracteristica_seq = seq[caracteristica.location.start: caracteristica.location.end].reverse_complement()\n",
    "print(caracteristica_seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "O exemplo acima não é tão ruim (a menos que você tenha que trabalhar com as características compostas (joins), isso torna um pouco chato o trabalho). Em vez disso, o objeto **SeqFeature** possui o método **extract** que faz todo o serviço acima para você:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AGCCTTTGCCGTC\n"
     ]
    }
   ],
   "source": [
    "caracteristica_seq = caracteristica.extract(seq)\n",
    "print(caracteristica_seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "O tamanho do **SeqFeature** ou a localização devem ser iguais a descrição do intervalo selecionado:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13\n"
     ]
    }
   ],
   "source": [
    "print(len(caracteristica_seq))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13\n"
     ]
    }
   ],
   "source": [
    "print(len(caracteristica))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13\n"
     ]
    }
   ],
   "source": [
    "print(len(caracteristica.location))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Para os objetos **FeatureLocation** simples, o comprimento é a diferença entre a posição de começo e a final. Porém, o comprimento de **CompoundLocation** é a soma das regiões constituintes."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4.4 Comparação"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Os objetos **SeqRecord** podem ser um pouco complexos, mas veja um exemplo simples:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "registro_1 = SeqRecord(Seq('ACGT'), id='teste')\n",
    "registro_2 = SeqRecord(Seq('ACGT'), id='teste')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "O que acontece quando tentamos comparar dois registros 'idênticos'?:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "ename": "NotImplementedError",
     "evalue": "SeqRecord comparison is deliberately not implemented. Explicitly compare the attributes of interest.",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNotImplementedError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-60-a805f41a6bb8>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0mregistro_1\u001b[0m \u001b[1;33m==\u001b[0m \u001b[0mregistro_2\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;32mD:\\Jupyter\\lib\\site-packages\\Bio\\SeqRecord.py\u001b[0m in \u001b[0;36m__eq__\u001b[1;34m(self, other)\u001b[0m\n\u001b[0;32m    792\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0m__eq__\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mother\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    793\u001b[0m         \u001b[1;34m\"\"\"Define the equal-to operand (not implemented).\"\"\"\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 794\u001b[1;33m         \u001b[1;32mraise\u001b[0m \u001b[0mNotImplementedError\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0m_NO_SEQRECORD_COMPARISON\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    795\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    796\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0m__ne__\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mother\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mNotImplementedError\u001b[0m: SeqRecord comparison is deliberately not implemented. Explicitly compare the attributes of interest."
     ]
    }
   ],
   "source": [
    "registro_1 == registro_2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Surpreendentemente, as versões mais antigas do Biopython usavam a comparação de objetos do Python para o **SeqRecord**, retornando *True* se as variáveis apontassem o mesmo objeto na memória. A partir da versão 1.67 do Biopython, o **SeqRecord** compara somente se estiver explícito a informação desejada para comparação, fora isso é retornado o erro acima.\n",
    "\n",
    "Para evitar isso, basta passar quais são os atributos de interesse para a comparação:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "registro_1.id == registro_2.id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "registro_1.seq == registro_2.seq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Esteja ciente de que compara objetos complexos rapidamente torna-se complicado (veja a **Seção 3.10** do Capítulo 3: https://github.com/GTL98/biopython_br/tree/main/Cap%203%20-%20Objetos%20de%20Sequ%C3%AAncia)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4.5 Referências"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Outra anotação comum relacionada à sequência é a referência do artigo ou outro trabalho de publicação em que essa sequência em questão foi usada. Há uma maneira bem simples de represenrae a referência no Biopython, existe a classe `Bio.SeqFeature.Reference` que armazena as informações relevantes sobre a referência como um atributo de um objeto.\n",
    "\n",
    "Os atributos incluem coisas que esperamos encontrar em uma referência, como a revista, título e autores. Além disso, pode ser armazenado o medline_id, pubmed_id e comentário sobre a referência. Tudo isso pode ser acessado como atributos do objeto.\n",
    "\n",
    "A referência possui também o objeto de localização específica da sequência que foi usada. Por exemplo, você pode ter um artigo que trabalha com um local particular do gene e dem BAC (cromossomo bacteriano artificial) e quer especificar que refere-se somente e está posição estudada. O argumento **location** é potencialmente uma localização confusa, como visto na **Seção 4.3.2**.\n",
    "\n",
    "Qualquer objeto de referência está armazenado em uma lista no objeto **SeqRecord** em um dicionário do argumento **annotations**, podendo ser acessado com a chave 'references'. Isso é tudo. As referências são fáceis de trabalhar e certamente ajudarão muito nos casos qe você encontrar."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4.6 O método de formatação"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "O método **format** do **SeqRecord** retora ma string contendo o seu registro formatado usando um dos arquivos de saída suportatos pelo `Bio.SeqIO`, como o arquivo FASTA:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "registro = SeqRecord(\n",
    "Seq(\n",
    "'MMYQQGCFAGGTVLRLAKDLAENNRGARVLVVCSEITAVTFRGPSETHLDSMVGQALFGD'\n",
    "'GAGAVIVGSDPDLSVERPLYELVWTGATLLPDSEGAIDGHLREVGLTFHLLKDVPGLISK'\n",
    "'NIEKSLKEAFTPLGISDWNSTFWIAHPGGPAILDQVEAKLGLKEEKMRATREVLSEYGNM'\n",
    "'SSAC'),\n",
    "id='gi|14150838|gb|AAK54648.1|AF376133_1',\n",
    "description='chalcone synthase [Cucumis sativus]')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Deverá retornar:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">gi|14150838|gb|AAK54648.1|AF376133_1 chalcone synthase [Cucumis sativus]\n",
      "MMYQQGCFAGGTVLRLAKDLAENNRGARVLVVCSEITAVTFRGPSETHLDSMVGQALFGD\n",
      "GAGAVIVGSDPDLSVERPLYELVWTGATLLPDSEGAIDGHLREVGLTFHLLKDVPGLISK\n",
      "NIEKSLKEAFTPLGISDWNSTFWIAHPGGPAILDQVEAKLGLKEEKMRATREVLSEYGNM\n",
      "SSAC\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(registro.format('fasta'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Esse método **format** necessita de um argumento, em minúsculo e que seja suportado pelo `Bio.SeqIO` como um formato de saída (veja mais no Capítulo 5: https://github.com/GTL98/biopython_br/tree/main/Cap%205%20-%20Sequ%C3%AAncias%20Input_Output). Entretanto, muitos formatos de arquivo suportados pelo `Bio.SeqIO` podem requerer mais de um registro para serem escritos (geralmente em casos de arquivos com sequências de alinhamento) e não funcionaram pelo método **format** (veja a **Seção 5.5.4**)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4.7 Fatiando um SeqRecord"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Você pode fatias um objeto **SeqRecord** para obter um novo objeto **SeqRecord** que cobre uma parte da sequência. O que é importante aqui é que as anotações por letra também são fatiadas e qualquer caracterítica que caia dentro da nova sequência é preservada (com suas localizações ajustadas).\n",
    "\n",
    "Por exemplo, vamos usar o arquivo do GenBank já visto:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SeqRecord(seq=Seq('TGTAACGAACGGTGCAATAGTGATCCACACCCAACGCCTGAAATCAGATCCAGG...CTG'), id='NC_005816.1', name='NC_005816', description='Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, complete sequence', dbxrefs=['BioProject:PRJNA224116', 'BioSample:SAMN02602970', 'Assembly:GCF_000007885.1'])"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "registro = SeqIO.read('NC_005816.1.gb', 'genbank')\n",
    "registro"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9609"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(registro)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(registro.features)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Nesse exemplo, nós vamos focar no gene **pst** (YP_RS22230). Se você olhar diretamente para o arquivo GenBank, você encontrar esse gene/CDS na localização 4815..5888, ou na 4814:5888 na contagem Python. Se você olhar para o arquivo GenBank, verá que essas informaçõe sestão na décima e décima primeira poição do arquivo. Pela contagem Python são as entradas 9 e 10 da lista **features**:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "type: gene\n",
      "location: [4814:5888](-)\n",
      "qualifiers:\n",
      "    Key: gene, Value: ['pst']\n",
      "    Key: locus_tag, Value: ['YP_RS22230']\n",
      "    Key: old_locus_tag, Value: ['pPCP06', 'YP_pPCP06']\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(registro.features[9])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "type: CDS\n",
      "location: [4814:5888](-)\n",
      "qualifiers:\n",
      "    Key: EC_number, Value: ['3.2.1.17']\n",
      "    Key: codon_start, Value: ['1']\n",
      "    Key: gene, Value: ['pst']\n",
      "    Key: inference, Value: ['COORDINATES: similar to AA sequence:RefSeq:WP_001285181.1']\n",
      "    Key: locus_tag, Value: ['YP_RS22230']\n",
      "    Key: note, Value: ['Derived by automated computational analysis using gene prediction method: Protein Homology.']\n",
      "    Key: old_locus_tag, Value: ['pPCP06', 'YP_pPCP06']\n",
      "    Key: product, Value: ['pesticin']\n",
      "    Key: protein_id, Value: ['WP_002218509.1']\n",
      "    Key: transl_table, Value: ['11']\n",
      "    Key: translation, Value: ['MSDTMVVNGSGGVPAFLFSGSTLSSYRPNFEANSITIALPHYVDLPGRSNFKLMYIMGFPIDTEMEKDSEYSNKIRQESKISKTEGTVSYEQKITVETGQEKDGVKVYRVMVLEGTIAESIEHLDKKENEDILNNNRNRIVLADNTVINFDNISQLKEFLRRSVNIVDHDIFSSNGFEGFNPTSHFPSNPSSDYFNSTGVTFGSGVDLGQRSKQDLLNDGVPQYIADRLDGYYMLRGKEAYDKVRTAPLTLSDNEAHLLSNIYIDKFSHKIEGLFNDANIGLRFSDLPLRTRTALVSIGYQKGFKLSRTAPTVWNKVIAKDWNGLVNAFNNIVDGMSDRRKREGALVQKDIDSGLLK']\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(registro.features[10])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Vamos agora fatiar a sequência pai de 4800 até 5900 (incluindo o gene/CDS **pst**) e veremos quantas características retornam:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SeqRecord(seq=Seq('TGAGGGGCTCATGGTTATTTTAACAATCCACTATCGATATCTTTTTGCACCAGA...TCT'), id='NC_005816.1', name='NC_005816', description='Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, complete sequence', dbxrefs=[])"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sub_registro = registro[4800:5900]\n",
    "sub_registro"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1100"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(sub_registro)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(sub_registro.features)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Nosso sub registro possui duas característicasm o gene e a CDS para YP_RS22230:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "type: gene\n",
      "location: [14:1088](-)\n",
      "qualifiers:\n",
      "    Key: gene, Value: ['pst']\n",
      "    Key: locus_tag, Value: ['YP_RS22230']\n",
      "    Key: old_locus_tag, Value: ['pPCP06', 'YP_pPCP06']\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(sub_registro.features[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "type: CDS\n",
      "location: [14:1088](-)\n",
      "qualifiers:\n",
      "    Key: EC_number, Value: ['3.2.1.17']\n",
      "    Key: codon_start, Value: ['1']\n",
      "    Key: gene, Value: ['pst']\n",
      "    Key: inference, Value: ['COORDINATES: similar to AA sequence:RefSeq:WP_001285181.1']\n",
      "    Key: locus_tag, Value: ['YP_RS22230']\n",
      "    Key: note, Value: ['Derived by automated computational analysis using gene prediction method: Protein Homology.']\n",
      "    Key: old_locus_tag, Value: ['pPCP06', 'YP_pPCP06']\n",
      "    Key: product, Value: ['pesticin']\n",
      "    Key: protein_id, Value: ['WP_002218509.1']\n",
      "    Key: transl_table, Value: ['11']\n",
      "    Key: translation, Value: ['MSDTMVVNGSGGVPAFLFSGSTLSSYRPNFEANSITIALPHYVDLPGRSNFKLMYIMGFPIDTEMEKDSEYSNKIRQESKISKTEGTVSYEQKITVETGQEKDGVKVYRVMVLEGTIAESIEHLDKKENEDILNNNRNRIVLADNTVINFDNISQLKEFLRRSVNIVDHDIFSSNGFEGFNPTSHFPSNPSSDYFNSTGVTFGSGVDLGQRSKQDLLNDGVPQYIADRLDGYYMLRGKEAYDKVRTAPLTLSDNEAHLLSNIYIDKFSHKIEGLFNDANIGLRFSDLPLRTRTALVSIGYQKGFKLSRTAPTVWNKVIAKDWNGLVNAFNNIVDGMSDRRKREGALVQKDIDSGLLK']\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(sub_registro.features[1])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Observe que as localizações estão ajustadas para refletir na nova sequência pai!\n",
    "\n",
    "Embora o Biopython tenha feito algo sensato e, esperamos, intuitivo com as características (e com as anotações por letra), para outra anotaçõa é impossível saber se isso ainda aplica-se ao sub registro ou não. Para evitar suposições, com exceção do tipo molecular, os argumentos **annotations** e **dbxrefs** são omitidos no sub registro e cabe a você a tarefa de passar todas as informações relevantes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'molecule_type': 'DNA'}"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sub_registro.annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sub_registro.dbxrefs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Você gostaria de preservar as outras entradas como o organismo? Cuidado ao copiar todo o dicionário de anotações, pois neste caso a sequência parcial já não é mais um DNA circular, é uma DNA linar:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "sub_registro.annotations['topology'] = 'linear'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A mesma ideia pode ser aplicada para **id**, **name** e **description**, mas praticamente essas informações são preservadas:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'NC_005816.1'"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sub_registro.id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'NC_005816'"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sub_registro.name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, complete sequence'"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sub_registro.description"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Isso ilustra bem o problema, porém, o novo sub registro *não é* a sequência completa do plasmídeo, o que torna a descrição errada!. Vamos consertar isso e, em seguida, analisar o sub registro como um arquivo GenBank usando o método **format** apresentado na **Seção 4.6**: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "sub_registro.description = 'Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, partial'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LOCUS       NC_005816               1100 bp    DNA     linear   UNK 01-JAN-1980\n",
      "DEFINITION  Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, partial.\n",
      "ACCESSION   NC_005816\n",
      "VERSION     NC_005816.1\n",
      "...\n"
     ]
    }
   ],
   "source": [
    "print(sub_registro.format('genbank')[:205] + '...')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Para os exemplo com arquivos FASTQ, onde é feito a anotação por letra, veja as seções **20.1.7** e **20.1.8** do Capítulo 20 (***link do capítulo 20***)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4.8 Adicionar objetos SeqRecord"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Você pode adicionar objetos **SeqRecord** juntos, gerando um novo objeto **SeqRecord**. O que é importante aqui é que qualquer anotação por letra também são adicionadas, todas as características são preservadas (com as suas localizações ajustadas) e qualquer outra anotação se mantém (como nome, descrição e identificação).\n",
    "\n",
    "Como exemplo, vamos usar o arquivo `example.fastq` presente no diretório GitHub deste mesmo documento:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "registro = next(SeqIO.parse('example.fastq', 'fastq'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "25"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(registro)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CCCTTCTTGTCTTCAGCGTTTCTCC\n"
     ]
    }
   ],
   "source": [
    "print(registro.seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[26, 26, 18, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 22, 26, 26, 26, 26, 26, 26, 26, 23, 23]\n"
     ]
    }
   ],
   "source": [
    "print(registro.letter_annotations['phred_quality'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Vamos supor que esses dados são do sequenciador Roche 454 e que a informação mostrada como TTT deveria ser apenas TT. Nós podemos fazer um registro novo fatiando o **SeqRecord** antes do terceiro T:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CCCTTCTTGTCTTCAGCGTT\n"
     ]
    }
   ],
   "source": [
    "esquerda = registro[:20]\n",
    "print(esquerda.seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[26, 26, 18, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 22, 26, 26, 26, 26]\n"
     ]
    }
   ],
   "source": [
    "print(esquerda.letter_annotations['phred_quality'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CTCC\n"
     ]
    }
   ],
   "source": [
    "direita = registro[21:]\n",
    "print(direita.seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[26, 26, 23, 23]\n"
     ]
    }
   ],
   "source": [
    "print(direita.letter_annotations['phred_quality'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Agora vamos adicionar as duas partes e uma só:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "edicao = esquerda + direita"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "24"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(edicao)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[26, 26, 18, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 22, 26, 26, 26, 26, 26, 26, 23, 23]\n"
     ]
    }
   ],
   "source": [
    "print(edicao.letter_annotations['phred_quality'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Fácil e intuitivo? Espero que sim! Você pode fazer de uma maneira mais curta:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "24"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edicao = registro[:20] + registro[21:]\n",
    "len(edicao)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Agora, para um exemplo com as características, vamos usar um arquivo GenBank. Suponha que você tenha um genoma circular:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "registro = SeqIO.read('NC_005816.1.gb', 'genbank')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SeqRecord(seq=Seq('TGTAACGAACGGTGCAATAGTGATCCACACCCAACGCCTGAAATCAGATCCAGG...CTG'), id='NC_005816.1', name='NC_005816', description='Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, complete sequence', dbxrefs=['BioProject:PRJNA224116', 'BioSample:SAMN02602970', 'Assembly:GCF_000007885.1'])"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "registro"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9609"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(registro)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(registro.features)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['BioProject:PRJNA224116',\n",
       " 'BioSample:SAMN02602970',\n",
       " 'Assembly:GCF_000007885.1']"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "registro.dbxrefs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['molecule_type', 'topology', 'data_file_division', 'date', 'accessions', 'sequence_version', 'keywords', 'source', 'organism', 'taxonomy', 'references', 'comment', 'structured_comment'])"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "registro.annotations.keys()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Você pode alterar a origem da seguinte maneira:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SeqRecord(seq=Seq('GATACGCAGTCATATTTTTTACACAATTCTCTAATCCCGACAAGGTCGTAGGTC...GGA'), id='NC_005816.1', name='NC_005816', description='Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, complete sequence', dbxrefs=[])"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mudado = registro[2000:] + registro[:2000]\n",
    "mudado"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Observe que agora muitas anotações não estão do mesmo modo como no registor original, todas as anotações (menos o tipo molecular) e uma das características foram perdidas:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "18"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(mudado.features)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mudado.dbxrefs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['molecule_type'])"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mudado.annotations.keys()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Isso ocorre porque o passo de fatiar o **SeqRecord** é cuidadosa nas anotações que deverão ser preservadas (propagar anotações de modo errado pode causar grandes erros). Se você quer manter as referências da comparação com o banco de dados ou o dicionário com as anotações, você deve fazer isso de modo explícito:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [],
   "source": [
    "mudado.dbxrefs = registro.dbxrefs[:]\n",
    "mudado.annotations = registro.annotations.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['BioProject:PRJNA224116',\n",
       " 'BioSample:SAMN02602970',\n",
       " 'Assembly:GCF_000007885.1']"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mudado.dbxrefs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['molecule_type', 'topology', 'data_file_division', 'date', 'accessions', 'sequence_version', 'keywords', 'source', 'organism', 'taxonomy', 'references', 'comment', 'structured_comment'])"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mudado.annotations.keys()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Vale ressaltar que em um exemplo como este, você provavelmente deverá alterar os identificadores de registro, uma vez que as referências do NCBI referem-se à sequência original não modificada."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4.9 Objetos SeqRecord reverso complementar"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Uma das novidades da versão 1.57 do Biopython foi a inclusão do método **reverse_complement** dos objetos **SeqRecord**. Isso tenta equilibrar a facilidade de uso com preocupações do que fazer sobre a anotação com um registro reverso complementar.\n",
    "\n",
    "Para a sequência, será usado o método de reverso complementar do objeto **Seq**. Quaiquer características são transferidas com a localização e a fita recalculada, assim como a anotação por letra também é copiada, mas de modo reverso. Todavia, a transferência da maioria das anotações é problemática.\n",
    "\n",
    "Por exemplo, se a identificação do registro for um acesso, esse acesso não deve realmente aplicar-se a sequência reversa complementar e trasnferir o identificador por padrão poderá facilmente corromper os dados em análises downstream. Logo, por padrão, a identificação, nome, descrição, anotação e dbxrefs não são transferidas por padrão.\n",
    "\n",
    "O método **reverse_complement** do objeto **SeqRecord** pega uma série de argumentos opcionais para as propriedades do registro. Configurar esses argumentos como *True* indica copiar os valores antigos, enquanto *False* indica para deixar os valores antigos e usar os valores padrão. Você pode alternativamente fornecer o novo desejado.\n",
    "\n",
    "Considere o exemplo a seguir:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NC_005816.1 9609 19 3 13\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "registro = SeqIO.read('NC_005816.1.gb', 'genbank')\n",
    "print(registro.id, len(registro), len(registro.features), len(registro.dbxrefs), len(registro.annotations))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Aqui você pode pegar a sequência reversa complementar e especificar um novo identificador, mas note que é deixado de lado o dbxrefs e as a notações (mas não as características):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TESTE 9609 19 0 0\n"
     ]
    }
   ],
   "source": [
    "registro_rc = registro.reverse_complement(id='TESTE')\n",
    "print(registro_rc.id, len(registro_rc), len(registro_rc.features), len(registro_rc.dbxrefs), len(registro_rc.annotations))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Só isso, mais nada!"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
