{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Capítulo 5 - Sequências Input/Output"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Neste capítulo, discuriremos mais sobre o módulo `Bio.SeqIO`, cjo teve uma breve introdução no Capítulo 2 (https://github.com/GTL98/biopython_br/tree/main/Cap%202%20-%20O%20que%20podemos%20fazer%20com%20o%20Biopython) e usamos algumas de suas funções no Capítulo 4 (https://github.com/GTL98/biopython_br/tree/main/Cap%204%20-%20Objetos%20de%20anota%C3%A7%C3%A3o%20de%20sequ%C3%AAncia). Esse módulo fornece uma interface simples para trabalhar com variados formatos de arquivos de sequência em um caminho uniforme. Para saber mais, acesse o link https://biopython.org/wiki/SeqIO e a documentação online https://biopython.org/docs/1.79/api/Bio.SeqIO.html:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on package Bio.SeqIO in Bio:\n",
      "\n",
      "NAME\n",
      "    Bio.SeqIO - Sequence input/output as SeqRecord objects.\n",
      "\n",
      "DESCRIPTION\n",
      "    Bio.SeqIO is also documented at SeqIO_ and by a whole chapter in our tutorial:\n",
      "    \n",
      "      - `HTML Tutorial`_\n",
      "      - `PDF Tutorial`_\n",
      "    \n",
      "    .. _SeqIO: http://biopython.org/wiki/SeqIO\n",
      "    .. _`HTML Tutorial`: http://biopython.org/DIST/docs/tutorial/Tutorial.html\n",
      "    .. _`PDF Tutorial`: http://biopython.org/DIST/docs/tutorial/Tutorial.pdf\n",
      "    \n",
      "    Input\n",
      "    -----\n",
      "    The main function is Bio.SeqIO.parse(...) which takes an input file handle\n",
      "    (or in recent versions of Biopython alternatively a filename as a string),\n",
      "    and format string.  This returns an iterator giving SeqRecord objects:\n",
      "    \n",
      "    >>> from Bio import SeqIO\n",
      "    >>> for record in SeqIO.parse(\"Fasta/f002\", \"fasta\"):\n",
      "    ...     print(\"%s %i\" % (record.id, len(record)))\n",
      "    gi|1348912|gb|G26680|G26680 633\n",
      "    gi|1348917|gb|G26685|G26685 413\n",
      "    gi|1592936|gb|G29385|G29385 471\n",
      "    \n",
      "    Note that the parse() function will invoke the relevant parser for the\n",
      "    format with its default settings.  You may want more control, in which case\n",
      "    you need to create a format specific sequence iterator directly.\n",
      "    \n",
      "    Some of these parsers are wrappers around low-level parsers which build up\n",
      "    SeqRecord objects for the consistent SeqIO interface. In cases where the\n",
      "    run-time is critical, such as large FASTA or FASTQ files, calling these\n",
      "    underlying parsers will be much faster - in this case these generator\n",
      "    functions which return tuples of strings:\n",
      "    \n",
      "    >>> from Bio.SeqIO.FastaIO import SimpleFastaParser\n",
      "    >>> from Bio.SeqIO.QualityIO import FastqGeneralIterator\n",
      "    \n",
      "    \n",
      "    Input - Single Records\n",
      "    ----------------------\n",
      "    If you expect your file to contain one-and-only-one record, then we provide\n",
      "    the following 'helper' function which will return a single SeqRecord, or\n",
      "    raise an exception if there are no records or more than one record:\n",
      "    \n",
      "    >>> from Bio import SeqIO\n",
      "    >>> record = SeqIO.read(\"Fasta/f001\", \"fasta\")\n",
      "    >>> print(\"%s %i\" % (record.id, len(record)))\n",
      "    gi|3318709|pdb|1A91| 79\n",
      "    \n",
      "    This style is useful when you expect a single record only (and would\n",
      "    consider multiple records an error).  For example, when dealing with GenBank\n",
      "    files for bacterial genomes or chromosomes, there is normally only a single\n",
      "    record.  Alternatively, use this with a handle when downloading a single\n",
      "    record from the internet.\n",
      "    \n",
      "    However, if you just want the first record from a file containing multiple\n",
      "    record, use the next() function on the iterator:\n",
      "    \n",
      "    >>> from Bio import SeqIO\n",
      "    >>> record = next(SeqIO.parse(\"Fasta/f002\", \"fasta\"))\n",
      "    >>> print(\"%s %i\" % (record.id, len(record)))\n",
      "    gi|1348912|gb|G26680|G26680 633\n",
      "    \n",
      "    The above code will work as long as the file contains at least one record.\n",
      "    Note that if there is more than one record, the remaining records will be\n",
      "    silently ignored.\n",
      "    \n",
      "    \n",
      "    Input - Multiple Records\n",
      "    ------------------------\n",
      "    For non-interlaced files (e.g. Fasta, GenBank, EMBL) with multiple records\n",
      "    using a sequence iterator can save you a lot of memory (RAM).  There is\n",
      "    less benefit for interlaced file formats (e.g. most multiple alignment file\n",
      "    formats).  However, an iterator only lets you access the records one by one.\n",
      "    \n",
      "    If you want random access to the records by number, turn this into a list:\n",
      "    \n",
      "    >>> from Bio import SeqIO\n",
      "    >>> records = list(SeqIO.parse(\"Fasta/f002\", \"fasta\"))\n",
      "    >>> len(records)\n",
      "    3\n",
      "    >>> print(records[1].id)\n",
      "    gi|1348917|gb|G26685|G26685\n",
      "    \n",
      "    If you want random access to the records by a key such as the record id,\n",
      "    turn the iterator into a dictionary:\n",
      "    \n",
      "    >>> from Bio import SeqIO\n",
      "    >>> record_dict = SeqIO.to_dict(SeqIO.parse(\"Fasta/f002\", \"fasta\"))\n",
      "    >>> len(record_dict)\n",
      "    3\n",
      "    >>> print(len(record_dict[\"gi|1348917|gb|G26685|G26685\"]))\n",
      "    413\n",
      "    \n",
      "    However, using list() or the to_dict() function will load all the records\n",
      "    into memory at once, and therefore is not possible on very large files.\n",
      "    Instead, for *some* file formats Bio.SeqIO provides an indexing approach\n",
      "    providing dictionary like access to any record. For example,\n",
      "    \n",
      "    >>> from Bio import SeqIO\n",
      "    >>> record_dict = SeqIO.index(\"Fasta/f002\", \"fasta\")\n",
      "    >>> len(record_dict)\n",
      "    3\n",
      "    >>> print(len(record_dict[\"gi|1348917|gb|G26685|G26685\"]))\n",
      "    413\n",
      "    >>> record_dict.close()\n",
      "    \n",
      "    Many but not all of the supported input file formats can be indexed like\n",
      "    this. For example \"fasta\", \"fastq\", \"qual\" and even the binary format \"sff\"\n",
      "    work, but alignment formats like \"phylip\", \"clustalw\" and \"nexus\" will not.\n",
      "    \n",
      "    In most cases you can also use SeqIO.index to get the record from the file\n",
      "    as a raw string (not a SeqRecord). This can be useful for example to extract\n",
      "    a sub-set of records from a file where SeqIO cannot output the file format\n",
      "    (e.g. the plain text SwissProt format, \"swiss\") or where it is important to\n",
      "    keep the output 100% identical to the input). For example,\n",
      "    \n",
      "    >>> from Bio import SeqIO\n",
      "    >>> record_dict = SeqIO.index(\"Fasta/f002\", \"fasta\")\n",
      "    >>> len(record_dict)\n",
      "    3\n",
      "    >>> print(record_dict.get_raw(\"gi|1348917|gb|G26685|G26685\").decode())\n",
      "    >gi|1348917|gb|G26685|G26685 human STS STS_D11734.\n",
      "    CGGAGCCAGCGAGCATATGCTGCATGAGGACCTTTCTATCTTACATTATGGCTGGGAATCTTACTCTTTC\n",
      "    ATCTGATACCTTGTTCAGATTTCAAAATAGTTGTAGCCTTATCCTGGTTTTACAGATGTGAAACTTTCAA\n",
      "    GAGATTTACTGACTTTCCTAGAATAGTTTCTCTACTGGAAACCTGATGCTTTTATAAGCCATTGTGATTA\n",
      "    GGATGACTGTTACAGGCTTAGCTTTGTGTGAAANCCAGTCACCTTTCTCCTAGGTAATGAGTAGTGCTGT\n",
      "    TCATATTACTNTAAGTTCTATAGCATACTTGCNATCCTTTANCCATGCTTATCATANGTACCATTTGAGG\n",
      "    AATTGNTTTGCCCTTTTGGGTTTNTTNTTGGTAAANNNTTCCCGGGTGGGGGNGGTNNNGAAA\n",
      "    <BLANKLINE>\n",
      "    >>> print(record_dict[\"gi|1348917|gb|G26685|G26685\"].format(\"fasta\"))\n",
      "    >gi|1348917|gb|G26685|G26685 human STS STS_D11734.\n",
      "    CGGAGCCAGCGAGCATATGCTGCATGAGGACCTTTCTATCTTACATTATGGCTGGGAATC\n",
      "    TTACTCTTTCATCTGATACCTTGTTCAGATTTCAAAATAGTTGTAGCCTTATCCTGGTTT\n",
      "    TACAGATGTGAAACTTTCAAGAGATTTACTGACTTTCCTAGAATAGTTTCTCTACTGGAA\n",
      "    ACCTGATGCTTTTATAAGCCATTGTGATTAGGATGACTGTTACAGGCTTAGCTTTGTGTG\n",
      "    AAANCCAGTCACCTTTCTCCTAGGTAATGAGTAGTGCTGTTCATATTACTNTAAGTTCTA\n",
      "    TAGCATACTTGCNATCCTTTANCCATGCTTATCATANGTACCATTTGAGGAATTGNTTTG\n",
      "    CCCTTTTGGGTTTNTTNTTGGTAAANNNTTCCCGGGTGGGGGNGGTNNNGAAA\n",
      "    <BLANKLINE>\n",
      "    >>> record_dict.close()\n",
      "    \n",
      "    Here the original file and what Biopython would output differ in the line\n",
      "    wrapping. Also note that the get_raw method will return a bytes object,\n",
      "    hence the use of decode to turn it into a string.\n",
      "    \n",
      "    Also note that the get_raw method will preserve the newline endings. This\n",
      "    example FASTQ file uses Unix style endings (b\"\\n\" only),\n",
      "    \n",
      "    >>> from Bio import SeqIO\n",
      "    >>> fastq_dict = SeqIO.index(\"Quality/example.fastq\", \"fastq\")\n",
      "    >>> len(fastq_dict)\n",
      "    3\n",
      "    >>> raw = fastq_dict.get_raw(\"EAS54_6_R1_2_1_540_792\")\n",
      "    >>> raw.count(b\"\\n\")\n",
      "    4\n",
      "    >>> raw.count(b\"\\r\\n\")\n",
      "    0\n",
      "    >>> b\"\\r\" in raw\n",
      "    False\n",
      "    >>> len(raw)\n",
      "    78\n",
      "    >>> fastq_dict.close()\n",
      "    \n",
      "    Here is the same file but using DOS/Windows new lines (b\"\\r\\n\" instead),\n",
      "    \n",
      "    >>> from Bio import SeqIO\n",
      "    >>> fastq_dict = SeqIO.index(\"Quality/example_dos.fastq\", \"fastq\")\n",
      "    >>> len(fastq_dict)\n",
      "    3\n",
      "    >>> raw = fastq_dict.get_raw(\"EAS54_6_R1_2_1_540_792\")\n",
      "    >>> raw.count(b\"\\n\")\n",
      "    4\n",
      "    >>> raw.count(b\"\\r\\n\")\n",
      "    4\n",
      "    >>> b\"\\r\\n\" in raw\n",
      "    True\n",
      "    >>> len(raw)\n",
      "    82\n",
      "    >>> fastq_dict.close()\n",
      "    \n",
      "    Because this uses two bytes for each new line, the file is longer than\n",
      "    the Unix equivalent with only one byte.\n",
      "    \n",
      "    \n",
      "    Input - Alignments\n",
      "    ------------------\n",
      "    You can read in alignment files as alignment objects using Bio.AlignIO.\n",
      "    Alternatively, reading in an alignment file format via Bio.SeqIO will give\n",
      "    you a SeqRecord for each row of each alignment:\n",
      "    \n",
      "    >>> from Bio import SeqIO\n",
      "    >>> for record in SeqIO.parse(\"Clustalw/hedgehog.aln\", \"clustal\"):\n",
      "    ...     print(\"%s %i\" % (record.id, len(record)))\n",
      "    gi|167877390|gb|EDS40773.1| 447\n",
      "    gi|167234445|ref|NP_001107837. 447\n",
      "    gi|74100009|gb|AAZ99217.1| 447\n",
      "    gi|13990994|dbj|BAA33523.2| 447\n",
      "    gi|56122354|gb|AAV74328.1| 447\n",
      "    \n",
      "    \n",
      "    Output\n",
      "    ------\n",
      "    Use the function Bio.SeqIO.write(...), which takes a complete set of\n",
      "    SeqRecord objects (either as a list, or an iterator), an output file handle\n",
      "    (or in recent versions of Biopython an output filename as a string) and of\n",
      "    course the file format::\n",
      "    \n",
      "      from Bio import SeqIO\n",
      "      records = ...\n",
      "      SeqIO.write(records, \"example.faa\", \"fasta\")\n",
      "    \n",
      "    Or, using a handle::\n",
      "    \n",
      "        from Bio import SeqIO\n",
      "        records = ...\n",
      "        with open(\"example.faa\", \"w\") as handle:\n",
      "          SeqIO.write(records, handle, \"fasta\")\n",
      "    \n",
      "    You are expected to call this function once (with all your records) and if\n",
      "    using a handle, make sure you close it to flush the data to the hard disk.\n",
      "    \n",
      "    \n",
      "    Output - Advanced\n",
      "    -----------------\n",
      "    The effect of calling write() multiple times on a single file will vary\n",
      "    depending on the file format, and is best avoided unless you have a strong\n",
      "    reason to do so.\n",
      "    \n",
      "    If you give a filename, then each time you call write() the existing file\n",
      "    will be overwritten. For sequential files formats (e.g. fasta, genbank) each\n",
      "    \"record block\" holds a single sequence.  For these files it would probably\n",
      "    be safe to call write() multiple times by re-using the same handle.\n",
      "    \n",
      "    However, trying this for certain alignment formats (e.g. phylip, clustal,\n",
      "    stockholm) would have the effect of concatenating several multiple sequence\n",
      "    alignments together.  Such files are created by the PHYLIP suite of programs\n",
      "    for bootstrap analysis, but it is clearer to do this via Bio.AlignIO instead.\n",
      "    \n",
      "    Worse, many fileformats have an explicit header and/or footer structure\n",
      "    (e.g. any XMl format, and most binary file formats like SFF). Here making\n",
      "    multiple calls to write() will result in an invalid file.\n",
      "    \n",
      "    \n",
      "    Conversion\n",
      "    ----------\n",
      "    The Bio.SeqIO.convert(...) function allows an easy interface for simple\n",
      "    file format conversions. Additionally, it may use file format specific\n",
      "    optimisations so this should be the fastest way too.\n",
      "    \n",
      "    In general however, you can combine the Bio.SeqIO.parse(...) function with\n",
      "    the Bio.SeqIO.write(...) function for sequence file conversion. Using\n",
      "    generator expressions or generator functions provides a memory efficient way\n",
      "    to perform filtering or other extra operations as part of the process.\n",
      "    \n",
      "    \n",
      "    File Formats\n",
      "    ------------\n",
      "    When specifying the file format, use lowercase strings.  The same format\n",
      "    names are also used in Bio.AlignIO and include the following:\n",
      "    \n",
      "        - abi     - Applied Biosystem's sequencing trace format\n",
      "        - abi-trim - Same as \"abi\" but with quality trimming with Mott's algorithm\n",
      "        - ace     - Reads the contig sequences from an ACE assembly file.\n",
      "        - cif-atom - Uses Bio.PDB.MMCIFParser to determine the (partial) protein\n",
      "          sequence as it appears in the structure based on the atomic coordinates.\n",
      "        - cif-seqres - Reads a macromolecular Crystallographic Information File\n",
      "          (mmCIF) file to determine the complete protein sequence as defined by the\n",
      "          _pdbx_poly_seq_scheme records.\n",
      "        - embl    - The EMBL flat file format. Uses Bio.GenBank internally.\n",
      "        - fasta   - The generic sequence file format where each record starts with\n",
      "          an identifier line starting with a \">\" character, followed by\n",
      "          lines of sequence.\n",
      "        - fasta-2line - Stricter interpretation of the FASTA format using exactly\n",
      "          two lines per record (no line wrapping).\n",
      "        - fastq   - A \"FASTA like\" format used by Sanger which also stores PHRED\n",
      "          sequence quality values (with an ASCII offset of 33).\n",
      "        - fastq-sanger - An alias for \"fastq\" for consistency with BioPerl and EMBOSS\n",
      "        - fastq-solexa - Original Solexa/Illumnia variant of the FASTQ format which\n",
      "          encodes Solexa quality scores (not PHRED quality scores) with an\n",
      "          ASCII offset of 64.\n",
      "        - fastq-illumina - Solexa/Illumina 1.3 to 1.7 variant of the FASTQ format\n",
      "          which encodes PHRED quality scores with an ASCII offset of 64\n",
      "          (not 33). Note as of version 1.8 of the CASAVA pipeline Illumina\n",
      "          will produce FASTQ files using the standard Sanger encoding.\n",
      "        - gck     - Gene Construction Kit's format.\n",
      "        - genbank - The GenBank or GenPept flat file format.\n",
      "        - gb      - An alias for \"genbank\", for consistency with NCBI Entrez Utilities\n",
      "        - ig      - The IntelliGenetics file format, apparently the same as the\n",
      "          MASE alignment format.\n",
      "        - imgt    - An EMBL like format from IMGT where the feature tables are more\n",
      "          indented to allow for longer feature types.\n",
      "        - nib     - UCSC's nib file format for nucleotide sequences, which uses one\n",
      "          nibble (4 bits) to represent each nucleotide, and stores two nucleotides in\n",
      "          one byte.\n",
      "        - pdb-seqres -  Reads a Protein Data Bank (PDB) file to determine the\n",
      "          complete protein sequence as it appears in the header (no dependencies).\n",
      "        - pdb-atom - Uses Bio.PDB to determine the (partial) protein sequence as\n",
      "          it appears in the structure based on the atom coordinate section of the\n",
      "          file (requires NumPy for Bio.PDB).\n",
      "        - phd     - Output from PHRED, used by PHRAP and CONSED for input.\n",
      "        - pir     - A \"FASTA like\" format introduced by the National Biomedical\n",
      "          Research Foundation (NBRF) for the Protein Information Resource\n",
      "          (PIR) database, now part of UniProt.\n",
      "        - seqxml  - SeqXML, simple XML format described in Schmitt et al (2011).\n",
      "        - sff     - Standard Flowgram Format (SFF), typical output from Roche 454.\n",
      "        - sff-trim - Standard Flowgram Format (SFF) with given trimming applied.\n",
      "        - snapgene - SnapGene's native format.\n",
      "        - swiss   - Plain text Swiss-Prot aka UniProt format.\n",
      "        - tab     - Simple two column tab separated sequence files, where each\n",
      "          line holds a record's identifier and sequence. For example,\n",
      "          this is used as by Aligent's eArray software when saving\n",
      "          microarray probes in a minimal tab delimited text file.\n",
      "        - qual    - A \"FASTA like\" format holding PHRED quality values from\n",
      "          sequencing DNA, but no actual sequences (usually provided\n",
      "          in separate FASTA files).\n",
      "        - uniprot-xml - The UniProt XML format (replacement for the SwissProt plain\n",
      "          text format which we call \"swiss\")\n",
      "        - xdna        - DNA Strider's and SerialCloner's native format.\n",
      "    \n",
      "    Note that while Bio.SeqIO can read all the above file formats, it cannot\n",
      "    write to all of them.\n",
      "    \n",
      "    You can also use any file format supported by Bio.AlignIO, such as \"nexus\",\n",
      "    \"phylip\" and \"stockholm\", which gives you access to the individual sequences\n",
      "    making up each alignment as SeqRecords.\n",
      "\n",
      "PACKAGE CONTENTS\n",
      "    AbiIO\n",
      "    AceIO\n",
      "    FastaIO\n",
      "    GckIO\n",
      "    IgIO\n",
      "    InsdcIO\n",
      "    Interfaces\n",
      "    NibIO\n",
      "    PdbIO\n",
      "    PhdIO\n",
      "    PirIO\n",
      "    QualityIO\n",
      "    SeqXmlIO\n",
      "    SffIO\n",
      "    SnapGeneIO\n",
      "    SwissIO\n",
      "    TabIO\n",
      "    TwoBitIO\n",
      "    UniprotIO\n",
      "    XdnaIO\n",
      "    _index\n",
      "    _twoBitIO\n",
      "\n",
      "FUNCTIONS\n",
      "    convert(in_file, in_format, out_file, out_format, molecule_type=None)\n",
      "        Convert between two sequence file formats, return number of records.\n",
      "        \n",
      "        Arguments:\n",
      "         - in_file - an input handle or filename\n",
      "         - in_format - input file format, lower case string\n",
      "         - out_file - an output handle or filename\n",
      "         - out_format - output file format, lower case string\n",
      "         - molecule_type - optional molecule type to apply, string containing\n",
      "           \"DNA\", \"RNA\" or \"protein\".\n",
      "        \n",
      "        **NOTE** - If you provide an output filename, it will be opened which will\n",
      "        overwrite any existing file without warning.\n",
      "        \n",
      "        The idea here is that while doing this will work::\n",
      "        \n",
      "            from Bio import SeqIO\n",
      "            records = SeqIO.parse(in_handle, in_format)\n",
      "            count = SeqIO.write(records, out_handle, out_format)\n",
      "        \n",
      "        it is shorter to write::\n",
      "        \n",
      "            from Bio import SeqIO\n",
      "            count = SeqIO.convert(in_handle, in_format, out_handle, out_format)\n",
      "        \n",
      "        Also, Bio.SeqIO.convert is faster for some conversions as it can make some\n",
      "        optimisations.\n",
      "        \n",
      "        For example, going from a filename to a handle:\n",
      "        \n",
      "        >>> from Bio import SeqIO\n",
      "        >>> from io import StringIO\n",
      "        >>> handle = StringIO(\"\")\n",
      "        >>> SeqIO.convert(\"Quality/example.fastq\", \"fastq\", handle, \"fasta\")\n",
      "        3\n",
      "        >>> print(handle.getvalue())\n",
      "        >EAS54_6_R1_2_1_413_324\n",
      "        CCCTTCTTGTCTTCAGCGTTTCTCC\n",
      "        >EAS54_6_R1_2_1_540_792\n",
      "        TTGGCAGGCCAAGGCCGATGGATCA\n",
      "        >EAS54_6_R1_2_1_443_348\n",
      "        GTTGCTTCTGGCGTGGGTGGGGGGG\n",
      "        <BLANKLINE>\n",
      "        \n",
      "        Note some formats like SeqXML require you to specify the molecule type\n",
      "        when it cannot be determined by the parser:\n",
      "        \n",
      "        >>> from Bio import SeqIO\n",
      "        >>> from io import BytesIO\n",
      "        >>> handle = BytesIO()\n",
      "        >>> SeqIO.convert(\"Quality/example.fastq\", \"fastq\", handle, \"seqxml\", \"DNA\")\n",
      "        3\n",
      "    \n",
      "    index(filename, format, alphabet=None, key_function=None)\n",
      "        Indexes a sequence file and returns a dictionary like object.\n",
      "        \n",
      "        Arguments:\n",
      "         - filename - string giving name of file to be indexed\n",
      "         - format   - lower case string describing the file format\n",
      "         - alphabet - no longer used, leave as None\n",
      "         - key_function - Optional callback function which when given a\n",
      "           SeqRecord identifier string should return a unique key for the\n",
      "           dictionary.\n",
      "        \n",
      "        This indexing function will return a dictionary like object, giving the\n",
      "        SeqRecord objects as values.\n",
      "        \n",
      "        As of Biopython 1.69, this will preserve the ordering of the records in\n",
      "        file when iterating over the entries.\n",
      "        \n",
      "        >>> from Bio import SeqIO\n",
      "        >>> records = SeqIO.index(\"Quality/example.fastq\", \"fastq\")\n",
      "        >>> len(records)\n",
      "        3\n",
      "        >>> list(records)  # make a list of the keys\n",
      "        ['EAS54_6_R1_2_1_413_324', 'EAS54_6_R1_2_1_540_792', 'EAS54_6_R1_2_1_443_348']\n",
      "        >>> print(records[\"EAS54_6_R1_2_1_540_792\"].format(\"fasta\"))\n",
      "        >EAS54_6_R1_2_1_540_792\n",
      "        TTGGCAGGCCAAGGCCGATGGATCA\n",
      "        <BLANKLINE>\n",
      "        >>> \"EAS54_6_R1_2_1_540_792\" in records\n",
      "        True\n",
      "        >>> print(records.get(\"Missing\", None))\n",
      "        None\n",
      "        >>> records.close()\n",
      "        \n",
      "        If the file is BGZF compressed, this is detected automatically. Ordinary\n",
      "        GZIP files are not supported:\n",
      "        \n",
      "        >>> from Bio import SeqIO\n",
      "        >>> records = SeqIO.index(\"Quality/example.fastq.bgz\", \"fastq\")\n",
      "        >>> len(records)\n",
      "        3\n",
      "        >>> print(records[\"EAS54_6_R1_2_1_540_792\"].seq)\n",
      "        TTGGCAGGCCAAGGCCGATGGATCA\n",
      "        >>> records.close()\n",
      "        \n",
      "        When you call the index function, it will scan through the file, noting\n",
      "        the location of each record. When you access a particular record via the\n",
      "        dictionary methods, the code will jump to the appropriate part of the\n",
      "        file and then parse that section into a SeqRecord.\n",
      "        \n",
      "        Note that not all the input formats supported by Bio.SeqIO can be used\n",
      "        with this index function. It is designed to work only with sequential\n",
      "        file formats (e.g. \"fasta\", \"gb\", \"fastq\") and is not suitable for any\n",
      "        interlaced file format (e.g. alignment formats such as \"clustal\").\n",
      "        \n",
      "        For small files, it may be more efficient to use an in memory Python\n",
      "        dictionary, e.g.\n",
      "        \n",
      "        >>> from Bio import SeqIO\n",
      "        >>> records = SeqIO.to_dict(SeqIO.parse(\"Quality/example.fastq\", \"fastq\"))\n",
      "        >>> len(records)\n",
      "        3\n",
      "        >>> list(records)  # make a list of the keys\n",
      "        ['EAS54_6_R1_2_1_413_324', 'EAS54_6_R1_2_1_540_792', 'EAS54_6_R1_2_1_443_348']\n",
      "        >>> print(records[\"EAS54_6_R1_2_1_540_792\"].format(\"fasta\"))\n",
      "        >EAS54_6_R1_2_1_540_792\n",
      "        TTGGCAGGCCAAGGCCGATGGATCA\n",
      "        <BLANKLINE>\n",
      "        \n",
      "        As with the to_dict() function, by default the id string of each record\n",
      "        is used as the key. You can specify a callback function to transform\n",
      "        this (the record identifier string) into your preferred key. For example:\n",
      "        \n",
      "        >>> from Bio import SeqIO\n",
      "        >>> def make_tuple(identifier):\n",
      "        ...     parts = identifier.split(\"_\")\n",
      "        ...     return int(parts[-2]), int(parts[-1])\n",
      "        >>> records = SeqIO.index(\"Quality/example.fastq\", \"fastq\",\n",
      "        ...                       key_function=make_tuple)\n",
      "        >>> len(records)\n",
      "        3\n",
      "        >>> list(records)  # make a list of the keys\n",
      "        [(413, 324), (540, 792), (443, 348)]\n",
      "        >>> print(records[(540, 792)].format(\"fasta\"))\n",
      "        >EAS54_6_R1_2_1_540_792\n",
      "        TTGGCAGGCCAAGGCCGATGGATCA\n",
      "        <BLANKLINE>\n",
      "        >>> (540, 792) in records\n",
      "        True\n",
      "        >>> \"EAS54_6_R1_2_1_540_792\" in records\n",
      "        False\n",
      "        >>> print(records.get(\"Missing\", None))\n",
      "        None\n",
      "        >>> records.close()\n",
      "        \n",
      "        Another common use case would be indexing an NCBI style FASTA file,\n",
      "        where you might want to extract the GI number from the FASTA identifier\n",
      "        to use as the dictionary key.\n",
      "        \n",
      "        Notice that unlike the to_dict() function, here the key_function does\n",
      "        not get given the full SeqRecord to use to generate the key. Doing so\n",
      "        would impose a severe performance penalty as it would require the file\n",
      "        to be completely parsed while building the index. Right now this is\n",
      "        usually avoided.\n",
      "        \n",
      "        See Also: Bio.SeqIO.index_db() and Bio.SeqIO.to_dict()\n",
      "    \n",
      "    index_db(index_filename, filenames=None, format=None, alphabet=None, key_function=None)\n",
      "        Index several sequence files and return a dictionary like object.\n",
      "        \n",
      "        The index is stored in an SQLite database rather than in memory (as in the\n",
      "        Bio.SeqIO.index(...) function).\n",
      "        \n",
      "        Arguments:\n",
      "         - index_filename - Where to store the SQLite index\n",
      "         - filenames - list of strings specifying file(s) to be indexed, or when\n",
      "           indexing a single file this can be given as a string.\n",
      "           (optional if reloading an existing index, but must match)\n",
      "         - format   - lower case string describing the file format\n",
      "           (optional if reloading an existing index, but must match)\n",
      "         - alphabet - no longer used, leave as None.\n",
      "         - key_function - Optional callback function which when given a\n",
      "           SeqRecord identifier string should return a unique\n",
      "           key for the dictionary.\n",
      "        \n",
      "        This indexing function will return a dictionary like object, giving the\n",
      "        SeqRecord objects as values:\n",
      "        \n",
      "        >>> from Bio import SeqIO\n",
      "        >>> files = [\"GenBank/NC_000932.faa\", \"GenBank/NC_005816.faa\"]\n",
      "        >>> def get_gi(name):\n",
      "        ...     parts = name.split(\"|\")\n",
      "        ...     i = parts.index(\"gi\")\n",
      "        ...     assert i != -1\n",
      "        ...     return parts[i+1]\n",
      "        >>> idx_name = \":memory:\" #use an in memory SQLite DB for this test\n",
      "        >>> records = SeqIO.index_db(idx_name, files, \"fasta\", key_function=get_gi)\n",
      "        >>> len(records)\n",
      "        95\n",
      "        >>> records[\"7525076\"].description\n",
      "        'gi|7525076|ref|NP_051101.1| Ycf2 [Arabidopsis thaliana]'\n",
      "        >>> records[\"45478717\"].description\n",
      "        'gi|45478717|ref|NP_995572.1| pesticin [Yersinia pestis biovar Microtus str. 91001]'\n",
      "        >>> records.close()\n",
      "        \n",
      "        In this example the two files contain 85 and 10 records respectively.\n",
      "        \n",
      "        BGZF compressed files are supported, and detected automatically. Ordinary\n",
      "        GZIP compressed files are not supported.\n",
      "        \n",
      "        See Also: Bio.SeqIO.index() and Bio.SeqIO.to_dict(), and the Python module\n",
      "        glob which is useful for building lists of files.\n",
      "    \n",
      "    parse(handle, format, alphabet=None)\n",
      "        Turn a sequence file into an iterator returning SeqRecords.\n",
      "        \n",
      "        Arguments:\n",
      "         - handle   - handle to the file, or the filename as a string\n",
      "           (note older versions of Biopython only took a handle).\n",
      "         - format   - lower case string describing the file format.\n",
      "         - alphabet - no longer used, should be None.\n",
      "        \n",
      "        Typical usage, opening a file to read in, and looping over the record(s):\n",
      "        \n",
      "        >>> from Bio import SeqIO\n",
      "        >>> filename = \"Fasta/sweetpea.nu\"\n",
      "        >>> for record in SeqIO.parse(filename, \"fasta\"):\n",
      "        ...    print(\"ID %s\" % record.id)\n",
      "        ...    print(\"Sequence length %i\" % len(record))\n",
      "        ID gi|3176602|gb|U78617.1|LOU78617\n",
      "        Sequence length 309\n",
      "        \n",
      "        For lazy-loading file formats such as twobit, for which the file contents\n",
      "        is read on demand only, ensure that the file remains open while extracting\n",
      "        sequence data.\n",
      "        \n",
      "        If you have a string 'data' containing the file contents, you must\n",
      "        first turn this into a handle in order to parse it:\n",
      "        \n",
      "        >>> data = \">Alpha\\nACCGGATGTA\\n>Beta\\nAGGCTCGGTTA\\n\"\n",
      "        >>> from Bio import SeqIO\n",
      "        >>> from io import StringIO\n",
      "        >>> for record in SeqIO.parse(StringIO(data), \"fasta\"):\n",
      "        ...     print(\"%s %s\" % (record.id, record.seq))\n",
      "        Alpha ACCGGATGTA\n",
      "        Beta AGGCTCGGTTA\n",
      "        \n",
      "        Use the Bio.SeqIO.read(...) function when you expect a single record\n",
      "        only.\n",
      "    \n",
      "    read(handle, format, alphabet=None)\n",
      "        Turn a sequence file into a single SeqRecord.\n",
      "        \n",
      "        Arguments:\n",
      "         - handle   - handle to the file, or the filename as a string\n",
      "           (note older versions of Biopython only took a handle).\n",
      "         - format   - string describing the file format.\n",
      "         - alphabet - no longer used, should be None.\n",
      "        \n",
      "        This function is for use parsing sequence files containing\n",
      "        exactly one record.  For example, reading a GenBank file:\n",
      "        \n",
      "        >>> from Bio import SeqIO\n",
      "        >>> record = SeqIO.read(\"GenBank/arab1.gb\", \"genbank\")\n",
      "        >>> print(\"ID %s\" % record.id)\n",
      "        ID AC007323.5\n",
      "        >>> print(\"Sequence length %i\" % len(record))\n",
      "        Sequence length 86436\n",
      "        \n",
      "        If the handle contains no records, or more than one record,\n",
      "        an exception is raised.  For example:\n",
      "        \n",
      "        >>> from Bio import SeqIO\n",
      "        >>> record = SeqIO.read(\"GenBank/cor6_6.gb\", \"genbank\")\n",
      "        Traceback (most recent call last):\n",
      "            ...\n",
      "        ValueError: More than one record found in handle\n",
      "        \n",
      "        If however you want the first record from a file containing\n",
      "        multiple records this function would raise an exception (as\n",
      "        shown in the example above).  Instead use:\n",
      "        \n",
      "        >>> from Bio import SeqIO\n",
      "        >>> record = next(SeqIO.parse(\"GenBank/cor6_6.gb\", \"genbank\"))\n",
      "        >>> print(\"First record's ID %s\" % record.id)\n",
      "        First record's ID X55053.1\n",
      "        \n",
      "        Use the Bio.SeqIO.parse(handle, format) function if you want\n",
      "        to read multiple records from the handle.\n",
      "    \n",
      "    to_dict(sequences, key_function=None)\n",
      "        Turn a sequence iterator or list into a dictionary.\n",
      "        \n",
      "        Arguments:\n",
      "         - sequences  - An iterator that returns SeqRecord objects,\n",
      "           or simply a list of SeqRecord objects.\n",
      "         - key_function - Optional callback function which when given a\n",
      "           SeqRecord should return a unique key for the dictionary.\n",
      "        \n",
      "        e.g. key_function = lambda rec : rec.name\n",
      "        or,  key_function = lambda rec : rec.description.split()[0]\n",
      "        \n",
      "        If key_function is omitted then record.id is used, on the assumption\n",
      "        that the records objects returned are SeqRecords with a unique id.\n",
      "        \n",
      "        If there are duplicate keys, an error is raised.\n",
      "        \n",
      "        Since Python 3.7, the default dict class maintains key order, meaning\n",
      "        this dictionary will reflect the order of records given to it. For\n",
      "        CPython and PyPy, this was already implemented for Python 3.6, so\n",
      "        effectively you can always assume the record order is preserved.\n",
      "        \n",
      "        Example usage, defaulting to using the record.id as key:\n",
      "        \n",
      "        >>> from Bio import SeqIO\n",
      "        >>> filename = \"GenBank/cor6_6.gb\"\n",
      "        >>> format = \"genbank\"\n",
      "        >>> id_dict = SeqIO.to_dict(SeqIO.parse(filename, format))\n",
      "        >>> print(list(id_dict))\n",
      "        ['X55053.1', 'X62281.1', 'M81224.1', 'AJ237582.1', 'L31939.1', 'AF297471.1']\n",
      "        >>> print(id_dict[\"L31939.1\"].description)\n",
      "        Brassica rapa (clone bif72) kin mRNA, complete cds\n",
      "        \n",
      "        A more complex example, using the key_function argument in order to\n",
      "        use a sequence checksum as the dictionary key:\n",
      "        \n",
      "        >>> from Bio import SeqIO\n",
      "        >>> from Bio.SeqUtils.CheckSum import seguid\n",
      "        >>> filename = \"GenBank/cor6_6.gb\"\n",
      "        >>> format = \"genbank\"\n",
      "        >>> seguid_dict = SeqIO.to_dict(SeqIO.parse(filename, format),\n",
      "        ...               key_function = lambda rec : seguid(rec.seq))\n",
      "        >>> for key, record in sorted(seguid_dict.items()):\n",
      "        ...     print(\"%s %s\" % (key, record.id))\n",
      "        /wQvmrl87QWcm9llO4/efg23Vgg AJ237582.1\n",
      "        BUg6YxXSKWEcFFH0L08JzaLGhQs L31939.1\n",
      "        SabZaA4V2eLE9/2Fm5FnyYy07J4 X55053.1\n",
      "        TtWsXo45S3ZclIBy4X/WJc39+CY M81224.1\n",
      "        l7gjJFE6W/S1jJn5+1ASrUKW/FA X62281.1\n",
      "        uVEYeAQSV5EDQOnFoeMmVea+Oow AF297471.1\n",
      "        \n",
      "        This approach is not suitable for very large sets of sequences, as all\n",
      "        the SeqRecord objects are held in memory. Instead, consider using the\n",
      "        Bio.SeqIO.index() function (if it supports your particular file format).\n",
      "        \n",
      "        Since Python 3.6, the default dict class maintains key order, meaning\n",
      "        this dictionary will reflect the order of records given to it. As of\n",
      "        Biopython 1.72, on older versions of Python we explicitly use an\n",
      "        OrderedDict so that you can always assume the record order is preserved.\n",
      "    \n",
      "    write(sequences, handle, format)\n",
      "        Write complete set of sequences to a file.\n",
      "        \n",
      "        Arguments:\n",
      "         - sequences - A list (or iterator) of SeqRecord objects, or a single\n",
      "           SeqRecord.\n",
      "         - handle    - File handle object to write to, or filename as string.\n",
      "         - format    - lower case string describing the file format to write.\n",
      "        \n",
      "        Note if providing a file handle, your code should close the handle\n",
      "        after calling this function (to ensure the data gets flushed to disk).\n",
      "        \n",
      "        Returns the number of records written (as an integer).\n",
      "\n",
      "FILE\n",
      "    d:\\jupyter\\lib\\site-packages\\bio\\seqio\\__init__.py\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "help(SeqIO)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "O \"problma\" é que você tem que trabalhar com objetos **SeqRecords** (visto no Capítulo 4: https://github.com/GTL98/biopython_br/tree/main/Cap%204%20-%20Objetos%20de%20anota%C3%A7%C3%A3o%20de%20sequ%C3%AAncia) que contêm o objeto **Seq** (visto no Capítulo 3: https://github.com/GTL98/biopython_br/tree/main/Cap%203%20-%20Objetos%20de%20Sequ%C3%AAncia) mais as anotações como a identificação e descrição. Observe quando estamos trabalhando com grandes arquivos FASTA ou FASTQ, a sobrecarga que trabalhar com todos esses objetos deixará o script lento. Nesse caso, considere usar os analisadores **SimpleFastaParser** e **FastqGeneralIterator** que retornarão uma tupla de strings para cada registro (ver a **Seção 5.6**)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5.1 Analisar ou ler as sequências"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A função **parse** do módulo `Bio.SeqIO` é usada para ler uma sequência em objetos **SeqRecord**. Essa função necessita de dois argumentos:\n",
    "1. O primeiro argumento é um *identificador* para ler os dados ou um nome de arquivo. Um identificador típico abre o arquivo para leitura, mas pode ser a saída de um programa de linha de comando, ou um conjunto de dados baixados da internet (ver a **Seção 5.3**). Veja a **Seção 24.1** para mais informações sobre os identificadores (Capítulo 24: ***link do capítulo 24***).\n",
    "\n",
    "2. O segundo argumento é uma string de caractéres minúsculos para especificar o formato de arquivo. Para saber todos os formatos suportados, acesse: https://biopython.org/wiki/SeqIO.\n",
    "\n",
    "A função **parse** retorna um *iterador* que contém objetos **SeqRecord**. Iteradores são usados nos loops (mostrados ao decorrer deste capítulo).\n",
    "\n",
    "Muitas vezes você encontrará arquivos com um único registro. Para esta situação, use a função **read** do módulo `Bio.SeqIO` que possui os mesmos argumentos. Use essa função desde que haja somente um registro no arquivo, caso contrário, será retornado um erro."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5.1.1 Lendo os arquivos de sequência"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "De modo geral, a função **parse** é usada para ler os objetos **SeqRecord** em um loop como o seguinte:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gi|2765658|emb|Z78533.1|CIZ78533\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGATGAGACCGTGG...CGC')\n",
      "740\n",
      "gi|2765657|emb|Z78532.1|CCZ78532\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAACAG...GGC')\n",
      "753\n",
      "gi|2765656|emb|Z78531.1|CFZ78531\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAGCAG...TAA')\n",
      "748\n",
      "gi|2765655|emb|Z78530.1|CMZ78530\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAAACAACAT...CAT')\n",
      "744\n",
      "gi|2765654|emb|Z78529.1|CLZ78529\n",
      "Seq('ACGGCGAGCTGCCGAAGGACATTGTTGAGACAGCAGAATATACGATTGAGTGAA...AAA')\n",
      "733\n",
      "gi|2765652|emb|Z78527.1|CYZ78527\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAGTAG...CCC')\n",
      "718\n",
      "gi|2765651|emb|Z78526.1|CGZ78526\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAGTAG...TGT')\n",
      "730\n",
      "gi|2765650|emb|Z78525.1|CAZ78525\n",
      "Seq('TGTTGAGATAGCAGAATATACATCGAGTGAATCCGGAGGACCTGTGGTTATTCG...GCA')\n",
      "704\n",
      "gi|2765649|emb|Z78524.1|CFZ78524\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATAGTAG...AGC')\n",
      "740\n",
      "gi|2765648|emb|Z78523.1|CHZ78523\n",
      "Seq('CGTAACCAGGTTTCCGTAGGTGAACCTGCGGCAGGATCATTGTTGAGACAGCAG...AAG')\n",
      "709\n",
      "gi|2765647|emb|Z78522.1|CMZ78522\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAGCAG...GAG')\n",
      "700\n",
      "gi|2765646|emb|Z78521.1|CCZ78521\n",
      "Seq('GTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAGTAGAATATATGATCGAGT...ACC')\n",
      "726\n",
      "gi|2765645|emb|Z78520.1|CSZ78520\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAGCAG...TTT')\n",
      "753\n",
      "gi|2765644|emb|Z78519.1|CPZ78519\n",
      "Seq('ATATGATCGAGTGAATCTGGTGGACTTGTGGTTACTCAGCTCGCCATAGGCTTT...TTA')\n",
      "699\n",
      "gi|2765643|emb|Z78518.1|CRZ78518\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGGAGGATCATTGTTGAGATAGTAG...TCC')\n",
      "658\n",
      "gi|2765642|emb|Z78517.1|CFZ78517\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAGTAG...AGC')\n",
      "752\n",
      "gi|2765641|emb|Z78516.1|CPZ78516\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAGTAT...TAA')\n",
      "726\n",
      "gi|2765640|emb|Z78515.1|MXZ78515\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGCTGAGACCGTAG...AGC')\n",
      "765\n",
      "gi|2765639|emb|Z78514.1|PSZ78514\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGGACCTTCGGGAGGATCATTTTTGAAGCCCCCA...CTA')\n",
      "755\n",
      "gi|2765638|emb|Z78513.1|PBZ78513\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACCGCCA...GAG')\n",
      "742\n",
      "gi|2765637|emb|Z78512.1|PWZ78512\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGGACCTTCGGGAGGATCATTTTTGAAGCCCCCA...AGC')\n",
      "762\n",
      "gi|2765636|emb|Z78511.1|PEZ78511\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTTCGGAAGGATCATTGTTGAGACCCCCA...GGA')\n",
      "745\n",
      "gi|2765635|emb|Z78510.1|PCZ78510\n",
      "Seq('CTAACCAGGGTTCCGAGGTGACCTTCGGGAGGATTCCTTTTTAAGCCCCCGAAA...TTA')\n",
      "750\n",
      "gi|2765634|emb|Z78509.1|PPZ78509\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACCGCCA...GGA')\n",
      "731\n",
      "gi|2765633|emb|Z78508.1|PLZ78508\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACCGCCA...TGA')\n",
      "741\n",
      "gi|2765632|emb|Z78507.1|PLZ78507\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACCCCCA...TGA')\n",
      "740\n",
      "gi|2765631|emb|Z78506.1|PLZ78506\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACCGCAA...TGA')\n",
      "727\n",
      "gi|2765630|emb|Z78505.1|PSZ78505\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACCGCCA...TTT')\n",
      "711\n",
      "gi|2765629|emb|Z78504.1|PKZ78504\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTTCGGAAGGATCATTGTTGAGACCGCAA...TAA')\n",
      "743\n",
      "gi|2765628|emb|Z78503.1|PCZ78503\n",
      "Seq('CGTAACCAGGTTTCCGTAGGTGAACCTCCGGAAGGATCCTTGTTGAGACCGCCA...TAA')\n",
      "727\n",
      "gi|2765627|emb|Z78502.1|PBZ78502\n",
      "Seq('CGTAACCAGGTTTCCGTAGGTGAACCTCCGGAAGGATCATTGTTGAGACCGCCA...CGC')\n",
      "757\n",
      "gi|2765626|emb|Z78501.1|PCZ78501\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACCGCAA...AGA')\n",
      "770\n",
      "gi|2765625|emb|Z78500.1|PWZ78500\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGCTCATTGTTGAGACCGCAA...AAG')\n",
      "767\n",
      "gi|2765624|emb|Z78499.1|PMZ78499\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAGGGATCATTGTTGAGATCGCAT...ACC')\n",
      "759\n",
      "gi|2765623|emb|Z78498.1|PMZ78498\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAAGGTCATTGTTGAGATCACAT...AGC')\n",
      "750\n",
      "gi|2765622|emb|Z78497.1|PDZ78497\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AGC')\n",
      "788\n",
      "gi|2765621|emb|Z78496.1|PAZ78496\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCGCAT...AGC')\n",
      "774\n",
      "gi|2765620|emb|Z78495.1|PEZ78495\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTCCGGAAGGATCATTGTTGAGATCACAT...GTG')\n",
      "789\n",
      "gi|2765619|emb|Z78494.1|PNZ78494\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGGTCGCAT...AAG')\n",
      "688\n",
      "gi|2765618|emb|Z78493.1|PGZ78493\n",
      "Seq('CGTAACAAGGATTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCGCAT...CCC')\n",
      "719\n",
      "gi|2765617|emb|Z78492.1|PBZ78492\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCGCAT...ATA')\n",
      "743\n",
      "gi|2765616|emb|Z78491.1|PCZ78491\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCGCAT...AGC')\n",
      "737\n",
      "gi|2765615|emb|Z78490.1|PFZ78490\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TGA')\n",
      "728\n",
      "gi|2765614|emb|Z78489.1|PDZ78489\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...GGC')\n",
      "740\n",
      "gi|2765613|emb|Z78488.1|PTZ78488\n",
      "Seq('CTGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACGCAATAATTGATCGA...GCT')\n",
      "696\n",
      "gi|2765612|emb|Z78487.1|PHZ78487\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TAA')\n",
      "732\n",
      "gi|2765611|emb|Z78486.1|PBZ78486\n",
      "Seq('CGTCACGAGGTTTCCGTAGGTGAATCTGCGGGAGGATCATTGTTGAGATCACAT...TGA')\n",
      "731\n",
      "gi|2765610|emb|Z78485.1|PHZ78485\n",
      "Seq('CTGAACCTGGTGTCCGAAGGTGAATCTGCGGATGGATCATTGTTGAGATATCAT...GTA')\n",
      "735\n",
      "gi|2765609|emb|Z78484.1|PCZ78484\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGGGGAAGGATCATTGTTGAGATCACAT...TTT')\n",
      "720\n",
      "gi|2765608|emb|Z78483.1|PVZ78483\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...GCA')\n",
      "740\n",
      "gi|2765607|emb|Z78482.1|PEZ78482\n",
      "Seq('TCTACTGCAGTGACCGAGATTTGCCATCGAGCCTCCTGGGAGCTTTCTTGCTGG...GCA')\n",
      "629\n",
      "gi|2765606|emb|Z78481.1|PIZ78481\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TGA')\n",
      "572\n",
      "gi|2765605|emb|Z78480.1|PGZ78480\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TGA')\n",
      "587\n",
      "gi|2765604|emb|Z78479.1|PPZ78479\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AGT')\n",
      "700\n",
      "gi|2765603|emb|Z78478.1|PVZ78478\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTCCGGAAGGATCAGTGTTGAGATCACAT...GGC')\n",
      "636\n",
      "gi|2765602|emb|Z78477.1|PVZ78477\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TGC')\n",
      "716\n",
      "gi|2765601|emb|Z78476.1|PGZ78476\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...CCC')\n",
      "592\n",
      "gi|2765600|emb|Z78475.1|PSZ78475\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...GGT')\n",
      "716\n",
      "gi|2765599|emb|Z78474.1|PKZ78474\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACGT...CTT')\n",
      "733\n",
      "gi|2765598|emb|Z78473.1|PSZ78473\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AGG')\n",
      "626\n",
      "gi|2765597|emb|Z78472.1|PLZ78472\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AGC')\n",
      "737\n",
      "gi|2765596|emb|Z78471.1|PDZ78471\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AGC')\n",
      "740\n",
      "gi|2765595|emb|Z78470.1|PPZ78470\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...GTT')\n",
      "574\n",
      "gi|2765594|emb|Z78469.1|PHZ78469\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...GTT')\n",
      "594\n",
      "gi|2765593|emb|Z78468.1|PAZ78468\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCGCAT...GTT')\n",
      "610\n",
      "gi|2765592|emb|Z78467.1|PSZ78467\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TGA')\n",
      "730\n",
      "gi|2765591|emb|Z78466.1|PPZ78466\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...CCC')\n",
      "641\n",
      "gi|2765590|emb|Z78465.1|PRZ78465\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TGC')\n",
      "702\n",
      "gi|2765589|emb|Z78464.1|PGZ78464\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAGCGGAAGGGTCATTGTTGAGATCACATAATA...AGC')\n",
      "733\n",
      "gi|2765588|emb|Z78463.1|PGZ78463\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGTTCATTGTTGAGATCACAT...AGC')\n",
      "738\n",
      "gi|2765587|emb|Z78462.1|PSZ78462\n",
      "Seq('CGTCACGAGGTCTCCGGATGTGACCCTGCGGAAGGATCATTGTTGAGATCACAT...CAT')\n",
      "736\n",
      "gi|2765586|emb|Z78461.1|PWZ78461\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTCCGGAAGGATCATTGTTGAGATCACAT...TAA')\n",
      "732\n",
      "gi|2765585|emb|Z78460.1|PCZ78460\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTCCGGAAGGATCATTGTTGAGATCACAT...TTA')\n",
      "745\n",
      "gi|2765584|emb|Z78459.1|PDZ78459\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TTT')\n",
      "744\n",
      "gi|2765583|emb|Z78458.1|PHZ78458\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TTG')\n",
      "738\n",
      "gi|2765582|emb|Z78457.1|PCZ78457\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTCCGGAAGGATCATTGTTGAGATCACAT...GAG')\n",
      "739\n",
      "gi|2765581|emb|Z78456.1|PTZ78456\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AGC')\n",
      "740\n",
      "gi|2765580|emb|Z78455.1|PJZ78455\n",
      "Seq('CGTAACCAGGTTTCCGTAGGTGGACCTTCGGGAGGATCATTTTTGAGATCACAT...GCA')\n",
      "745\n",
      "gi|2765579|emb|Z78454.1|PFZ78454\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AAC')\n",
      "695\n",
      "gi|2765578|emb|Z78453.1|PSZ78453\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...GCA')\n",
      "745\n",
      "gi|2765577|emb|Z78452.1|PBZ78452\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...GCA')\n",
      "743\n",
      "gi|2765576|emb|Z78451.1|PHZ78451\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGTACCTCCGGAAGGATCATTGTTGAGATCACAT...AGC')\n",
      "730\n",
      "gi|2765575|emb|Z78450.1|PPZ78450\n",
      "Seq('GGAAGGATCATTGCTGATATCACATAATAATTGATCGAGTTAAGCTGGAGGATC...GAG')\n",
      "706\n",
      "gi|2765574|emb|Z78449.1|PMZ78449\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TGC')\n",
      "744\n",
      "gi|2765573|emb|Z78448.1|PAZ78448\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AGG')\n",
      "742\n",
      "gi|2765572|emb|Z78447.1|PVZ78447\n",
      "Seq('CGTAACAAGGATTCCGTAGGTGAACCTGCGGGAGGATCATTGTTGAGATCACAT...AGC')\n",
      "694\n",
      "gi|2765571|emb|Z78446.1|PAZ78446\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTCCGGAAGGATCATTGTTGAGATCACAT...CCC')\n",
      "712\n",
      "gi|2765570|emb|Z78445.1|PUZ78445\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TGT')\n",
      "715\n",
      "gi|2765569|emb|Z78444.1|PAZ78444\n",
      "Seq('CGTAACAAGGTTTCCGTAGGGTGAACTGCGGAAGGATCATTGTTGAGATCACAT...ATT')\n",
      "688\n",
      "gi|2765568|emb|Z78443.1|PLZ78443\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AGG')\n",
      "784\n",
      "gi|2765567|emb|Z78442.1|PBZ78442\n",
      "Seq('GTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACATAATAATTGATCGAGT...AGT')\n",
      "721\n",
      "gi|2765566|emb|Z78441.1|PSZ78441\n",
      "Seq('GGAAGGTCATTGCCGATATCACATAATAATTGATCGAGTTAATCTGGAGGATCT...GAG')\n",
      "703\n",
      "gi|2765565|emb|Z78440.1|PPZ78440\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGGACCTCCGGGAGGATCATTGTTGAGATCACAT...GCA')\n",
      "744\n",
      "gi|2765564|emb|Z78439.1|PBZ78439\n",
      "Seq('CATTGTTGAGATCACATAATAATTGATCGAGTTAATCTGGAGGATCTGTTTACT...GCC')\n",
      "592\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "for registro_seq in SeqIO.parse('ls_orchid.fasta', 'fasta'):\n",
    "    print(registro_seq.id)\n",
    "    print(repr(registro_seq.seq))\n",
    "    print(len(registro_seq))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "O exemplo acima é o mesmo apresentado na **Seção 2.4** do Capítulo 2 (https://github.com/GTL98/biopython_br/tree/main/Cap%202%20-%20O%20que%20podemos%20fazer%20com%20o%20Biopython), onde carregamos as sequências de DNA no formato FASTA. Se você quiser carregar um arquivo no formato GenBank, basta mudar a extensão do arquivo e o segundo parâmetro:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Z78533.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGATGAGACCGTGG...CGC')\n",
      "740\n",
      "Z78532.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAACAG...GGC')\n",
      "753\n",
      "Z78531.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAGCAG...TAA')\n",
      "748\n",
      "Z78530.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAAACAACAT...CAT')\n",
      "744\n",
      "Z78529.1\n",
      "Seq('ACGGCGAGCTGCCGAAGGACATTGTTGAGACAGCAGAATATACGATTGAGTGAA...AAA')\n",
      "733\n",
      "Z78527.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAGTAG...CCC')\n",
      "718\n",
      "Z78526.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAGTAG...TGT')\n",
      "730\n",
      "Z78525.1\n",
      "Seq('TGTTGAGATAGCAGAATATACATCGAGTGAATCCGGAGGACCTGTGGTTATTCG...GCA')\n",
      "704\n",
      "Z78524.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATAGTAG...AGC')\n",
      "740\n",
      "Z78523.1\n",
      "Seq('CGTAACCAGGTTTCCGTAGGTGAACCTGCGGCAGGATCATTGTTGAGACAGCAG...AAG')\n",
      "709\n",
      "Z78522.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAGCAG...GAG')\n",
      "700\n",
      "Z78521.1\n",
      "Seq('GTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAGTAGAATATATGATCGAGT...ACC')\n",
      "726\n",
      "Z78520.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAGCAG...TTT')\n",
      "753\n",
      "Z78519.1\n",
      "Seq('ATATGATCGAGTGAATCTGGTGGACTTGTGGTTACTCAGCTCGCCATAGGCTTT...TTA')\n",
      "699\n",
      "Z78518.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGGAGGATCATTGTTGAGATAGTAG...TCC')\n",
      "658\n",
      "Z78517.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAGTAG...AGC')\n",
      "752\n",
      "Z78516.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAGTAT...TAA')\n",
      "726\n",
      "Z78515.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGCTGAGACCGTAG...AGC')\n",
      "765\n",
      "Z78514.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGGACCTTCGGGAGGATCATTTTTGAAGCCCCCA...CTA')\n",
      "755\n",
      "Z78513.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACCGCCA...GAG')\n",
      "742\n",
      "Z78512.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGGACCTTCGGGAGGATCATTTTTGAAGCCCCCA...AGC')\n",
      "762\n",
      "Z78511.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTTCGGAAGGATCATTGTTGAGACCCCCA...GGA')\n",
      "745\n",
      "Z78510.1\n",
      "Seq('CTAACCAGGGTTCCGAGGTGACCTTCGGGAGGATTCCTTTTTAAGCCCCCGAAA...TTA')\n",
      "750\n",
      "Z78509.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACCGCCA...GGA')\n",
      "731\n",
      "Z78508.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACCGCCA...TGA')\n",
      "741\n",
      "Z78507.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACCCCCA...TGA')\n",
      "740\n",
      "Z78506.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACCGCAA...TGA')\n",
      "727\n",
      "Z78505.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACCGCCA...TTT')\n",
      "711\n",
      "Z78504.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTTCGGAAGGATCATTGTTGAGACCGCAA...TAA')\n",
      "743\n",
      "Z78503.1\n",
      "Seq('CGTAACCAGGTTTCCGTAGGTGAACCTCCGGAAGGATCCTTGTTGAGACCGCCA...TAA')\n",
      "727\n",
      "Z78502.1\n",
      "Seq('CGTAACCAGGTTTCCGTAGGTGAACCTCCGGAAGGATCATTGTTGAGACCGCCA...CGC')\n",
      "757\n",
      "Z78501.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACCGCAA...AGA')\n",
      "770\n",
      "Z78500.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGCTCATTGTTGAGACCGCAA...AAG')\n",
      "767\n",
      "Z78499.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAGGGATCATTGTTGAGATCGCAT...ACC')\n",
      "759\n",
      "Z78498.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAAGGTCATTGTTGAGATCACAT...AGC')\n",
      "750\n",
      "Z78497.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AGC')\n",
      "788\n",
      "Z78496.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCGCAT...AGC')\n",
      "774\n",
      "Z78495.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTCCGGAAGGATCATTGTTGAGATCACAT...GTG')\n",
      "789\n",
      "Z78494.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGGTCGCAT...AAG')\n",
      "688\n",
      "Z78493.1\n",
      "Seq('CGTAACAAGGATTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCGCAT...CCC')\n",
      "719\n",
      "Z78492.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCGCAT...ATA')\n",
      "743\n",
      "Z78491.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCGCAT...AGC')\n",
      "737\n",
      "Z78490.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TGA')\n",
      "728\n",
      "Z78489.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...GGC')\n",
      "740\n",
      "Z78488.1\n",
      "Seq('CTGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACGCAATAATTGATCGA...GCT')\n",
      "696\n",
      "Z78487.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TAA')\n",
      "732\n",
      "Z78486.1\n",
      "Seq('CGTCACGAGGTTTCCGTAGGTGAATCTGCGGGAGGATCATTGTTGAGATCACAT...TGA')\n",
      "731\n",
      "Z78485.1\n",
      "Seq('CTGAACCTGGTGTCCGAAGGTGAATCTGCGGATGGATCATTGTTGAGATATCAT...GTA')\n",
      "735\n",
      "Z78484.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGGGGAAGGATCATTGTTGAGATCACAT...TTT')\n",
      "720\n",
      "Z78483.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...GCA')\n",
      "740\n",
      "Z78482.1\n",
      "Seq('TCTACTGCAGTGACCGAGATTTGCCATCGAGCCTCCTGGGAGCTTTCTTGCTGG...GCA')\n",
      "629\n",
      "Z78481.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TGA')\n",
      "572\n",
      "Z78480.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TGA')\n",
      "587\n",
      "Z78479.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AGT')\n",
      "700\n",
      "Z78478.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTCCGGAAGGATCAGTGTTGAGATCACAT...GGC')\n",
      "636\n",
      "Z78477.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TGC')\n",
      "716\n",
      "Z78476.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...CCC')\n",
      "592\n",
      "Z78475.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...GGT')\n",
      "716\n",
      "Z78474.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACGT...CTT')\n",
      "733\n",
      "Z78473.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AGG')\n",
      "626\n",
      "Z78472.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AGC')\n",
      "737\n",
      "Z78471.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AGC')\n",
      "740\n",
      "Z78470.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...GTT')\n",
      "574\n",
      "Z78469.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...GTT')\n",
      "594\n",
      "Z78468.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCGCAT...GTT')\n",
      "610\n",
      "Z78467.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TGA')\n",
      "730\n",
      "Z78466.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...CCC')\n",
      "641\n",
      "Z78465.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TGC')\n",
      "702\n",
      "Z78464.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAGCGGAAGGGTCATTGTTGAGATCACATAATA...AGC')\n",
      "733\n",
      "Z78463.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGTTCATTGTTGAGATCACAT...AGC')\n",
      "738\n",
      "Z78462.1\n",
      "Seq('CGTCACGAGGTCTCCGGATGTGACCCTGCGGAAGGATCATTGTTGAGATCACAT...CAT')\n",
      "736\n",
      "Z78461.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTCCGGAAGGATCATTGTTGAGATCACAT...TAA')\n",
      "732\n",
      "Z78460.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTCCGGAAGGATCATTGTTGAGATCACAT...TTA')\n",
      "745\n",
      "Z78459.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TTT')\n",
      "744\n",
      "Z78458.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TTG')\n",
      "738\n",
      "Z78457.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTCCGGAAGGATCATTGTTGAGATCACAT...GAG')\n",
      "739\n",
      "Z78456.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AGC')\n",
      "740\n",
      "Z78455.1\n",
      "Seq('CGTAACCAGGTTTCCGTAGGTGGACCTTCGGGAGGATCATTTTTGAGATCACAT...GCA')\n",
      "745\n",
      "Z78454.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AAC')\n",
      "695\n",
      "Z78453.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...GCA')\n",
      "745\n",
      "Z78452.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...GCA')\n",
      "743\n",
      "Z78451.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGTACCTCCGGAAGGATCATTGTTGAGATCACAT...AGC')\n",
      "730\n",
      "Z78450.1\n",
      "Seq('GGAAGGATCATTGCTGATATCACATAATAATTGATCGAGTTAAGCTGGAGGATC...GAG')\n",
      "706\n",
      "Z78449.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TGC')\n",
      "744\n",
      "Z78448.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AGG')\n",
      "742\n",
      "Z78447.1\n",
      "Seq('CGTAACAAGGATTCCGTAGGTGAACCTGCGGGAGGATCATTGTTGAGATCACAT...AGC')\n",
      "694\n",
      "Z78446.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTCCGGAAGGATCATTGTTGAGATCACAT...CCC')\n",
      "712\n",
      "Z78445.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TGT')\n",
      "715\n",
      "Z78444.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGGTGAACTGCGGAAGGATCATTGTTGAGATCACAT...ATT')\n",
      "688\n",
      "Z78443.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AGG')\n",
      "784\n",
      "Z78442.1\n",
      "Seq('GTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACATAATAATTGATCGAGT...AGT')\n",
      "721\n",
      "Z78441.1\n",
      "Seq('GGAAGGTCATTGCCGATATCACATAATAATTGATCGAGTTAATCTGGAGGATCT...GAG')\n",
      "703\n",
      "Z78440.1\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGGACCTCCGGGAGGATCATTGTTGAGATCACAT...GCA')\n",
      "744\n",
      "Z78439.1\n",
      "Seq('CATTGTTGAGATCACATAATAATTGATCGAGTTAATCTGGAGGATCTGTTTACT...GCC')\n",
      "592\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "for registro_seq in SeqIO.parse('ls_orchid.gbk', 'genbank'):\n",
    "    print(registro_seq.id)\n",
    "    print(repr(registro_seq.seq))\n",
    "    print(len(registro_seq))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "De modo silimar, se você quiser ler um outro formato de arquivo, assumindo que seja suportado pela função, basta mudar o segundo argumento, por exemplo: 'swiss' para arquivos SwissProt ou 'embl' para arquivos de texto EMBL. Na wiki do Biopython (https://biopython.org/wiki/SeqIO) há uma lista completa com os formatos de arquivos suportados, bem como a documentação *built in* (https://biopython.org/docs/1.79/api/Bio.SeqIO.html).\n",
    "\n",
    "Outra maneira de usar os iteradores em Python é com *list comprehension*. Por exemplo, se você quiser extrair uma lista de identificadores de um arquivo, você pode facilmente fazer isso com *list comprehension*:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Z78533.1',\n",
       " 'Z78532.1',\n",
       " 'Z78531.1',\n",
       " 'Z78530.1',\n",
       " 'Z78529.1',\n",
       " 'Z78527.1',\n",
       " 'Z78526.1',\n",
       " 'Z78525.1',\n",
       " 'Z78524.1',\n",
       " 'Z78523.1',\n",
       " 'Z78522.1',\n",
       " 'Z78521.1',\n",
       " 'Z78520.1',\n",
       " 'Z78519.1',\n",
       " 'Z78518.1',\n",
       " 'Z78517.1',\n",
       " 'Z78516.1',\n",
       " 'Z78515.1',\n",
       " 'Z78514.1',\n",
       " 'Z78513.1',\n",
       " 'Z78512.1',\n",
       " 'Z78511.1',\n",
       " 'Z78510.1',\n",
       " 'Z78509.1',\n",
       " 'Z78508.1',\n",
       " 'Z78507.1',\n",
       " 'Z78506.1',\n",
       " 'Z78505.1',\n",
       " 'Z78504.1',\n",
       " 'Z78503.1',\n",
       " 'Z78502.1',\n",
       " 'Z78501.1',\n",
       " 'Z78500.1',\n",
       " 'Z78499.1',\n",
       " 'Z78498.1',\n",
       " 'Z78497.1',\n",
       " 'Z78496.1',\n",
       " 'Z78495.1',\n",
       " 'Z78494.1',\n",
       " 'Z78493.1',\n",
       " 'Z78492.1',\n",
       " 'Z78491.1',\n",
       " 'Z78490.1',\n",
       " 'Z78489.1',\n",
       " 'Z78488.1',\n",
       " 'Z78487.1',\n",
       " 'Z78486.1',\n",
       " 'Z78485.1',\n",
       " 'Z78484.1',\n",
       " 'Z78483.1',\n",
       " 'Z78482.1',\n",
       " 'Z78481.1',\n",
       " 'Z78480.1',\n",
       " 'Z78479.1',\n",
       " 'Z78478.1',\n",
       " 'Z78477.1',\n",
       " 'Z78476.1',\n",
       " 'Z78475.1',\n",
       " 'Z78474.1',\n",
       " 'Z78473.1',\n",
       " 'Z78472.1',\n",
       " 'Z78471.1',\n",
       " 'Z78470.1',\n",
       " 'Z78469.1',\n",
       " 'Z78468.1',\n",
       " 'Z78467.1',\n",
       " 'Z78466.1',\n",
       " 'Z78465.1',\n",
       " 'Z78464.1',\n",
       " 'Z78463.1',\n",
       " 'Z78462.1',\n",
       " 'Z78461.1',\n",
       " 'Z78460.1',\n",
       " 'Z78459.1',\n",
       " 'Z78458.1',\n",
       " 'Z78457.1',\n",
       " 'Z78456.1',\n",
       " 'Z78455.1',\n",
       " 'Z78454.1',\n",
       " 'Z78453.1',\n",
       " 'Z78452.1',\n",
       " 'Z78451.1',\n",
       " 'Z78450.1',\n",
       " 'Z78449.1',\n",
       " 'Z78448.1',\n",
       " 'Z78447.1',\n",
       " 'Z78446.1',\n",
       " 'Z78445.1',\n",
       " 'Z78444.1',\n",
       " 'Z78443.1',\n",
       " 'Z78442.1',\n",
       " 'Z78441.1',\n",
       " 'Z78440.1',\n",
       " 'Z78439.1']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "identificadores = [registro_seq.id for registro_seq in SeqIO.parse('ls_orchid.gbk', 'genbank')]\n",
    "identificadores"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Há mais exemplos usando a função **parse** com *list comprehension* na **Seção 20.2**, como pltar o tamanho das sequências ou o conteúdo GC. (Capítulo 20: ***link do capítulo 20***)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5.1.2 Iterando sobre os registros em um arquivo de sequência"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Nos exemplos acima, nós usamos um loop para iterar sobre os registros um por um. Você pode usar o loop com todos os objetos Python (incluindo listas, tuplas e strings) que suporte iteração.\n",
    "\n",
    "O objeto retornado pelo `Bio.SeqIO` é atualmente um interador que retorna objetos **SeqRecord**. Você consegue ver cada registro, mas somente uma vez. O ponto positivo disso é qe você salva grandes quantidades de memória quando trabalha com arquivos grandes.\n",
    "\n",
    "Ao invés de usar um loop, você pode usar a função **next** em um iterador para percorrer as entradas desse modo:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gi|2765658|emb|Z78533.1|CIZ78533\n",
      "gi|2765658|emb|Z78533.1|CIZ78533 C.irapeanum 5.8S rRNA gene and ITS1 and ITS2 DNA\n",
      "gi|2765657|emb|Z78532.1|CCZ78532\n",
      "gi|2765657|emb|Z78532.1|CCZ78532 C.californicum 5.8S rRNA gene and ITS1 and ITS2 DNA\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "iterador_registro = SeqIO.parse('ls_orchid.fasta', 'fasta')\n",
    "\n",
    "primeiro_registro = next(iterador_registro)\n",
    "print(primeiro_registro.id)\n",
    "print(primeiro_registro.description)\n",
    "\n",
    "segundo_registro = next(iterador_registro)\n",
    "print(segundo_registro.id)\n",
    "print(segundo_registro.description)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note que se você usar a função **next** e não obtiver mais resultados, uma exceção é retornada (**StopIteration**).\n",
    "\n",
    "Mais um caso especial a considerar é quando o seu arquivo possui vários registros, mas você quer somente o primeiro. Nessa situação, o código é bem conciso:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGATGAGACCGTGG...CGC'), id='Z78533.1', name='Z78533', description='C.irapeanum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[])"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "primeiro_registro = next(SeqIO.parse('ls_orchid.gbk', 'genbank'))\n",
    "primeiro_registro"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Um pequeno aviso. Usar a função **next** como mostrado, ignorará reistros adicionais do arquivo. Se o seu arquivo possuir somente um registro, como arquivos GenBank com um único cromossomo, é aconselhável que seja usada a função **read**. Isso verificará se não há registros extras inesperados presentes."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5.1.3 Obtendo uma lista de registros em um arquivo de sequência"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Na seção anterior, nós falamos sobre o fato da função **parse** retornar um iterador **SeqRecord** que mostra os registros um a um. Muitas vezes, você precisa ser capaz de acessar os registros em qualquer ordem. Uma lista Python é perfeita para isso. Pode os transformar um registro iterado em ma lista de objetos **SeqRecord** usando a função **list** nativa do Python:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Encontrado 94 registros\n",
      "Último registro:\n",
      "Seq('CATTGTTGAGATCACATAATAATTGATCGAGTTAATCTGGAGGATCTGTTTACT...GCC')\n",
      "592\n",
      "Primeiro registro:\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGATGAGACCGTGG...CGC')\n",
      "740\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "registros = list(SeqIO.parse('ls_orchid.gbk', 'genbank'))\n",
    "\n",
    "print(f'Encontrado {len(registros)} registros')\n",
    "\n",
    "print('Último registro:')\n",
    "ultimo_registro = registros[-1]  # usamos alguns truques de lista\n",
    "print(repr(ultimo_registro.seq))\n",
    "print(len(ultimo_registro))\n",
    "\n",
    "print('Primeiro registro:')\n",
    "primeiro_registro = registros[0]  # lembre-se que o Python começa a contar do zero\n",
    "print(repr(primeiro_registro.seq))\n",
    "print(len(primeiro_registro))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "É óbvio que você pode continuar usando um loop para os objetos **SeqRecord**, mas usar uma lista é mais flexível do que um iterador (por exemplo, você pode determinar o número de registros o comprimento da lista), mas você precisa de mais memória pois manterá todos os registros na memória de uma só vez."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5.1.4 Extraindo dados"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "O objeto **SeqRecord** e suas estruturas de anotação foram descritas em detalhes no Capítulo 4 (https://github.com/GTL98/biopython_br/tree/main/Cap%204%20-%20Objetos%20de%20anota%C3%A7%C3%A3o%20de%20sequ%C3%AAncia). Como um exemplo de como as anotações são armazenadas, camos olhar a saída da análise do primeiro registor de um arquivo GenBank:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: Z78533.1\n",
      "Name: Z78533\n",
      "Description: C.irapeanum 5.8S rRNA gene and ITS1 and ITS2 DNA\n",
      "Number of features: 5\n",
      "/molecule_type=DNA\n",
      "/topology=linear\n",
      "/data_file_division=PLN\n",
      "/date=30-NOV-2006\n",
      "/accessions=['Z78533']\n",
      "/sequence_version=1\n",
      "/gi=2765658\n",
      "/keywords=['5.8S ribosomal RNA', '5.8S rRNA gene', 'internal transcribed spacer', 'ITS1', 'ITS2']\n",
      "/source=Cypripedium irapeanum\n",
      "/organism=Cypripedium irapeanum\n",
      "/taxonomy=['Eukaryota', 'Viridiplantae', 'Streptophyta', 'Embryophyta', 'Tracheophyta', 'Spermatophyta', 'Magnoliophyta', 'Liliopsida', 'Asparagales', 'Orchidaceae', 'Cypripedioideae', 'Cypripedium']\n",
      "/references=[Reference(title='Phylogenetics of the slipper orchids (Cypripedioideae: Orchidaceae): nuclear rDNA ITS sequences', ...), Reference(title='Direct Submission', ...)]\n",
      "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGATGAGACCGTGG...CGC')\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "iterador_registro = SeqIO.parse('ls_orchid.gbk', 'genbank')\n",
    "primeiro_registro = next(iterador_registro)\n",
    "print(primeiro_registro)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "O código acima retorna um sumário capaz de um humano ler e compreender o seu conteúdo. Como exemplo, vamos chamar o atributo **annotations**, que é um dicionário Python. As informações do dicionário das anotações são as mesmas que as mostradas acima. Você pode mostrá-las diretamente:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'molecule_type': 'DNA', 'topology': 'linear', 'data_file_division': 'PLN', 'date': '30-NOV-2006', 'accessions': ['Z78533'], 'sequence_version': 1, 'gi': '2765658', 'keywords': ['5.8S ribosomal RNA', '5.8S rRNA gene', 'internal transcribed spacer', 'ITS1', 'ITS2'], 'source': 'Cypripedium irapeanum', 'organism': 'Cypripedium irapeanum', 'taxonomy': ['Eukaryota', 'Viridiplantae', 'Streptophyta', 'Embryophyta', 'Tracheophyta', 'Spermatophyta', 'Magnoliophyta', 'Liliopsida', 'Asparagales', 'Orchidaceae', 'Cypripedioideae', 'Cypripedium'], 'references': [Reference(title='Phylogenetics of the slipper orchids (Cypripedioideae: Orchidaceae): nuclear rDNA ITS sequences', ...), Reference(title='Direct Submission', ...)]}\n"
     ]
    }
   ],
   "source": [
    "print(primeiro_registro.annotations)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Como qualquer dicionário Python, você pode selecionar somente as chaves:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['molecule_type', 'topology', 'data_file_division', 'date', 'accessions', 'sequence_version', 'gi', 'keywords', 'source', 'organism', 'taxonomy', 'references'])\n"
     ]
    }
   ],
   "source": [
    "print(primeiro_registro.annotations.keys())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ou os valores:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_values(['DNA', 'linear', 'PLN', '30-NOV-2006', ['Z78533'], 1, '2765658', ['5.8S ribosomal RNA', '5.8S rRNA gene', 'internal transcribed spacer', 'ITS1', 'ITS2'], 'Cypripedium irapeanum', 'Cypripedium irapeanum', ['Eukaryota', 'Viridiplantae', 'Streptophyta', 'Embryophyta', 'Tracheophyta', 'Spermatophyta', 'Magnoliophyta', 'Liliopsida', 'Asparagales', 'Orchidaceae', 'Cypripedioideae', 'Cypripedium'], [Reference(title='Phylogenetics of the slipper orchids (Cypripedioideae: Orchidaceae): nuclear rDNA ITS sequences', ...), Reference(title='Direct Submission', ...)]])\n"
     ]
    }
   ],
   "source": [
    "print(primeiro_registro.annotations.values())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "De modo geral, os valores de **annotations** são strings ou uma lista de strings. Um caso especial ocorre quando uma referência no arquivo é armazenada em algum objeto de referência.\n",
    "\n",
    "Suponha que você queira extrair uma lista das espécies presentes no arquivo *ls_orchid.gbk*. A informação que você deseja, *Cypripedium irapeanum*, está armazenada nas chaves **source** e **organism** do dicionário **annotations**:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cypripedium irapeanum\n"
     ]
    }
   ],
   "source": [
    "print(primeiro_registro.annotations['source'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ou:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cypripedium irapeanum\n"
     ]
    }
   ],
   "source": [
    "print(primeiro_registro.annotations['organism'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Geralmente, a chave **organism** é usado para o nome científico (em Latim, e.g. *Aedes aegypti*) enquanto a chave **source** é usado para o nome comum (e.g. Mosquito-da-dengue). Nesse exemplo acima, ambas as chaves possuem o mesmo valor.\n",
    "\n",
    "Agora, vamos passar por todos os registros, construindo uma lista das espécies de cada sequência de orquídea:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Cypripedium irapeanum', 'Cypripedium californicum', 'Cypripedium fasciculatum', 'Cypripedium margaritaceum', 'Cypripedium lichiangense', 'Cypripedium yatabeanum', 'Cypripedium guttatum', 'Cypripedium acaule', 'Cypripedium formosanum', 'Cypripedium himalaicum', 'Cypripedium macranthon', 'Cypripedium calceolus', 'Cypripedium segawai', 'Cypripedium parviflorum var. pubescens', 'Cypripedium reginae', 'Cypripedium flavum', 'Cypripedium passerinum', 'Mexipedium xerophyticum', 'Phragmipedium schlimii', 'Phragmipedium besseae', 'Phragmipedium wallisii', 'Phragmipedium exstaminodium', 'Phragmipedium caricinum', 'Phragmipedium pearcei', 'Phragmipedium longifolium', 'Phragmipedium lindenii', 'Phragmipedium lindleyanum', 'Phragmipedium sargentianum', 'Phragmipedium kaiteurum', 'Phragmipedium czerwiakowianum', 'Phragmipedium boissierianum', 'Phragmipedium caudatum', 'Phragmipedium warszewiczianum', 'Paphiopedilum micranthum', 'Paphiopedilum malipoense', 'Paphiopedilum delenatii', 'Paphiopedilum armeniacum', 'Paphiopedilum emersonii', 'Paphiopedilum niveum', 'Paphiopedilum godefroyae', 'Paphiopedilum bellatulum', 'Paphiopedilum concolor', 'Paphiopedilum fairrieanum', 'Paphiopedilum druryi', 'Paphiopedilum tigrinum', 'Paphiopedilum hirsutissimum', 'Paphiopedilum barbigerum', 'Paphiopedilum henryanum', 'Paphiopedilum charlesworthii', 'Paphiopedilum villosum', 'Paphiopedilum exul', 'Paphiopedilum insigne', 'Paphiopedilum gratrixianum', 'Paphiopedilum primulinum', 'Paphiopedilum victoria', 'Paphiopedilum victoria', 'Paphiopedilum glaucophyllum', 'Paphiopedilum supardii', 'Paphiopedilum kolopakingii', 'Paphiopedilum sanderianum', 'Paphiopedilum lowii', 'Paphiopedilum dianthum', 'Paphiopedilum parishii', 'Paphiopedilum haynaldianum', 'Paphiopedilum adductum', 'Paphiopedilum stonei', 'Paphiopedilum philippinense', 'Paphiopedilum rothschildianum', 'Paphiopedilum glanduliferum', 'Paphiopedilum glanduliferum', 'Paphiopedilum sukhakulii', 'Paphiopedilum wardii', 'Paphiopedilum ciliolare', 'Paphiopedilum dayanum', 'Paphiopedilum hennisianum', 'Paphiopedilum callosum', 'Paphiopedilum tonsum', 'Paphiopedilum javanicum', 'Paphiopedilum fowliei', 'Paphiopedilum schoseri', 'Paphiopedilum bougainvilleanum', 'Paphiopedilum hookerae', 'Paphiopedilum papuanum', 'Paphiopedilum mastersianum', 'Paphiopedilum argus', 'Paphiopedilum venustum', 'Paphiopedilum acmodontum', 'Paphiopedilum urbanianum', 'Paphiopedilum appletonianum', 'Paphiopedilum lawrenceanum', 'Paphiopedilum bullenianum', 'Paphiopedilum superbiens', 'Paphiopedilum purpuratum', 'Paphiopedilum barbatum']\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "todas_especies = []\n",
    "for registro_seq in SeqIO.parse('ls_orchid.gbk', 'genbank'):\n",
    "    todas_especies.append(registro_seq.annotations['organism'])\n",
    "print(todas_especies)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Outra forma de escrever o mesmo código é usando a *list comprehension*:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Cypripedium irapeanum', 'Cypripedium californicum', 'Cypripedium fasciculatum', 'Cypripedium margaritaceum', 'Cypripedium lichiangense', 'Cypripedium yatabeanum', 'Cypripedium guttatum', 'Cypripedium acaule', 'Cypripedium formosanum', 'Cypripedium himalaicum', 'Cypripedium macranthon', 'Cypripedium calceolus', 'Cypripedium segawai', 'Cypripedium parviflorum var. pubescens', 'Cypripedium reginae', 'Cypripedium flavum', 'Cypripedium passerinum', 'Mexipedium xerophyticum', 'Phragmipedium schlimii', 'Phragmipedium besseae', 'Phragmipedium wallisii', 'Phragmipedium exstaminodium', 'Phragmipedium caricinum', 'Phragmipedium pearcei', 'Phragmipedium longifolium', 'Phragmipedium lindenii', 'Phragmipedium lindleyanum', 'Phragmipedium sargentianum', 'Phragmipedium kaiteurum', 'Phragmipedium czerwiakowianum', 'Phragmipedium boissierianum', 'Phragmipedium caudatum', 'Phragmipedium warszewiczianum', 'Paphiopedilum micranthum', 'Paphiopedilum malipoense', 'Paphiopedilum delenatii', 'Paphiopedilum armeniacum', 'Paphiopedilum emersonii', 'Paphiopedilum niveum', 'Paphiopedilum godefroyae', 'Paphiopedilum bellatulum', 'Paphiopedilum concolor', 'Paphiopedilum fairrieanum', 'Paphiopedilum druryi', 'Paphiopedilum tigrinum', 'Paphiopedilum hirsutissimum', 'Paphiopedilum barbigerum', 'Paphiopedilum henryanum', 'Paphiopedilum charlesworthii', 'Paphiopedilum villosum', 'Paphiopedilum exul', 'Paphiopedilum insigne', 'Paphiopedilum gratrixianum', 'Paphiopedilum primulinum', 'Paphiopedilum victoria', 'Paphiopedilum victoria', 'Paphiopedilum glaucophyllum', 'Paphiopedilum supardii', 'Paphiopedilum kolopakingii', 'Paphiopedilum sanderianum', 'Paphiopedilum lowii', 'Paphiopedilum dianthum', 'Paphiopedilum parishii', 'Paphiopedilum haynaldianum', 'Paphiopedilum adductum', 'Paphiopedilum stonei', 'Paphiopedilum philippinense', 'Paphiopedilum rothschildianum', 'Paphiopedilum glanduliferum', 'Paphiopedilum glanduliferum', 'Paphiopedilum sukhakulii', 'Paphiopedilum wardii', 'Paphiopedilum ciliolare', 'Paphiopedilum dayanum', 'Paphiopedilum hennisianum', 'Paphiopedilum callosum', 'Paphiopedilum tonsum', 'Paphiopedilum javanicum', 'Paphiopedilum fowliei', 'Paphiopedilum schoseri', 'Paphiopedilum bougainvilleanum', 'Paphiopedilum hookerae', 'Paphiopedilum papuanum', 'Paphiopedilum mastersianum', 'Paphiopedilum argus', 'Paphiopedilum venustum', 'Paphiopedilum acmodontum', 'Paphiopedilum urbanianum', 'Paphiopedilum appletonianum', 'Paphiopedilum lawrenceanum', 'Paphiopedilum bullenianum', 'Paphiopedilum superbiens', 'Paphiopedilum purpuratum', 'Paphiopedilum barbatum']\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "todas_especies = [\n",
    "    registro_seq.annotations['organism']\n",
    "    for registro_seq in SeqIO.parse('ls_orchid.gbk', 'genbank')\n",
    "]\n",
    "print(todas_especies)"
   ]
  },
  {
   "attachments": {
    "image-2.png": {
     "image/png": "iVBORw0KGgoAAAANSUhEUgAABNgAAAB6CAYAAABtLe6KAAAgAElEQVR4Ae2ajW4cS66k/f4vPQteTMx+DpOZzKrqltRiAhf8iwiywpbPrLB//jNvHBgHxoFxYBwYB8aBcWAcGAfGgXFgHBgHxoFxYBwYBy478Ocy85sT//z58x/+3zc/d847dIB/tpF/+vtt3/vpf55Pft/83XjSzdEaB8aBcWAcGAfGgXFgHBgHxoFx4JoDH/ubifl/dF77C/FTWL/tz/e3fe9P+Xv4He6cvxvf4U9hbhgHxoFxYBwYB8aBcWAcGAfGgd/uwPyC7bf/Dfih3//bfqnw2773h/61/JKz5+/Gl9g+S8eBcWAcGAfGgXFgHBgHxoFxYBz4y4Fv/ws2/T8e/7q6UYin2KAM5Ac5oD9XxR90+qVT9Z2Kl0SG9JEO6O+E4kd+5HzUr3Jg/i7/qj/u+dhxYBwYB8aBcWAcGAc+xoFv/Qs2/Y9sxRPXxVE84Q72+zugP1fF73ixblO8c6M0FO9oDfezHNDfCcXP+rr5mt/mgP4eK/6275/vHQfGgXFgHHidA/pvi+LrNo3yT3NAfycUf9r9c+/3cWB+wWZ/FvqhYjTI/0piOvn/iIuk0nFKhcv6zs3qjBc9fxUu6ztXdYZVT5hdFF5xh/+KuW5TvHODNBQrLc27kTpdjnDkMtc8i8Qxz7BVjzzlFTb6u3eHu9PuzLP9HZ4wzle/E52ruuJqvovk77A+f4rb1SGOud+1qslTfooXL+IdLnXu5NUNdzQ7XN/b4QxmHHAH/O/RqnbuE/WdfXe4T9x+orG6VbNMTzPGDq6DCc1540DlAP/OPf13xbWrmrdVmF2fGpGv8I7N6oyf4T655x48+a2uXdXcWWF2fWpEvsI7Nqszfoa728v2eK/a4bios+e4DqbSyrjq5ds1/QZRRpyeIp7iji/cKrrGCpvNnK86w3pPWEWfr2pxPK44ml3hvIK7u8Pn36GWD4p3bpKGYqWleTdSp8sRjtzI1e/Er+D6zpObM+7d3sqnE23X6XCdk9WZTobLeuRm81XvKW7orPb4jHvfxfWdJ3sz7hM99yWrn9iz0tDOFWZm48DKAf0d6sSVzumss08Y11Z/F533lfXuVs55J/vMiYmcs8iz55gKl3Gn9/sc8L8vTzrg2lXNnRVm15fGDqe58B41z6JjP712D578Xteuau6sMLu+NHY4zYX3qPkqOudOvdrjM9/jc9U7nM+jFpcxw616+X+twAjxn/hoyu4bHFvV7kOFq/rOj7rCet+5Pl/Vzv2Kvav7qll2t3rOUf87xSdv7Go5blfTrx3W53e4ocXn2quavMhXWJ89yXWtk9rv8vqO1o7ru6o606mw3ifXZ7v6q7hxF9/uTs7Ji5yzXf4k17VO692tmp/qDn4ceLcD+rvaiU/d1tnlGO72WVWT89V5dWPV1727eYVTn1Fa0VMecd44kDnAvyNP/z1x7armXRVm15fGDqe58IyaVZHY35C7D09+s2tXNXdWmF1fGjuc5sIzaraL5NzNd7t8zn0+U01M5Oor+pwY5oE/eUu0lp+KnhzwKixv79wvPO9Rj5HzXU5e5P58nmGco9q56neic9+x13dmd3Yw4p1gxXl3fPLGp7Soc+oHuZHzrWbCrTCrmfhV9FuEc80Ml/WC3+Fqz5VI/WzfiSa1qu+RnmMrfNWXzt2bKx32O7l/DzmrmXArzGomfhWDmz3XzHBZL7Q63Gxnt9fVr+7r7hncOPAOB/zv83fZWd3l/ernrOq/4/uyHX43MT7j7dmMc+k4Tn1FzqPntXATxwE5wL8jkb/q3dnT5Tqu+p5VXxrhg3LFV3nzXXX13YqvulP6iid7xFGsuJozZtiYV09cztVj5PxOTs3I/fmcmGzGubQcp74i59HzWrhd/Pd6MCga+elzflVLt5qrL1wniqO44wQue+IrZpiqJ86rtSv9zl3v4tKLamcHo286wYoT0XmqiVGumaLzM5x6jg0NPekpqr+KwiqusNVM3IhX3orPWaW/wqxmV24V547uHa72r2Lo693ZdcI9weq2Kj6hRY1qz6q/4nMWefZWmNUs0+r27uje4e7ue6X2arfv9fqEK2xXw3FeS89jhav6zlfteNWaZ1GYKmac6Dk+6wUmexmXuGqe9dmTRtbT7GqkZuTveJ2dFabqf+XdnZt2mGrufdb8ZvYj98d5zLx2/JM1d63ybGeFz7DRc3zWC8zuuY7qHe/qXPpVrHQdHzjvRb17Gcd7O42r8zt7utwubvUNoaH3hJ60dtF3eV3xM5z3ot69jOO9ncbV+Z09XW4Xt/qG0MjeE9pXdavd3mfNXexH7o/zmHnt+Kr+V9mQFM4OMfj/lc7Z1dLo4oRfRddaYVezqzo73m6+uilmV/lXebrnKn/H83nUq+f4FTZmjs9q18gw7FW60iG2k4uXRednmFXv1fydvs+j5vM5Z3fyO7p3uKc339nV5XZx3dvv6r2av9P3edR8PufsTn5H9w53dfOrdFc7NfPdXguXxQzrPdXOV38XnRf1juNz1/B5VjvnZO8dbtziz+/rzp3ndeh4L+onnus+obnT2O30Ob/VZ7tdT86z3d5T7XvVV+zOhc8iNXzOWeScex2zVzzu7OS84RQvbocnjDiMmq0i8U/kq12cZbs43+Wv4Geapz2/+4Tf5XZx3d1P61V7fU9VZ/wKm/Vfwc80T3t+6wm/y+3iTnYL+yrtrm6G8x5r3R2R/cj9cd7BO1/1v8qaIHJZdgyg/5dWeO+7VjZnz/esavJ8z4rns6s65Llm1JzrPu+p3+F3uY7LbnnF3tUev2m1X144R/0sOnZVk7/CrWbSWGGymXhZdHyGWfXIX+GqWYdPjPLQU87oezircufsatfZ4Tm/w6VON7+zr8vt4t5xc+zgPd2dxHX4xCj33exX+sJ4JL6T3+Hf4e5ue6X26e6TWxy7q3nLDss5eZFz1snJ7+CFIe907x1u7OfTPYqcRa6+ouaqT6P4d2Jn5x39ipvtDWzVl042j947XrW76vMmx9yZUUs67EXuT3P21cvwxF3Nqc8d3ucsdmXzque3VbiqT36Fyfrk3c0z/arnuypc1Se/wlR9cp/Mfd+JdpfrONUnu4gVX5GzJ3Ppd6Lv7XCIIZ/9Tk7uk7nvPtHuch2n+mRXhZWWYoU77UtPseJrrhg45Yrek5bmiuorZn31InZfG0lx5dkSzRQdo76iz1kLo8jZLhdHcYfP5uIqZpisJ7xiByNsFV2jwmV9crP5qkdu5Cusz5zb5Wc873V2ibPC+ixqPZ9Fv9vLsNTuzHVHhuVsl/PmHTabk+/f4HjHZrVzos5wVS/j7zQqjvrVrui/4/n+k51dbhfX3X1Hj9zuPuLIj3z1HJvVGT/DVb2MH70KH/3du8Pdafvcd/n8nfXJLY5VrXtVK6ofMXrVE17RceozEsO+cs1VK6ofUT1Gn7NmTk7k/nzuGJ+Tv5oFrppn/W6P+6/kvmdVX9FfcVa7NMv4mlUx4zzR2+3zOXf6bFWveDEjV1j2IufjrNMn5mrOnX5PaK7mJzPXdu5uzu9z7mrmusSe5iut1U2xx+eu5XPetppl2uQ+me/uWO064TrW69Uen93hutaqjj3V293gc9fyOfesZoHbzal1J7+z54TrWK+vfMMTGtXernaG2/W0M8NlM/UiksP+Kq//hhcsLonc39059XZaxHp+hyutqxrkScsjMZ38Dp/czi5iyI2cs13uXNVXeeJnd3DG3HdxprzCZP1uL7QzrHYqdjBdLWkydvXJ8ZwaPstq4rP8Coc6GT96xGR5xbvLXel2Z35vl5fdXnHv7Mg0r+pd5fEGarBf5cRnecbLcFUv40evwqtf8e5yV7rZTPcoZph39XSD4mqvMIzEsx959+14u3nsyTBZz2/qYJxT7SPOdaPm83l3FriKm/W7Pe6/kvueVX1Ff8dZ7YtZ9Xa8FbfS3PWzneT4fDVzbHWv46TJfvRYuxZn4u84xF3JudPvWe3e8VbcbOa7K/2qz2/vYIh/It/t9HnUfD7XrOprHrGDIf5qfmfPKdfxWd35Dud1OE9jdjf4PGo+n2tW9TWP2MEQfzW/s+eU6/isPvkO559wd9iudobLerGPfa9jxudYzdh3jjAe/1b2aVGvFq1mIbebc+UJlrzTPc69w+/e7Lio+XZzYj1fcVez0NnNfRfrDjfDZD3qZrlzMkz0HNeppeXYTK/qrfrS72JOcNR2ns86debBiuf4ql5p+CzTcEzUGc57Ge8ut9I86XfvzDS73C4u25H1ruqRl+nueuRHvnuOr+qdDueZBufKM5z3hPXouKx2ztXata/qPME7ueUEW93mGlntXMf4POoM471O3dHOdJznmJP5VW7G6/b8vidrvyHqJ1+mn/VWOzO8eivelZl0FU80xFnFTM/xwrAfPdaR83HGvvN8dqfmTr/H93LuvE7NOx3P2Sv3+p4rtd+e1a7rmO58xwudDsb3Xanv7LnKdR7rzjcQH/mrn+/Lar/BMd35jhc6HYzvu1Lf2XOV6zzW3W8gJ/KnX1c/w2W9uI99r2PG59jujDjlfyuru4hc7ocFzefErGbZSsdnmKr3HbjVbdHv3NfBVDsqbtWnTgdDPPMV92QW2NVbaZHnuE4tvmOj3+1VWGkrZnqaMXZxFYf9k5x7dzxiI+fzmc+JzXLnZ5is57yTvXe42S2rnu9aYX3W5XZxrl/VV/TIqXR3/RMNYiPn85nPic1y52eYrOe8k713uNkt6rmu+l8RT245wfq3OHdV77g+j9r1sp5jspra2bzqkZftPpn7ji4343V7vuPpOrvjiR0rXZ9F3XlXeR3twLh+l7fiUjPT4zxyPfaf0pf2E5H3KZeuasbVjLgsFzeizzlbzZ3XqV37at3ZJYzvUF+xOxde0XlRa6aYYZ7oSV/xRFMcxROusOIyalZFYiN/1fM9q9pvcGx3vuOFTgfj+67Ud/bc4epW14h695yzw1+Zd3dkuKwXN7Dvdcz0HKe+4m4unOL/V1aniBRWXkD/+hhhs1jx1XeO+p14lXuVFzedcDvYDqbyouJWfep0MMQzr7hVv8MlRnlHL7BdnHQZM263192d6fEG5V3cVbx4jCc7O9gOhvuZ/0Qu79/l7/i+Ozuy+0/1TvF3d3b2dTDZHdH7idxXfEulebV/4usJlvc4L2o+n3MW+W5eYTo836XauVHz+ZyzyO/Mr3IzXrfn9z9dZ3fc3dHR7GCyO67yMi3v3dGuuFVfu1dzn7Gu+MRkuXhPxEw/63GXzznr5Dt+Na/6nZ13ML43aj6fcxb51fmO19H2W67WnVsq7Ttcap7qnOK5q5v7jqj5fM5Z5FfnO15H22+5WnduqbTvcKl5onOC5Y7TvLsnw2U97fcZ6w6GeOXiVfHvv9UFSmKKBez/2sKs4orPmWtwtsuvcK9weAf57Gc5sZFnr4PJeNGruFWfOh0M8cwrbtXvcIlR3tELbBcnXcaM2+11d2d6vEF5F5fh1TuN3LnjEht59jqYjBe9n8DljdV3VH1yIz95J9wT7O6GUy3id9rV/ESD2Miz18FkvOj9RG71LXe/x3Xpjc92NbmRr94Jljo73t157Mo0sh7vWuU77ivnK22fRa3ns+h3e9J4VczuqHYRW2GiT1zk2etgnuRlWt67elPorLicnewkz3PpeH9Xi3c37vbEPHvOyzCr3o5fzav+atcTs93eV813uvFtHcw7PFjteOrGU51T/Oobqtlux6vmO924t4Opvuukf2fPHS5v7Op0cdS+mnd2VZiqH7f4jLVuZa+Ti1fF/L8EQPsSjNL0FJ+K/Ld5R+uUe4r3u6/wVxyfRd19znXeau6zp/a6rt8UdQcjXhfruKir5zPnVjdmuArL3RWPGOVXseKfxpN9od3BdzDZnTtezKvnXMd6TZ0d9yqWPOW+S/1OPOE6NursVX1iXYszz4n1WbemxlP3nWrq1h1vdZ9zHeu1dkbccYk9zbvaq/ueuNHvWH3HCZY6K57PovbnmO7ceZm2tHzmXOEi+sy5GYb83dz1yfUZd/us2pPhuOPp/GSfY6OunmMzXIVRv8NZ3ZDxdz3tVtzhORdH8YlZaEjPo/S9v6vFuxt9T1fPeVFXL5s537nV3PuZtrRWM2G60feS57Nsr2PIj7yae9+1d3Pfc6f2XSdaXa5wmbZmjBmOPWIjf8Vb7fBZdoNj/MZq7n3X3s19z53ad51odbnCZdqaMXZwGebJHu+J3J/PifFZh7viu57Xru/1v9cDcSoWVOesaqz6v3SF9ZlzvT7BO3ZX+66oycnmWY+cXe78HZ7z78DlPZ3cb/baNXzO2rGresWLmXOrXtZ3rtfc7XkX67hd7XtUk6feKhLfyV2rwxHmu3B5h25j5Nxz4jq581k7n7Msd3xVP8Wt9Kt+tjd6xFcY9onv5OT6vh3/u3D9jtN6952ar3SFYTzFk8vcdTiLvPuct6tdd4f3Ofk+W9VXeaHpz/eczJ27q6XtuOh3e9K4Gn3Pql7tyHgVPsOuetRZ4XxG3hP5Hf0dt5pXfX2Pz1XHXHnE1SNuh13pcOaaq5q8yFdYn+24J3PXXtWue7Ve7chmvscxJ3Pn7mrXvlPvdmm+2yGcYoXXvBMzjQ6PmEzjtEe9Tu76zjmZO3dXu/aderdL890O4RQrvOadmGl0eMRkGqc96nVy6jues8h9rtpnzmMtjiJnni//6ySBiN1HTienbgcvDHlZLpxihlFPmG4UT9F56neic7M608lwWS/jRi/Dei/jOqaq73BDc/d87yne+aqpo55izJQrVr2sL04WuTfLnZNhTneGZva6u5zrvFV9leu8qFd7OHuaSz3uUc6558J0o/NZuwZnVe6crM64GS7rkZvNVz1ylTte/V103qp2rRWWM+dFzfkqf5qb6Z32VvdqttIUhvEUTy5z1+Es8pPn3FXtuiusz5wbtWOq2rkVLuvvuHfn2U71pK1aMfrKFaueNK5G6e/iTj/jrzgZvupRp8J4n5yn8js7dtxqXvX1TT5XHXPlEVePuB12pcOZa+5qciPf4TXf8e7Otcej696pXXtV+x7Hnsydu6td+06926X5bodwihVe8128y5d+pXPal14nurZzTubO3dWufafe7dJ8t0M4xQqv+S7e5Uu/0jnpS6sTXdc5u7nwgVMecfWI22JXQlq6w/jcD9jV4u9wnItTRWIjXz3H7mrXIt5n3ZoayldcYbK44vnslJ/h1XPtqhY+ixXH+871eVU7T3WG10wxMMoVq570hKuicLvo/ArvuF2d6ZCTzXc98j2vuI5jXXHYJ95z4rLc8awzfNU74RHbyaud0Xf+CsuZ81gTx5yYVX6FIz1ylWsW8coj3/NKz3GsKw77xHtOXJY7nnWGf7rHfZ53dpGzwxO7y13L8T7f1c6POp73XcfnWe2crM540Vu9jBN477vG3flqR6Wd9bs9v/+09j2s72h1udzneaXhONYV54k+90R+8jrcDJP1fK9joo7HvnO8PsE6t6qp2ckznYqXYaPneMft5sI7TrXmT0fpM8YO1pH7uztf7dhp+y0ntWtX9U7TeSu8Y1mveDEjtpPv9E7m2b7sJtd03ul8tWOn7btOateu6p2m81Z4x7Je8WJGbCff6XXmd/Y4N9vnmKjjsZ/x2Oti//1XjSqHOZfq6Eyii8u43d47dnRvGdzzDvy2P9/f9r3P/435XMX5u/G5f7bzZd/TgfmZ+55/LnPVOPBKB7o/913cK28d7XFgHBgHxoGvc2B+wfZ13s/mGw78tv8B89u+98ZfjV9Hnb8bv+6PfD74ix2Yn7kv/gOY9ePAFzjQ/bnv4r7gE2blODAOjAPjwBscePsv2N71H5537XnDn9GsSBz4bX++v+17kz/yaRUOzN+NwphpjwMvcmB+5l5k7MiOA9/Ygc7PfQfzjT9xThsHxoFxYBx4wIFHf8EW9/h/XFb1A/eXEr63BM7gRzrw2/58f9v3/si/lF909Pzd+CLjZ+2vdWB+5n7tH/18+C93wH/2V/Uvt2o+fxwYB8aBX+vA479gk5Nf/R8d36+7Jn6GA7/tz/e3fe9n/C19z1fM3433+DxbxgE5MD9zcmLiOPA7HfB/A1j/Tkfmq8eBcWAcGAfkwMt+waYFE8eBcWAcGAfGgXFgHBgHxoFxYBwYB8aBcWAcGAfGgU92YH7B9sl/uvNt48A4MA6MA+PAODAOjAPjwDgwDowD48A4MA6MAy93YH7B9nKLZ8E4MA6MA+PAODAOjAPjwDgwDowD48A4MA6MA+PAJzswv2D75D/d+bZxYBwYB8aBcWAcGAfGgXFgHBgHxoFxYBwYB8aBlzswv2B7ucWzYBwYB8aBcWAcGAfGgXFgHBgHxoFxYBwYB8aBceCTHZhfsH3yn+582zgwDowD48A4MA6MA+PAODAOjAPjwDgwDowD48DLHZhfsL3c4lkwDowD48A4MA6MA+PAODAOjAPjwDgwDowD48A48MkOzC/YPvlPd75tHBgHxoFxYBwYB8aBcWAcGAfGgXFgHBgHxoFx4OUOzC/YXm7xLBgHxoFxYBwYB8aBcWAcGAfGgXFgHBgHxoFxYBz4ZAfmF2yf/Kc73zYOjAPjwDgwDowD48A4MA6MA+PAODAOjAPjwDjwcgdav2D78+fPf7L/61yX8dQjX73TSA3lmYZmu3iHG9oZP3q7l/F2HM3vcEMj46unHVUULovkZPNOb3ffSuMul/crr/Zpvoqv4IYmX7Vj16eG8oyjWSdmfPV2fOGySG427/SowXzFJc7zjOeYqr7DDc2MH73O63Ir3K5f3bDiVRz1K67mq1hxo797V7kZb7dL8zvc0Mj46mlHFYXLYsWpdlb4TLvTO9Wr8OpnOzXzmGE7vdDp4DKM36A6w6onTBaFYcxw0SPmJL+iV3GqO1Z4n61udyzrjMf5Ls/40eu8V3B9d7Vj16/uX/EqjvpXuRlPmrt4ws2wnV52Q8bLcFnvq7hxS7Y7ep1XcXf8jFfty7Cd3qlehWd/tZc4zzOeY1Z1xlePPPVOIzWYr3SIm3wc+A0OLP9VXP2waFaZpHknhkYHl2Gy/V3ck9xsp/eyfeo5Nuruu8rNeFXPb6lwWT+4Wb/T+0ouv7l7KznKX82V/l2vdK+idBk1W0Xid7nr7PCcB5f1SX53b4fvmKrO7q6w7Gc87xHP3HFZfYrfaYRehql63N/lOkd1tSPri6OYYaqeOIwZlvNVfpWb8aqe769wWd+5UXdxFTbje8/3+jyrncP6BJ9hO713f69u4ncq14xRM4/EnOR3dJwbdbY7w3kv41U95756b7av2uk33+FKq7tLeEbfz9kuf5JbfYPvqOrs1hNsxvdepueYqLvvK7jZTu9V9ztuVWcaGT7DRS/Ddnqud4UjjQ5XGHEYNWPkvMqJ3+WhscNUc99f4bK+c6ceBz7VgfJf9OwHI+tlxmS4VS80VvPVbLXfdTOs97TrhCvOLvou1uKe7BX/ClecbtQuv6/Dv8KR7ldy9c26ZReFZ9xxNCdHuWbdGLwu1nHaySiM6xLjuTjdSH6XI5zfpX4n3t1LPu9gHnd0nu494Yqzi9n+HUdzctU7jXc0rnDJUf5VN/v+qHmL5qso/AlXnG7k/i5HOHKVc6Y8YvY4P8mp1eWR47k0oq88YvY4P8ld+5SrW054gc2eNPymFVacbqRWlyMcuco5Ux5x9Yjr5JmWeDFTHnH1iFvlmcYKz9kdrnRCQ/lp5P6v4uoG7ffv0XwVT7jCnsZsvzRefbPvvrJXnF30Xf5tO37Msyee662w4nQjta5wxO9yhROPkTPlEVePuE4eWh1chuEd2XzVI3fyceCTHUh/YrMfjsyEwPnLuBlOvN2MeuKsouO9fppLfeXZjphVTzxhvK540Xes1xmXGOYZVjs0I5655h4DUz3yV7iM/y6u76nuzPpPc7Md8mY34y3irKLjvc64xDDPsNELjB7xzDX3SG4262gEjzjmrqk62+s8r8XNomO93nGEr3DeF57RMVHHfPeoscM7VnW1g3rCMmY8cjQnh7nmHqlBPHPnqCaXPXGjpzzDiqPoWK+FYySGOTHMA6NHPHPNPZKrmfO8Fm4Vycl2kOvYCl/1Q4saWc19VU6N1a6Mf8J1rOpMN3rZLeJo5nWlxT450uGcuWNVE8M80xNHM6/JV04Mc809Spt953lNrHJilGvGGDN/wjM6JuouN8NJbzfb3UAdYpVr7pF7hfXoHNXksid+9JRnWHEUHeu1cKtITmdnaJGT1d197+DyVuXZfdm3C+8x4+tbfEauMOo5tqqFV6xwrr/Cx8yf8B4dp3qnEThqieeRGOaOU53t5ayjEXjimEvL42qvY6ceB366A//+C5H80Jx8JH/I7v4wXdEiJ+72evUtjvU64xIT+ZVHjeB7vdJ0rNcZl5jIT94dru+5o/Uu7rv2POnNE1r87tDz2nc4JvAnj/qnXN9zonWC9T2qqRE9r4XLomO93nECf/Kof8r1PSdaJ9g7e74LV3fwu6PntXBZdKzXO07gTx71T7naQ43oeS3cKpKzu+MEW+2kxjtu9ju4/xO/9+T73BvV1Oj+GZGz81V7PFLjHXu57/TmO9zVd+/uuLP3Dlc3U6P7Z/QEN9PYeZVx3nnzFa/I6X5f9p2n3Ezj1KtMY3fHT/zeOzfLI8UTrROs9CeOA7/FgX/+F/idHxjnRn3nuV5Hi5zAs97dQ2yHS/xOe3U7dTp7qXXKJV459Va58Iwr/G5GnchP3ju479iRfbPvPfXGNV3P51lNTsxZZ/f4PMNkezLtE26m6bdkmCf3cl+mW+13rNeZD9yVzbu7TrmZbvcWx53sdm52R9Vz7p29J1zewxuiz3qnSWyHS7xy3rLKhWdc4auZ81lH3nldThe320mdwLKOvPOucKTb5Tque5v2KFIneqy7ml2O47r6ulWROp2biVcurZMobsQre6/u0r4un3cq73IznEUoc8YAACAASURBVDQUM0z0NGessN4nR7ljOrW4EeOxVq/SIfaUK01q7PZlnNO93PdqLnd1vy37RulodhLF1X7W6u30upwuLtvn3O5trkWdmLHONH2eYXzHqna9Cuu4u3urPdMfB36qA//8r0f/oTn5sDvcbM8VPXGop17E1ROOGPUyLmfZnDqrXDrEqLfTFa7LFV6RvF0ujuIOv5tLR3GH51wcRc52uTiKFV5zxQqX9cVRzDBVTxzFCtftS0exw8uw6kX0x1k2dzzrO1zqKO/qdXHSraJ0OFcv4uoJR4x6GZezbE4dz+9wXSvqrl4Xd2fHd+LyFn171ovZ6p1yhVdcaftMHEWfd+uMr17EziN+xenidjulQ5x6q/0VvssRn7tW3C5OulWUDufqrfZX+BWHuisctbNcOpypl+lyls2ps8qlQ4x6mS5n2Zw6nn8V1++IuntLF3dnR8ZlTzdkvZit3h2udKWhqP4qZlj1Iq6ecMSo9zSXujtt3hP5HS61pJP1ujdJQ5FazDVX5GyXi6O4w1fzjK9eRH+cZXPH7+quXhe32zfzceBTHfjnp/XOD82K6zPWlbnERL57xBNb9TuYFZezyK88apBf9TuYFZezyPl8xjpwrE+53KN8pSdMFd/BfceO7PtWe33GOtOKHjGR7x7xxFb93Q7yPL/L5X3Ksx2aMa5wPmNdaXT6HUy1K7icRX7y7nCzPV29Fc5nrF/5vdzj+W6v41nTp9P+XS73Rc7nM9aBY33K1R5qqOfa7Fc5dfwWcro4cjynBmdVnxjmxEd+8rrcFc5nrHnLaZ9c5tSJvHornM9YU++0H1xyIufzGesK1+kHhlq+lxpZ/iqu67LO7jj5DmpFzucz1rsdxHpe7ej0OxjuI77KiY9894gntup3MK/iUjfyk7fi+ow1d5z2yWVOncir18Vl/BXXZ6ypddoPLjmR8/mMNXHMiXG9Ls41WFNj8nHgkx34+6dx88O6M4I/RP6D6TPWlS4xrpdxiOec/UqHmC6XnEqXWllODc7Zr7SJ6XLJcV2fsQ591qdc3qd8pSdMFd/BfceO7PtWe33GOtOKHjGR7x7xxLLvOiczYnf3Ees5b2P+BM41WFe7qn5ws9fRdC45Pst2sHeHSx3lXb0VzmesYw/ryE/eiusz1ru9xHrO+zir+oHJ3hUuOa7rM9axn/UpV/dTQ72dNnHKqeO3CHNFl1zl3KXeFW3qrG7mDuVd7grnM9ba499V9YO7e9Rf4Vc4n7Hm/k4/MHzk7GbEVhpVf6dN3i7nHa57h+u6rCtdYla3rHA+Yx17WUfO5zPWFa7qu7ZwHc2KK42I1DnFn+pw1zu43Nf5tu5Nrsu60qj6nbuov8J3cbxF+YrrM9biR+z0A8NHzm5GLDWYE+N6XZxrsKbG5OPAJzvw90+q/YCvfrgyU/hD5Fyfsc60okeM62Uc4n2+mvmuLpeanftc9+pe6XC/eorVjH2/2Wes/dZTru5ipL7rEZfl7+C+Y8fpt/lNrDOt6BET+e4R79hqxr7v8Bnr3X3Eeu63qX4C5xqstcdvZ3832807+wJz8qh5ys32dPVWOJ+xdo9idvKo5Vyfsd7tJdZz3scZ+67vs9280mU/cj6fsfZ9p1ztcU31XZ/9LKeO30J8F0eO59Q4ma2wq5udFzVvWHFXOJ+x5s6q73eQk+XUibx6K5zPWFOv6gemmrEfOZ/PWFc49iPvcHyva3hNzSe5rsvab1BNzOqWFc5nrGMP68j5fMa6wrEfecURbjVfzcRXJDby3SPesatZYFfz1ewql5qRn7wV12esuaPq776HGo4Nzepx3wqX8Vdcn7GmVtUPTDVjP3I+n7EmjjkxrtfFuQZrakw+DnyyA3//NNoP8eqHKzOFP0Qr7tO4uIWau9t8fpVLXuSnj/yMu5qvZqFVzdmPvHoZLutl/Kdxd3bc4Xa/486OO9zufV1c3ELs7jbOyYu8ehku62X874bjPbt7fX6VS17kJ+8ON9vT1buD63Kfvq+7t8Kxv7vN51e55EVevQyX9TJ+hWN/x8vm7FEr8up1cR1+hqF+NmeP2NXN5Cjvcu/iyNduxt28wgavetS8giM/21HN2f+Oe3ffsrr5DveuL777jt4dbtxBvt+1m9/h+i5qRb56xGa41Xw1C63VfDVbccmL/OR1uRWO/Wzvbk4OsZFXr4vL+F1uhWN/p885eZFX77vjqrunPw78VAf++Wns/hBmH9zlPo2LW1xzV/P+Hdbn4lZ9zXfR+buaejusz8Wt+porZrisJzzj0zhqK+/uEJ6xy+3iqK38Hdzuji4ubnfsrj79XtfLdkrTY8Z1zDv1/J5dzVt3WJ+LW/U1X8U73Ey3q3cH1+U+fV93b4Xz/q7m/Tusz8Wt+porZrisJzxjhfP+rqam5871ueouTniPzt/VzmftXM52eZd7F+f8Xb2627kV9i7O+btadzhOfY8Vzvu7WrqOU78T38Ht7ngH7u4O5+9q/hnssD4n1/M7WOd6zV0+29VPcH0HNXd5l1vhvL+rV/c4t8J2cRm/y61w3t/VusFx6nv87ji/d+px4Kc78M8v2OKDuj+I2cd3uB3M6R2uuat5+w7r8xWXs13uuruaejusz1dczpRX/KovXsQO5gRHbeXdHcIznnBPsNwR+au5Xf0uLrvZuV7zm1cz4SpM1Rcvu40z5h0t4TvYCuP9Xa2dEXdYn6+4nO3yle6O6/MTrQ62wlR9vyerO9wKU/W5p8J4f1evNO9wqavc9XZ9zSN2uY7zmpqevwq72+N7vXY+6xMseZGfcDvYCuP9Xe13snYuZ553sBXG+7uaux3LmfIK4/1dLb2IjuVsl7+a29Xv4rrfW+lVffpUYby/q1eaJ1zqdL9fnN0en4uX7XGs16/iUneXr24St8J4f1dLL4vOzTDqnWDFUexwK4z3d7V2RnQsZ8o7mK7WiWZ3rzQnjgOf4kDrF2zxA5K9rO8/TB1Mph0916pwGda5XlPLZ7t6x+VceWj62+3xOfk+29U7LueRu57m3o/an2N8rrqLE57xXVzfE3X2sv6rua6f3RW9Li7DOtdr7vRZ1P4co7n3T7jSUHQt9bPo2JO9GXfV4/4VLpvtuJwrDx1/O23hM65miq6lfhYdm+k7Rjrez7iBzfodrmNO9p5wHctaO/UdnO3yHZfzTF/zbI9mio6p+o7zWrws3sEGN3tZ3/fs6kxXPeeq34knXMdG7c8xmnt/V4uXRedmGPUcG7U/x2ju/V0tXsQMy3mG0TzjrnriZZrBy17Wz3Y8yXX9TDt6XVyG7XyX9vqeu9xMTz3trG4WLovkeu54n7N27K7+rlzepTy+xV/2fTuM5hl31RMvi87LMOo5NursZf0O1zHS9v6uFi9ihuU8w/hctWupn0XHRu3PMT6fehz4VAf+/Wn475f6D0VVZ8ZU2KqfaUTP8R1chan0uOOUKzw1Vrnwfgv7nrveXa70XXdXi+f7d7yYV8+5FS7rv5Pru6q6c+cruNLM9kdPc8UOrsLs9LSjG7mnyxGOXOaaK3KW5cJ1Y2gQm2mqR1zkd7mVru9RLTyjZrtITpY7P8Ow5/hdfYVLjvLdHp+LF9Fnu9o51PLcte5ype+6u1o837/jxdw51PLc9Xyuuour8M5XLXxE9SKuHnErbBeX7TrlOn5Xf4fv3d3o86dudt1d/e69nb8P1c13uNLMNNyDwO6e9LqRel2OcH4ftTwXR/Eu1/VdL/ZUTzesMJXeV3H1Ldy/yoVnXOGzmXtALc+d73PVXVyFd75q4Rk168bgEkstz4mLnM9nu5pc5s7jLMsdv6szjemNA5/owN8/ofaFux+UmFevwxWmq9HBVZjoa5+i9065xEtzFSs8+567XszZczxr4iL35/NV/SRXWr5P/U58N9f3ZXV1d4b13h2utLoaHVyFib72KTpW/U58kist36v+KjpnVYcO5ye6d7ncxRuqnHjmFZ594rOc2Mg7zzmr2vVWWM2co1rzThRHscMRJjjKI64eccKyd8olnjq7nLzId3jOHe9arMmLvHpdHPnOyeoKz77nruNz1V2c8IxXuM5Z1bGLc+72nLjIq9fFke+cVR08zqnjOXGR+/P5qg4u567FmrjI/fk8q52jOsN6T1iPjlvVzlXtHPVX0Tmr2nVWWJ8Flz3XYk1c5He51Fae7dCMkTj2PSfuiZup57tYE6e9q7njM474GbbqBYczaWSRuMir18WR75ysJp55hq16weOMOp4TF7k/n69q56p2jvqr6JxVvdKZ2TjwSQ78+xNqX3f3B+UO37l22v9K4v7XLBLHel3Q/q+9w3Luuety7jOvHeu141l3sMR4Tq0sdzzrDM8esZGfvK/g+k7Wu9uJ9XzHjblzWO/4xEZePeIqjPodLDGeS6eKjmddcdQnNvKT51zW1Kn6xCh3rNfCZXGH5dzzTI89x7MmrsqJj/zkOZf1Soc4z1c8zZzDWpgqEus5OZyxn+WO9TrjqNfBEuO5dKroeNbkVH1ilHewxETefc5j7Rqr2RUs9SI/eU9yqcUbqj4xyjtYYiI/ec5lTZ2qT4zyDpYYz6UTkTP2s3yH5dzzTI89x7MmrsqJ97ziqH+KFy+ic1kTl+XEek48Z+xnuWO9zjjqdbDERF494iqM+o71WrgsOtbrjKPeDsu559JYReewJq/qE6O8gyUm8u5zHuuOBvGek88Z+1newRLjeabJ3il+xaUWcZOPA7/Bgf6/NL/BjfnGcWAcGAfGgXFgHBgHxoFxYBwYB8aBcWAcGAfGgXHg0IH5BduhYQMfB8aBcWAcGAfGgXFgHBgHxoFxYBwYB8aBcWAcGAfowPyCjW5MPg6MA+PAODAOjAPjwDgwDowD48A4MA6MA+PAODAOHDowv2A7NGzg48A4MA6MA+PAODAOjAPjwDgwDowD48A4MA6MA+MAHZhfsNGNyceBcWAcGAfGgXFgHBgHxoFxYBwYB8aBcWAcGAfGgUMH5hdsh4YNfBwYB8aBcWAcGAfGgXFgHBgHxoFxYBwYB8aBcWAcoAPzCza6Mfk4MA6MA+PAODAOjAPjwDgwDowD48A4MA6MA+PAOHDowPyC7dCwgY8D48A4MA6MA+PAODAOjAPjwDgwDowD48A4MA6MA3RgfsFGNyYfB8aBcWAcGAfGgXFgHBgHxoFxYBwYB8aBcWAcGAcOHZhfsB0aNvBxYBwYB8aBcWAcGAfGgXFgHBgHxoFxYBwYB8aBcYAOzC/Y6Mbk48A4MA6MA+PAODAOjAPjwDgwDowD48A4MA6MA+PAoQPzC7ZDwwY+DowD48A4MA6MA+PAODAOjAPjwDgwDowD48A4MA7QgfkFG92YfBwYB8aBcWAcGAfGgXFgHBgHxoFxYBwYB8aBcWAcOHRgfsF2aNjAx4FxYBwYB8aBcWAcGAfGgXFgHBgHxoFxYBwYB8YBOjC/YKMbk48D48A4MA6MA+PAODAOjAPjwDgwDowD48A4MA6MA4cObH/B9ufPn//4/3V3OC/q6mXYTu9Ur8KzX+0lpsorbvR37yo34+12cZ7x1SMuy4XLYoZX7xQvnmLGj17nvYLb2V3tde4Kt5pV397lrHCrWbY3w2e4qpfxo+evwu36rqO64mlexYxXYbN+xlcvw7MnnEdiIvd5t3Yd1RVf8ypmvAqb9TN+9DrvFdzO7mrvb+Pu/owynypOhu30Qq+DyzBXbqk46md7ouevwu36rqO64mm+it+NG/fwVfft+tRgvuIRl+Vd7gq3mmU7o3eFU/GqHVm/2pthvfcObrVj1/dbVa94wmQx42W46GXYTu8rudm3rG4mfoVbze5+L29QvtvXwe00VvPVTLs9XuFI46u42j9xHPgUB/7+XyXJV2U/bAksbZ1wM2yn54uvcKTxaq70tU9R/U4UhzHjcV7lGa/quUaFy/rOjbqLc27G855zVDsuq4X1mGGr3l1upbvrf5e9cUd2q9+X1RnPe+T5rFtTI/IOzzmsMz7nVZ7xqp5rVDj2yWH/JKdG5B2uc1hnfM6rPON576u4usP3q9+Jv4Xr38k684lz5hm20wuNDi7DcP+pzhUuOdk9nR41ujc7R/WVfe/i6rbuNwrPqFsVOdvl4iju8JwHh/VJrn2Kd7jVHdJexc7eiv9ObmdXhvHbM0zVc27UGTbDVdiM772v5PJb/K5VPTf/+/+ppfKLHp/69l24fsfU48CnOND+BVt8MH/IOwYI3+EKexp5R5dLjvJ3cLVDOyOq143kKhfX9TTPojjdSI0uRzhylXOmPOLqEbfKM40VnrM7XOlQQ71uDG4X67jvsFc36Db/Hs2zKM4ukrvDVvMrGuR4rj3RVx5x9Yjr5NTq4H1/l+O4u3vJj1z6zP3WiiNuFZ3nOypetX+Fz2bcn81Xvd/C5Xd6Ln+irzxi9jg/yV37lKtbTniB5etyr3Bc+4oGOcpdt6qFZ6yw3idHuWN2dfB2mGqunVc07nKrm3b9u3vJ53czjxtWb3ej5pmGZrv4FHe3p5pzf4Wp+uQqFzZq5RGzx/lJ7trv5Oo7TnYGdm6+9gu2Kz7f+TN6giuNiePAJzqQ/2v+3y/lD2y0vF4Z4livV1zNyIl89Rxb4bP+09xsh27nLNvLuTiK2YwagfNaXEZimBPDPDB6xDPX3CO5mjnPa+EYiVHOufKY+ROe0TFRd7kZTnqccR9zYT2Sm806GsEjjrlrqn5qL/W01+8RxqPwjI6RVtZnjxqRr55jK3zVD21qZHW2nxzmGVaamhGvXDPGmO2e+IorvDCMGT7m1XOu1xmPGOUVzvvCMzom6pj7I4e541RTg3jmwnr8RK5/Y9T8Tp+7T147PqvJWe26y/U9qjPd6PEWYRkzHjnZXLo7HXGJU64ZY7ZXeEZylL+Dm+1Y7edsd3+FFU9zj7xJWI/OUU2ueopdjcA7VrW0PGZ7xdHMa9eo9lY471NfuWO0w/vCMzqm4jqOGpGvnmNVV5xMTxzNvK602CdHOpyv8ndxfY/q6rbVd4irWGlkfXEUM4x6wnjU3GPgqtfVyPgnXMeqznSjF3M9YT1q7vEprutOPQ58mgP//6cs+TL+wMXY64Tyv5Zjvf4fcJGQE/nqnWBd5ydy9Q28PXpeC8dITOQn7w5Xe6gRPa+FYyQm8pP3E7n+fSffcIL1PV5f0SIn9Lz2HY4J/J3HfTutE2x1EzX8Wzqc3Y2uwX2n3DtaT+ylxnf3iree+jxc/5u2r+lZ9++Gq1LjlX9m79rj3+f1yR0n2Dt7vgv3zh1f5dVX3ay9/O7oeS0cIzGRn7yv4vqNJ3ecYH2PampEz2vhVpGcyE/eu7h39vj33NE64Z5g/Uav72idcE+wX3Wj7516HPh0B5b/KvOHNoxgHfnqEXvKlS41Vvu6OOkyPsld3cidkfveEy61qJPpEpvNT/Zyl3LX79Tiajdr9aizmxPr+ZPc7Dbfp9r3nnClweh6nDF33Lv2VjdE328itjN3/K7e7RO/ixO+itQJDOvI/fk8wzhHtXPVvxq7el3c7g7qBJZ15P52c8ezfpKb3cZdzH3vb+PSi5OcvgWPddfDKxzd2OU6rnvb1W/SfVn0WzLM3b3dHdnuJ7knPt+5xW8+2XuH+1U3cy/vjz7rzIfdnNqefxXX7+h8pzh+c+aJsKtInZP91KTG6R3v4PqO0xv5rVc9kobfor5Hx336zXe+9w7XfZ96HPhEB/79f9HgK/UDhNZf/8Fl3/M7XGlJQ1F9j5or+nxVi6O4wvpMHEWfr2pxFFfY1SzjqxfRH2fZ3PGs73AznayX3XRn70/k0hfl3e/o4qS7i1f0xKG2ehH9cZbNHb+ru3pdXHcfcdRmP3LOIj95d7jZnq5eF5ftYE86WS9m/oRX9PmqFkdxhfWZOIo+X9XiKK6wPhNH0eerWhzFFdZn4ij6fFWLo7jCrmYZX72InUd8lyPdLreLky7jHS51lHf1ujjpMv5ELu9X3v2OLk66jHe41FHe1evipFtF6XCuXkR/nGVzx7P+Ki5vUN69pYuTbhWlw7l6ETuP+C5Huu/g3tmhOxnv6HW5XRzvWuV39LrcLi6786u42S3TGwc+zYHyX3L+4PGjq34H0+FWOsGtHnVXuIz/Kq7rso47WEfO5zPWFa7TDwy1Tvfe4eo+aqjnd7HvM7/ZsV5z35Nc12W9u5lYz/1+1U/gXIO19ngkJvLdI57Yqh8Yzjo7qJvlXb0uLtuhHjXUi1j1fRY4PvI833Gp082zHRm3i8u46lFDvYhV32eBO3nUfZLruqx3NxPr+Sdyu39e9IKcqk8Mc+IjP3ld7grnM9ZxC+vI776uXheX3fMduX4T6+wbokdM5NVb4XzGereDWM+v3ELOSs9nrCuNTj8w1Ir85H0VN7uxe8sK5zPW3HnaJ5c5dSI/ee/grnb4jHX1HcREfvK63BXOZ6yrW4iJ/OR1uSucz1jHLawj5/MZ67tc7pl8HPhUB/7+icJX+g+TRuxHnj1iOGe/4l7Bn+pe2UGO8tVen7EOPuvI+XzGusJV/Z32iud7We90ia12VP2dNnm7nHe47h2u67IOXda+12esq5uIcT1yVjifsaYGc2Ii3z3iiWXfdVYzanTzrl4Xt9pLDeLYj5zvZEZsaLB2Xe7o5l29Lm61lxrEsR8532pGXJa/iuu6rOMO1pHz+Yz1J3L57avcfRCW/ch37xRPvS53hfMZ69jFOvK7r6vXxWX3fEeu38Q6+4boERN59VY4n7He7SDW8yu3kLPS8xnrSqPqB5ePWj4jLsu/invnltXNPmPNnZ1+YHaPOh089d7BXe3wGWveyZyYyE9el7vC+Yx1dQsxkZ+8LneF8xnruIV15Hw+Y32Xyz2TjwOf6sDfP1H4Sv9hwuivH0r2ld/hSiMidSKvXheX8V/FdV3Wu28j1nN+A2fsuz5n5ETO5zPWrnnK1R7XVN/1q77vJS7Lue9Jruuy9m/xvcR6nn3DTo+clZ7PWFODOTGR7x7xjq1m7Hd2uK7XXb0uzvVZU4P9yKsZ+5Hz+Yy1azqXOt2c+iu9Lm61lxqOq2bsR37yXsV1XdZxH+vI+XzG+hO5/PZV7j4Qu5oRFzmxkZ+8LneF8xnru/dl30L9yKvXxWX878j1m1hn3xA9YiKv3grnM9a7HcR6fuUWclZ6PmNdabAfeYcTmJNHzXdysxu7t6xwPmPNnVU/MKsZNRwbvJPHPa/irnb4jHX1HcREfvK63BXOZ6yrW4iJ/OR1uSucz1jHLawj5/MZ67tc7pl8HPhUB/7+ifrvV/oPkn/8ar6ahc5uzl3ERl69Li7jv4Ob7ch6J/eRv+NxTl7k1ctwWS/jVzj2dzzOyYv85L2Dm+3Ietndn4KLb+O37L6Vc/Iiv/u6el1cdQ/5Gaaasx959TJc1qv4nX5Xr4urdpKfYao5+5GfvHdwsx1ZL7s7w2W9T+Fm3xE9fnOG2c3JITbyk9fl3sF1ud27u3pdXLb3u3O7970D944dd/6MqvvY3+lzTl7kJ++ruNmN3Vvu4sjf3ZHN2aNW5CfvHdzujqdxmQ9P73ha76tuvvMdd7jZ905vHPg0B9J/lf0HZ1fTlB3W5+R63sV2ca4f9Tu42Y6sd3Kf83e1tB2nvscMl/WcF3WF8/6ulrbj1O/Ed3CzHVkvu/dTcPFt/i27Wn44Tv2rsavXxVV3OH9XS8dx6nvMcFnPeSd1V6+Lq3Y7f1dLx3Hqd+I7uNmOrJfdm+Gy3qdws++Inn/zrq50Mq0V1me+1+eq7+C6XO3axa5eF5ft++7c7n3vwL1jx50/o+o+7+9q3eA49Tvxq7jZbd1b7uKcv6uzW9Vzrvqd+A5ud8fTuOz7n97xtN5X3XznO+5ws++d3jjwaQ58xC/Y4g+l+8Oe/QG+mlvpV33eWGG8v6s7mh2M7yFHeYXx/q6WXkTHcrbLX82t9Ks+7+1gTr+/o9nB3N3rO7y+4gM5Vb7a45wT7I7rWl6Tv5oJV2Gqvngn8UTrBOs3OHdXk+9Yznb5q7mVftXnvRWm6v90Lu9n7t+7q8n13Lk+X9Un3A62wlT91W3V7ETrBOv7vjO3e1sXF9/ewVaYqk9PO5juHdLtaFYY7+9q7Ty9kbyv5N65w71xrdV3OXdXZ9rqOVf9TnwXt7Ong4lv6uKy7z/hdrAdzE+5ufMtFabq88+ggyF+8nHgUxz42F+wxQ919rK+/wOQYUIr63e4jtFd3u/or7iZnnri6TvUV+Q8w2guPKNmipxFruf9XS1exAzLuXLuY8/5mjFe5Vba3u/o8x7mrsWZ5479yr3ZLerxbvUYOVce892jxg7v2Aqf9TPuqse7MxznkTtGc+9Hnb2qT6xrcea5Yyv9rJ9xVz3uznCcK+/uFZ7xKtdvk6b3O/qfzo3v6/iQeceefMoicdmujKPeCdex2S7HVHsybmCrvnSE4R7OPCdOuWOk6X3hGR3zVVzeFHn1ujh9xw5fzb0ftT/H+Fx1Fxd4x57szbirnu7r7hWOPPV8j2OE877zos5e1SfWtTjz3LGZvmOk4f1dLV4WnZthqt67uL4nan+O8bnqLk54xhOuY6P25xifq+7ihGc84To2an+O0dz77+Jq/8Rx4FMd+OenkD9sq48mTj+Q7J1yMzz1tCPDqef4qhaescJ6nxzljtnV4kXcYX3uHGp5nnGF8dmuFi/iDutz51DL84wrjM+qWnjGCut9cpQ7ZleLF3GH9Tm5zLs4cRy/q8Xz6DyfqyZOvSwSFzmfz6qanCx3XoZhz/FVXXHY99y1OPfZrr7CJSfLfWeGYc/xVV1x2PfctTj3WVWTo7zCel94Rsfs6uH+2f6bV3nEvufuu89Vd3HCM55yHb+rV7sqLjlZ7rwMw57jq5oc5RXW+8IzOqaqyVFeYau+eB4d73OvHb+ryd9hfU4u8y5OHMfv6uARI50sEhc5n8+qmhzlFdb7wjM6pqrJyXLnZRj2HL+rg0sM3wjv4AAAGKJJREFUtTwnLvLqdXEZ/51c37Wrs3uj57wKl/VPuY7f1dnOn3Tz7vt8zu/12a4md/Jx4JMd+Odfb/5wrD6cuMjjsXfKzfDU044Mx55zspp45hnWe8Qzd9yqJi/yFdZnjnct1hl3NXc8a/Ii52yXO961WLsWZ67jWNXOUa35KgrrccXx2ZNcae12CMfonFVNHnPncMacOPY9Jy5yfz7Paud47RyfZ7Vzspo8ztn3nLjI/fl8VV/hOsdr3+fzrHZOVpPHOfueExe5P59ntXNUZ1jvCevRcat6uL/nF2zxZ736u+CzK383nOP1bofjo3ZOVme8787Vd3Rvr3DsS7MTyYu8wxHGuao1V1R/FYXtxNAh7kTXsdSpcueorvDsC+uRmCp3jtfO83lWO2dVB5/zTE894iKvXheX8d/N9X2rOrs3es6pcFn/Ctc5qzrb+dNuXn2fz/x7fb6qnTv1OPCpDvzzrzd/MHYf7VivV/wOlpjIu895rHcaxHq+48bcOax3fGI9J5cz9rO8gyXG80yTPcezrnDsZ3mlISznngtTRcezrjjsE+85cVnueNYZnj1iIz95zmW90yF2tZe4E80MSy3PM7z3rnBCw3msVzt85vVKR1hiPBcmi45lneG9R3zk3ec81q6xmp1iqeW5a3nteNaOzWriPc/w7DmeNXFZTqznGZ49x7MmLsuJ9TzDq+dY1sIormbCKHawxER+8p7kUmt1A3Ger3iaXeEE13mspV1FYj2vOOo7nrUwq0i85ytezE7x1HMua+KynFjPMzx7p/gVl1oVjv0srzSE5dxzYaroeNYVR31iPRdmFa9wpOdc1sJErPrEKO9giYn85H0XLu/Y3U/sV34v7/gNN9/53hPuzsuZjwM/zYGzf5V/2tfNvePAODAOjAPjwDgwDowD48A4MA6MA+PAODAOjAPjwIsdmF+wvdjgkR8HxoFxYBwYB8aBcWAcGAfGgXFgHBgHxoFxYBz4bAfmF2yf/ec7XzcOjAPjwDgwDowD48A4MA6MA+PAODAOjAPjwDjwYgfmF2wvNnjkx4FxYBwYB8aBcWAcGAfGgXFgHBgHxoFxYBwYBz7bgfkF22f/+c7XjQPjwDgwDowD48A4MA6MA+PAODAOjAPjwDgwDrzYgfkF24sNHvlxYBwYB8aBcWAcGAfGgXFgHBgHxoFxYBwYB8aBz3ZgfsH22X++83XjwDgwDowD48A4MA6MA+PAODAOjAPjwDgwDowDL3ZgfsH2YoNHfhwYB8aBcWAcGAfGgXFgHBgHxoFxYBwYB8aBceCzHZhfsH32n+983TgwDowD48A4MA6MA+PAODAOjAPjwDgwDowD48CLHZhfsL3Y4JEfB8aBcWAcGAfGgXFgHBgHxoFxYBwYB8aBcWAc+GwH5hdsn/3nO183DowD48A4MA6MA+PAODAOjAPjwDgwDowD48A48GIH5hdsLzZ45MeBcWAcGAfGgXFgHBgHxoFxYBwYB8aBcWAcGAc+24H5Bdtn//nO140D48A4MA6MA+PAODAOjAPjwDgwDowD48A4MA682IH5BduLDR75cWAcGAfGgXFgHBgHxoFxYBwYB8aBcWAcGAfGgc92YH7B9tl/vvN148A4MA6MA+PAODAOjAPjwDgwDowD48A4MA6MAy92oPULtj9//vwn+7/ObRkvep1XcTv8intnr3OrHbu+66iueJqvYsWN/u59BTfbubuT84yvHnFZLpzHDOs957Amlv2TnBrMKw1iqvwqN+Od7sg02LuiV3HUp77nwlTR8aorPPvCZpG4LM846mV49oTLYheXcdULDeWn8S6X9ytf3SBMFjNehst6d7ihl/Gj13ldboXb9asbKp7jK9yu7zpRZ5wMV/UyvnoVR33hPGq+is5RveJoJqxHzVfROaxXvJgR6/lV7o73VXv9+6I+eRlfPeqodxp3vqz07nJ5v/LdPuGy+N242Y3ee8XNviOrq70Zlr2Mx/kqv8MN3Yyv3mrvirvj3eHqNsZqHzEneaaX8TPc6tsyDfau6FWc6o4VnjPepZzzyceB3+bA8n9l6IdkFSvDVhzN7nArDfVX8c5eclc7VjNqRL7CauYc1Zp3ojiKHY4w4iiq34niMGY8zqs841U916hw7DtHNTG7PDg7TDXXPsUKx76wHompcuewzjicM8+wnR41Iu9whHk31/ep1j2dKI5ihyOMOIrqd2JwOrgM85VcfevpDeQpr75N81W8ys143qv2Oi6ryc3mnR41Ij/ldPAZxvdWuzOc9zL9qneF6xzV1Q72hfVITJU7R3WFz/riKGaYqieOYoVjX1iPxOzyK1znsM72cV7lGa/qhUY12/W/kstv393JOXmn97+b6/tU83t2uTiKO3zMq3eHG5oZv9rl/avcjFf1Ojud6xzVjstqYbN4gs+wnd532du5VZi7NztfuoyOmXoc+E0OlP8F4A/JKs/MWuE5u8OVDjXU20VylO84mgsfUb3TeEWDHOXv2Ksd2hlRvW4kV7m4rqd5FsXpRmpc4Yjf5QoXPOWnUTtPNMhR3t0rfBalETPlEbPH+UlOrROe3/EOLm9V/o692qGdEdXrxiscaX8lV9+sW7pRPEZx/XuIqfIrXHF2Mdu542hOrnqn8YrGFY7fRQ3lwkStPOLqEdfJqdXBV/u/O1f3XfneO1zuUy69bhQv4hUO+dRgHrqr190rnGur34lfyZUHnTuJEe/K7e/mcp9yfksnFy9iBx+Y7N3hSk8afovmq3iFK043cv8Vjvh3uK4RNfU0Z+T8JKeGcvGjVh4xe5yf5NQ64e3u6NzM3cQzr/Y4d+px4BMdSH/asx/U7OOzH56nudkO3cLZ03u1g5H72Gfud3DmuWMr/ayfcTOcdnL2VVzeohuip5w3CqtIDHPNPVKLeOWOj5oczYX3qLnHTEOYrkbgHVvpZv073OxWv0eYXfQ7VnjHqq44MdcT1qPmHndcx0dNjua+T7XmHqkhrEfnqH6KKz3F7n7hGd/F9T2qeQvzmPsTRzOvHc/asV4Tq5wY5Zoxxsyf8IyOiTrjOo4aO7xjK3zV527X4qzKyQmM1xmPGOYZVpqaEa9cM8aY+ROe0TFRv4Ob7dAtnPFW5sJ63HEdHzU5mnMXc809UoN45Y6PmhyfiyeM146XHnHKM6zwq5n4EU8eea/k+h7V1a28RViPX8HNdvJWzf1W1Zp7pIawjI6PmhzNyVGuGWPG1Vw8YbwWLouO9XrH6eKlQ7xyzRhj5k94RsdEnXGFc67Xwq0iOatd1CAn+l4TW+Xk7PY6VvVK22fiaJfXjmftWK+JnXwc+C0O/Puvmv1joB+2riH8wfoNXPfl5PtPsHf2fBeu7uB3R89r4RiJifzkfRXXbzy54wR7Z49zVXN/9LwWbhXJiXz1TrCuM1x3pF//BO/u3CgnqBE9r4XLomO93nECf/Kof8r1PSdaJ1jf4/UVLXJCz2vf4ZjAnzzqD3ft3E/0Sl/E26PntXCMxER+593Rehf3XXvcx9nrjuxrehZor1cKjvU64xIT+cn7Kq5u5P7oeS3cKpITeeeRE3ivTzV2e6m/w1a7qXF68x1udc/0x4Gf7sA//1rwB+X0B/VJ7snuJ/fe/QPt3tLFZfc4N+ru+you7+MN0WedfYvPMwz1mTuXs13u3JO9mbbrZZjodXEZ/w6XetS5ehM1Iq+e41ZY13iS69qr+sm97/re7Hv8OzJM1XsH13eceMW7qRN91jtNYjtc4nfavLGj7fhd3b2li9vt0/yKHjmhwzpyfz7PMM5R7Vz1O/G7cO987x1uxyNh3Kt37dX+iLzB6+we4pVT7zSXhuIJXxzFV3ClzdjdQ47yr+B2dwZOdzJ2+eREfvLucLmHOtFnvbuJ2A6XeOW8ZZULr7jC+kwcRZ93a/EjxmOt3k7rLufVe/2+7nf5d1Pn9OY7XL9j6nHgUxz4578Q/EE5/UH9bdzsL0HXgy7uzo7vxOUt+vasFzN/wiv6fFWLo7jC+kwcRZ+f1tJRrPiaK1a4rC+OYobp9DK+ehE7j/gVp4vLdg43c6Xf++7+3bmPLkgn68Vs9U65wiuutH0mjqLPT2vpKFZ8zRUrXLcvHcUOL8OqF9EfZ9nc8ayHSzfW+U/0il+k+7Ne9vdGeEXyruTSUTzREEfxFVxpK75iR6apfYoZpuqJo1jhsr44ihmm6omjWOGyvjiKGabTy/jqRVw94YhRL+Nyls2p4/lXcXmHbsh63e+RhiK1qjzDqhex84hfcbq43U7pEKfean/ghbvCJWfyceCTHPjnJ10/KIonHyuO4lNc6WUxdnj/qb0nOsJ2b+nipMu44vqMdWiwjpzPZ6zvcrXHNXf9u3u5L/KTt+L6jHW1g5jIq9fFZfw7XOlRQ72IVZ8Y5sRHXr0VzmesQ4915Hw+Y73jUifLqfXkXtdlvbuZWM+zb9jpVRz1uzuEZ+xyVzifsa52dfodTLUruJxFfvLucLM9Xb0uLtuR9U71iKde1Q8MZ5Hz+Yz1jkudLKeW783w7L2K67qsYz/ryPl8xnrHpU6WU+ude3UL96sXser77PRm7lDOXa4nTBXfwV3t8BnruJl15Hw+Y/1KLm/Ict7x5M3ZLvZWe4lb5dQgrup3MCsuZ5Hz+Yx14Fg7lzpZfocrPWqoF7HqE8Oc+Mh3j3hiqz4xzIlf7V3hfMa62tXpdzDVLnInHwc+2YF//rXgD0XkJ+9VXNdlHfexjvzk3eFme7p6XdzpDtdlHVqsI+fzGeu7XO1xzay/u0uc7CbXZ+261MnyFddnrDOt7Na7uIzPOyK/8qhBPvsd7S5+hfMZ67iNdeR8PmO941Iny6n15F7XZb27mVjPs2/Y6VUc9bs7hGfsclc4n7GudlX94Gavo+lccnyW7WDvDpc6yrt6XZx0d/FUj3hqsx8538mM2NBg7brckeXfkes3sd59L7Ge77iZP+xlepr7jPXdvdkO9XbavCNyPp+xJo45Ma5HXJa/g7va4TPWcS/ryPl8xvqVXN6Q5bzjyZuzXeyt9hK3yqlBHPv+TcIRo15E9p17MiN2p8v9WU4tvynDZz1qcM5+R/sO/h17V/f5jHV1W9UPbvY6mhU305veOPApDvzzE8MfltMfildxXZd1/EGwjvzk3eFme7p6XdzpDtdlHVqsI+fzGeu7XO1xTfVdv+o/eTN3ZDlvPd17qkf8ai9xWX6HKz1qqKe4mgmjSGzk1VvhfMY69FhHzucz1jsudbKcWk/udV3Wu5uJ9Tz7hp1exVG/u0N4xi53hfMZ62oX+5FXHOFW82rGfuQn7w4329PV6+KyHVnvVI9416tm7EfO5zPWgWPtXOpk+Xfk+k2sd99LrOc7buYPe5me5j5jfXdvtkM9Rd+X9QPDR47nxDHv4shR/g7uaofPWMeNrCPn8xnrV3J5Q5bzjidvznaxt9pL3CqnhuNWs8Cu5tWM/cj5fMba9zmXOllOrVOu9KihnuJqJowisZ1biJeG4momjCKxq70rnM9Ya0/Eqr+b7eYrXe6ffBz4RAf+/hfTftDih+Pk8YfpVdxsR9br3n2Hm+3o6nVxT+/o7s1wWe/kPvJ3PM7Ji7x6GS7rVXzvd7lfhfN7o+7eknGdn2Gon83ZIzby6t3BvYOb3f2OvdmOrPf0fZle1uvecofb3VHh2N/d4fOrXPIiP3l3uNmerl4Xl+3Ieid6xO60OCcv8upluKxX8b3/3bnZfVnPvyvqDJf1Mm7W63IzXNbLdlQ98jNMNWc/8uo9jcv2dHfc4XZ3ZLis173lHdzuLV1c9+auXoaretydYVbz1Sy0qjn7kVcvw2W9iu/9O9zQIt+1O3NyqBX56hGb4XZzcohd7b2LI5/7la/mq1nwd3PtmDgOfKID//xrwR+IyE/eO7jZjqzXvfsON9vR1evint7R3Zvhst7Jfc7f1dJ2nPoeM1zWc15Vd7lfhcvu7t6ScaPn/F1d6WRaFdZ3nODewc3uecfebEfWe/q+TC/rdW+5w+3uqHDe39W8dYf1ubhVX/NVvMPNdLt6XVy2I+ud6Dl2V2uf49T3mOGynvOq+rtzs/uyXvZ9GS7rZdys1+VmuKyX7ah6zt/V0nGc+h6fxrl+1N0dd7jdHRku63VveQe3e0sX1725q5fhqp7v3tXU2WF9Lm7V11wxw2U94XfxDje0nb+rV/c49wTrXK9PtCpsV7PCeX9X844d1ufkTj4OfLoD6W/Q7vxQvJpb6Vf9zh/gHa7rn2idYK/sqfSrPndUmKp/hetaXnc0O5iVLvlZ3uF2MKHdxZ1i/e6TPTuua3ntfNZPYyu9qt+5pcOlDvMOt8JU/Y7+HS71I+9oOUf1u7idPRXG+7ta35Z5c4dL3V3ue3b41fxE6wS72hmzEy3H7mrudixnyitM1RdvFb8zt7qt6vM7K0zVJ7fKO9wKU/WrXew7d1evuJwpdz31PXZxzov6XdzOngpT9fk9FabqP8WlDvPvvJd3Mvebd/WruNRV7rfs+pqvYqW54mjm3F0tXhadm2HUc+yuFi+Lzs0w6nWwFcb7u1o7I+6wPid38nHg0x1o/YItfkiyl/X9ByrDhFbW73Ado7u8n+nf2VtxtV/R71A/i459+mbX1w3ez/Y65lVc38NaOyOyr5zzDKO58IyaMcbcHznKdxifqxZfUf0sCsNY4bxPjnLHRB0zf8J3o/NZuwZnnjs2an+O0dz7T3NjT0ezg3nXze6J9nrs4pwX9bu4vidqf47R3Pu7WryIO6zPd1zOlYeGP9fNMMGp+tRzLc48d2ylX/Wp51qcee7YXU1+huU8csdo7v2os5f1vzPXb9M3eb/zXSfcwHY0O5hX7c08UE879R3qK3KeYXyuWnxF9TtRHMUORxhxFNXPojCMjuMscj3vc1Zhqv7T3NjT0exgTm6+s/eEG3dX/6d7pVfhsv6Oy3mmr/lOmzjlik9zMz31tDOLwihmGPWE6UbxsugaGUY9x0btzzGae39Xixdxh/U5uZOPA5/uwL8/hf/9Yv/BqOrMoArr/TtcaVFDvV0kR/mOo7nwVRROscKpL9wuCs+44/j8K7m8hXd4TlzkfD7b1Ve45Cjf7fG5eB67OPEcX9XCM1ZY71cc9j1faRDbxYnj+F0tXsQd1udXuOQod91dLV7EHdbnT3FXOpzt8tV9T3N9166O/cSs7iEu8rtc7XLdqhaescJ6n5wsv4t3vupsF3vCKXLGXPOIq0ecY322q7lnh9WcHOWa7aLwjDuO5uQo16wbxYvY5Qh3hUuOcul1o3gR73Ko5blrc+6zXU0uc+dxtsvfyfVdu5q377A+fzeX+5T7TbtavIg7rObkKNdsF4X3fex77pp3udJ33V0tnu9f8chRvsJzJrzvY99z8iOv3hVcpRX9K3qr+7TLdXe13yKdLLrWXW62Y3rjwKc5UP+rkvxD4D9kUVcvw3rvDldarqH+KjpH9YqjmbBVFE6xwrEv7CoSz3zF8Rl5kft8Vd/lUtu1WBMXuT+fr+orXOeoXu3xmTgeuzjynJPVxDPPsN6r8Ox7vtIgtotbcVyDNXmRc7bLr3Cdo3q3i3NxFDnb5eIo7vCci+ORmMhP3ru5vm9Vx3dwvvou4iK/y+Uu185q4plnWO8Rn+Wn+NBwTlZnu9hzDmfMiWPfc+Ii9+fzVX2F6xzVqz2aCetR81V0juoVx2fiKPp8VYujuMJqJqxHzTvxCpcc7mDfc+Ii9+fzVe1c1c5RvxPfzfV9q9rvX2F99m6u71Ptd61qcRRXWM2E9aj5KpJDHPueExd5PPYcz5o4cVdzx7MmL3LOqtw5qis8+8L6LvY9Jz/y6l3BVVrRv6K3uo+7XHtV+y3U8dx17nJdf+px4BMdqP9V+e/X+g8W650hxHq+48bcOaxXfOI8X/Hu7JTu6b6KRx1hVpF4z1e8mDme9VPcq5rVfup5XnHufutd/smd/AbnsSYuy4n13PGc+8zrDpaYyE+ec1nvdIj1fMV1LOsVTzPiPRemio5nXXHUJ9ZzYap4iqfOd+Hyjuo+9rPcNbzOOOrtsJx7Lo0qOp51xWGf+Mi7z3msOxrEr/YSt9PtYInxfKXvWNYrXsyI9fyV3K/a7d/Ieve9d27mHs99L+c+87qDJcZz1/P6FE/+d+HyDt6X5cR6nuHZczxr4jwnznPHZrVzWGd49YjzXJgqOp61c1azHfYO17Wjpp7nGV49x7IWporEeu4czn3mdQdLTOTVI67CqN/BErPaK01G57KucOxnuWt4nXHUO8GKM3Ec+OkO1P9a/PQvm/vHgXFgHBgHxoFxYBwYB8aBcWAcGAfGgXFgHBgHxoE3ODC/YHuDybNiHBgHxoFxYBwYB8aBcWAcGAfGgXFgHBgHxoFx4HMdmF+wfe6f7XzZODAOjAPjwDgwDowD48A4MA6MA+PAODAOjAPjwBsc+H9x/u30CFamzQAAAABJRU5ErkJggg=="
    }
   },
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Beleza! O resultado obtido é rápido porque as anotações em arquivos GenBank estão dispóstos de forma padronizada.\n",
    "\n",
    "Agora, suponha que você queira extrair a lista das espécies de um arquivo FASTA, ao invés de um arquivo GenBank. A má notícia é que você deverá escrever mais um pouco de código para extrair a informação desejada da linha **description** do registro (isso se a informação estiver na primeira posição no arquivo!). O nosso exemplo começa assim:\n",
    "\n",
    "![image-2.png](attachment:image-2.png)**...**\n",
    "\n",
    "Vamos ver como é disposta a infomação presente em **description**:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gi|2765658|emb|Z78533.1|CIZ78533 C.irapeanum 5.8S rRNA gene and ITS1 and ITS2 DNA\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "registro = SeqIO.parse('ls_orchid.fasta', 'fasta')\n",
    "primeiro_registro = next(registro)\n",
    "print(primeiro_registro.description)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Se você verificar manualmente esse arquivo, verá que o nome da espécie está na segunda posição quando separado nos espaços em branco. Isso significa que deveremos quebrar essa informação do registro nos espaços em branco, sendo a posição 1 que essa informação ocupa:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['C.irapeanum', 'C.californicum', 'C.fasciculatum', 'C.margaritaceum', 'C.lichiangense', 'C.yatabeanum', 'C.guttatum', 'C.acaule', 'C.formosanum', 'C.himalaicum', 'C.macranthum', 'C.calceolus', 'C.segawai', 'C.pubescens', 'C.reginae', 'C.flavum', 'C.passerinum', 'M.xerophyticum', 'P.schlimii', 'P.besseae', 'P.wallisii', 'P.exstaminodium', 'P.caricinum', 'P.pearcei', 'P.longifolium', 'P.lindenii', 'P.lindleyanum', 'P.sargentianum', 'P.kaiteurum', 'P.czerwiakowianum', 'P.boissierianum', 'P.caudatum', 'P.warszewiczianum', 'P.micranthum', 'P.malipoense', 'P.delenatii', 'P.armeniacum', 'P.emersonii', 'P.niveum', 'P.godefroyae', 'P.bellatulum', 'P.concolor', 'P.fairrieanum', 'P.druryi', 'P.tigrinum', 'P.hirsutissimum', 'P.barbigerum', 'P.henryanum', 'P.charlesworthii', 'P.villosum', 'P.exul', 'P.insigne', 'P.gratrixianum', 'P.primulinum', 'P.victoria', 'P.victoria', 'P.glaucophyllum', 'P.supardii', 'P.kolopakingii', 'P.sanderianum', 'P.lowii', 'P.dianthum', 'P.parishii', 'P.haynaldianum', 'P.adductum', 'P.stonei', 'P.philippinense', 'P.rothschildianum', 'P.glanduliferum', 'P.glanduliferum', 'P.sukhakulii', 'P.wardii', 'P.ciliolare', 'P.dayanum', 'P.hennisianum', 'P.callosum', 'P.tonsum', 'P.javanicum', 'P.fowliei', 'P.schoseri', 'P.bougainvilleanum', 'P.hookerae', 'P.papuanum', 'P.mastersianum', 'P.argus', 'P.venustum', 'P.acmodontum', 'P.urbanianum', 'P.appletonianum', 'P.lawrenceanum', 'P.bullenianum', 'P.superbiens', 'P.purpuratum', 'P.barbatum']\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "todas_especies = []\n",
    "for registro_seq in SeqIO.parse('ls_orchid.fasta', 'fasta'):\n",
    "    todas_especies.append(registro_seq.description.split()[1])\n",
    "print(todas_especies)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Também é possível usar a *list comprehensions*:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['C.irapeanum', 'C.californicum', 'C.fasciculatum', 'C.margaritaceum', 'C.lichiangense', 'C.yatabeanum', 'C.guttatum', 'C.acaule', 'C.formosanum', 'C.himalaicum', 'C.macranthum', 'C.calceolus', 'C.segawai', 'C.pubescens', 'C.reginae', 'C.flavum', 'C.passerinum', 'M.xerophyticum', 'P.schlimii', 'P.besseae', 'P.wallisii', 'P.exstaminodium', 'P.caricinum', 'P.pearcei', 'P.longifolium', 'P.lindenii', 'P.lindleyanum', 'P.sargentianum', 'P.kaiteurum', 'P.czerwiakowianum', 'P.boissierianum', 'P.caudatum', 'P.warszewiczianum', 'P.micranthum', 'P.malipoense', 'P.delenatii', 'P.armeniacum', 'P.emersonii', 'P.niveum', 'P.godefroyae', 'P.bellatulum', 'P.concolor', 'P.fairrieanum', 'P.druryi', 'P.tigrinum', 'P.hirsutissimum', 'P.barbigerum', 'P.henryanum', 'P.charlesworthii', 'P.villosum', 'P.exul', 'P.insigne', 'P.gratrixianum', 'P.primulinum', 'P.victoria', 'P.victoria', 'P.glaucophyllum', 'P.supardii', 'P.kolopakingii', 'P.sanderianum', 'P.lowii', 'P.dianthum', 'P.parishii', 'P.haynaldianum', 'P.adductum', 'P.stonei', 'P.philippinense', 'P.rothschildianum', 'P.glanduliferum', 'P.glanduliferum', 'P.sukhakulii', 'P.wardii', 'P.ciliolare', 'P.dayanum', 'P.hennisianum', 'P.callosum', 'P.tonsum', 'P.javanicum', 'P.fowliei', 'P.schoseri', 'P.bougainvilleanum', 'P.hookerae', 'P.papuanum', 'P.mastersianum', 'P.argus', 'P.venustum', 'P.acmodontum', 'P.urbanianum', 'P.appletonianum', 'P.lawrenceanum', 'P.bullenianum', 'P.superbiens', 'P.purpuratum', 'P.barbatum']\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "todas_especies =[\n",
    "    registro_seq.description.split()[1]\n",
    "    for registro_seq in SeqIO.parse('ls_orchid.fasta', 'fasta')\n",
    "]\n",
    "print(todas_especies)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "De modo geral, extrair informações da linha de descrição de um arquivo FASTA não é uma tarefa muito legal. Você pode conseguir suas sequências de maneira bem anotadas em arquivos GenBank e EMBL, uma vez que esse tipo de anotação de informação é muito mais fácil de trabalhar."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5.1.5 Modificando os dados"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Na seção anterior, vimos como extrair os dados de um **SeqRecord**. Outra tarefa comum é alterar esses dados. Os atributos do **SeqRecord** podem ser modificados diretamente:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'gi|2765658|emb|Z78533.1|CIZ78533'"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "iterador_registro = SeqIO.parse('ls_orchid.fasta', 'fasta')\n",
    "primeiro_registro = next(iterador_registro)\n",
    "primeiro_registro.id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'novo_id'"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "primeiro_registro.id = 'novo_id'\n",
    "primeiro_registro.id"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Observe que se você quiser mudar a forma de como o arquivo FASTA é gerado quando gravado em um arquivo (ver a **Seção 5.5**), você deve modificar os atributos **id** e **description**. Para garantir o mesmo comportamento, é aconcelhável incluir o **id** seuido de um espaço em branco e depois o **description**:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">novo_id nova descrição\n",
      "CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGATGAGACCGTGGAATAAA\n",
      "CGATCGAGTGAATCCGGAGGACCGGTGTACTCAGCTCACCGGGGGCATTGCTCCCGTGGT\n",
      "GACCCTGATTTGTTGTTGGGCCGCCTCGGGAGCGTCCATGGCGGGTTTGAACCT\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "iterador_registro = SeqIO.parse('ls_orchid.fasta', 'fasta')\n",
    "primeiro_registro = next(iterador_registro)\n",
    "primeiro_registro.id = 'novo_id'\n",
    "primeiro_registro.description = primeiro_registro.id + ' ' + 'nova descrição'\n",
    "print(primeiro_registro.format('fasta')[:200])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5.2 Analisando sequências de arquivos compactos"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Na seção anterior vimos a análise dos dados de uma sequência. Em vez de usar o nome do arquivo, você pode fornecer ao `Bio.SeqIO` um identificador (ver a **Seção 24.1** para mais informações. Capítulo 24: ***link do capítulo 24***). Nessa seção usaremos identificadores para analisar sequências de arquivos compactos.\n",
    "\n",
    "Como você viu acima, podemos usar as funções **read** ou **parse** com o nome do arquivo para, por exemplo, calcular o tamanho total das sequências no arquivo GenBank:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "67518\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "print(sum(len(r) for r in SeqIO.parse('ls_orchid.gbk', 'genbank')))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Aqui usamos um identificador de arquivo, com a instância **with** para fechar o identificador automaticamente quando terminar a análise:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "67518\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "with open('ls_orchid.gbk') as identificador:\n",
    "    print(sum(len(r) for r in SeqIO.parse(identificador, 'genbank')))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ou podemos usar ainda o modo manual de se abrir e fechar o identificador (materiais mais antigos usam essa forma de escrita):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "67518\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "identificador = open('ls_orchid.gbk')\n",
    "print(sum(len(r) for r in SeqIO.parse(identificador, 'genbank')))\n",
    "identificador.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Suponha que você tenha um arquivo compacto, o que fazer? Esse tipo de distibição de arquivo é muito comum em Linux. Sabendo disso, vamos usar a biblioteca `gzip` (que já vem instalada com o Python) para abrir arquivos compactos e lê-los (o arquivo está no mesmo diretório desde documento):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "67518\n"
     ]
    }
   ],
   "source": [
    "import gzip\n",
    "from Bio import SeqIO\n",
    "with gzip.open('ls_orchid_gzip.gbk.gz', 'rt') as identificador:\n",
    "    print(sum(len(r) for r in SeqIO.parse(identificador, 'genbank')))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "De modo semelhante, podemos usar a biblioteca `bzip2` (já vem instalada com o Python). O arquivo para exemplo está no mesmo diretório desde documento:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "67518\n"
     ]
    }
   ],
   "source": [
    "import bz2\n",
    "from Bio import SeqIO\n",
    "with bz2.open('ls_orchid_bzip2.gbk.bz2', 'rt') as identificador:\n",
    "    print(sum(len(r) for r in SeqIO.parse(identificador, 'genbank')))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "O código acima mostra uma variante do gzip (GNU Zip) chamada de BGZF (Blocked GNU Zip Format), que pode ser tratado como um arquivo gzip comum para leitura, mas possui vantagens para acesso aleatório; falaremos disso mais tarde na **Seção 5.4.4**."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5.3 Analisando sequência da Internet"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Na seção anterior, vimos a análise de sequências armazendas em arquivos (usando o próprio nome ou um identificador) e em arquivos compactados (usando identificador). Agora usaremos o `Bio.SeqIO` com um outro tipo de identificador, a Internet para baixar e analisar as sequências em banco de dados online.\n",
    "\n",
    "Note que você *pode* fazer o download e a análise dos dados em um objeto **SeqRecord** de uma vez, mas isso não diz que é uma boa ideia. De modo geral, você deve baixar as sequências uma vez e salvá-la em um arquivo para usá-la mais vezes."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5.3.1 Análise de registros GenBank da Internet"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Na **Seção 9.6** do Capítulo 9 (***link do capítulo 9***) falaremos sobre a interface Entrez EFetch com mais detalhes, mas agora vamos apenas nos conectar ao NCBI e pegar várias sequências do gênero *Opuntia* (Figo da Índia) do GenBank.\n",
    "\n",
    "Antes de tudo, vamos buscar somente um registro. Se você não se importa com o download de anotações e características, baixa um arquivo FASTA é uma boa escolha, já que são compactos. Lembre-se, quando você estiver trabalhando com apenas um registro, use a função **read** de `Bio.SeqIO`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AF191665.1 com 0 características.\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "Entrez.email = 'teste@exemplo.com'  # sempre diga ao NCBI quem é você\n",
    "with Entrez.efetch(db='nucleotide', rettype='fasta', retmode='text', id='6273291') as identificador:\n",
    "    registro_seq = SeqIO.read(identificador, 'fasta')\n",
    "print(f'{registro_seq.id} com {len(registro_seq.features)} características.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "O NCBI sempre permitirá que você selecione o tipo de arquivo que deseja trabalhar. Desde o outono de 2009, a API do Entrez EFetch permitia que você use o 'genbank' como o tipo retornado dos dados. Contudo, o NCBI insiste em usar o tipo retornado de dados como 'gb' ('gp' para proteínas), como descito no *EFetch for Sequence and other Molecular Biology Databases* (https://www.ncbi.nlm.nih.gov/books/NBK3837/). Como resultado, desde a versão 1.50 do Biopython; 'gb' pode ser usado como substituto para 'genbank' em `Bio.SeqIO`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AF191665.1 com 3 características.\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "Entrez.email = 'teste@exemplo.com'\n",
    "with Entrez.efetch(db='nucleotide', rettype='gb', retmode='text', id='6273291') as identificador:\n",
    "    registro_seq = SeqIO.read(identificador, 'gb')\n",
    "print(f'{registro_seq.id} com {len(registro_seq.features)} características.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Observe que agora temos três características.\n",
    "\n",
    "Agora vamos buscar por vários registros. Como o identificador agora terá múltiplos registros, usaremos a função **parse** de `Bio.SeqIO`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AF191665.1 Opuntia marenae rpl16 gene; chloroplast gene for c...\n",
      "Tamanho da sequência: 902, 3, de: chloroplast Grusonia marenae\n",
      "\n",
      "\n",
      "AF191664.1 Opuntia clavata rpl16 gene; chloroplast gene for c...\n",
      "Tamanho da sequência: 899, 3, de: chloroplast Grusonia clavata\n",
      "\n",
      "\n",
      "AF191663.1 Opuntia bradtiana rpl16 gene; chloroplast gene for...\n",
      "Tamanho da sequência: 899, 3, de: chloroplast Grusonia bradtiana\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "Entrez.email = 'teste@exemplo.com'\n",
    "with Entrez.efetch(db='nucleotide', rettype='gb', retmode='text', id='6273291,6273290,6273289') as identificador:\n",
    "    for registro_seq in SeqIO.parse(identificador, 'gb'):\n",
    "        print(f'{registro_seq.id} {registro_seq.description[:50]}...')\n",
    "        print(f'Tamanho da sequência: {len(registro_seq)}, {len(registro_seq.features)}, '\n",
    "              f'de: {registro_seq.annotations[\"source\"]}')\n",
    "        print('\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Veja o Capítulo 9 (***link do capítulo 9***) para mais informações sobe o módulo `Bio.Entrez` e leiaa **Seção 9.1** sobre as diretrizes do NCBI para usar o Entrez."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5.3.2 Análise de sequências SwissProt da Internet"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Agora vamos usar um identificador para baixar um arquivo SwissProt do ExPASy, que serão mais detalhados no Capítulo 10 (***link do capítulo 10***). Como mencionado acima, quando você esteja trabalhando somente com um registro, use a função **read** de `Bio.SeqIO`: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "O23729\n",
      "CHS3_BROFI\n",
      "RecName: Full=Chalcone synthase 3; EC=2.3.1.74; AltName: Full=Naringenin-chalcone synthase 3;\n",
      "Seq('MAPAMEEIRQAQRAEGPAAVLAIGTSTPPNALYQADYPDYYFRITKSEHLTELK...GAE')\n",
      "Tamanho 394\n",
      "['Acyltransferase', 'Flavonoid biosynthesis', 'Transferase']\n"
     ]
    }
   ],
   "source": [
    "from Bio import ExPASy\n",
    "from Bio import SeqIO\n",
    "with ExPASy.get_sprot_raw('O23729') as identificador:\n",
    "    registro_seq = SeqIO.read(identificador, 'swiss')\n",
    "print(registro_seq.id)\n",
    "print(registro_seq.name)\n",
    "print(registro_seq.description)\n",
    "print(repr(registro_seq.seq))\n",
    "print(f'Tamanho {len(registro_seq)}')\n",
    "print(registro_seq.annotations['keywords'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5.4 Arquivos de sequência como dicionários"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Fazer um loop através de um iterador **SeqIO.parse** pode esgotar o arquivo. Para arquivos auto indexador, como o formato de arquivo twoBit, o valor retornado pelo iterador pode ser usado como o valor em um dicionário, permitindo acesso aleatório ao conteúdo da sequência. Como nesse caso a análise é feita sob demanda, o arquivo deve permanecer aberto a todo momento enquanto os dados da sequência estiverem sendo acessados (o arquivo usado de exemplo está no mesmo diretório que este documento):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "identificador = open('sequence.bigendian.2bit', 'rb')\n",
    "registros = SeqIO.parse(identificador, 'twobit')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['seq11111', 'seq222', 'seq3333', 'seq4', 'seq555', 'seq6'])"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "registros.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SeqRecord(seq=Seq('TTGATCGGTGACAAATTTTTTACAAAGAACTGTAGGACTTGCTACTTCTCCCTC...ACA'), id='seq222', name='<unknown name>', description='<unknown description>', dbxrefs=[])"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "registros['seq222']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('TTGATCGGTGACAAATTTTTTACAAAGAACTGTAGGACTTGCTACTTCTCCCTC...ACA')"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "registros['seq222'].seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "cannot retrieve sequence: file is closed",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32mD:\\Jupyter\\lib\\site-packages\\IPython\\core\\formatters.py\u001b[0m in \u001b[0;36m__call__\u001b[1;34m(self, obj)\u001b[0m\n\u001b[0;32m    700\u001b[0m                 \u001b[0mtype_pprinters\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mtype_printers\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    701\u001b[0m                 deferred_pprinters=self.deferred_printers)\n\u001b[1;32m--> 702\u001b[1;33m             \u001b[0mprinter\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mpretty\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    703\u001b[0m             \u001b[0mprinter\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mflush\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    704\u001b[0m             \u001b[1;32mreturn\u001b[0m \u001b[0mstream\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mgetvalue\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mD:\\Jupyter\\lib\\site-packages\\IPython\\lib\\pretty.py\u001b[0m in \u001b[0;36mpretty\u001b[1;34m(self, obj)\u001b[0m\n\u001b[0;32m    392\u001b[0m                         \u001b[1;32mif\u001b[0m \u001b[0mcls\u001b[0m \u001b[1;32mis\u001b[0m \u001b[1;32mnot\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;31m \u001b[0m\u001b[0;31m\\\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    393\u001b[0m                                 \u001b[1;32mand\u001b[0m \u001b[0mcallable\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mcls\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m__dict__\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mget\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'__repr__'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 394\u001b[1;33m                             \u001b[1;32mreturn\u001b[0m \u001b[0m_repr_pprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mcycle\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    395\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    396\u001b[0m             \u001b[1;32mreturn\u001b[0m \u001b[0m_default_pprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mcycle\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mD:\\Jupyter\\lib\\site-packages\\IPython\\lib\\pretty.py\u001b[0m in \u001b[0;36m_repr_pprint\u001b[1;34m(obj, p, cycle)\u001b[0m\n\u001b[0;32m    698\u001b[0m     \u001b[1;34m\"\"\"A pprint that just redirects to the normal repr function.\"\"\"\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    699\u001b[0m     \u001b[1;31m# Find newlines and replace them with p.break_()\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 700\u001b[1;33m     \u001b[0moutput\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mrepr\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    701\u001b[0m     \u001b[0mlines\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0moutput\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0msplitlines\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    702\u001b[0m     \u001b[1;32mwith\u001b[0m \u001b[0mp\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mgroup\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mD:\\Jupyter\\lib\\site-packages\\Bio\\Seq.py\u001b[0m in \u001b[0;36m__repr__\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    315\u001b[0m             \u001b[1;31m# there is a stop codon at the end of a sequence.\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    316\u001b[0m             \u001b[1;31m# Note total length is 54+3+3=60\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 317\u001b[1;33m             \u001b[0mstart\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mdata\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;36m54\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mdecode\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"ASCII\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    318\u001b[0m             \u001b[0mend\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mdata\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;33m-\u001b[0m\u001b[1;36m3\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mdecode\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"ASCII\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    319\u001b[0m             \u001b[1;32mreturn\u001b[0m \u001b[1;34mf\"{self.__class__.__name__}('{start}...{end}')\"\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mD:\\Jupyter\\lib\\site-packages\\Bio\\SeqIO\\TwoBitIO.py\u001b[0m in \u001b[0;36m__getitem__\u001b[1;34m(self, key)\u001b[0m\n\u001b[0;32m    128\u001b[0m         \u001b[1;32mexcept\u001b[0m \u001b[0mValueError\u001b[0m \u001b[1;32mas\u001b[0m \u001b[0mexception\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    129\u001b[0m             \u001b[1;32mif\u001b[0m \u001b[0mstr\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mexception\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;33m==\u001b[0m \u001b[1;34m\"seek of closed file\"\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 130\u001b[1;33m                 \u001b[1;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"cannot retrieve sequence: file is closed\"\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;32mfrom\u001b[0m \u001b[1;32mNone\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    131\u001b[0m             \u001b[1;32mraise\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    132\u001b[0m         \u001b[0mdata\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mnumpy\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mfromfile\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mstream\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m\"uint8\"\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mcount\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mbyteSize\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mValueError\u001b[0m: cannot retrieve sequence: file is closed"
     ]
    }
   ],
   "source": [
    "identificador.close()\n",
    "registros['seq222'].seq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Para outros formatos de arquivo, o `Bio.SeqIO` fornece três funções relacionadas que permitem acesso ao dicionário com os dados de um arquivo com múltiplas sequências. Isso gera uma troca entre flexibilidade e uso de memória. Em resumo:\n",
    "- `Bio.SeqIO.to_dict()`: é mais flexível mas demanda de mais memória (veja a **Seção 5.4.1**). Essa função basicamente constrói um dicionário Python com cada entrada mantida como um objeto **SeqRecord** na memória, permitindo você modificar os regitros;\n",
    "\n",
    "- `Bio.SeqIO.index()`: é um meio termo, atua como um leitor de dicionários e analisador de sequências em objetos **SeqRecord** sob demanda (veja a **Seção 5.4.2**) e;\n",
    "\n",
    "- `Bio.SeqIO.index_db()`: também atua como um leitor de dicionário, mas armazena os identificadores e os *offsets* em um arquivo no disco (como um banco de dados SQLite3), o que indica que tem requisitos de memória bem baixos (veja a **Seção 5.4.3**), mas será um pouco mais lento.\n",
    "\n",
    "Na **Seção 5.4.5** você terá uma visão mais geral."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5.4.1 Arquivos de sequência como dicionários (na memória)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A próxima coisa que faremos com nossos onipresentes arquivos de orquídeas é mostrar como indexá-los e acessá-los como um banco de dados usando dicionários Python (como hash em Perl). Isso é uma mão na roda para moderar arquivos grandes onde você necessita acessar somente alguns dados do arquivo, é um jeito gambiarra de se fazer, mas funciona muito bem. Para lidar com arquivos maiores onde a memória torna-se um problema, veja a **Seção 5.4.2**.\n",
    "\n",
    "Você pode usar o `Bio.SeqIO.to_dict` para fazer um dicionário com os registros (na memória). Por padrão, isso é usado para cada identificador do registro (i.e. o argumento **id**) com a chave. Vamos usar em nosso arquivo GenBank:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "dic_orquidea = SeqIO.to_dict(SeqIO.parse('ls_orchid.gbk', 'gb'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "É apenas requerido um argmento para a função **to_dict**, uma lista ou gerador fornecido pelos objetos **SeqRecord**. Aqui nós usamos a saída da função **parse**. Como o nome sugere, é a função **to_dict** retorna um dicionário Python.\n",
    "\n",
    "Desde que a variável **dic_orquidea** é um dicionário Python, podemos olhar para todas as chaves acessíveis:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "94"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(dic_orquidea)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Z78533.1',\n",
       " 'Z78532.1',\n",
       " 'Z78531.1',\n",
       " 'Z78530.1',\n",
       " 'Z78529.1',\n",
       " 'Z78527.1',\n",
       " 'Z78526.1',\n",
       " 'Z78525.1',\n",
       " 'Z78524.1',\n",
       " 'Z78523.1',\n",
       " 'Z78522.1',\n",
       " 'Z78521.1',\n",
       " 'Z78520.1',\n",
       " 'Z78519.1',\n",
       " 'Z78518.1',\n",
       " 'Z78517.1',\n",
       " 'Z78516.1',\n",
       " 'Z78515.1',\n",
       " 'Z78514.1',\n",
       " 'Z78513.1',\n",
       " 'Z78512.1',\n",
       " 'Z78511.1',\n",
       " 'Z78510.1',\n",
       " 'Z78509.1',\n",
       " 'Z78508.1',\n",
       " 'Z78507.1',\n",
       " 'Z78506.1',\n",
       " 'Z78505.1',\n",
       " 'Z78504.1',\n",
       " 'Z78503.1',\n",
       " 'Z78502.1',\n",
       " 'Z78501.1',\n",
       " 'Z78500.1',\n",
       " 'Z78499.1',\n",
       " 'Z78498.1',\n",
       " 'Z78497.1',\n",
       " 'Z78496.1',\n",
       " 'Z78495.1',\n",
       " 'Z78494.1',\n",
       " 'Z78493.1',\n",
       " 'Z78492.1',\n",
       " 'Z78491.1',\n",
       " 'Z78490.1',\n",
       " 'Z78489.1',\n",
       " 'Z78488.1',\n",
       " 'Z78487.1',\n",
       " 'Z78486.1',\n",
       " 'Z78485.1',\n",
       " 'Z78484.1',\n",
       " 'Z78483.1',\n",
       " 'Z78482.1',\n",
       " 'Z78481.1',\n",
       " 'Z78480.1',\n",
       " 'Z78479.1',\n",
       " 'Z78478.1',\n",
       " 'Z78477.1',\n",
       " 'Z78476.1',\n",
       " 'Z78475.1',\n",
       " 'Z78474.1',\n",
       " 'Z78473.1',\n",
       " 'Z78472.1',\n",
       " 'Z78471.1',\n",
       " 'Z78470.1',\n",
       " 'Z78469.1',\n",
       " 'Z78468.1',\n",
       " 'Z78467.1',\n",
       " 'Z78466.1',\n",
       " 'Z78465.1',\n",
       " 'Z78464.1',\n",
       " 'Z78463.1',\n",
       " 'Z78462.1',\n",
       " 'Z78461.1',\n",
       " 'Z78460.1',\n",
       " 'Z78459.1',\n",
       " 'Z78458.1',\n",
       " 'Z78457.1',\n",
       " 'Z78456.1',\n",
       " 'Z78455.1',\n",
       " 'Z78454.1',\n",
       " 'Z78453.1',\n",
       " 'Z78452.1',\n",
       " 'Z78451.1',\n",
       " 'Z78450.1',\n",
       " 'Z78449.1',\n",
       " 'Z78448.1',\n",
       " 'Z78447.1',\n",
       " 'Z78446.1',\n",
       " 'Z78445.1',\n",
       " 'Z78444.1',\n",
       " 'Z78443.1',\n",
       " 'Z78442.1',\n",
       " 'Z78441.1',\n",
       " 'Z78440.1',\n",
       " 'Z78439.1']"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(dic_orquidea.keys())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "No Python 3, os métodos do dicionário (como **keys** e **values**) são iteradores ao invés de listas. Se você realmente quiser, pode até olhar todos os registros de uma só vez:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGATGAGACCGTGG...CGC'), id='Z78533.1', name='Z78533', description='C.irapeanum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAACAG...GGC'), id='Z78532.1', name='Z78532', description='C.californicum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAGCAG...TAA'), id='Z78531.1', name='Z78531', description='C.fasciculatum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAAACAACAT...CAT'), id='Z78530.1', name='Z78530', description='C.margaritaceum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('ACGGCGAGCTGCCGAAGGACATTGTTGAGACAGCAGAATATACGATTGAGTGAA...AAA'), id='Z78529.1', name='Z78529', description='C.lichiangense 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAGTAG...CCC'), id='Z78527.1', name='Z78527', description='C.yatabeanum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAGTAG...TGT'), id='Z78526.1', name='Z78526', description='C.guttatum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TGTTGAGATAGCAGAATATACATCGAGTGAATCCGGAGGACCTGTGGTTATTCG...GCA'), id='Z78525.1', name='Z78525', description='C.acaule 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATAGTAG...AGC'), id='Z78524.1', name='Z78524', description='C.formosanum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACCAGGTTTCCGTAGGTGAACCTGCGGCAGGATCATTGTTGAGACAGCAG...AAG'), id='Z78523.1', name='Z78523', description='C.himalaicum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAGCAG...GAG'), id='Z78522.1', name='Z78522', description='C.macranthum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('GTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAGTAGAATATATGATCGAGT...ACC'), id='Z78521.1', name='Z78521', description='C.calceolus 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAGCAG...TTT'), id='Z78520.1', name='Z78520', description='C.segawai 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('ATATGATCGAGTGAATCTGGTGGACTTGTGGTTACTCAGCTCGCCATAGGCTTT...TTA'), id='Z78519.1', name='Z78519', description='C.pubescens 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGGAGGATCATTGTTGAGATAGTAG...TCC'), id='Z78518.1', name='Z78518', description='C.reginae 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAGTAG...AGC'), id='Z78517.1', name='Z78517', description='C.flavum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAGTAT...TAA'), id='Z78516.1', name='Z78516', description='C.passerinum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGCTGAGACCGTAG...AGC'), id='Z78515.1', name='Z78515', description='M.xerophyticum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGGACCTTCGGGAGGATCATTTTTGAAGCCCCCA...CTA'), id='Z78514.1', name='Z78514', description='P.schlimii 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACCGCCA...GAG'), id='Z78513.1', name='Z78513', description='P.besseae 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGGACCTTCGGGAGGATCATTTTTGAAGCCCCCA...AGC'), id='Z78512.1', name='Z78512', description='P.wallisii 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTTCGGAAGGATCATTGTTGAGACCCCCA...GGA'), id='Z78511.1', name='Z78511', description='P.exstaminodium 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CTAACCAGGGTTCCGAGGTGACCTTCGGGAGGATTCCTTTTTAAGCCCCCGAAA...TTA'), id='Z78510.1', name='Z78510', description='P.caricinum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACCGCCA...GGA'), id='Z78509.1', name='Z78509', description='P.pearcei 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACCGCCA...TGA'), id='Z78508.1', name='Z78508', description='P.longifolium 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACCCCCA...TGA'), id='Z78507.1', name='Z78507', description='P.lindenii 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACCGCAA...TGA'), id='Z78506.1', name='Z78506', description='P.lindleyanum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACCGCCA...TTT'), id='Z78505.1', name='Z78505', description='P.sargentianum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTTCGGAAGGATCATTGTTGAGACCGCAA...TAA'), id='Z78504.1', name='Z78504', description='P.kaiteurum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACCAGGTTTCCGTAGGTGAACCTCCGGAAGGATCCTTGTTGAGACCGCCA...TAA'), id='Z78503.1', name='Z78503', description='P.czerwiakowianum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACCAGGTTTCCGTAGGTGAACCTCCGGAAGGATCATTGTTGAGACCGCCA...CGC'), id='Z78502.1', name='Z78502', description='P.boissierianum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACCGCAA...AGA'), id='Z78501.1', name='Z78501', description='P.caudatum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGCTCATTGTTGAGACCGCAA...AAG'), id='Z78500.1', name='Z78500', description='P.warszewiczianum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAGGGATCATTGTTGAGATCGCAT...ACC'), id='Z78499.1', name='Z78499', description='P.micranthum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAAGGTCATTGTTGAGATCACAT...AGC'), id='Z78498.1', name='Z78498', description='P.malipoense 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AGC'), id='Z78497.1', name='Z78497', description='P.delenatii 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCGCAT...AGC'), id='Z78496.1', name='Z78496', description='P.armeniacum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTCCGGAAGGATCATTGTTGAGATCACAT...GTG'), id='Z78495.1', name='Z78495', description='P.emersonii 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGGTCGCAT...AAG'), id='Z78494.1', name='Z78494', description='P.niveum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGATTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCGCAT...CCC'), id='Z78493.1', name='Z78493', description='P.godefroyae 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCGCAT...ATA'), id='Z78492.1', name='Z78492', description='P.bellatulum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCGCAT...AGC'), id='Z78491.1', name='Z78491', description='P.concolor 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TGA'), id='Z78490.1', name='Z78490', description='P.fairrieanum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...GGC'), id='Z78489.1', name='Z78489', description='P.druryi 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CTGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACGCAATAATTGATCGA...GCT'), id='Z78488.1', name='Z78488', description='P.tigrinum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TAA'), id='Z78487.1', name='Z78487', description='P.hirsutissimum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTCACGAGGTTTCCGTAGGTGAATCTGCGGGAGGATCATTGTTGAGATCACAT...TGA'), id='Z78486.1', name='Z78486', description='P.barbigerum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CTGAACCTGGTGTCCGAAGGTGAATCTGCGGATGGATCATTGTTGAGATATCAT...GTA'), id='Z78485.1', name='Z78485', description='P.henryanum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGGGGAAGGATCATTGTTGAGATCACAT...TTT'), id='Z78484.1', name='Z78484', description='P.charlesworthii 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...GCA'), id='Z78483.1', name='Z78483', description='P.villosum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('TCTACTGCAGTGACCGAGATTTGCCATCGAGCCTCCTGGGAGCTTTCTTGCTGG...GCA'), id='Z78482.1', name='Z78482', description='P.exul 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TGA'), id='Z78481.1', name='Z78481', description='P.insigne 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TGA'), id='Z78480.1', name='Z78480', description='P.gratrixianum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AGT'), id='Z78479.1', name='Z78479', description='P.primulinum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTCCGGAAGGATCAGTGTTGAGATCACAT...GGC'), id='Z78478.1', name='Z78478', description='P.victoria 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TGC'), id='Z78477.1', name='Z78477', description='P.victoria 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...CCC'), id='Z78476.1', name='Z78476', description='P.glaucophyllum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...GGT'), id='Z78475.1', name='Z78475', description='P.supardii 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACGT...CTT'), id='Z78474.1', name='Z78474', description='P.kolopakingii 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AGG'), id='Z78473.1', name='Z78473', description='P.sanderianum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AGC'), id='Z78472.1', name='Z78472', description='P.lowii 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AGC'), id='Z78471.1', name='Z78471', description='P.dianthum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...GTT'), id='Z78470.1', name='Z78470', description='P.parishii 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...GTT'), id='Z78469.1', name='Z78469', description='P.haynaldianum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCGCAT...GTT'), id='Z78468.1', name='Z78468', description='P.adductum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TGA'), id='Z78467.1', name='Z78467', description='P.stonei 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...CCC'), id='Z78466.1', name='Z78466', description='P.philippinense 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TGC'), id='Z78465.1', name='Z78465', description='P.rothschildianum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAGCGGAAGGGTCATTGTTGAGATCACATAATA...AGC'), id='Z78464.1', name='Z78464', description='P.glanduliferum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGTTCATTGTTGAGATCACAT...AGC'), id='Z78463.1', name='Z78463', description='P.glanduliferum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTCACGAGGTCTCCGGATGTGACCCTGCGGAAGGATCATTGTTGAGATCACAT...CAT'), id='Z78462.1', name='Z78462', description='P.sukhakulii 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTCCGGAAGGATCATTGTTGAGATCACAT...TAA'), id='Z78461.1', name='Z78461', description='P.wardii 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTCCGGAAGGATCATTGTTGAGATCACAT...TTA'), id='Z78460.1', name='Z78460', description='P.ciliolare 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TTT'), id='Z78459.1', name='Z78459', description='P.dayanum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TTG'), id='Z78458.1', name='Z78458', description='P.hennisianum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTCCGGAAGGATCATTGTTGAGATCACAT...GAG'), id='Z78457.1', name='Z78457', description='P.callosum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AGC'), id='Z78456.1', name='Z78456', description='P.tonsum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACCAGGTTTCCGTAGGTGGACCTTCGGGAGGATCATTTTTGAGATCACAT...GCA'), id='Z78455.1', name='Z78455', description='P.javanicum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AAC'), id='Z78454.1', name='Z78454', description='P.fowliei 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...GCA'), id='Z78453.1', name='Z78453', description='P.schoseri 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...GCA'), id='Z78452.1', name='Z78452', description='P.bougainvilleanum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGTACCTCCGGAAGGATCATTGTTGAGATCACAT...AGC'), id='Z78451.1', name='Z78451', description='P.hookerae 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('GGAAGGATCATTGCTGATATCACATAATAATTGATCGAGTTAAGCTGGAGGATC...GAG'), id='Z78450.1', name='Z78450', description='P.papuanum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TGC'), id='Z78449.1', name='Z78449', description='P.mastersianum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AGG'), id='Z78448.1', name='Z78448', description='P.argus 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGATTCCGTAGGTGAACCTGCGGGAGGATCATTGTTGAGATCACAT...AGC'), id='Z78447.1', name='Z78447', description='P.venustum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTCCGGAAGGATCATTGTTGAGATCACAT...CCC'), id='Z78446.1', name='Z78446', description='P.acmodontum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TGT'), id='Z78445.1', name='Z78445', description='P.urbanianum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGGTGAACTGCGGAAGGATCATTGTTGAGATCACAT...ATT'), id='Z78444.1', name='Z78444', description='P.appletonianum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AGG'), id='Z78443.1', name='Z78443', description='P.lawrenceanum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('GTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACATAATAATTGATCGAGT...AGT'), id='Z78442.1', name='Z78442', description='P.bullenianum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('GGAAGGTCATTGCCGATATCACATAATAATTGATCGAGTTAATCTGGAGGATCT...GAG'), id='Z78441.1', name='Z78441', description='P.superbiens 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CGTAACAAGGTTTCCGTAGGTGGACCTCCGGGAGGATCATTGTTGAGATCACAT...GCA'), id='Z78440.1', name='Z78440', description='P.purpuratum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('CATTGTTGAGATCACATAATAATTGATCGAGTTAATCTGGAGGATCTGTTTACT...GCC'), id='Z78439.1', name='Z78439', description='P.barbatum 5.8S rRNA gene and ITS1 and ITS2 DNA', dbxrefs=[])]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(dic_orquidea.values())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Você pode acessar um único objeto **SeqRecord** através das chaves e manipulação que já fizemos:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "P.supardii 5.8S rRNA gene and ITS1 and ITS2 DNA\n"
     ]
    }
   ],
   "source": [
    "registro_seq = dic_orquidea['Z78475.1']\n",
    "print(registro_seq.description)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...GGT')"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "registro_seq.seq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "É muito fácil criar um 'banco de dados' na memória com os registros do arquivo GenBank. Agora vamos fazer a mesma coisa com o arquivo FASTA.\n",
    "\n",
    "Vale ressaltar que se você tiver o conhecimeto básico em Python pode fazer um dicionário como este 'na mão'. Contudo, os métodos de construção de um dicionário não lidam muito bem com chaves repetidas. Usar a função **to_dict** permite que seja verificada a duplicidade de chaves e retornar uma exceção caso exista essa duplicidade."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 5.4.1.1 Especificando as chaves do dicionário"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Usaremos o mesmo código acima, mas com um arquivo FASTA:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "dic_orquidea = SeqIO.to_dict(SeqIO.parse('ls_orchid.fasta', 'fasta'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Agora as chaves são assim:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['gi|2765658|emb|Z78533.1|CIZ78533', 'gi|2765657|emb|Z78532.1|CCZ78532', 'gi|2765656|emb|Z78531.1|CFZ78531', 'gi|2765655|emb|Z78530.1|CMZ78530', 'gi|2765654|emb|Z78529.1|CLZ78529', 'gi|2765652|emb|Z78527.1|CYZ78527', 'gi|2765651|emb|Z78526.1|CGZ78526', 'gi|2765650|emb|Z78525.1|CAZ78525', 'gi|2765649|emb|Z78524.1|CFZ78524', 'gi|2765648|emb|Z78523.1|CHZ78523', 'gi|2765647|emb|Z78522.1|CMZ78522', 'gi|2765646|emb|Z78521.1|CCZ78521', 'gi|2765645|emb|Z78520.1|CSZ78520', 'gi|2765644|emb|Z78519.1|CPZ78519', 'gi|2765643|emb|Z78518.1|CRZ78518', 'gi|2765642|emb|Z78517.1|CFZ78517', 'gi|2765641|emb|Z78516.1|CPZ78516', 'gi|2765640|emb|Z78515.1|MXZ78515', 'gi|2765639|emb|Z78514.1|PSZ78514', 'gi|2765638|emb|Z78513.1|PBZ78513', 'gi|2765637|emb|Z78512.1|PWZ78512', 'gi|2765636|emb|Z78511.1|PEZ78511', 'gi|2765635|emb|Z78510.1|PCZ78510', 'gi|2765634|emb|Z78509.1|PPZ78509', 'gi|2765633|emb|Z78508.1|PLZ78508', 'gi|2765632|emb|Z78507.1|PLZ78507', 'gi|2765631|emb|Z78506.1|PLZ78506', 'gi|2765630|emb|Z78505.1|PSZ78505', 'gi|2765629|emb|Z78504.1|PKZ78504', 'gi|2765628|emb|Z78503.1|PCZ78503', 'gi|2765627|emb|Z78502.1|PBZ78502', 'gi|2765626|emb|Z78501.1|PCZ78501', 'gi|2765625|emb|Z78500.1|PWZ78500', 'gi|2765624|emb|Z78499.1|PMZ78499', 'gi|2765623|emb|Z78498.1|PMZ78498', 'gi|2765622|emb|Z78497.1|PDZ78497', 'gi|2765621|emb|Z78496.1|PAZ78496', 'gi|2765620|emb|Z78495.1|PEZ78495', 'gi|2765619|emb|Z78494.1|PNZ78494', 'gi|2765618|emb|Z78493.1|PGZ78493', 'gi|2765617|emb|Z78492.1|PBZ78492', 'gi|2765616|emb|Z78491.1|PCZ78491', 'gi|2765615|emb|Z78490.1|PFZ78490', 'gi|2765614|emb|Z78489.1|PDZ78489', 'gi|2765613|emb|Z78488.1|PTZ78488', 'gi|2765612|emb|Z78487.1|PHZ78487', 'gi|2765611|emb|Z78486.1|PBZ78486', 'gi|2765610|emb|Z78485.1|PHZ78485', 'gi|2765609|emb|Z78484.1|PCZ78484', 'gi|2765608|emb|Z78483.1|PVZ78483', 'gi|2765607|emb|Z78482.1|PEZ78482', 'gi|2765606|emb|Z78481.1|PIZ78481', 'gi|2765605|emb|Z78480.1|PGZ78480', 'gi|2765604|emb|Z78479.1|PPZ78479', 'gi|2765603|emb|Z78478.1|PVZ78478', 'gi|2765602|emb|Z78477.1|PVZ78477', 'gi|2765601|emb|Z78476.1|PGZ78476', 'gi|2765600|emb|Z78475.1|PSZ78475', 'gi|2765599|emb|Z78474.1|PKZ78474', 'gi|2765598|emb|Z78473.1|PSZ78473', 'gi|2765597|emb|Z78472.1|PLZ78472', 'gi|2765596|emb|Z78471.1|PDZ78471', 'gi|2765595|emb|Z78470.1|PPZ78470', 'gi|2765594|emb|Z78469.1|PHZ78469', 'gi|2765593|emb|Z78468.1|PAZ78468', 'gi|2765592|emb|Z78467.1|PSZ78467', 'gi|2765591|emb|Z78466.1|PPZ78466', 'gi|2765590|emb|Z78465.1|PRZ78465', 'gi|2765589|emb|Z78464.1|PGZ78464', 'gi|2765588|emb|Z78463.1|PGZ78463', 'gi|2765587|emb|Z78462.1|PSZ78462', 'gi|2765586|emb|Z78461.1|PWZ78461', 'gi|2765585|emb|Z78460.1|PCZ78460', 'gi|2765584|emb|Z78459.1|PDZ78459', 'gi|2765583|emb|Z78458.1|PHZ78458', 'gi|2765582|emb|Z78457.1|PCZ78457', 'gi|2765581|emb|Z78456.1|PTZ78456', 'gi|2765580|emb|Z78455.1|PJZ78455', 'gi|2765579|emb|Z78454.1|PFZ78454', 'gi|2765578|emb|Z78453.1|PSZ78453', 'gi|2765577|emb|Z78452.1|PBZ78452', 'gi|2765576|emb|Z78451.1|PHZ78451', 'gi|2765575|emb|Z78450.1|PPZ78450', 'gi|2765574|emb|Z78449.1|PMZ78449', 'gi|2765573|emb|Z78448.1|PAZ78448', 'gi|2765572|emb|Z78447.1|PVZ78447', 'gi|2765571|emb|Z78446.1|PAZ78446', 'gi|2765570|emb|Z78445.1|PUZ78445', 'gi|2765569|emb|Z78444.1|PAZ78444', 'gi|2765568|emb|Z78443.1|PLZ78443', 'gi|2765567|emb|Z78442.1|PBZ78442', 'gi|2765566|emb|Z78441.1|PSZ78441', 'gi|2765565|emb|Z78440.1|PPZ78440', 'gi|2765564|emb|Z78439.1|PBZ78439'])\n"
     ]
    }
   ],
   "source": [
    "print(dic_orquidea.keys())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Se você observou bem, essas strings são as mesmas que as vistas na **Seção 2.4.1** do Capítulo 2 (https://github.com/GTL98/biopython_br/tree/main/Cap%202%20-%20O%20que%20podemos%20fazer%20com%20o%20Biopython). Suponha que você prefere ter as chaves de um modo diferente, como só os números de acesso. Sabendo disso, a função **to_dict** possui o parâmetro **key_function**, onde podemos definir uma função que será usada nas chaves do resgistro. Vale ressaltar que o código abaixo serve para o arquivo de exemplo, fique atento as atualizações do arquivo FASTA para mudar a função sempre que mudar algo no arquivo FASTA:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "def num_acesso(registro):\n",
    "    \"\"\"Entrgar um objeto SeqRecord e retornar o número de acesso como uma string\n",
    "    \n",
    "    e.g. 'gi|2765613|emb|Z78488.1|PTZ78488' --> 'Z78488.1'\n",
    "    \"\"\"\n",
    "    partes = registro.id.split('|')\n",
    "    assert len(partes) == 5 and partes[0] == 'gi' and partes[2] == 'emb'\n",
    "    return partes[3]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Então podemos entregar essa função para **to_dict** para a construção do dicionário:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "dic_orquidea = SeqIO.to_dict(SeqIO.parse('ls_orchid.fasta', 'fasta'), key_function=num_acesso)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finalmente, como desejado, as novas chaves do dicionário:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['Z78533.1', 'Z78532.1', 'Z78531.1', 'Z78530.1', 'Z78529.1', 'Z78527.1', 'Z78526.1', 'Z78525.1', 'Z78524.1', 'Z78523.1', 'Z78522.1', 'Z78521.1', 'Z78520.1', 'Z78519.1', 'Z78518.1', 'Z78517.1', 'Z78516.1', 'Z78515.1', 'Z78514.1', 'Z78513.1', 'Z78512.1', 'Z78511.1', 'Z78510.1', 'Z78509.1', 'Z78508.1', 'Z78507.1', 'Z78506.1', 'Z78505.1', 'Z78504.1', 'Z78503.1', 'Z78502.1', 'Z78501.1', 'Z78500.1', 'Z78499.1', 'Z78498.1', 'Z78497.1', 'Z78496.1', 'Z78495.1', 'Z78494.1', 'Z78493.1', 'Z78492.1', 'Z78491.1', 'Z78490.1', 'Z78489.1', 'Z78488.1', 'Z78487.1', 'Z78486.1', 'Z78485.1', 'Z78484.1', 'Z78483.1', 'Z78482.1', 'Z78481.1', 'Z78480.1', 'Z78479.1', 'Z78478.1', 'Z78477.1', 'Z78476.1', 'Z78475.1', 'Z78474.1', 'Z78473.1', 'Z78472.1', 'Z78471.1', 'Z78470.1', 'Z78469.1', 'Z78468.1', 'Z78467.1', 'Z78466.1', 'Z78465.1', 'Z78464.1', 'Z78463.1', 'Z78462.1', 'Z78461.1', 'Z78460.1', 'Z78459.1', 'Z78458.1', 'Z78457.1', 'Z78456.1', 'Z78455.1', 'Z78454.1', 'Z78453.1', 'Z78452.1', 'Z78451.1', 'Z78450.1', 'Z78449.1', 'Z78448.1', 'Z78447.1', 'Z78446.1', 'Z78445.1', 'Z78444.1', 'Z78443.1', 'Z78442.1', 'Z78441.1', 'Z78440.1', 'Z78439.1'])\n"
     ]
    }
   ],
   "source": [
    "print(dic_orquidea.keys())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Nada muito complicado, espero!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 5.4.1.2 Indexar um dicionário usando a soma de verificação SEGUID"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Como outro exemplo do trabalho com dicionários do objetos **SeqRecord**, vamos usar a função de soma de verificação SEGUID. Essa é uma soma de verificação recente e as colições devem ser bem raras (i.e. duas sequências diferentes com a mesma soma de verificação), uma melhoria na soma de verificação CRC64.\n",
    "\n",
    "Vamos trabalhar novamente com o nosso arquivo GenBank:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Z78533.1 JUEoWn6DPhgZ9nAyowsgtoD9TTo\n",
      "Z78532.1 MN/s0q9zDoCVEEc+k/IFwCNF2pY\n",
      "Z78531.1 xN45pACrTnmBH8a8Y9cWSgoLrwE\n",
      "Z78530.1 yMhI5UUQfFOPcoJXb9B19XUyYlY\n",
      "Z78529.1 s1Pnjq9zoSHoI/CG9jQr4GyeMZY\n",
      "Z78527.1 MRf6S1OYhtbdPVS845oCmLTqMgo\n",
      "Z78526.1 QCDzCtL6AKuc+h4UQDD6wFjz3Vs\n",
      "Z78525.1 Zh/FImuuDRmmM/5fXaCYAHS7wo0\n",
      "Z78524.1 hKw0C1fPNpi2KUM6iV0/8IadTX0\n",
      "Z78523.1 71sZ82r6eAOBpwZs2solBr/biUs\n",
      "Z78522.1 /8BBi+Dm0kpdHcYREmyKKbQSzKQ\n",
      "Z78521.1 VrTEM/VCul51xFo2OBDsZKgiFTI\n",
      "Z78520.1 //QmLx77H51zTbd8LR+71pXY9ew\n",
      "Z78519.1 rb1CrF1Gbr6UDMLeRLPOWZaKIBc\n",
      "Z78518.1 38aHi4XYVQUZkydCrxSVn2SYmCM\n",
      "Z78517.1 swq+/4vgyWJssxu11Lcx+2IxNC4\n",
      "Z78516.1 +CFmEqzVnCluPe39ytjiLRL7dvU\n",
      "Z78515.1 ZG15PBF2qqDhj75CGTDiRAuT8Is\n",
      "Z78514.1 Rvx7kJnPQJKgBZzNZ5dx+CnksgY\n",
      "Z78513.1 IjG6hja6MS3M/CzKI9AuxFnjFxc\n",
      "Z78512.1 EqkQVKx2qD2+qLp/W6hwQ1GjKo8\n",
      "Z78511.1 cnm0fCmTFbHMlTN6t7i/ee4ydDc\n",
      "Z78510.1 GVd6GFv7uF0uq66u4gZB95scKFY\n",
      "Z78509.1 tS2JwZWXudMwdlyJBXKxjm/G1Ik\n",
      "Z78508.1 e85/JrwCy5T7J+zptVbhBw8govc\n",
      "Z78507.1 +pQOTxhX/lW2fFWnZ7hmBDbLz9w\n",
      "Z78506.1 AUA6LlNQUBe4Vtui414zw6evNQY\n",
      "Z78505.1 LTPf5nwmqYaoaNYD7nqb8QZZvKM\n",
      "Z78504.1 /EOQScJF+tmjObpwV1ujB8sUuFI\n",
      "Z78503.1 IA/IHU4qHwSC53tnBLuiVc/QSw4\n",
      "Z78502.1 BEeHOaq0TqDvtvpjOiv9BzdriuY\n",
      "Z78501.1 ddme0ctIDntTIfqbt+nUux6cmAE\n",
      "Z78500.1 XpWsuWowv3eOeXHfvx+U7pYPFjI\n",
      "Z78499.1 TAdFQoCPhjyGewA0jUVQlqaaJ+8\n",
      "Z78498.1 mOL1AF6RYoy+kgF+kqPEj36V21U\n",
      "Z78497.1 0/IKS3K4h6llrMitD1+DhRoXh1E\n",
      "Z78496.1 RbfK+HNz5NPcb27AVhz8LQAmoaY\n",
      "Z78495.1 EDrD6GbdkndGpQ5uCbs+7G1Y+BM\n",
      "Z78494.1 UTwhnUE1auIOvnpZncJ66UKK+tA\n",
      "Z78493.1 EKpDickgW5VN+ZC0REMXwXibwFw\n",
      "Z78492.1 AoEtEBoiFDQ62fRhZzm6iu2yqxU\n",
      "Z78491.1 iOtnBkEnCdtxp/sUlZD87/O9K+4\n",
      "Z78490.1 jazbRXVqeDy/X6b8TExuqZFBndw\n",
      "Z78489.1 IbyzgyMk3PCE+ZjlENmclwc9p6g\n",
      "Z78488.1 zoC93s1ZqKuZmJdIClLyUxOy3+M\n",
      "Z78487.1 iC/DUEJ8v1TjmDsiH7isULxYkCo\n",
      "Z78486.1 rHoLsPA87sRSfavlupM1lPM/i4A\n",
      "Z78485.1 QM54knrAHva3f8VMOW1PCmaA5gs\n",
      "Z78484.1 hqVogP9q5ylo64FkDTYcVwjjls4\n",
      "Z78483.1 x+5s3bSieMINp7e9513bWB0WYmI\n",
      "Z78482.1 XnJOy06wOb+5QIhfX9A4V7qb92U\n",
      "Z78481.1 JmaG2IWTkH34eWK+KCCip9eAGLY\n",
      "Z78480.1 D27nfpSx9o2+7GPjTjY6UHU/nPo\n",
      "Z78479.1 HaIM62dIBcplTdxEcQcP7SKQbxc\n",
      "Z78478.1 r8JKJUvNKACsIu88s0qYb29JFN8\n",
      "Z78477.1 vkGng4nfa/dvmrf3WVV5tyjLoHg\n",
      "Z78476.1 zh9Dp1+pKUQNAThekUrXzMz89rk\n",
      "Z78475.1 BhWYrGw+l+hSTmzKOoWpimcF2uA\n",
      "Z78474.1 cTnWtDcURdi6BZxnjg85lE03jmM\n",
      "Z78473.1 pyzdQlyyX2kmBkt1S2iwm3ycY0Y\n",
      "Z78472.1 a3PV4ZvUIgERjhTpwWUoHhxWsWM\n",
      "Z78471.1 RRQ1Lz3nMbZMKw2JASe4JuNW5Wg\n",
      "Z78470.1 K8wBHUqmWhq4TQug6yV2bPT6aM4\n",
      "Z78469.1 4haNqGOzkb8noLLZk+u/VuSkfCI\n",
      "Z78468.1 YDmyRkyPgmL841LHj+XcTzOkXdo\n",
      "Z78467.1 +jCRlZLZLp3XjI5x3m2g4pIROmM\n",
      "Z78466.1 jfFcaXxqmXpgP+VFAb9JhdUfyHw\n",
      "Z78465.1 yk4sNU0F0caUXOKYX7kcJPPO3Mo\n",
      "Z78464.1 fwfYA8BS6/bloU6y4+L1zOmkKj0\n",
      "Z78463.1 8t+iwYVJSWQ/wo/ArXqtYZuUy+Q\n",
      "Z78462.1 PEJCjbUAJrMGxwaMMsYxtt4t5YE\n",
      "Z78461.1 stsMhWdzjr9epky+4mbxvHM8ogQ\n",
      "Z78460.1 SM1fL9+9O3CypA8fHC1S+ABggbg\n",
      "Z78459.1 6IN7iuaDyk6DuLftwslcPBASyts\n",
      "Z78458.1 g5n6cv2y17Y7i5R3vN4N1LxvRSY\n",
      "Z78457.1 zUvpKJfH6skjVFBiCA+R3c+6MCE\n",
      "Z78456.1 e7V4imTShLvzhDLzIH/pQErKZek\n",
      "Z78455.1 KnRQrK1fWDMVemltKPF3xjJ6lUs\n",
      "Z78454.1 4Gnab8LwTdw9zEtSTKSgYX5EPA8\n",
      "Z78453.1 0+mvhvhsuxsW78QKyeZugpinzRc\n",
      "Z78452.1 6ABkQeqBFxRcaV+94WG/w9Pafqs\n",
      "Z78451.1 +Mx9joE9+a5ZgObEGq/NAupqsyo\n",
      "Z78450.1 sSYXtzIvxY4NpoRex23VCWQ3uAc\n",
      "Z78449.1 /7boTKvbRdKIPuUeMZaljyK8zCI\n",
      "Z78448.1 klX7lthrUMgsCSzT02FN/rTnpgY\n",
      "Z78447.1 kg4ILddCRIwMms/EO4nRqXu3VmA\n",
      "Z78446.1 zi1FYc6X85tcZBDo2fpiOEX7YDo\n",
      "Z78445.1 Qy5l9GBO+69+MFDnRAHeoOLKzjw\n",
      "Z78444.1 yB/N9GkdbOdA8x94hPZwunFmqQA\n",
      "Z78443.1 p7CGx55KuAhTySRUwJzHo/e56/U\n",
      "Z78442.1 RL79qd+aTH4kpek8q3nUwQdTiEg\n",
      "Z78441.1 4UjNimWz3PcmI0F0BOC5IksDrOg\n",
      "Z78440.1 kFkgs9JK3psvuvwaD0rp9iCRcdY\n",
      "Z78439.1 H+JfaShya/4yyAj7IbMqgNkxdxQ\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.SeqUtils.CheckSum import seguid\n",
    "for registro in SeqIO.parse('ls_orchid.gbk', 'gb'):\n",
    "    print(registro.id, seguid(registro.seq))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Agora, vamos chamar a função **to_dict** com o parâmetro **key_function** transformando um objeto **SeqRecord** em string. Nós não podemos usar a função **seguid** diretamente porque ela espera que seja entregada um objeto **Seq** ou uma string. Porém, podemos usar a função **lambda** para criar uma função 'única' e forncecer à **to_dict**:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.SeqUtils.CheckSum import seguid\n",
    "dic_seguid = SeqIO.to_dict(SeqIO.parse('ls_orchid.gbk', 'gb'), lambda rec: seguid(rec.seq))\n",
    "registro = dic_seguid['MN/s0q9zDoCVEEc+k/IFwCNF2pY']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Z78532.1\n"
     ]
    }
   ],
   "source": [
    "print(registro.id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C.californicum 5.8S rRNA gene and ITS1 and ITS2 DNA\n"
     ]
    }
   ],
   "source": [
    "print(registro.description)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Foi retornado o registro Z78532.1, a segunda entrada do arquivo."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5.4.2 Arquivos de sequência como dicionários (arquivos indexados)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Como mostrado nos exemplos anteriores, usar a função **to_dict** de `Bio.SeqIO` é bem flexível. Todavia, devido tudo ser colocado na memória, o tamanho do arquivo que você pode trabalhar será limitado pela memória RAM. De modo geral, esse modo é usado somente para pequenos e médios arquivos.\n",
    "\n",
    "Para grandes arquivos, você deve considerar o uso da função **index** de `Bio.SeqIO` que trabalhar um pouco diferente. Embora ainda retorne um dicionário como um objeto, essa função *não* guarda *tudo* na memória. Ao invés disso, a função apenas registra onde cada registro está no arquivo quando perguntada por um registro em questão ou uma análise por demanda.\n",
    "\n",
    "Como exemplo, vamos usar nosso velho conhecido arquivo GenBank:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "dic_orquidea = SeqIO.index('ls_orchid.gbk', 'gb')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "94"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(dic_orquidea)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Z78533.1',\n",
       " 'Z78532.1',\n",
       " 'Z78531.1',\n",
       " 'Z78530.1',\n",
       " 'Z78529.1',\n",
       " 'Z78527.1',\n",
       " 'Z78526.1',\n",
       " 'Z78525.1',\n",
       " 'Z78524.1',\n",
       " 'Z78523.1',\n",
       " 'Z78522.1',\n",
       " 'Z78521.1',\n",
       " 'Z78520.1',\n",
       " 'Z78519.1',\n",
       " 'Z78518.1',\n",
       " 'Z78517.1',\n",
       " 'Z78516.1',\n",
       " 'Z78515.1',\n",
       " 'Z78514.1',\n",
       " 'Z78513.1',\n",
       " 'Z78512.1',\n",
       " 'Z78511.1',\n",
       " 'Z78510.1',\n",
       " 'Z78509.1',\n",
       " 'Z78508.1',\n",
       " 'Z78507.1',\n",
       " 'Z78506.1',\n",
       " 'Z78505.1',\n",
       " 'Z78504.1',\n",
       " 'Z78503.1',\n",
       " 'Z78502.1',\n",
       " 'Z78501.1',\n",
       " 'Z78500.1',\n",
       " 'Z78499.1',\n",
       " 'Z78498.1',\n",
       " 'Z78497.1',\n",
       " 'Z78496.1',\n",
       " 'Z78495.1',\n",
       " 'Z78494.1',\n",
       " 'Z78493.1',\n",
       " 'Z78492.1',\n",
       " 'Z78491.1',\n",
       " 'Z78490.1',\n",
       " 'Z78489.1',\n",
       " 'Z78488.1',\n",
       " 'Z78487.1',\n",
       " 'Z78486.1',\n",
       " 'Z78485.1',\n",
       " 'Z78484.1',\n",
       " 'Z78483.1',\n",
       " 'Z78482.1',\n",
       " 'Z78481.1',\n",
       " 'Z78480.1',\n",
       " 'Z78479.1',\n",
       " 'Z78478.1',\n",
       " 'Z78477.1',\n",
       " 'Z78476.1',\n",
       " 'Z78475.1',\n",
       " 'Z78474.1',\n",
       " 'Z78473.1',\n",
       " 'Z78472.1',\n",
       " 'Z78471.1',\n",
       " 'Z78470.1',\n",
       " 'Z78469.1',\n",
       " 'Z78468.1',\n",
       " 'Z78467.1',\n",
       " 'Z78466.1',\n",
       " 'Z78465.1',\n",
       " 'Z78464.1',\n",
       " 'Z78463.1',\n",
       " 'Z78462.1',\n",
       " 'Z78461.1',\n",
       " 'Z78460.1',\n",
       " 'Z78459.1',\n",
       " 'Z78458.1',\n",
       " 'Z78457.1',\n",
       " 'Z78456.1',\n",
       " 'Z78455.1',\n",
       " 'Z78454.1',\n",
       " 'Z78453.1',\n",
       " 'Z78452.1',\n",
       " 'Z78451.1',\n",
       " 'Z78450.1',\n",
       " 'Z78449.1',\n",
       " 'Z78448.1',\n",
       " 'Z78447.1',\n",
       " 'Z78446.1',\n",
       " 'Z78445.1',\n",
       " 'Z78444.1',\n",
       " 'Z78443.1',\n",
       " 'Z78442.1',\n",
       " 'Z78441.1',\n",
       " 'Z78440.1',\n",
       " 'Z78439.1']"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(dic_orquidea.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "P.supardii 5.8S rRNA gene and ITS1 and ITS2 DNA\n"
     ]
    }
   ],
   "source": [
    "registro_seq = dic_orquidea['Z78475.1']\n",
    "print(registro_seq.description)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...GGT')"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "registro_seq.seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "dic_orquidea.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Observe que a função **index** não necessita de um identificador, somente o nome do arquivo. Há boas razões para isso, mas é um pouco técnica a explicação. O segundo argumento é o formato de arquivo. Você pode usar diversos formatos de arquivos, incluindo FASTA e FASTQ (veja mais na **Seção 20.1.11** no Capítulo 20: ***link do capítulo 20***). Contudo, arquivos de alinhamento como PHYLIP ou Clustal não são suportados. Finalmente, como argumento opcional, você pode forncer uma função chave.\n",
    "\n",
    "Abaixo está o mesmo código, mas usando um arquivo FASTA:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "dic_orquidea = SeqIO.index('ls_orchid.fasta', 'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "94"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(dic_orquidea)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['gi|2765658|emb|Z78533.1|CIZ78533',\n",
       " 'gi|2765657|emb|Z78532.1|CCZ78532',\n",
       " 'gi|2765656|emb|Z78531.1|CFZ78531',\n",
       " 'gi|2765655|emb|Z78530.1|CMZ78530',\n",
       " 'gi|2765654|emb|Z78529.1|CLZ78529',\n",
       " 'gi|2765652|emb|Z78527.1|CYZ78527',\n",
       " 'gi|2765651|emb|Z78526.1|CGZ78526',\n",
       " 'gi|2765650|emb|Z78525.1|CAZ78525',\n",
       " 'gi|2765649|emb|Z78524.1|CFZ78524',\n",
       " 'gi|2765648|emb|Z78523.1|CHZ78523',\n",
       " 'gi|2765647|emb|Z78522.1|CMZ78522',\n",
       " 'gi|2765646|emb|Z78521.1|CCZ78521',\n",
       " 'gi|2765645|emb|Z78520.1|CSZ78520',\n",
       " 'gi|2765644|emb|Z78519.1|CPZ78519',\n",
       " 'gi|2765643|emb|Z78518.1|CRZ78518',\n",
       " 'gi|2765642|emb|Z78517.1|CFZ78517',\n",
       " 'gi|2765641|emb|Z78516.1|CPZ78516',\n",
       " 'gi|2765640|emb|Z78515.1|MXZ78515',\n",
       " 'gi|2765639|emb|Z78514.1|PSZ78514',\n",
       " 'gi|2765638|emb|Z78513.1|PBZ78513',\n",
       " 'gi|2765637|emb|Z78512.1|PWZ78512',\n",
       " 'gi|2765636|emb|Z78511.1|PEZ78511',\n",
       " 'gi|2765635|emb|Z78510.1|PCZ78510',\n",
       " 'gi|2765634|emb|Z78509.1|PPZ78509',\n",
       " 'gi|2765633|emb|Z78508.1|PLZ78508',\n",
       " 'gi|2765632|emb|Z78507.1|PLZ78507',\n",
       " 'gi|2765631|emb|Z78506.1|PLZ78506',\n",
       " 'gi|2765630|emb|Z78505.1|PSZ78505',\n",
       " 'gi|2765629|emb|Z78504.1|PKZ78504',\n",
       " 'gi|2765628|emb|Z78503.1|PCZ78503',\n",
       " 'gi|2765627|emb|Z78502.1|PBZ78502',\n",
       " 'gi|2765626|emb|Z78501.1|PCZ78501',\n",
       " 'gi|2765625|emb|Z78500.1|PWZ78500',\n",
       " 'gi|2765624|emb|Z78499.1|PMZ78499',\n",
       " 'gi|2765623|emb|Z78498.1|PMZ78498',\n",
       " 'gi|2765622|emb|Z78497.1|PDZ78497',\n",
       " 'gi|2765621|emb|Z78496.1|PAZ78496',\n",
       " 'gi|2765620|emb|Z78495.1|PEZ78495',\n",
       " 'gi|2765619|emb|Z78494.1|PNZ78494',\n",
       " 'gi|2765618|emb|Z78493.1|PGZ78493',\n",
       " 'gi|2765617|emb|Z78492.1|PBZ78492',\n",
       " 'gi|2765616|emb|Z78491.1|PCZ78491',\n",
       " 'gi|2765615|emb|Z78490.1|PFZ78490',\n",
       " 'gi|2765614|emb|Z78489.1|PDZ78489',\n",
       " 'gi|2765613|emb|Z78488.1|PTZ78488',\n",
       " 'gi|2765612|emb|Z78487.1|PHZ78487',\n",
       " 'gi|2765611|emb|Z78486.1|PBZ78486',\n",
       " 'gi|2765610|emb|Z78485.1|PHZ78485',\n",
       " 'gi|2765609|emb|Z78484.1|PCZ78484',\n",
       " 'gi|2765608|emb|Z78483.1|PVZ78483',\n",
       " 'gi|2765607|emb|Z78482.1|PEZ78482',\n",
       " 'gi|2765606|emb|Z78481.1|PIZ78481',\n",
       " 'gi|2765605|emb|Z78480.1|PGZ78480',\n",
       " 'gi|2765604|emb|Z78479.1|PPZ78479',\n",
       " 'gi|2765603|emb|Z78478.1|PVZ78478',\n",
       " 'gi|2765602|emb|Z78477.1|PVZ78477',\n",
       " 'gi|2765601|emb|Z78476.1|PGZ78476',\n",
       " 'gi|2765600|emb|Z78475.1|PSZ78475',\n",
       " 'gi|2765599|emb|Z78474.1|PKZ78474',\n",
       " 'gi|2765598|emb|Z78473.1|PSZ78473',\n",
       " 'gi|2765597|emb|Z78472.1|PLZ78472',\n",
       " 'gi|2765596|emb|Z78471.1|PDZ78471',\n",
       " 'gi|2765595|emb|Z78470.1|PPZ78470',\n",
       " 'gi|2765594|emb|Z78469.1|PHZ78469',\n",
       " 'gi|2765593|emb|Z78468.1|PAZ78468',\n",
       " 'gi|2765592|emb|Z78467.1|PSZ78467',\n",
       " 'gi|2765591|emb|Z78466.1|PPZ78466',\n",
       " 'gi|2765590|emb|Z78465.1|PRZ78465',\n",
       " 'gi|2765589|emb|Z78464.1|PGZ78464',\n",
       " 'gi|2765588|emb|Z78463.1|PGZ78463',\n",
       " 'gi|2765587|emb|Z78462.1|PSZ78462',\n",
       " 'gi|2765586|emb|Z78461.1|PWZ78461',\n",
       " 'gi|2765585|emb|Z78460.1|PCZ78460',\n",
       " 'gi|2765584|emb|Z78459.1|PDZ78459',\n",
       " 'gi|2765583|emb|Z78458.1|PHZ78458',\n",
       " 'gi|2765582|emb|Z78457.1|PCZ78457',\n",
       " 'gi|2765581|emb|Z78456.1|PTZ78456',\n",
       " 'gi|2765580|emb|Z78455.1|PJZ78455',\n",
       " 'gi|2765579|emb|Z78454.1|PFZ78454',\n",
       " 'gi|2765578|emb|Z78453.1|PSZ78453',\n",
       " 'gi|2765577|emb|Z78452.1|PBZ78452',\n",
       " 'gi|2765576|emb|Z78451.1|PHZ78451',\n",
       " 'gi|2765575|emb|Z78450.1|PPZ78450',\n",
       " 'gi|2765574|emb|Z78449.1|PMZ78449',\n",
       " 'gi|2765573|emb|Z78448.1|PAZ78448',\n",
       " 'gi|2765572|emb|Z78447.1|PVZ78447',\n",
       " 'gi|2765571|emb|Z78446.1|PAZ78446',\n",
       " 'gi|2765570|emb|Z78445.1|PUZ78445',\n",
       " 'gi|2765569|emb|Z78444.1|PAZ78444',\n",
       " 'gi|2765568|emb|Z78443.1|PLZ78443',\n",
       " 'gi|2765567|emb|Z78442.1|PBZ78442',\n",
       " 'gi|2765566|emb|Z78441.1|PSZ78441',\n",
       " 'gi|2765565|emb|Z78440.1|PPZ78440',\n",
       " 'gi|2765564|emb|Z78439.1|PBZ78439']"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(dic_orquidea.keys())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 5.4.2.1 Especificando as chaves do dicionário"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Suponha que você queira usar as mesmas chaves usadas anteriormente, como fazer? Assim como fizemos na função **to_dict** na **Seção 5.4.1.1**, você precisará escrever uma função para modificar a identificação no arquivo FASTA para a chave que você quer. Para isso, vamos usar a função **num_acesso** que fizemos anteriormente com algumas modificações junto com a função **index** dessa vez:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "def obter_chave(identificador):\n",
    "    \"\"\"Entrgar um objeto SeqRecord e retornar o número de acesso como uma string\n",
    "    \n",
    "    e.g. 'gi|2765613|emb|Z78488.1|PTZ78488' --> 'Z78488.1'\n",
    "    \"\"\"\n",
    "    partes = identificador.split('|')\n",
    "    assert len(partes) == 5 and partes[0] == 'gi' and partes[2] == 'emb'\n",
    "    return partes[3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "dic_orquidea = SeqIO.index('ls_orchid.fasta', 'fasta', key_function=obter_chave)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Z78533.1',\n",
       " 'Z78532.1',\n",
       " 'Z78531.1',\n",
       " 'Z78530.1',\n",
       " 'Z78529.1',\n",
       " 'Z78527.1',\n",
       " 'Z78526.1',\n",
       " 'Z78525.1',\n",
       " 'Z78524.1',\n",
       " 'Z78523.1',\n",
       " 'Z78522.1',\n",
       " 'Z78521.1',\n",
       " 'Z78520.1',\n",
       " 'Z78519.1',\n",
       " 'Z78518.1',\n",
       " 'Z78517.1',\n",
       " 'Z78516.1',\n",
       " 'Z78515.1',\n",
       " 'Z78514.1',\n",
       " 'Z78513.1',\n",
       " 'Z78512.1',\n",
       " 'Z78511.1',\n",
       " 'Z78510.1',\n",
       " 'Z78509.1',\n",
       " 'Z78508.1',\n",
       " 'Z78507.1',\n",
       " 'Z78506.1',\n",
       " 'Z78505.1',\n",
       " 'Z78504.1',\n",
       " 'Z78503.1',\n",
       " 'Z78502.1',\n",
       " 'Z78501.1',\n",
       " 'Z78500.1',\n",
       " 'Z78499.1',\n",
       " 'Z78498.1',\n",
       " 'Z78497.1',\n",
       " 'Z78496.1',\n",
       " 'Z78495.1',\n",
       " 'Z78494.1',\n",
       " 'Z78493.1',\n",
       " 'Z78492.1',\n",
       " 'Z78491.1',\n",
       " 'Z78490.1',\n",
       " 'Z78489.1',\n",
       " 'Z78488.1',\n",
       " 'Z78487.1',\n",
       " 'Z78486.1',\n",
       " 'Z78485.1',\n",
       " 'Z78484.1',\n",
       " 'Z78483.1',\n",
       " 'Z78482.1',\n",
       " 'Z78481.1',\n",
       " 'Z78480.1',\n",
       " 'Z78479.1',\n",
       " 'Z78478.1',\n",
       " 'Z78477.1',\n",
       " 'Z78476.1',\n",
       " 'Z78475.1',\n",
       " 'Z78474.1',\n",
       " 'Z78473.1',\n",
       " 'Z78472.1',\n",
       " 'Z78471.1',\n",
       " 'Z78470.1',\n",
       " 'Z78469.1',\n",
       " 'Z78468.1',\n",
       " 'Z78467.1',\n",
       " 'Z78466.1',\n",
       " 'Z78465.1',\n",
       " 'Z78464.1',\n",
       " 'Z78463.1',\n",
       " 'Z78462.1',\n",
       " 'Z78461.1',\n",
       " 'Z78460.1',\n",
       " 'Z78459.1',\n",
       " 'Z78458.1',\n",
       " 'Z78457.1',\n",
       " 'Z78456.1',\n",
       " 'Z78455.1',\n",
       " 'Z78454.1',\n",
       " 'Z78453.1',\n",
       " 'Z78452.1',\n",
       " 'Z78451.1',\n",
       " 'Z78450.1',\n",
       " 'Z78449.1',\n",
       " 'Z78448.1',\n",
       " 'Z78447.1',\n",
       " 'Z78446.1',\n",
       " 'Z78445.1',\n",
       " 'Z78444.1',\n",
       " 'Z78443.1',\n",
       " 'Z78442.1',\n",
       " 'Z78441.1',\n",
       " 'Z78440.1',\n",
       " 'Z78439.1']"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(dic_orquidea.keys())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Fica fácil quando você sabe como fazer né?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 5.4.2.2 Obter os dados brutos para um registro"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Um dicionário vindo do `Bio.SeqIO.index()` entrega para você cada entrada de um objeto **SeqRecord**. Entretanto, algumas vezes é melhor obter os dados brutos originais diretamente do arquivo. Para isso, basta usar o método **get_raw** que precisa somente de um argumento (o identificador do registro) e te retorna uma string de bytes (extraída do arquivo sem modificação).\n",
    "\n",
    "Um bom exemplo é extrair um subconjunto de registros vindos de um grande arquivo onde também o `Bio.SeqIO.write()` não suporta (ainda) o formato de saída (e.g. um texto de um arquivo SwissProt) ou  onde você precisa preservar um texto exatamente como está (e.g. saídas do Biopython com arquivos GenBank e EMBL ainda não preservam os últimos bits de informação).\n",
    "\n",
    "Vamos supor que você tenha baixado todo o UniProt no formato de arquivo SwissProt do site FTP (https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.dat.gz). Vamos descompactá-lo e usar o arquivo para extrair os diversos registros:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "uniprot = SeqIO.index('uniprot_sprot.dat', 'swiss')\n",
    "with open('selected.dat', 'wb') as identificador_saida:\n",
    "    for acc in ['P33487', 'P19801', 'P13689', 'Q8JZQ5', 'Q9TRC7']:\n",
    "        identificador_saida.write(uniprot.get_raw(acc))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "O arquivo gerado pelo código acima está no mesmo diretório que este documento.\n",
    "\n",
    "Observe que o modo de abertura de arquivo foi feita com o modo de escrita em binário. Isso é importante porque o método **get_raw** retorna uma string em bytes.\n",
    "\n",
    "Na **Seção 20.1.5** do Capítulo 20 (***link do capítulo 20***) há um exemplo completo de como usar a função **index** para classificar um grande arquivo de sequência, sem carregar tudo na memória de uma vez."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5.4.3 Arquivos de sequência como dicionários (arquivos indexados de banco de dados)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**pag 60**"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
